{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc2f91d7-5a60-4c4f-a8da-f683dc8600a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#check how many peptides\n",
    "#then skip the segments calculation and gnomAD query, as it was alredy done\n",
    "#retreave positions of PSMs that were left with hybrid filtering\n",
    "#take new numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80788e2e-6742-4922-b932-e62d9e5a4c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from explore_import import  *\n",
    "import ionbot_preprocess as io\n",
    "import data_preprocess as dt\n",
    "import hpp_checker as hpp\n",
    "\n",
    "import asyncio\n",
    "import pickle\n",
    "import time\n",
    "import os\n",
    "import itertools \n",
    "import time\n",
    "import re\n",
    "import ast\n",
    "from scipy.spatial import distance\n",
    "from gql import gql, Client\n",
    "from gql.transport.aiohttp import AIOHTTPTransport\n",
    "import asyncio\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "\n",
    "from Download_UnimodDB import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "35c027d3-b3d6-4e7e-8a81-5eef58e18d60",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_peptide_sequences=pickle.load(open('./../all_peptide_sequences_OP_7302.pkl', 'rb'))\n",
    "is_unique=pickle.load(open('./../is_unique_OP_7302.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "19f8ea8e-8524-440c-9506-80c4689b8688",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_peptide_sequences_byprot={e:k for k,v in all_peptide_sequences.items() for e in v}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a002d77-f234-44a8-b624-aa58ee90321d",
   "metadata": {},
   "outputs": [],
   "source": [
    "OPfasta_file='./../ionbot_openprot/openprot_2_0_0-human-ensembl106+refseq+uniprot2022_06_01.fasta'\n",
    "input_file = open(OPfasta_file)\n",
    "OPfasta = SeqIO.to_dict(SeqIO.parse(OPfasta_file, \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "279b8897-0cfa-429e-92a1-2724d2919a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://ftp.ensembl.org/pub/release-106/fasta/homo_sapiens/dna/\n",
    "ENSfasta_file=\"./ionbot_openprot/Homo_sapiens.GRCh38.dna.primary_assembly.fa\"\n",
    "input_file = open(ENSfasta_file)\n",
    "ENSfasta = SeqIO.to_dict(SeqIO.parse(ENSfasta_file, \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "178859cc-169b-4b2e-a3a4-0a47e4744e73",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('combined_first_datasets.pickle', 'rb') as handle:\n",
    "    combined_first_datasets=pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "403d8b84-07f5-4405-b439-9c628118a347",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Download_Unimod_Dict_names():\n",
    "    unimod = Download_UnimodDB()\n",
    "    condensed_unimod = {}\n",
    "    for uni_id,df in unimod.groupby(\"unimod_id\").__iter__():\n",
    "        condensed_unimod[uni_id] = {}\n",
    "        condensed_unimod[uni_id]['residues'] = \"\".join([_ for _ in df.residue if '-' not in _])\n",
    "        condensed_unimod[uni_id]['mono_mass'] = [_ for _ in set(df.mono_mass)][0]\n",
    "        condensed_unimod[uni_id]['code_name'] = [_ for _ in set(df.code_name)][0]\n",
    "        condensed_unimod[uni_id]['full_name'] = [_ for _ in set(df.full_name)][0]\n",
    "        \n",
    "    return condensed_unimod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "195206d2-9590-4951-ad03-752415cbd7b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3-letter to 1-letter amino acid mapping\n",
    "aa_map = {\n",
    "    \"Ala\": \"A\", \"Cys\": \"C\", \"Asp\": \"D\", \"Glu\": \"E\", \"Phe\": \"F\",\n",
    "    \"Gly\": \"G\", \"His\": \"H\", \"Ile\": \"I\", \"Lys\": \"K\", \"Leu\": \"L\",\n",
    "    \"Met\": \"M\", \"Asn\": \"N\", \"Pro\": \"P\", \"Gln\": \"Q\", \"Arg\": \"R\",\n",
    "    \"Ser\": \"S\", \"Thr\": \"T\", \"Val\": \"V\", \"Trp\": \"W\", \"Tyr\": \"Y\",\n",
    "    \"Xle\": \"L\",  # Xle usually refers to Leu/Ile, mapped to Leu (\"L\") here\n",
    "    \"CamCys\": \"C\", \"MetOx\": \"M\"  # Optional: include PTMs if needed\n",
    "}\n",
    "\n",
    "# Your list of mutations\n",
    "mutations = ['His2Asn', 'His2Asp', 'Ala->Ser', 'Ala->Thr', 'Ala->Asp', ...]  # etc.\n",
    "\n",
    "# Function to extract 1-letter code of the target amino acid\n",
    "def extract_target_aa(mutation):\n",
    "    if \"->\" in mutation:\n",
    "        target = mutation.split(\"->\")[1]\n",
    "    elif \"2\" in mutation:\n",
    "        target = mutation.split(\"2\")[1]\n",
    "    else:\n",
    "        return None  # or raise error if unexpected format\n",
    "\n",
    "    return aa_map.get(target, \"?\")  # return '?' if unknown\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "765c3bbd-22f1-4e40-9e49-23713433352e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_peptide_position_io(s):\n",
    "    match = re.search(r\"\\(\\((\\d+)-(\\d+)\\)\\)\", s)    \n",
    "    return (int(match.group(1)) , int(match.group(2))) if match else None\n",
    "\n",
    "def filter_nested_intervals(intervals):\n",
    "    if len(intervals)==1: return True\n",
    "    # Sort by start ascending, and end descending to prioritize wider intervals\n",
    "    sorted_intervals = sorted(intervals, key=lambda x: (x[0], -x[1]))\n",
    "    result = []\n",
    "\n",
    "    for i, (start_i, end_i) in enumerate(sorted_intervals):\n",
    "        is_nested = False\n",
    "        for j, (start_j, end_j) in enumerate(sorted_intervals):\n",
    "            if i != j and start_j <= start_i and end_i <= end_j:\n",
    "                is_nested = True\n",
    "                break\n",
    "        if not is_nested:\n",
    "            result.append((start_i, end_i))\n",
    "    result_bool=[i in result for i in intervals]\n",
    "    #print(result_bool)\n",
    "    return result_bool\n",
    "def filter_short_overlaps(i,overlapping_intervals):\n",
    "    indexes,intervals = overlapping_intervals\n",
    "    for indx,interval in zip(indexes,intervals):\n",
    "        #print(indx,interval)\n",
    "        if i in indx and len(indx)==1:\n",
    "            #print(indx)\n",
    "            return True\n",
    "        elif i in indx and len(indx)>1:\n",
    "            #print(indx,interval,abs(interval[0]-interval[1]))\n",
    "            return abs(interval[0]-interval[1])>=18\n",
    "        else:\n",
    "            continue\n",
    "def check_2_pep(group):\n",
    "    return len(group)>=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "27931240-30ca-48de-b376-3c62a6f98503",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/210214903.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  unimod_df_subs[\"subs_to_residue\"]=unimod_df_subs.code_name.apply(lambda x: extract_target_aa(x))\n"
     ]
    }
   ],
   "source": [
    "unimod = Download_Unimod_Dict_names()\n",
    "unimod_df=pd.DataFrame.from_dict(unimod, orient='index')\n",
    "#mark substitutions\n",
    "unimod_df[\"isSubstitution\"]=unimod_df.full_name.apply(lambda x: \"substitution\" in x)\n",
    "unimod_df_subs=unimod_df[unimod_df[\"isSubstitution\"]]\n",
    "#to what residue is change\n",
    "unimod_df_subs[\"subs_to_residue\"]=unimod_df_subs.code_name.apply(lambda x: extract_target_aa(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "67074618-dd7a-4bc5-af75-eb8048dbbf89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1565"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(unimod)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab534622-fc2d-4c93-94d4-28d8a7f000c4",
   "metadata": {},
   "source": [
    "#### select peptides with potential substitutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "83cbf7a9-bd9a-4cbb-b7b6-41c6cfbe4acd",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_datasets=pd.DataFrame()\n",
    "for dataset_name,subdict in combined_first_datasets.items():\n",
    "    df=subdict[\"openprot\"]\n",
    "    df[\"dataset\"]=dataset_name\n",
    "    comb_datasets=pd.concat([comb_datasets,df])\n",
    "comb_datasets.reset_index(drop=True,inplace=True)\n",
    "comb_datasets[\"peptide_length\"]=comb_datasets.database_peptide.apply(lambda x: len(x))\n",
    "comb_datasets[\"peptide_position\"]=comb_datasets.proteins.apply(lambda x: [get_peptide_position_io(s) for s in x.split(\"||\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "263212e6-404f-4d0e-a884-6ac07fefd8df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "117062"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#total psms with mass shift)\n",
    "len(comb_datasets[comb_datasets.modifications!=\"Unmodified\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5e74f33f-4da2-4cf8-9497-c653a7d36543",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18655\n"
     ]
    }
   ],
   "source": [
    "#take only peptides unique to class\n",
    "\n",
    "comb_datasets_clun=comb_datasets[comb_datasets.peptide_class.isin([\"unique_to_Noncanon\",\"unique_to_Canon\"])]\n",
    "print(len(comb_datasets_clun))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7a93e855-05c4-4019-8b7f-999cd3120641",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12292\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/1104008463.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  peps_wsub=comb_datasets_clun.groupby(\"database_peptide\").apply(lambda x: any(x.isSubstitution))\n"
     ]
    }
   ],
   "source": [
    "#identify peptides that have at least 1 PSM with a substitution mass shift, filter those that do not\n",
    "\n",
    "peps_wsub=comb_datasets_clun.groupby(\"database_peptide\").apply(lambda x: any(x.isSubstitution))\n",
    "comb_datasets_clun=comb_datasets_clun[comb_datasets_clun.database_peptide.isin(peps_wsub[peps_wsub].index)]\n",
    "print(len(comb_datasets_clun))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f0acaed9-996b-4eba-95f8-a2888edee5ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#now check that the expected residues of these mass shifts are present in the peptides (without concidering predicted PTM position)\n",
    "#coution false also can meed Unmodified or non-substitution ptm\n",
    "\n",
    "comb_datasets_clun[\"isSubstitution_residue\"]=comb_datasets_clun.apply(lambda x: False if x[\"modifications_masses\"]==\"Unmodified\" else\n",
    "                                                                                      any( any(aa in unimod_df_subs[unimod_df_subs.mono_mass==mass].residues.tolist() for aa in x[\"database_peptide\"]) for _,_,mass in x[\"modifications_masses\"]),axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cccdc43a-57f2-42ab-82c5-b0cbdd0c23aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12292\n"
     ]
    }
   ],
   "source": [
    "#filter out mass shifts of potential substitutuins, where peptide does not contain the expected residue\n",
    "\n",
    "peps_wsub_res=comb_datasets_clun[(comb_datasets_clun.isSubstitution)&(comb_datasets_clun.isSubstitution_residue)].database_peptide.unique()\n",
    "comb_datasets_clun=comb_datasets_clun[comb_datasets_clun.database_peptide.isin(peps_wsub_res)]\n",
    "print(len(comb_datasets_clun))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "325bf1e1-12bc-4868-ba7f-1b03c237f7bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/4134410897.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  a=comb_datasets_clun.groupby(\"database_peptide\").apply(lambda x: Counter(x.isSubstitution_residue)[True]/len(x))\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAogAAAGxCAYAAADh1EdCAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjEsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvc2/+5QAAAAlwSFlzAAAPYQAAD2EBqD+naQAASiNJREFUeJzt3XucTXX////nNmfTzGaGmTEZx0SOiYgOyCkZk06UmtQlkaIJZVwd0LfLRIVKpboq5RB1hVTCUEYu5Dh1OZR0IS7GSGOPYYzDvH9/+M36eM+BGZkTj/vttm83e+3XWuu93nvvtZ77PWstLmOMEQAAAPD/q1DaDQAAAEDZQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwFLkgDh79mw1atRIAQEBcrlcSk5OLoZm5W/Lli0aPXq0du7cmee1Bx98ULVq1SqxtpyPP//8U/fcc4/CwsLkcrnUs2fPAmvbt28vl8vlPAICAtSsWTNNmjRJ2dnZTp0xRrNmzdKNN96osLAw+fv7q3r16uratav++c9/WsvMWdaDDz6Y7zpfeOEFpya/Pi4pLpdLo0ePdp6f7X1v3769GjduXHKNK4OOHj2q0aNHa9myZRd82cuWLZPL5SqWZZ9rnf/6178uyPLeeustTZ06Nc/0nTt3yuVyWa+tXLlSo0eP1qFDh857feV9P3UpK8pnpSCjR4+Wy+W68I0rBuV1X3sh21Jc39ez7Uvat2+v9u3bn9dyS5QpgtTUVOPj42N69Ohhli1bZlatWmWOHDlSlEX8JZ999pmRZL777rs8r23fvt1s2LChxNpyPuLi4oyvr6+ZPn26WbVqlfnll18KrG3Xrp2pU6eOWbVqlVm1apX54osvTLdu3Ywk8/TTTzt1I0aMMJJM//79zRdffGG+/fZb8+GHH5rY2FgTHR1tLVOSCQoKMhUrVjTp6enWa9nZ2aZ27domODjYSDI7duy4oNteFKtWrTK7d+92np/tfW/Xrp1p1KhRCbau7Dlw4ICRZEaNGnXBl/3dd98V2PfFJWedn3322QVZXqNGjUy7du3yTD927JhZtWqVSU1Ndaa9/PLLf/nzX973U5eyonxWCjJq1ChTxENrqSmv+9oL2Zbi+r6ebV+yefNms3nz5vNabknyLkqY3LZtm06cOKH7779f7dq1O2vt0aNHVbFixaKl1b+gbt26Jbau87Vp0ybVrVtX9913X6HqAwICdN111znPu3XrpgYNGmjy5Ml68cUXdfLkSU2aNEkPPPCA3n33XWveBx980BppzHHbbbfp888/16xZs9S/f39n+rfffqsdO3aof//+eu+9985zCy+MM7cZKC5+fn4l/lkrD/up4lbSx4YLoTQ+KyXhYtymC6m4vq8NGzYsluVecIVNkn379jWSrEfOL62+ffuawMBA89NPP5nOnTubyy67zFx33XXGGGMWL15sYmJizOWXX278/PxM3bp1zSOPPGIOHDiQZx1bt24199xzjwkLCzO+vr4mKirKxMbGmmPHjpkPP/wwz/olmQ8//NBpQ82aNa3lZWZmmvj4eFOrVi3j4+NjIiMjzaBBg0xaWppVV7NmTdO9e3fzzTffmObNmxt/f39Tv3598/777xeqbw4ePGgeffRRExkZaXx8fEzt2rXN3//+d3Ps2DFjjDE7duzIt+1nG5Up6BfS3XffbSSZ//3vf87I0YgRIwrVTknmscceM3369HHenxx9+vQx119/fb6/ejZs2GC6d+9uqlatanx9fU21atXMrbfeav3yzG3y5MnG5XKZ/fv3O9NeeeUVI8kMGjTImXbq1ClTqVIlM3ToUKudOaNh53rfc/ppzZo15oYbbjABAQGmdu3aJiEhwZw6deqcfZKdnW3efPNN06xZM+Pv728qVapk7rzzTvPbb785NZ988omRZN544w1r3ueff95UqFDBLF682Bjzf+/zuHHjzIsvvmiioqKMn5+fadGihVmyZEmedW/bts3ce++9Tr82aNDATJ48OU9dWlqaGTp0qKldu7bx9fU1VatWNd26dTNbt24t8LPVt2/fIq9n69atpmvXriYgIMCEhoaaAQMGmPnz5xdqBDE1NdX079/fVK9e3fj6+poqVaqYtm3bmsTERKemZs2aVrtytGvXzhq1yRlBnDZtmnnyySdNeHi48ff3NzfddFOeX/O//fab6d27t6lWrZrx9fU1YWFh5uabbzYbN2501pm7b3L2Ezl9l/NZyhn5Keh7eubn8kxnbldZ2E8dOXLEDBs2zNSqVcv4+fmZypUrmxYtWpiZM2fmafuZctq+ePFi8+CDD5rKlSubihUrmujoaOv7kCMxMdHcfPPNJigoyAQEBJi2bdvm+Zzn9On69evNnXfeaSpVqmQiIiJKpQ0bNmwwt99+uwkKCjLBwcHmvvvus0YEi/JZyfHVV1+ZZs2aGV9fX1OrVi3z8ssv5zuCWJj9jDEX9772008/Na1atTLBwcHOvA899JDzes76c4+45fdXjJy2LF++3LRu3dr4+/ubyMhI8+yzz5qTJ09a87/11lumadOmJjAw0Fx22WWmfv36ZuTIkYXa5tzf14I+B7n78lz7ktz7PGPOnSPOXM9jjz1mPv74Y9OgQQMTEBBgmjZtar788kurrjD75HMpdEDcvn27efPNN40kM3bsWLNq1SpniLRv377Gx8fH1KpVyyQkJJilS5eaRYsWGWOMefvtt01CQoKZP3++SUpKMh999JFp1qyZqV+/vjl+/Liz/OTkZHPZZZeZWrVqmSlTppilS5ea6dOnm169epn09HSTmppqxo4daySZN9980/nTa84XPPcbmZ2dbbp27Wq8vb3Nc889ZxYvXmxeeeUVExgYaJo3b251es2aNU316tVNw4YNzccff2wWLVrkBLGkpKSz9ktmZqbz4XvllVfM4sWLzXPPPWe8vb3Nrbfeaoz5vz9PNG/e3PqzscfjKXC5BQXEa665xnh7e5ujR48aY4y54oorTFBQkHn11VfN1q1bTXZ2doHLzPlgLV261EgyW7ZsMcacDiD+/v7mgw8+yBMQMzIyTGhoqGnZsqX59NNPTVJSkpk9e7YZOHCgM39+fv75ZyPJOijdcsstJiAgwNSrV8+Z9sMPPxhJZsGCBVY7c75o53rf27VrZ0JDQ029evXMlClTTGJiohk0aJCRZD766KMC25ejf//+xsfHxwwbNswsXLjQzJw50zRo0MCEh4eblJQUp27gwIHG19fXrF271hhjzNKlS02FChXMs88+69Tk7DyioqLMDTfcYD7//HPz2WefmWuvvdb4+PiYlStXOrWbN282brfbNGnSxHz88cdm8eLFZtiwYaZChQpm9OjRTl16erpp1KiRCQwMNC+88IJZtGiR+fzzz80TTzxhvv32W3Ps2DGzcOFCI8n069fP6Z/t27cXaT0pKSkmLCzMXH755ebDDz80CxYsMPfdd5+pUaNGoQJi165dTdWqVc27775rli1bZubNm2eef/55M2vWLKemqAExKirK3HbbbebLL78006dPN1dccYUJDg62Dqr169c3V1xxhZk2bZpJSkoyn3/+uRk2bJjT3g0bNpg6deqY5s2bO32TEzJz7+x3795tBg8ebCSZOXPm5PmeFiYgloX91IABA0zFihXNhAkTzHfffWe++uor89JLL+X5gZNbzsEyKirK/O1vfzPffPONeffdd01YWJiJioqyAuu0adOMy+UyPXv2NHPmzDFffvmliY6ONl5eXlZAyzlQ1qxZ04wYMcIkJiaaefPmlVobnnrqKbNo0SIzYcIEp49zjkNF+awYY8ySJUuMl5eXueGGG8ycOXOc73rOd+ZMhdnPXMz72pUrVxqXy2Xuueces2DBAutUqNzvfWEDYmhoqImMjDSvv/66WbRokRkyZIhzjMuR8+N+8ODBZvHixWbJkiVmypQpZsiQIYXa5vMNiOfal+Te5xUmR5y5nlq1aplWrVqZTz/91CxYsMC0b9/eeHt7W/vGwuyTz6VIJ0oUdG5QzujiBx98cNb5s7OzzYkTJ8yuXbuMJPPFF184r918882mUqVKZz3H42znCuR+I3MOmuPHj7fqZs+ebSSZd99915lWs2ZN4+/vb3bt2uVMy8zMNCEhIWbAgAFn3aYpU6YYSebTTz+1po8bN875JZyjKOdN5NSeOHHCnDhxwuzdu9fEx8cbSebuu+926tasWePskKTT5xhGR0ebjz/+OE9YzPny5JxvOHz4cGOMMW+++aa57LLLzOHDh/MExHXr1hlJZ92pF6R69ermb3/7mzHGmKysLBMYGOicM5nT1//4xz+Mj4+PycjIsNp55oH4XOfFSDI//PCDNb1hw4ama9euZ23fqlWrjCTz6quvWtN3795tAgICrHM9jx07Zpo3b25q165ttmzZYsLDw027du2sX6s5O4/IyEiTmZnpTE9PTzchISGmU6dOzrSuXbua6tWr5/mR8Pjjjxt/f3/z559/GmOMeeGFF4yks/7qO9s5iIVdz4gRI4zL5TLJyclWXefOnQsVEC+77DITFxd31pqiBsRrrrnG+gzv3LnT+Pj4mIcfftgYY8wff/xhJJlJkyaddb0FnVeW387+bOcNFSYgGlP6+6nGjRubnj175ln3ueQcoG+//XZr+r///W8jybz44ovGmNMjlCEhIaZHjx5W3alTp0yzZs1Mq1atnGk54ez5558v9TY8+eSTVu2MGTOMJDN9+nRnWlE+K61bty7wu35mQCzsfuZi3tfmjGgeOnSowJqiBsTcGcKY00G8QoUKzjY//vjjplKlSmdtW1G+r4UNiMacfV+Se59XlBwhyYSHh1vXEaSkpJgKFSqYhIQEZ1ph9snnckFvc3PnnXfmmZaamqqBAwcqKipK3t7e8vHxUc2aNSVJW7dulXT6nJSkpCT16tVLVatWvSBt+fbbbyUpzxW7d999twIDA7V06VJr+tVXX60aNWo4z/39/XXllVdq165d51xPYGCg7rrrLmt6znpzr6coNm/eLB8fH/n4+CgyMlKvvvqq7rvvPuscwWuvvVbbt2/XwoUL9fe//11t2rTR0qVL9cADDygmJkanP0+2nCuZp02bppMnT+r9999Xr169dNlll+WpveKKK1S5cmWNGDFCU6ZM0ZYtWwrd/o4dO2rJkiWSTl/RdfToUQ0dOlRVqlRRYmKiJGnJkiVq06aNAgMDi9o9joiICLVq1cqa1rRp03O+d1999ZVcLpfuv/9+nTx50nlERESoWbNm1pW7fn5++vTTT3Xw4EFdc801Msbok08+kZeXV57l3nHHHfL393eeBwUFqUePHlq+fLlOnTqlY8eOaenSpbr99ttVsWJFa9233nqrjh07ptWrV0uSvvnmG1155ZXq1KlTkfulKOv57rvv1KhRIzVr1sxaRp8+fQq1rlatWmnq1Kl68cUXtXr1ap04caLI7c2tT58+1tWgNWvWVNu2bfXdd99JkkJCQlS3bl29/PLLmjBhgjZu3JjvebdlWXHsp1q1aqVvvvlG8fHxWrZsmTIzM4vUptznSLdt21Y1a9Z0+n3lypX6888/1bdvX+szlZ2drVtuuUVr167VkSNHrGXkd2wo6TbkXmavXr3k7e3tLLMojhw5orVr1xb4XT9TYfczF/O+9tprr5V0us8//fRT/e9//zvvNuQICgpSTEyMNa1Pnz7Kzs7W8uXLJZ3+Lhw6dEj33nuvvvjiC/3xxx9/eb3Foag5okOHDgoKCnKeh4eHKywsLM9+4K/uky9YQKxYsaKCg4OtadnZ2erSpYvmzJmjp59+WkuXLtWaNWucg1LOjistLU2nTp1S9erVL1RzdPDgQXl7e+cJnC6XSxERETp48KA1PTQ0NM8y/Pz8zrlzPXjwoCIiIvLc1iAsLEze3t551lMUdevW1dq1a7Vu3Tpt2rRJhw4d0vTp0+V2u606Hx8fde3aVf/4xz+0aNEi7d69W+3bt9dXX32lb775Jt9lP/TQQzpw4IDGjh2rDRs2qF+/fvnWud1uJSUl6eqrr9bf//53NWrUSJGRkRo1atQ5P3CdOnXS77//rl9//VVLlixR8+bNFRYWpptvvllLlixRZmamVq5ceV7h50zn+97t379fxhiFh4c7QTznsXr16jw7kyuuuEI33nijjh07pvvuu0/VqlXLd7kRERH5Tjt+/LgyMjJ08OBBnTx5Um+88Uae9d56662S5Kz7wIED5/29KMp6cj7HhdmW/MyePVt9+/bVP//5T7Vp00YhISF64IEHlJKScl5tL2jdZ353XS6Xli5dqq5du2r8+PG65pprVLVqVQ0ZMkSHDx8+7/WWpOLYT73++usaMWKE5s2bpw4dOigkJEQ9e/bUr7/+Wqg2navf9+/fL0m666678nyuxo0bJ2OM/vzzT2v+gr4rJdmG3Mv09vZWaGjoee2j09LSlJ2dXajvTGH3Mxfzvvamm27SvHnzdPLkST3wwAOqXr26GjdurE8++eS82xIeHp5nWk7f57ynsbGx+uCDD7Rr1y7deeedCgsLU+vWrZ3QXFYUNUcU5n24EPvkIl3FfDb53fdp06ZN+vHHHzV16lT17dvXmb59+3arLiQkRF5eXtqzZ8+Fao5CQ0N18uRJHThwwNr5GmOUkpLi/KK5EOv54YcfZIyx+iA1NVUnT55UlSpVznvZ/v7+atmy5Xm1KS4uTsuWLdOmTZucMHCmqKgoderUSWPGjFH9+vXVtm3bApfXpEkTzZo1S8YY/fTTT5o6dapeeOEFBQQEKD4+vsD5OnbsKOn0L9fExER17tzZmf7ss89q+fLlysrK+ss7rfNVpUoVuVwuff/99/Lz88vzeu5p//znP/X111+rVatWmjx5snr37q3WrVvnmS+/L2BKSop8fX112WWXycfHR15eXoqNjdVjjz2Wb9tq164tSapatep5fy8qV65c6PWEhoYW2O7CqFKliiZNmqRJkybp999/1/z58xUfH6/U1FQtXLhQ0unPc1ZWVp55//jjj3y/JwW158ydY82aNfX+++9LOn2XhU8//VSjR4/W8ePHNWXKlEK1vbD8/Pzybf9f+RFYHPupwMBAjRkzRmPGjNH+/fud0cQePXro559/Puf8BfX7FVdcIUnOe/XGG28UeBVs7oN3Ue8LWBxtSElJ0eWXX+48P3nypA4ePJjvwfZcKleuLJfLVajvTFH2MxfrvlY6fQeN2267TVlZWVq9erUSEhLUp08f1apVS23atHFGYnN/xwoa9cv5kXCmnL4/8z196KGH9NBDD+nIkSNavny5Ro0apejoaG3bts35a2ZhFdTGv7IPyGnvhc4Rhdknn0ux/k8qORua+0vxzjvvWM8DAgLUrl07ffbZZ2cdAs5ZTmH+ZJLzhZk+fbo1/fPPP9eRI0ec1/+qjh07KiMjQ/PmzbOmf/zxx1Y7isOJEycK/GDm/Pk+MjKywPmHDRumHj166LnnnivU+lwul5o1a6aJEyeqUqVK2rBhw1nrq1WrpoYNG+rzzz/X+vXrnZ1W586ddeDAAU2YMEHBwcHnPAgW5X0viujoaBlj9L///U8tW7bM82jSpIlT+5///EdDhgzRAw88oO+//15NmzZV7969lZaWlme5c+bM0bFjx5znhw8f1pdffqkbb7xRXl5eqlixojp06KCNGzeqadOm+a47ZwfXrVs3bdu2zflTZH4K6p+irKdDhw7avHmzfvzxR2sZM2fOLGKvSjVq1NDjjz+uzp07W5+RWrVq6aeffrJqt23bpl9++SXf5XzyySfWKRK7du3SypUrC7zB7JVXXqlnn31WTZo0sdZbmBGOM2ul/D9r+bX/22+/VUZGRqGXkVtx76fCw8P14IMP6t5779Uvv/yio0ePnnOeGTNmWM9XrlypXbt2Of1+/fXXq1KlStqyZUu+n6mWLVvK19f3L7W7ONqQe5mffvqpTp48aX2eCvtZCQwMVKtWrQr8rp+pKPuZHBfbvjb3Otq1a6dx48ZJkjZu3ChJzg2pc3/H5s+fn+9yDh8+nOe1mTNnqkKFCrrpppvy1AcGBqpbt2565plndPz4cW3evNlpj1S4bQ4PD5e/v3+eNn7xxRd5aou6HyjOHFHQPvlcLtgIYn4aNGigunXrKj4+XsYYhYSE6Msvv8x3eHfChAm64YYb1Lp1a8XHx+uKK67Q/v37NX/+fL3zzjsKCgpy7pz+7rvvKigoSP7+/qpdu3a+vwA7d+6srl27asSIEUpPT9f111+vn376SaNGjVLz5s0VGxt7QbbxgQce0Jtvvqm+fftq586datKkiVasWKGxY8fq1ltvLdZfbB6PR7Vq1dLdd9+tTp06KSoqShkZGVq2bJlee+01XXXVVbrjjjsKnL9Lly7q0qXLWdfx1Vdf6a233lLPnj1Vp04dGWM0Z84cHTp0yNkJnU3Hjh31xhtvKCAgQNdff72k06NWtWvX1uLFixUTEyNv77N/DIvyvhfF9ddfr0ceeUQPPfSQ1q1bp5tuukmBgYHat2+fVqxYoSZNmujRRx/VkSNH1KtXL9WuXVtvvfWWfH199emnn+qaa67RQw89lOdL7eXlpc6dO2vo0KHKzs7WuHHjlJ6erjFjxjg1r732mm644QbdeOONevTRR1WrVi0dPnxY27dv15dffukEwri4OM2ePVu33Xab4uPj1apVK2VmZiopKUnR0dHOuSg1a9bUF198oY4dOyokJERVqlRRrVq1irSeDz74QN27d9eLL76o8PBwzZgxo1AjTh6PRx06dFCfPn3UoEEDBQUFae3atVq4cKH1+YuNjdX999+vQYMG6c4779SuXbs0fvz4As87Tk1N1e23367+/fvL4/Fo1KhR8vf318iRIyWdPpA8/vjjuvvuu1WvXj35+vrq22+/1U8//WSNtuSMysyePVt16tSRv79/vgflnNqc96dv377y8fFR/fr1FRQUpNjYWD333HN6/vnn1a5dO23ZskWTJ0/Oc8pHae+nWrdurejoaDVt2lSVK1fW1q1bNW3aNLVp06ZQ9x9ct26dHn74Yd19993avXu3nnnmGV1++eUaNGiQJOmyyy7TG2+8ob59++rPP//UXXfdpbCwMB04cEA//vijDhw4oLfffrvI7S7uNsyZM0fe3t7q3LmzNm/erOeee07NmjVTr169nJqifFb+3//7f7rlllvUuXNnDRs2TKdOndK4ceMUGBho/Xm7sPuZi3lf+/zzz2vPnj3q2LGjqlevrkOHDum1116Tj4+Pc0/la6+9VvXr19fw4cN18uRJVa5cWXPnztWKFSvyXWZoaKgeffRR/f7777ryyiu1YMECvffee3r00Ued83T79+/v9Ee1atWUkpKihIQEud1uJywXZZtzziX94IMPVLduXTVr1kxr1qzJ94f02fYluV3oHFHYffI5FeWKlrNdxRwYGJjvPFu2bDGdO3c2QUFBpnLlyubuu+82v//+e75XBG7ZssXcfffdJjQ01Pj6+poaNWqYBx980LrVw6RJk0zt2rWNl5fXWe9XZMzpK/xGjBhhatasaXx8fEy1atXMo48+WuD9xXLL715F+Tl48KAZOHCgqVatmvH29jY1a9Y0I0eOzHP/ovO5ivlssrKyzCuvvGK6detmatSoYfz8/Iy/v7+56qqrzNNPP20OHjxo1SvXLQDyk/vKq59//tnce++9pm7duiYgIMC43W7TqlUrM3Xq1EJtxxdffGEkmc6dO1vT+/fvbySZ119/Pc88+X02CnrfC+qn/D4PBfnggw9M69atTWBgoAkICDB169Y1DzzwgFm3bp0xxpj777/fVKxYMc+d73Oufps4caIxxr4P4pgxY5z7TzVv3ty57dOZduzYYf72t7+Zyy+/3Pj4+JiqVauatm3bOldq5khLSzNPPPGEqVGjhvHx8TFhYWGme/fu5ueff3ZqlixZYpo3b278/PyMZN8HsbDryfmu+vv7m5CQENOvXz/n/TvbVczHjh0zAwcONE2bNnXucVa/fn0zatQo639ays7ONuPHjzd16tQx/v7+pmXLlubbb789630QhwwZYqpWrWr8/PzMjTfe6Lwnxhizf/9+8+CDD5oGDRo49zhr2rSpmThxonV1+c6dO02XLl1MUFCQc7uTM9+v3Fckjhw50kRGRpoKFSpY256VlWWefvppExUVZQICAky7du1McnJyvldnl+Z+Kj4+3rRs2dJUrlzZ+Pn5mTp16pgnn3zS/PHHH3nfvDOceQ/C2NhYU6lSJRMQEGBuvfVW8+uvv+apT0pKMt27dzchISHGx8fHXH755aZ79+7W8SHnCuL87ntb0m1Yv3696dGjh7nssstMUFCQuffee617BxpT9M/K/PnzTdOmTZ3j1UsvvVTg/6Ryrv3Mxbyv/eqrr0y3bt3M5Zdf7tyv9NZbbzXff/+9Vbdt2zbTpUsXExwcbKpWrWoGDx5svv7663yvYm7UqJFZtmyZadmypfHz8zPVqlUzf//7382JEyecuo8++sh06NDBhIeHG19fXxMZGWl69eplfvrpp0Jtc37b5vF4zMMPP2zCw8NNYGCg6dGjh9m5c2e+fVnQvqSg+yAWJkcUdBw/cz9U2H3yubj+/xUC+It27typ2rVr6+WXX9bw4cNLuzlAkUydOlUPPfSQ1q5de17nPpfVNowePVpjxozRgQMH/tI54cClpljPQQQAAED5Q0AEAACAhT8xAwAAwMIIIgAAACwERAAAAFgIiAAAALAU642yLzbZ2dnau3evgoKCivxfRwEAgNJhjNHhw4cVGRmpChUYGysMAmIR7N27V1FRUaXdDAAAcB52796t6tWrl3YzygUCYhHk/Bc5u3fvVnBwcCm3BgAAFEZ6erqioqLy/a/ukD8CYhHk/Fk5ODiYgAgAQDnD6WGFxx/iAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALCUiYC4fPly9ejRQ5GRkXK5XJo3b16emq1btyomJkZut1tBQUG67rrr9PvvvzuvZ2VlafDgwapSpYoCAwMVExOjPXv2WMtIS0tTbGys3G633G63YmNjdejQoWLeOgAAgPKlTATEI0eOqFmzZpo8eXK+r//222+64YYb1KBBAy1btkw//vijnnvuOfn7+zs1cXFxmjt3rmbNmqUVK1YoIyND0dHROnXqlFPTp08fJScna+HChVq4cKGSk5MVGxtb7NsHAABQnriMMaa0G3Eml8uluXPnqmfPns60e+65Rz4+Ppo2bVq+83g8HlWtWlXTpk1T7969JUl79+5VVFSUFixYoK5du2rr1q1q2LChVq9erdatW0uSVq9erTZt2ujnn39W/fr1z9m29PR0ud1ueTwe/qs9AADKCY7fRVcmRhDPJjs7W19//bWuvPJKde3aVWFhYWrdurX1Z+j169frxIkT6tKlizMtMjJSjRs31sqVKyVJq1atktvtdsKhJF133XVyu91OTW5ZWVlKT0+3HgAAABe7Mh8QU1NTlZGRoZdeekm33HKLFi9erNtvv1133HGHkpKSJEkpKSny9fVV5cqVrXnDw8OVkpLi1ISFheVZflhYmFOTW0JCgnO+otvtVlRU1AXeOgAAgLKnzAfE7OxsSdJtt92mJ598UldffbXi4+MVHR2tKVOmnHVeY4xcLpfz/Mx/F1RzppEjR8rj8TiP3bt3/4UtAQAAKB+8S7sB51KlShV5e3urYcOG1vSrrrpKK1askCRFRETo+PHjSktLs0YRU1NT1bZtW6dm//79eZZ/4MABhYeH57tuPz8/+fn5XahNAQCgRNWK/7q0m1BkO1/qXtpNgMrBCKKvr6+uvfZa/fLLL9b0bdu2qWbNmpKkFi1ayMfHR4mJic7r+/bt06ZNm5yA2KZNG3k8Hq1Zs8ap+eGHH+TxeJwaAAAAlJERxIyMDG3fvt15vmPHDiUnJyskJEQ1atTQU089pd69e+umm25Shw4dtHDhQn355ZdatmyZJMntdqtfv34aNmyYQkNDFRISouHDh6tJkybq1KmTpNMjjrfccov69++vd955R5L0yCOPKDo6ulBXMAMAAFwqykRAXLdunTp06OA8Hzp0qCSpb9++mjp1qm6//XZNmTJFCQkJGjJkiOrXr6/PP/9cN9xwgzPPxIkT5e3trV69eikzM1MdO3bU1KlT5eXl5dTMmDFDQ4YMca52jomJKfDeiwAAAJeqMncfxLKM+ygBAMoTzkE8jeN30ZX5cxABAABQsgiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFjKREBcvny5evToocjISLlcLs2bN6/A2gEDBsjlcmnSpEnW9KysLA0ePFhVqlRRYGCgYmJitGfPHqsmLS1NsbGxcrvdcrvdio2N1aFDhy78BgEAAJRjZSIgHjlyRM2aNdPkyZPPWjdv3jz98MMPioyMzPNaXFyc5s6dq1mzZmnFihXKyMhQdHS0Tp065dT06dNHycnJWrhwoRYuXKjk5GTFxsZe8O0BAAAoz7xLuwGS1K1bN3Xr1u2sNf/73//0+OOPa9GiRerevbv1msfj0fvvv69p06apU6dOkqTp06crKipKS5YsUdeuXbV161YtXLhQq1evVuvWrSVJ7733ntq0aaNffvlF9evXL56NAwAAKGfKxAjiuWRnZys2NlZPPfWUGjVqlOf19evX68SJE+rSpYszLTIyUo0bN9bKlSslSatWrZLb7XbCoSRdd911crvdTk1uWVlZSk9Ptx4AAAAXu3IREMeNGydvb28NGTIk39dTUlLk6+urypUrW9PDw8OVkpLi1ISFheWZNywszKnJLSEhwTlf0e12Kyoq6i9uCQAAQNlX5gPi+vXr9dprr2nq1KlyuVxFmtcYY82T3/y5a840cuRIeTwe57F79+6iNR4AAKAcKvMB8fvvv1dqaqpq1Kghb29veXt7a9euXRo2bJhq1aolSYqIiNDx48eVlpZmzZuamqrw8HCnZv/+/XmWf+DAAacmNz8/PwUHB1sPAACAi12ZD4ixsbH66aeflJyc7DwiIyP11FNPadGiRZKkFi1ayMfHR4mJic58+/bt06ZNm9S2bVtJUps2beTxeLRmzRqn5ocffpDH43FqAAAAUEauYs7IyND27dud5zt27FBycrJCQkJUo0YNhYaGWvU+Pj6KiIhwrjx2u93q16+fhg0bptDQUIWEhGj48OFq0qSJc1XzVVddpVtuuUX9+/fXO++8I0l65JFHFB0dzRXMAAAAZygTAXHdunXq0KGD83zo0KGSpL59+2rq1KmFWsbEiRPl7e2tXr16KTMzUx07dtTUqVPl5eXl1MyYMUNDhgxxrnaOiYk5570XAQAALjUuY4wp7UaUF+np6XK73fJ4PJyPCAAo82rFf13aTSiynS91P3dREXH8Lroyfw4iAAAAShYBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAABLmQiIy5cvV48ePRQZGSmXy6V58+Y5r504cUIjRoxQkyZNFBgYqMjISD3wwAPau3evtYysrCwNHjxYVapUUWBgoGJiYrRnzx6rJi0tTbGxsXK73XK73YqNjdWhQ4dKYAsBAADKjzIREI8cOaJmzZpp8uTJeV47evSoNmzYoOeee04bNmzQnDlztG3bNsXExFh1cXFxmjt3rmbNmqUVK1YoIyND0dHROnXqlFPTp08fJScna+HChVq4cKGSk5MVGxtb7NsHAABQnriMMaa0G3Eml8uluXPnqmfPngXWrF27Vq1atdKuXbtUo0YNeTweVa1aVdOmTVPv3r0lSXv37lVUVJQWLFigrl27auvWrWrYsKFWr16t1q1bS5JWr16tNm3a6Oeff1b9+vXP2bb09HS53W55PB4FBwdfkO0FAKC41Ir/urSbUGQ7X+p+wZfJ8bvoysQIYlF5PB65XC5VqlRJkrR+/XqdOHFCXbp0cWoiIyPVuHFjrVy5UpK0atUqud1uJxxK0nXXXSe32+3U5JaVlaX09HTrAQAAcLErdwHx2LFjio+PV58+fZxfASkpKfL19VXlypWt2vDwcKWkpDg1YWFheZYXFhbm1OSWkJDgnK/odrsVFRV1gbcGAACg7ClXAfHEiRO65557lJ2drbfeeuuc9cYYuVwu5/mZ/y6o5kwjR46Ux+NxHrt37z7/xgMAAJQT5SYgnjhxQr169dKOHTuUmJhonUMQERGh48ePKy0tzZonNTVV4eHhTs3+/fvzLPfAgQNOTW5+fn4KDg62HgAAABe7chEQc8Lhr7/+qiVLlig0NNR6vUWLFvLx8VFiYqIzbd++fdq0aZPatm0rSWrTpo08Ho/WrFnj1Pzwww/yeDxODQAAACTv0m6AJGVkZGj79u3O8x07dig5OVkhISGKjIzUXXfdpQ0bNuirr77SqVOnnHMGQ0JC5OvrK7fbrX79+mnYsGEKDQ1VSEiIhg8friZNmqhTp06SpKuuukq33HKL+vfvr3feeUeS9Mgjjyg6OrpQVzADAABcKspEQFy3bp06dOjgPB86dKgkqW/fvho9erTmz58vSbr66qut+b777ju1b99ekjRx4kR5e3urV69eyszMVMeOHTV16lR5eXk59TNmzNCQIUOcq51jYmLyvfciAADApazM3QexLOM+SgCA8oT7IJ7G8bvoysU5iAAAACg5BEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWMpEQFy+fLl69OihyMhIuVwuzZs3z3rdGKPRo0crMjJSAQEBat++vTZv3mzVZGVlafDgwapSpYoCAwMVExOjPXv2WDVpaWmKjY2V2+2W2+1WbGysDh06VMxbBwAAUL6UiYB45MgRNWvWTJMnT8739fHjx2vChAmaPHmy1q5dq4iICHXu3FmHDx92auLi4jR37lzNmjVLK1asUEZGhqKjo3Xq1Cmnpk+fPkpOTtbChQu1cOFCJScnKzY2tti3DwAAoDxxGWNMaTfiTC6XS3PnzlXPnj0lnR49jIyMVFxcnEaMGCHp9GhheHi4xo0bpwEDBsjj8ahq1aqaNm2aevfuLUnau3evoqKitGDBAnXt2lVbt25Vw4YNtXr1arVu3VqStHr1arVp00Y///yz6tevf862paeny+12y+PxKDg4uHg6AACAC6RW/Nel3YQi2/lS9wu+TI7fRVcmRhDPZseOHUpJSVGXLl2caX5+fmrXrp1WrlwpSVq/fr1OnDhh1URGRqpx48ZOzapVq+R2u51wKEnXXXed3G63U5NbVlaW0tPTrQcAAMDFrswHxJSUFElSeHi4NT08PNx5LSUlRb6+vqpcufJZa8LCwvIsPywszKnJLSEhwTlf0e12Kyoq6i9vDwAAQFlX5gNiDpfLZT03xuSZllvumvzqz7ackSNHyuPxOI/du3efR8sBAADKlzIfECMiIiQpzyhfamqqM6oYERGh48ePKy0t7aw1+/fvz7P8AwcO5BmdzOHn56fg4GDrAQAAcLEr8wGxdu3aioiIUGJiojPt+PHjSkpKUtu2bSVJLVq0kI+Pj1Wzb98+bdq0yalp06aNPB6P1qxZ49T88MMP8ng8Tg0AAAAk79JugCRlZGRo+/btzvMdO3YoOTlZISEhqlGjhuLi4jR27FjVq1dP9erV09ixY1WxYkX16dNHkuR2u9WvXz8NGzZMoaGhCgkJ0fDhw9WkSRN16tRJknTVVVfplltuUf/+/fXOO+9Ikh555BFFR0cX6gpmAACAS0WZCIjr1q1Thw4dnOdDhw6VJPXt21dTp07V008/rczMTA0aNEhpaWlq3bq1Fi9erKCgIGeeiRMnytvbW7169VJmZqY6duyoqVOnysvLy6mZMWOGhgwZ4lztHBMTU+C9FwEAAC5VZe4+iGUZ91ECAJQn3AfxNI7fRVfmz0EEAABAySIgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGApFwHx5MmTevbZZ1W7dm0FBASoTp06euGFF5Sdne3UGGM0evRoRUZGKiAgQO3bt9fmzZut5WRlZWnw4MGqUqWKAgMDFRMToz179pT05gAAAJRp5SIgjhs3TlOmTNHkyZO1detWjR8/Xi+//LLeeOMNp2b8+PGaMGGCJk+erLVr1yoiIkKdO3fW4cOHnZq4uDjNnTtXs2bN0ooVK5SRkaHo6GidOnWqNDYLAACgTPIu7QYUxqpVq3Tbbbepe/fukqRatWrpk08+0bp16ySdHj2cNGmSnnnmGd1xxx2SpI8++kjh4eGaOXOmBgwYII/Ho/fff1/Tpk1Tp06dJEnTp09XVFSUlixZoq5du5bOxgEAAJQx5WIE8YYbbtDSpUu1bds2SdKPP/6oFStW6NZbb5Uk7dixQykpKerSpYszj5+fn9q1a6eVK1dKktavX68TJ05YNZGRkWrcuLFTk1tWVpbS09OtBwAAwMWuXIwgjhgxQh6PRw0aNJCXl5dOnTqlf/zjH7r33nslSSkpKZKk8PBwa77w8HDt2rXLqfH19VXlypXz1OTMn1tCQoLGjBlzoTcHAACgTCsXI4izZ8/W9OnTNXPmTG3YsEEfffSRXnnlFX300UdWncvlsp4bY/JMy+1sNSNHjpTH43Eeu3fv/msbAgAAUA6UixHEp556SvHx8brnnnskSU2aNNGuXbuUkJCgvn37KiIiQtLpUcJq1ao586WmpjqjihERETp+/LjS0tKsUcTU1FS1bds23/X6+fnJz8+vuDYLAACgTCoXI4hHjx5VhQp2U728vJzb3NSuXVsRERFKTEx0Xj9+/LiSkpKc8NeiRQv5+PhYNfv27dOmTZsKDIgAAACXonIxgtijRw/94x//UI0aNdSoUSNt3LhREyZM0N/+9jdJp/+0HBcXp7Fjx6pevXqqV6+exo4dq4oVK6pPnz6SJLfbrX79+mnYsGEKDQ1VSEiIhg8friZNmjhXNQMAAKCcBMQ33nhDzz33nAYNGqTU1FRFRkZqwIABev75552ap59+WpmZmRo0aJDS0tLUunVrLV68WEFBQU7NxIkT5e3trV69eikzM1MdO3bU1KlT5eXlVRqbBQAAUCa5jDGmtBtRXqSnp8vtdsvj8Sg4OLi0mwMAwFnViv+6tJtQZDtf6n7Bl8nxu+jKxTmIAAAAKDkERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAs5SYg/u9//9P999+v0NBQVaxYUVdffbXWr1/vvG6M0ejRoxUZGamAgAC1b99emzdvtpaRlZWlwYMHq0qVKgoMDFRMTIz27NlT0psCAABQppWLgJiWlqbrr79ePj4++uabb7Rlyxa9+uqrqlSpklMzfvx4TZgwQZMnT9batWsVERGhzp076/Dhw05NXFyc5s6dq1mzZmnFihXKyMhQdHS0Tp06VQpbBQAAUDa5jDGmtBtxLvHx8fr3v/+t77//Pt/XjTGKjIxUXFycRowYIen0aGF4eLjGjRunAQMGyOPxqGrVqpo2bZp69+4tSdq7d6+ioqK0YMECde3a9ZztSE9Pl9vtlsfjUXBw8IXbQAAAikGt+K9LuwlFtvOl7hd8mRy/i65cjCDOnz9fLVu21N13362wsDA1b95c7733nvP6jh07lJKSoi5dujjT/Pz81K5dO61cuVKStH79ep04ccKqiYyMVOPGjZ2a3LKyspSenm49AAAALnblIiD+97//1dtvv6169epp0aJFGjhwoIYMGaKPP/5YkpSSkiJJCg8Pt+YLDw93XktJSZGvr68qV65cYE1uCQkJcrvdziMqKupCbxoAAECZUy4CYnZ2tq655hqNHTtWzZs314ABA9S/f3+9/fbbVp3L5bKeG2PyTMvtbDUjR46Ux+NxHrt37/5rGwIAAFAOlIuAWK1aNTVs2NCadtVVV+n333+XJEVEREhSnpHA1NRUZ1QxIiJCx48fV1paWoE1ufn5+Sk4ONh6AAAAXOzKRUC8/vrr9csvv1jTtm3bppo1a0qSateurYiICCUmJjqvHz9+XElJSWrbtq0kqUWLFvLx8bFq9u3bp02bNjk1AAAAkLxLuwGF8eSTT6pt27YaO3asevXqpTVr1ujdd9/Vu+++K+n0n5bj4uI0duxY1atXT/Xq1dPYsWNVsWJF9enTR5LkdrvVr18/DRs2TKGhoQoJCdHw4cPVpEkTderUqTQ3DwAAoEwpFwHx2muv1dy5czVy5Ei98MILql27tiZNmqT77rvPqXn66aeVmZmpQYMGKS0tTa1bt9bixYsVFBTk1EycOFHe3t7q1auXMjMz1bFjR02dOlVeXl6lsVkAAABlUrm4D2JZwX2UAADlCfdBPI3jd9GVi3MQAQAAUHIIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwlMuAmJCQIJfLpbi4OGeaMUajR49WZGSkAgIC1L59e23evNmaLysrS4MHD1aVKlUUGBiomJgY7dmzp4RbDwAAULaVu4C4du1avfvuu2ratKk1ffz48ZowYYImT56stWvXKiIiQp07d9bhw4edmri4OM2dO1ezZs3SihUrlJGRoejoaJ06daqkNwMAAKDMKlcBMSMjQ/fdd5/ee+89Va5c2ZlujNGkSZP0zDPP6I477lDjxo310Ucf6ejRo5o5c6YkyePx6P3339err76qTp06qXnz5po+fbr+85//aMmSJaW1SQAAAGVOuQqIjz32mLp3765OnTpZ03fs2KGUlBR16dLFmebn56d27dpp5cqVkqT169frxIkTVk1kZKQaN27s1OSWlZWl9PR06wEAAHCx8y7tBhTWrFmztGHDBq1duzbPaykpKZKk8PBwa3p4eLh27drl1Pj6+lojjzk1OfPnlpCQoDFjxlyI5gMAAJQb5WIEcffu3XriiSc0ffp0+fv7F1jncrms58aYPNNyO1vNyJEj5fF4nMfu3buL3ngAAIByplwExPXr1ys1NVUtWrSQt7e3vL29lZSUpNdff13e3t7OyGHukcDU1FTntYiICB0/flxpaWkF1uTm5+en4OBg6wEAAHCxKxcBsWPHjvrPf/6j5ORk59GyZUvdd999Sk5OVp06dRQREaHExERnnuPHjyspKUlt27aVJLVo0UI+Pj5Wzb59+7Rp0yanBgAAAOXkHMSgoCA1btzYmhYYGKjQ0FBnelxcnMaOHat69eqpXr16Gjt2rCpWrKg+ffpIktxut/r166dhw4YpNDRUISEhGj58uJo0aZLnohcAAIBLWbkIiIXx9NNPKzMzU4MGDVJaWppat26txYsXKygoyKmZOHGivL291atXL2VmZqpjx46aOnWqvLy8SrHlAAAAZYvLGGNKuxHlRXp6utxutzweD+cjAgDKvFrxX5d2E4ps50vdL/gyOX4XXbk4BxEAAAAlh4AIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALAQEAEAAGAhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALAREAAAAWAiIAAAAsBAQAQAAYCEgAgAAwEJABAAAgKVcBMSEhARde+21CgoKUlhYmHr27KlffvnFqjHGaPTo0YqMjFRAQIDat2+vzZs3WzVZWVkaPHiwqlSposDAQMXExGjPnj0luSkAAABlXrkIiElJSXrssce0evVqJSYm6uTJk+rSpYuOHDni1IwfP14TJkzQ5MmTtXbtWkVERKhz5846fPiwUxMXF6e5c+dq1qxZWrFihTIyMhQdHa1Tp06VxmYBAACUSS5jjCntRhTVgQMHFBYWpqSkJN10000yxigyMlJxcXEaMWKEpNOjheHh4Ro3bpwGDBggj8ejqlWratq0aerdu7ckae/evYqKitKCBQvUtWvXc643PT1dbrdbHo9HwcHBxbqNAAD8VbXivy7tJhTZzpe6X/BlcvwuunIxgpibx+ORJIWEhEiSduzYoZSUFHXp0sWp8fPzU7t27bRy5UpJ0vr163XixAmrJjIyUo0bN3ZqcsvKylJ6err1AAAAuNiVu4BojNHQoUN1ww03qHHjxpKklJQUSVJ4eLhVGx4e7ryWkpIiX19fVa5cucCa3BISEuR2u51HVFTUhd4cAACAMqfcBcTHH39cP/30kz755JM8r7lcLuu5MSbPtNzOVjNy5Eh5PB7nsXv37vNvOAAAQDlRrgLi4MGDNX/+fH333XeqXr26Mz0iIkKS8owEpqamOqOKEREROn78uNLS0gqsyc3Pz0/BwcHWAwAA4GJXLgKiMUaPP/645syZo2+//Va1a9e2Xq9du7YiIiKUmJjoTDt+/LiSkpLUtm1bSVKLFi3k4+Nj1ezbt0+bNm1yagAAACB5l3YDCuOxxx7TzJkz9cUXXygoKMgZKXS73QoICJDL5VJcXJzGjh2revXqqV69eho7dqwqVqyoPn36OLX9+vXTsGHDFBoaqpCQEA0fPlxNmjRRp06dSnPzAAAAypRyERDffvttSVL79u2t6R9++KEefPBBSdLTTz+tzMxMDRo0SGlpaWrdurUWL16soKAgp37ixIny9vZWr169lJmZqY4dO2rq1Kny8vIqqU0BAAAo88rlfRBLC/dRAgCUJ9wH8TSO30VXLs5BBAAAQMkhIAIAAMBCQAQAAICFgAgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAAi3dpNwDAxalW/Nel3YQi2/lS99JuAgCUCYwgAgAAwEJABAAAgIWACAAAAAsBEQAAABYCIgAAACwERAAAAFgIiAAAALBwH0T8JdzrDgCAiw8BESgHymMQBwCUXwREXHIIWwAAnB3nIAIAAMDCCCIAoMSVx5F8zl/GpYSAWIaUxx0mAAC4+BAQAaAc44clgOLAOYgAAACwMIIIAEAhMFqLSwkjiAAAALAQEAEAAGAhIAIAAMBCQAQAAIDlkrxI5a233tLLL7+sffv2qVGjRpo0aZJuvPHG0m4WgFLGRQgAcNolN4I4e/ZsxcXF6ZlnntHGjRt14403qlu3bvr9999Lu2kAAABlwiUXECdMmKB+/frp4Ycf1lVXXaVJkyYpKipKb7/9dmk3DQAAoEy4pP7EfPz4ca1fv17x8fHW9C5dumjlypV56rOyspSVleU893g8kqT09PRiaV921tFiWS4AAOVFcRxjc5ZpjLngy75YXVIB8Y8//tCpU6cUHh5uTQ8PD1dKSkqe+oSEBI0ZMybP9KioqGJrIwAAlzL3pOJb9uHDh+V2u4tvBReRSyog5nC5XNZzY0yeaZI0cuRIDR061HmenZ2tP//8U6GhofnW55aenq6oqCjt3r1bwcHBf73hOCf6vGTR3yWL/i559HnJKq7+Nsbo8OHDioyMvGDLvNhdUgGxSpUq8vLyyjNamJqammdUUZL8/Pzk5+dnTatUqVKR1xscHMyOpYTR5yWL/i5Z9HfJo89LVnH0NyOHRXNJXaTi6+urFi1aKDEx0ZqemJiotm3bllKrAAAAypZLagRRkoYOHarY2Fi1bNlSbdq00bvvvqvff/9dAwcOLO2mAQAAlAmXXEDs3bu3Dh48qBdeeEH79u1T48aNtWDBAtWsWfOCr8vPz0+jRo3K82dqFB/6vGTR3yWL/i559HnJor/LDpfhmm8AAACc4ZI6BxEAAADnRkAEAACAhYAIAAAACwERAAAAFgIiAAAALATEv+itt95S7dq15e/vrxYtWuj7778/a31SUpJatGghf39/1alTR1OmTCmhll4citLfc+bMUefOnVW1alUFBwerTZs2WrRoUQm29uJQ1M94jn//+9/y9vbW1VdfXbwNvMgUtb+zsrL0zDPPqGbNmvLz81PdunX1wQcflFBry7+i9veMGTPUrFkzVaxYUdWqVdNDDz2kgwcPllBry7fly5erR48eioyMlMvl0rx58845D8fMUmRw3mbNmmV8fHzMe++9Z7Zs2WKeeOIJExgYaHbt2pVv/X//+19TsWJF88QTT5gtW7aY9957z/j4+Jh//etfJdzy8qmo/f3EE0+YcePGmTVr1pht27aZkSNHGh8fH7Nhw4YSbnn5VdQ+z3Ho0CFTp04d06VLF9OsWbOSaexF4Hz6OyYmxrRu3dokJiaaHTt2mB9++MH8+9//LsFWl19F7e/vv//eVKhQwbz22mvmv//9r/n+++9No0aNTM+ePUu45eXTggULzDPPPGM+//xzI8nMnTv3rPUcM0sXAfEvaNWqlRk4cKA1rUGDBiY+Pj7f+qeffto0aNDAmjZgwABz3XXXFVsbLyZF7e/8NGzY0IwZM+ZCN+2idb593rt3b/Pss8+aUaNGERCLoKj9/c033xi3220OHjxYEs276BS1v19++WVTp04da9rrr79uqlevXmxtvFgVJiByzCxd/In5PB0/flzr169Xly5drOldunTRypUr851n1apVeeq7du2qdevW6cSJE8XW1ovB+fR3btnZ2Tp8+LBCQkKKo4kXnfPt8w8//FC//fabRo0aVdxNvKicT3/Pnz9fLVu21Pjx43X55Zfryiuv1PDhw5WZmVkSTS7Xzqe/27Ztqz179mjBggUyxmj//v3617/+pe7du5dEky85HDNL1yX3X+1dKH/88YdOnTql8PBwa3p4eLhSUlLynSclJSXf+pMnT+qPP/5QtWrViq295d359Hdur776qo4cOaJevXoVRxMvOufT57/++qvi4+P1/fffy9ub3UtRnE9///e//9WKFSvk7++vuXPn6o8//tCgQYP0559/ch7iOZxPf7dt21YzZsxQ7969dezYMZ08eVIxMTF64403SqLJlxyOmaWLEcS/yOVyWc+NMXmmnas+v+nIX1H7O8cnn3yi0aNHa/bs2QoLCyuu5l2UCtvnp06dUp8+fTRmzBhdeeWVJdW8i05RPuPZ2dlyuVyaMWOGWrVqpVtvvVUTJkzQ1KlTGUUspKL095YtWzRkyBA9//zzWr9+vRYuXKgdO3Zo4MCBJdHUSxLHzNLDT/zzVKVKFXl5eeX5pZmamprnF0+OiIiIfOu9vb0VGhpabG29GJxPf+eYPXu2+vXrp88++0ydOnUqzmZeVIra54cPH9a6deu0ceNGPf7445JOBxhjjLy9vbV48WLdfPPNJdL28uh8PuPVqlXT5ZdfLrfb7Uy76qqrZIzRnj17VK9evWJtc3l2Pv2dkJCg66+/Xk899ZQkqWnTpgoMDNSNN96oF198kRGtC4xjZuliBPE8+fr6qkWLFkpMTLSmJyYmqm3btvnO06ZNmzz1ixcvVsuWLeXj41Nsbb0YnE9/S6dHDh988EHNnDmT84SKqKh9HhwcrP/85z9KTk52HgMHDlT9+vWVnJys1q1bl1TTy6Xz+Yxff/312rt3rzIyMpxp27ZtU4UKFVS9evVibW95dz79ffToUVWoYB82vby8JP3fyBYuHI6ZpayULo65KOTcIuH99983W7ZsMXFxcSYwMNDs3LnTGGNMfHy8iY2NdepzLtl/8sknzZYtW8z777/PJftFUNT+njlzpvH29jZvvvmm2bdvn/M4dOhQaW1CuVPUPs+Nq5iLpqj9ffjwYVO9enVz1113mc2bN5ukpCRTr1498/DDD5fWJpQrRe3vDz/80Hh7e5u33nrL/Pbbb2bFihWmZcuWplWrVqW1CeXK4cOHzcaNG83GjRuNJDNhwgSzceNG57ZCHDPLFgLiX/Tmm2+amjVrGl9fX3PNNdeYpKQk57W+ffuadu3aWfXLli0zzZs3N76+vqZWrVrm7bffLuEWl29F6e927doZSXkeffv2LfmGl2NF/YyfiYBYdEXt761bt5pOnTqZgIAAU716dTN06FBz9OjREm51+VXU/n799ddNw4YNTUBAgKlWrZq57777zJ49e0q41eXTd999d9Z9MsfMssVlDOPiAAAA+D+cgwgAAAALAREAAAAWAiIAAAAsBEQAAABYCIgAAACwEBABAABgISACAADAQkAEAACAhYAIAAAACwERAAAAFgIiAAAALP8fOffMwOMcoi0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#how many PSMs of these peptides with potential substitutions have potential substitution\n",
    "\n",
    "a=comb_datasets_clun.groupby(\"database_peptide\").apply(lambda x: Counter(x.isSubstitution_residue)[True]/len(x))\n",
    "plt.hist(a);\n",
    "plt.title(\"fraction of PSMs with expected substitutions per peptides with substiutions\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6a7a15e6-194c-4c20-9228-be3d240efe9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Canon       1431\n",
       "unique_to_Noncanon     721\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#with at least one PSM with isoSAAV\n",
    "comb_datasets_clun.drop_duplicates(\"database_peptide\").peptide_class.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d2c65e24-827a-4caa-ac15-3be0bf3ef517",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Canon       992\n",
       "unique_to_Noncanon    560\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#by protein\n",
    "comb_datasets_clun.drop_duplicates(\"proteins_list\").peptide_class.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2683ca90-3c8c-4469-b830-d5be196997e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Canon       3458\n",
       "unique_to_Noncanon    1053\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#out of\n",
    "comb_datasets[comb_datasets.peptide_class.isin([\"unique_to_Noncanon\",\"unique_to_Canon\"])].drop_duplicates(\"database_peptide\").peptide_class.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "70b3dd1a-5729-4eb8-b0f4-eb175f8bd570",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.41382301908617697, 0.6847103513770181)"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1431/3458 ,  721/1053"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c32210d9-c3a1-4eb5-a0c0-7ea16243a7d4",
   "metadata": {},
   "source": [
    "## Skip"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "257ab6a1-95d3-413a-bd77-24827a71d9d4",
   "metadata": {},
   "source": [
    "#### creat all possible peptides with expected substitutions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "b9111aa8-443a-453a-8ead-86109cf20774",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#creat all possible peptides with expected substitutions\n",
    "\n",
    "columns=[\"database_peptide\",\"peptide_class\",\"leadprot\",'peptide_length', 'peptide_position','isSubstitution']\n",
    "list_columns=['dataset','spectrum_title', 'scan', 'spectrum_file','observed_retention_time', 'charge','modifications_masses']\n",
    "new_columns=[\"fraction_psm_wsub\",\"modifications_set\",\"residues_to_sub\",\"pos_to_sub\",\"sub_rule\",\"peptides_w1sub\",\"saav_info_pep\"]\n",
    "comb_clun_isosub=pd.DataFrame(columns=columns+list_columns+new_columns)\n",
    "for database_peptide,group in comb_datasets_clun.groupby(\"database_peptide\"):\n",
    "    #print(database_peptide)\n",
    "    #print(group[[\"database_peptide\",\"modifications_masses\"]])\n",
    "    mods=set([e for l in group.modifications_masses if l!=\"Unmodified\" for e in l ])\n",
    "    ##mod_res=set([aa for _,aa,mass in mods if mass in unimod_df_subs.mono_mass.unique()]) #.. what if seidue is incorrect\n",
    "    ##mod_res_poss=[i for i,aa in enumerate(database_peptide) if aa in mod_res]\n",
    "    #residues expected to be substituted -> #residues after substitution\n",
    "    exp_res=[list(unimod_df_subs[unimod_df_subs.mono_mass==mass][[\"residues\",\"subs_to_residue\"]].itertuples(index=False)) for _,aa,mass in mods if mass in unimod_df_subs.mono_mass.unique()]\n",
    "    exp_res=set([e for l in exp_res for e in l])\n",
    "    #same residue can be subst to several other because od diff mass shifts, so keep tuple\n",
    "    exp_res=[(list(exp_res)[i].residues,list(exp_res)[i].subs_to_residue) for i in range(len(exp_res))]\n",
    "    #filter where 1st residue is not in peptide\n",
    "    exp_res=[(aa,_) for aa,_ in exp_res if aa in database_peptide]\n",
    "    #positions of potential substiutions\n",
    "    mod_res=set([aa for aa,_ in exp_res])\n",
    "    mod_res_poss=[i for i,aa in enumerate(database_peptide) if aa in mod_res]\n",
    "    \n",
    "    exp_peptides=[database_peptide[:i] + aa_sub + database_peptide[i+1:] for i in mod_res_poss for aa,aa_sub in exp_res if database_peptide[i]==aa]\n",
    "    #first is new aa and second is ref aa\n",
    "    saav_info=[(i,a,b) for exp_pep in exp_peptides for i,(a,b) in enumerate(zip(exp_pep,database_peptide)) if a!=b]\n",
    "    \n",
    "    #print(database_peptide,mod_res,mod_res_poss,exp_res,exp_peptides)\n",
    "    tmp=group[columns].iloc[0]\n",
    "    for col in list_columns:\n",
    "        tmp[col]=group[col].tolist()\n",
    "    tmp[\"fraction_psm_wsub\"]=Counter(group.isSubstitution_residue)[True]/len(group)\n",
    "    tmp[\"modifications_set\"]=mods\n",
    "    tmp[\"residues_to_sub\"]=mod_res\n",
    "    tmp[\"pos_to_sub\"]=mod_res_poss\n",
    "    tmp[\"sub_rule\"]=exp_res\n",
    "    tmp[\"peptides_w1sub\"]=exp_peptides\n",
    "    tmp[\"saav_info_pep\"]=saav_info\n",
    "    comb_clun_isosub=pd.concat([comb_clun_isosub,pd.DataFrame(tmp).T])\n",
    "    #break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "79d9884d-127f-46a7-adae-3079ada4b046",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub.to_pickle(\"comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f30549b-99f9-4906-a9d8-ae5282136afd",
   "metadata": {},
   "source": [
    "#### identify the genomic coordinates of potential saavs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "60a1eed7-9041-464b-8a39-826958a4397a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protpos_ofaa_inpep(posinpep,peptide_position):\n",
    "        #adjust indexing from 1 \n",
    "        posinprot=peptide_position[0][0]+1+posinpep\n",
    "        return posinprot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "d452fc3b-5f02-4dd6-a273-2069505856b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify protein posisions of saav\n",
    "\n",
    "comb_clun_isosub[\"saav_info_prot\"]=comb_clun_isosub.apply(lambda x: [(get_protpos_ofaa_inpep(peppos,x[\"peptide_position\"]),subto, tosub) for peppos, subto, tosub in x[\"saav_info_pep\"]], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "8be25c1a-6df6-47e7-8347-e7e1776087b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub.to_pickle(\"comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "fa60e4cf-8a16-4830-8001-16903c272187",
   "metadata": {},
   "outputs": [],
   "source": [
    "#file was created in explore-3-noncanon-hpp.ipynb\n",
    "\n",
    "OPcoords=pd.read_pickle(\"./openprot_query/OPcoords.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "9250ecc1-f1fc-4ca7-a548-a675a4d5c100",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_OPcoords_accession_match(data):\n",
    "    if data.accession.isna().all():\n",
    "        #print(\"no accession\")\n",
    "        return False\n",
    "    strand=data.strand.tolist()[0]\n",
    "    if pd.isna(strand):\n",
    "        #print(\"no accession matched\")\n",
    "        return False\n",
    "    return True\n",
    "    \n",
    "def get_aa_triplet(protein,posav_aapos,OPcoords):#,givestrand=False, givechr=False, givegene=False):\n",
    "    #if not (\"_\" in protein or protein.startswith(\"ENSP\")): #uniprot\n",
    "    #    data=OPcoords[OPcoords.accessions==protein]\n",
    "    #else: #altprots, ensembl, refseq\n",
    "    #    data=OPcoords[OPcoords.accession==protein]\n",
    "    data=OPcoords[(OPcoords.accessions==protein)|(OPcoords.accession==protein)]\n",
    "    #print(data, posav_aapos)\n",
    "    is_acc=check_OPcoords_accession_match(data)\n",
    "    if is_acc:\n",
    "        strand=data.strand.tolist()[0]\n",
    "        if strand==-1:\n",
    "            try:\n",
    "                triplet=data.triplets.tolist()[0][posav_aapos-1-1] #-1\n",
    "            except IndexError:\n",
    "                print(data, posav_aapos)\n",
    "                return np.nan\n",
    "        else:\n",
    "            try:\n",
    "                triplet=data.triplets.tolist()[0][posav_aapos-1-1] #\n",
    "            except IndexError:\n",
    "                return np.nan\n",
    "        return triplet\n",
    "\n",
    "def get_protein_info(protein,OPcoords,givestrand=False, givechr=False, givegene=False):\n",
    "    #if not (\"_\" in protein or protein.startswith(\"ENSP\")): #uniprot\n",
    "    #    data=OPcoords[OPcoords.accessions==protein]\n",
    "    #else: #altprots, ensembl, refseq\n",
    "    #    data=OPcoords[OPcoords.accession==protein]\n",
    "    data=OPcoords[(OPcoords.accessions==protein)|(OPcoords.accession==protein)]\n",
    "    is_acc=check_OPcoords_accession_match(data)\n",
    "    if is_acc:\n",
    "        strand=data.strand.tolist()[0]\n",
    "        if givestrand:\n",
    "            return strand\n",
    "        if givechr:\n",
    "            return data.location_chr.tolist()[0]\n",
    "        if givegene:\n",
    "            return data.genes.tolist()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "id": "b2716049-f720-4c9c-bf04-71e0fa72bd88",
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify triplet of each substitution\n",
    "\n",
    "saav_info_triplet=comb_clun_isosub.apply(lambda x: [(get_aa_triplet(x[\"leadprot\"],protpos,OPcoords),subto, tosub) \n",
    "                                                    for protpos, subto, tosub in x[\"saav_info_prot\"]], axis=1)\n",
    "comb_clun_isosub[\"saav_info_triplet\"]=saav_info_triplet\n",
    "del saav_info_triplet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "2107d531-81ac-4018-b7d4-287fa523bb75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "158"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein=\"A0A2R8YEA7\"\n",
    "data=OPcoords[(OPcoords.accessions==protein)|(OPcoords.accession==protein)]\n",
    "len(data.triplets.iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "b623f327-6030-4290-bc93-9b0069f320db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub.to_pickle(\"comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "6c45975c-6a72-42b8-8f51-286e86ab78d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(547)"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many accessions were not found? -> can not identify their saav potential\n",
    "\n",
    "comb_clun_isosub.saav_info_triplet.apply(lambda x: not isinstance(x[0][0],list)).value_counts()[True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "f77dc560-9e27-43cd-89a8-41e860681f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate chromosome, strand and gene\n",
    "#note, OPcoords has genes only of uniprot sequences!\n",
    "\n",
    "comb_clun_isosub[\"strand\"]=comb_clun_isosub.apply(lambda x: get_protein_info(x[\"leadprot\"],OPcoords,givestrand=True), axis=1)\n",
    "comb_clun_isosub[\"chromosome\"]=comb_clun_isosub.apply(lambda x: get_protein_info(x[\"leadprot\"],OPcoords,givechr=True), axis=1)\n",
    "comb_clun_isosub[\"gene\"]=comb_clun_isosub.apply(lambda x: get_protein_info(x[\"leadprot\"],OPcoords,givegene=True), axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "0991153f-de81-43dc-bc88-17592c191a46",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub.to_pickle(\"comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 643,
   "id": "504f3344-5b40-4b81-8019-739d5f3feb28",
   "metadata": {},
   "outputs": [],
   "source": [
    "reverse_compliment_dna={\"A\":\"T\",\"T\":\"A\",\"C\":\"G\",\"G\":\"C\"}\n",
    "aminoacid_table={\n",
    "    'A': ['GCT', 'GCC', 'GCA', 'GCG'],  # Alanine\n",
    "    'R': ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],  # Arginine\n",
    "    'N': ['AAT', 'AAC'],  # Asparagine\n",
    "    'D': ['GAT', 'GAC'],  # Aspartic acid\n",
    "    'C': ['TGT', 'TGC'],  # Cysteine\n",
    "    'Q': ['CAA', 'CAG'],  # Glutamine\n",
    "    'E': ['GAA', 'GAG'],  # Glutamic acid\n",
    "    'G': ['GGT', 'GGC', 'GGA', 'GGG'],  # Glycine\n",
    "    'H': ['CAT', 'CAC'],  # Histidine\n",
    "    'I': ['ATT', 'ATC', 'ATA'],  # Isoleucine\n",
    "    'L': ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],  # Leucine\n",
    "    'K': ['AAA', 'AAG'],  # Lysine\n",
    "    'M': ['ATG'],  # Methionine (Start codon)\n",
    "    'F': ['TTT', 'TTC'],  # Phenylalanine\n",
    "    'P': ['CCT', 'CCC', 'CCA', 'CCG'],  # Proline\n",
    "    'S': ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],  # Serine\n",
    "    'T': ['ACT', 'ACC', 'ACA', 'ACG'],  # Threonine\n",
    "    'W': ['TGG'],  # Tryptophan\n",
    "    'Y': ['TAT', 'TAC'],  # Tyrosine\n",
    "    'V': ['GTT', 'GTC', 'GTA', 'GTG'],  # Valine\n",
    "    '*': ['TAA', 'TAG', 'TGA']  # Stop codons\n",
    "}\n",
    "\n",
    "aminoacid_table_rev={codon:aa for aa, codons in aminoacid_table.items() for codon in codons}\n",
    "\n",
    "def get_codon_fromfasta(poss,chromosome,ENSfasta): #strand\n",
    "    #print(poss)\n",
    "    if isinstance(poss,float): return np.nan\n",
    "    codon=[str(ENSfasta[chromosome].seq)[pos-1] for pos in poss]\n",
    "    #rev_codon=None\n",
    "    #if strand==-1:\n",
    "    #        rev_codon=[reverse_compliment_dna[bp] for bp in codon]\n",
    "    return codon #rev_codon\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "id": "1b383b01-9ff0-4c40-ac02-f038e78ec795",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#get reference codon sequence, caution strand\n",
    "\n",
    "comb_clun_isosub[\"saav_info_refcod\"]=comb_clun_isosub.apply(lambda x: [(get_codon_fromfasta(triplet,x[\"chromosome\"],ENSfasta),subto, tosub)  \n",
    "                                                                       for triplet, subto, tosub in x[\"saav_info_triplet\"]] \n",
    "                                                            if isinstance(x[\"saav_info_triplet\"][0][0],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "id": "59d369bd-101c-4eb3-b6a5-c037abc712c3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "comb_clun_isosub[\"saav_info_refcod_rev\"]=comb_clun_isosub.apply(lambda x: [([reverse_compliment_dna[bp] for bp in codon] \n",
    "                                                                            if isinstance(codon,list) else np.nan\n",
    "                                                                            ,subto, tosub) \n",
    "                                                                       for codon, subto, tosub in x[\"saav_info_refcod\"]]\n",
    "                                                                if x[\"strand\"]==-1 and isinstance(x[\"saav_info_refcod\"],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9156eaa-b9c7-4bdd-92a3-6c2ac0b12f10",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if isinstance(codon,list) else np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "id": "4fb054c6-734d-418a-ab11-c9d803e97142",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub.to_pickle(\"comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "id": "52239661-cebb-4b45-b4fb-32fca07b3798",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_isexpected_aa(row,i,OPcoords,ENSfasta):\n",
    "    #print(row[\"leadprot\"],row[\"saav_info_refcod\"][i],row[\"saav_info_triplet\"][i][0])\n",
    "    strand=row[\"strand\"]\n",
    "    codon= row[\"saav_info_refcod_rev\"][i][0] if strand ==-1 else row[\"saav_info_refcod\"][i][0]\n",
    "    is_expected_aa=\"\".join(codon) in aminoacid_table[row[\"saav_info_refcod\"][i][2]]\n",
    "    #correct weird cases - why?\n",
    "    if not is_expected_aa:\n",
    "        print(\"reference codon indeed encodes reference amino acid - \", is_expected_aa, row[\"leadprot\"])\n",
    "        #print(row)\n",
    "        protein=row[\"leadprot\"]\n",
    "        data=OPcoords[(OPcoords.accessions==protein)|(OPcoords.accession==protein)]\n",
    "        if strand==-1:\n",
    "            new_triplet=data.triplets.tolist()[0][row[\"saav_info_prot\"][i][0]-1]\n",
    "        else:\n",
    "            new_triplet=data.triplets.tolist()[0][row[\"saav_info_prot\"][i][0]-1]\n",
    "        new_codon=get_codon_fromfasta(new_triplet,row[\"chromosome\"],ENSfasta)\n",
    "        if strand==-1:\n",
    "            new_codon_ref=new_codon\n",
    "            new_codon=[reverse_compliment_dna[bp] for bp in new_codon]\n",
    "        new_is_expected_aa=\"\".join(new_codon) in aminoacid_table[row[\"saav_info_refcod\"][i][2]]\n",
    "        #if solved the case\n",
    "        if new_is_expected_aa:\n",
    "            print(\"new\", new_is_expected_aa, strand, codon, new_codon, row[\"saav_info_triplet\"][i][0], new_triplet)\n",
    "            dat={\"triplet\":new_triplet,\"refcod\": new_codon_ref if strand==-1 else new_codon ,\n",
    "                 \"refcod_rev\": new_codon if strand==-1 else np.nan , \"is_expected_aa\": new_is_expected_aa}\n",
    "            return dat\n",
    "         #not solvable cases - shift in codon and frame - weird\n",
    "        else:\n",
    "            print(\"new\", new_is_expected_aa, strand, codon, new_codon, row[\"saav_info_triplet\"][i][0], new_triplet)\n",
    "            #print(row)\n",
    "            return new_is_expected_aa\n",
    "    else:\n",
    "        return is_expected_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "fa9fc1b4-034c-4a13-93b3-12ff226b9e02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[39951384, 39951385, 39951386]"
      ]
     },
     "execution_count": 265,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein=\"IP_151829\"\n",
    "data=OPcoords[(OPcoords.accessions==protein)|(OPcoords.accession==protein)]\n",
    "data.triplets.tolist()[0][30-1-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 297,
   "id": "6aa9f75a-cb61-4652-93fa-2719dc832e5a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reference codon indeed encodes reference amino acid -  False H7C5E6\n",
      "new False -1.0 ['C', 'G', 'G'] ['G', 'G', 'G'] [36024452, 36024451, 36024450] [36024449, 36024448, 36024447]\n",
      "reference codon indeed encodes reference amino acid -  False H7C5E6\n",
      "new False -1.0 ['C', 'T', 'G'] ['C', 'G', 'T'] [36024446, 36024445, 36024444] [36024443, 36024442, 36024441]\n",
      "reference codon indeed encodes reference amino acid -  False H7C5E6\n",
      "new False -1.0 ['C', 'G', 'T'] ['C', 'C', 'G'] [36024443, 36024442, 36024441] [36024440, 36024439, 36024438]\n",
      "reference codon indeed encodes reference amino acid -  False H7C5E6\n",
      "new False -1.0 ['C', 'A', 'C'] ['A', 'A', 'C'] [36019703, 36019702, 36019701] [36019700, 36019699, 36019698]\n",
      "reference codon indeed encodes reference amino acid -  False H0Y9E9\n",
      "new False -1.0 ['T', 'T', 'A'] ['T', 'T', 'G'] [60108194, 60108193, 60108192] [60108191, 60108190, 60108189]\n",
      "reference codon indeed encodes reference amino acid -  False H0Y9E9\n",
      "new False -1.0 ['T', 'T', 'G'] ['G', 'G', 'T'] [60106033, 60106032, 60106031] [60106030, 60106029, 60106028]\n",
      "reference codon indeed encodes reference amino acid -  False H0YB39\n",
      "new False -1.0 ['G', 'T', 'T'] ['C', 'T', 'T'] [179620902, 179620901, 179620900] [179620899, 179620898, 179620897]\n",
      "reference codon indeed encodes reference amino acid -  False H0YB39\n",
      "new False -1.0 ['C', 'T', 'T'] ['C', 'T', 'C'] [179620899, 179620898, 179620897] [179620896, 179620895, 179620894]\n",
      "reference codon indeed encodes reference amino acid -  False H0YB39\n",
      "new False -1.0 ['C', 'T', 'C'] ['A', 'G', 'G'] [179620896, 179620895, 179620894] [179620893, 179620892, 179619360]\n",
      "reference codon indeed encodes reference amino acid -  False H0YB39\n",
      "new False -1.0 ['G', 'T', 'A'] ['C', 'G', 'G'] [179619353, 179619352, 179619351] [179619350, 179619349, 179619348]\n",
      "reference codon indeed encodes reference amino acid -  False H0YB39\n",
      "new False -1.0 ['C', 'A', 'C'] ['A', 'G', 'G'] [179619323, 179619322, 179619321] [179619320, 179619319, 179619318]\n",
      "reference codon indeed encodes reference amino acid -  False M0R076\n",
      "new False -1.0 ['T', 'C', 'A'] ['A', 'T', 'G'] [38840130, 38840129, 38840128] [38840127, 38840126, 38840125]\n",
      "reference codon indeed encodes reference amino acid -  False M0R076\n",
      "new False -1.0 ['T', 'C', 'T'] ['G', 'C', 'T'] [38840121, 38840120, 38840119] [38840118, 38840117, 38840116]\n",
      "reference codon indeed encodes reference amino acid -  False M0R076\n",
      "new False -1.0 ['G', 'C', 'A'] ['A', 'T', 'G'] [38840109, 38840108, 38840107] [38840106, 38840105, 38840104]\n",
      "reference codon indeed encodes reference amino acid -  False M0R076\n",
      "new False -1.0 ['G', 'A', 'G'] ['T', 'G', 'A'] [38837635, 38837634, 38837633] [38837632, 38837631, 38837630]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['C', 'C', 'C'] ['C', 'G', 'T'] [111755705, 111755706, 111755707] [111755708, 111755709, 111755710]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['C', 'C', 'A'] ['T', 'G', 'G'] [111755711, 111755712, 111755713] [111755714, 111755715, 111755716]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['C', 'C', 'T'] ['C', 'C', 'C'] [111755717, 111755718, 111755719] [111755720, 111755721, 111755722]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['C', 'C', 'C'] ['G', 'A', 'G'] [111755720, 111755721, 111755722] [111755723, 111755724, 111755725]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['C', 'G', 'A'] ['T', 'T', 'C'] [111755738, 111755739, 111755740] [111755741, 111755742, 111755743]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['T', 'C', 'A'] ['T', 'A', 'A'] [111781922, 111781923, 111781924] [111781925, 111781926, 111781927]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['C', 'C', 'G'] ['T', 'C', 'A'] [111781946, 111781947, 111781948] [111781949, 111781950, 111781951]\n",
      "reference codon indeed encodes reference amino acid -  False F8VP50\n",
      "new False 1.0 ['G', 'C', 'A'] ['G', 'G', 'A'] [111781952, 111781953, 111781954] [111781955, 111781956, 111781957]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y6G0\n",
      "new True 1.0 ['T', 'G', 'T'] ['G', 'G', 'G'] [41236495, 41236496, 41236497] [41236498, 41236499, 41236564]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y6G0\n",
      "new False 1.0 ['G', 'C', 'C'] ['T', 'T', 'T'] [41236565, 41236566, 41236567] [41236568, 41236569, 41236570]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y6G0\n",
      "new False 1.0 ['G', 'A', 'C'] ['A', 'T', 'A'] [41236589, 41236590, 41236591] [41236592, 41236593, 41236594]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y6G0\n",
      "new False 1.0 ['A', 'G', 'C'] ['T', 'G', 'C'] [41236598, 41236599, 41236600] [41236601, 41236602, 41236603]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y6G0\n",
      "new False 1.0 ['T', 'G', 'C'] ['T', 'G', 'T'] [41236601, 41236602, 41236603] [41236604, 41236605, 41236606]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y6G0\n",
      "new False 1.0 ['T', 'G', 'T'] ['T', 'T', 'T'] [41236604, 41236605, 41236606] [41236607, 41236608, 41236609]\n",
      "reference codon indeed encodes reference amino acid -  False M0QYL7\n",
      "new False -1.0 ['G', 'A', 'T'] ['G', 'G', 'A'] [38849874, 38849873, 38849872] [38849871, 38847434, 38847433]\n",
      "reference codon indeed encodes reference amino acid -  False M0QYL7\n",
      "new False -1.0 ['C', 'T', 'C'] ['C', 'C', 'C'] [38847402, 38847401, 38847400] [38847399, 38847398, 38847397]\n",
      "reference codon indeed encodes reference amino acid -  False M0R0I5\n",
      "new False -1.0 ['C', 'C', 'C'] ['C', 'A', 'G'] [6419218, 6419217, 6419216] [6419215, 6419214, 6419213]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y7C0\n",
      "new False 1.0 ['G', 'C', 'G'] ['G', 'C', 'G'] [173334, 173335, 173336] [173337, 173338, 173339]\n",
      "reference codon indeed encodes reference amino acid -  False A0A2R8Y7C0\n",
      "new False 1.0 ['C', 'A', 'G'] ['G', 'G', 'C'] [173385, 173386, 173387] [173388, 173389, 173390]\n",
      "reference codon indeed encodes reference amino acid -  False H0Y7F8\n",
      "new False -1.0 ['C', 'A', 'T'] ['G', 'A', 'T'] [125484483, 125484482, 125484481] [125484480, 125484479, 125484478]\n",
      "reference codon indeed encodes reference amino acid -  False H0Y7F8\n",
      "new False -1.0 ['C', 'A', 'C'] ['A', 'A', 'C'] [125484462, 125484461, 125484460] [125484459, 125484458, 125484457]\n",
      "reference codon indeed encodes reference amino acid -  False H0Y7F8\n",
      "new False -1.0 ['A', 'A', 'T'] ['A', 'T', 'T'] [125444591, 125444590, 125444589] [125444588, 125444587, 125444586]\n",
      "reference codon indeed encodes reference amino acid -  False H0Y7F8\n",
      "new False -1.0 ['T', 'A', 'A'] ['A', 'C', 'T'] [125444585, 125444584, 125444583] [125444582, 125444581, 125444580]\n"
     ]
    }
   ],
   "source": [
    "#check if it encodes an expected aa, if not -try to correct result\n",
    "#I notice that we need left shift on 1 codon with forward strand\n",
    "#and to 1 right in reverse\n",
    "\n",
    "comb_clun_isosub[\"is_expected_aa\"]=comb_clun_isosub.apply(lambda x: [check_isexpected_aa(x,i,OPcoords,ENSfasta)\n",
    "                                                                       for i in range(len(x[\"saav_info_refcod\"]))]\n",
    "                                                                if isinstance(x[\"saav_info_refcod\"],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 298,
   "id": "46ac42e0-8ad2-4e5f-8c3e-e5ac8aec522f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub.to_pickle(\"comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "id": "60d2f7c5-8d16-47fc-9afa-2b0172d991a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({True: 3027, False: 10})"
      ]
     },
     "execution_count": 299,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many peptides have False is_expected_aa - 10\n",
    "Counter(comb_clun_isosub.apply(lambda x: all(x[\"is_expected_aa\"]) if isinstance(x[\"saav_info_refcod\"],list) else True, axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1086,
   "id": "5bb26f66-240a-445b-a4fd-89ee9caa3c70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0032927230819888046"
      ]
     },
     "execution_count": 1086,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "10/3037"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "id": "563787c0-230a-4a5e-9d4d-2a35668d294d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3037"
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(comb_clun_isosub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "id": "e3578394-5152-4bce-b2d1-80a1794d9f29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "450"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many protein accessions were NOT found in OP\n",
    "len(comb_clun_isosub[comb_clun_isosub.is_expected_aa.isna()].leadprot.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1082,
   "id": "97e520af-be84-41d2-85a1-d316adf34680",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub[comb_clun_isosub.is_expected_aa.isna()].leadprot.unique()[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1080,
   "id": "3db16bca-0bc9-4251-84a8-68cfaf510883",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OPcoords[(OPcoords.accessions=='B4DHC4')|(OPcoords.accession=='B4DHC4')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1081,
   "id": "ac00e379-b05a-4b10-9610-1488e5245d7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub[comb_clun_isosub.leadprot=='B4DHC4']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "id": "8825925d-c890-40bf-96be-cae4dc85d639",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "450"
      ]
     },
     "execution_count": 334,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#are they really missing? - yes\n",
    "count=0\n",
    "for protein in comb_clun_isosub[comb_clun_isosub.is_expected_aa.isna()].leadprot.unique():\n",
    "    data=OPcoords[(OPcoords.accessions==protein)|(OPcoords.accession==protein)]\n",
    "    if all(data.transcript_translation_id.isna()):\n",
    "        count+=1\n",
    "count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "id": "b07186c2-4081-4927-b13c-d47b8efbe599",
   "metadata": {},
   "outputs": [],
   "source": [
    "#exclude peptides of proteins for which we dont have genomic coordinares (not found accession in OP database)\n",
    "\n",
    "comb_clun_isosub_filt=comb_clun_isosub[~comb_clun_isosub.is_expected_aa.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "id": "804f9b08-1804-49fa-8d99-b2bc2330ea20",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2490"
      ]
     },
     "execution_count": 318,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(comb_clun_isosub_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "id": "de86672e-78ae-4de6-8f6c-ad84fbfff8c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1851"
      ]
     },
     "execution_count": 322,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many protein accessions WERE found in OP\n",
    "len(comb_clun_isosub[~comb_clun_isosub.is_expected_aa.isna()].leadprot.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1072,
   "id": "655a967b-d00a-4f6d-b641-effd1c002e9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#450/1851"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 780,
   "id": "88b09f0a-7781-4cc4-9d66-22feb0a204f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({True: 1309, False: 542})"
      ]
     },
     "execution_count": 780,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=comb_clun_isosub[~comb_clun_isosub.is_expected_aa.isna()].leadprot.unique()\n",
    "b=[i.startswith((\"IP_\",\"II_\")) for i in a]\n",
    "Counter(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 781,
   "id": "34dec59f-5f38-4fdd-9504-5aa5edfd106a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({True: 1309, False: 992})"
      ]
     },
     "execution_count": 781,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=comb_clun_isosub.leadprot.unique()\n",
    "b=[i.startswith((\"IP_\",\"II_\")) for i in a]\n",
    "Counter(b)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0af8d10-317b-46a4-af37-c4c76bddb5ac",
   "metadata": {},
   "source": [
    "#### query gnomAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f5331624-a8c6-43c1-9dff-0329faaee5a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_continuous_intervals(nums):\n",
    "    nums = sorted(set(nums))  # Ensure sorted and unique\n",
    "    intervals = []\n",
    "    \n",
    "    start = prev = nums[0]\n",
    "    for num in nums[1:]:\n",
    "        if num == prev + 1:\n",
    "            prev = num\n",
    "        else:\n",
    "            intervals.append((start, prev))\n",
    "            start = prev = num\n",
    "    intervals.append((start, prev))  # Append last interval\n",
    "    return intervals\n",
    "def merge_close_intervals(intervals, max_gap=1000):\n",
    "    intervals.sort() # Sort intervals by start\n",
    "    merged = [intervals[0]]\n",
    "    for start, end in intervals[1:]:\n",
    "        last_start, last_end = merged[-1]\n",
    "        if start - last_end <= max_gap:\n",
    "            merged[-1] = (last_start, max(end, last_end))  # Merge the intervals\n",
    "        else:\n",
    "            merged.append((start, end)) \n",
    "    return merged\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "id": "d738a8a0-faab-4366-bb28-2564c7f45584",
   "metadata": {},
   "outputs": [],
   "source": [
    "#API for gnomAD\n",
    "\n",
    "async def gnomad_response(query):\n",
    "    transport = AIOHTTPTransport(url=\"https://gnomad.broadinstitute.org/api\")\n",
    "    client = Client(transport=transport, fetch_schema_from_transport=True)\n",
    "    result = await client.execute_async(query)\n",
    "    return result\n",
    "    \n",
    "async  def gnomad_api(**kwargs):\n",
    "    #https://gnomad.broadinstitute.org/downloads#api    \n",
    "    #request:\n",
    "    # Position    # VEP annotation\n",
    "    # Full allele count, and hemizygote/homozygote allele counts.    # Allele number\n",
    "    # Allele frequency    # Allele counts and numbers by population    # Filters and flags for exome data\n",
    "    chromosome = kwargs.get('chromosome', '1')\n",
    "    genome_start = kwargs.get('genome_start', '1')\n",
    "    genome_stop = kwargs.get('genome_stop', '1')\n",
    "    query = gql(\n",
    "        f'''\n",
    "        query VariantsInGene {{\n",
    "        region(chrom: \"{chromosome}\", start: {genome_start}, stop: {genome_stop}, reference_genome: GRCh38) {{\n",
    "            variants(dataset: gnomad_r4) {{\n",
    "              gene_symbol\n",
    "          \t\tgene_id\n",
    "              variant_id\n",
    "              pos\n",
    "              rsids\n",
    "              transcript_id\n",
    "              consequence\n",
    "              hgvsp\n",
    "               exome {{     \n",
    "                ac\n",
    "                an\n",
    "                af \n",
    "            \tflags\n",
    "              }}\n",
    "              genome {{     \n",
    "                ac\n",
    "                an\n",
    "                af \n",
    "            \tflags\n",
    "              }}\n",
    "            }}\n",
    "          }}\n",
    "        }}\n",
    "    '''\n",
    "    )\n",
    "    #populations {\n",
    "    #              id\n",
    "    #              ac\n",
    "    #              an\n",
    "    #              ac_hemi\n",
    "    #              ac_hom\n",
    "    #            }\n",
    "    result= await gnomad_response(query)\n",
    "    return result\n",
    "\n",
    "def get_varvep(result,vep='missense'):\n",
    "    a=result['gene']['variants']\n",
    "    return [ a[i] for i in range(len(a)) if vep in a[i]['consequence'] ]\n",
    "    \n",
    "def get_saav(hgvsp,prot_pos,ref_aa,new_aa):\n",
    "    exp_hgvsp=f\"p.{hgvp_map[ref_aa]}{prot_pos}{hgvp_map[new_aa]}\"\n",
    "    return hgvsp==exp_hgvsp\n",
    "\n",
    "async def check_saavs_bygene(chromosome,gencodon_sts): #gene,vep='missense'\n",
    "    genome_start,genome_stop=min(gencodon_sts),max(gencodon_sts)\n",
    "    #get SNPs from gnomAD\n",
    "    gnomad_result= await gnomad_api(chromosome=chromosome,genome_start=genome_start,genome_stop=genome_stop) #gene_symbol=gene)\n",
    "    is_saavs={}\n",
    "    for gencodon_st in gencodon_sts:\n",
    "        is_saav=[]\n",
    "        for snp in gnomad_result[\"region\"]['variants']:\n",
    "            if gencodon_st==snp['pos']: is_saav.append(snp)\n",
    "        is_saavs[gencodon_st]=is_saav\n",
    "    return is_saavs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "id": "77d468ea-6fa9-41ac-8718-f648cf838a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#creat a map of coordinates, to grep bigger ranges in one query\n",
    "\n",
    "gnomAD_qgr={}\n",
    "for chromosome, group in comb_clun_isosub_filt.groupby(\"chromosome\"):\n",
    "    all_triplets=group.saav_info_triplet.tolist()\n",
    "    all_coordinates=list(set([e for row in all_triplets for l in row for e in l[0]]))\n",
    "    chrom_intervals=get_continuous_intervals(all_coordinates)\n",
    "    gnomAD_qgr[chromosome]=chrom_intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "id": "64567f1c-ca06-4a37-916b-b39c28bb39ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5846"
      ]
     },
     "execution_count": 340,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#estimate, how many queries will be made to gnomAD -> too many\n",
    "\n",
    "len([1 for k,v in gnomAD_qgr.items() for l in v])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "id": "9f8a31b0-9128-4017-bb6d-1d4c81cebe32",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#merge intervals with max gap 10000\n",
    "\n",
    "gnomAD_qgr_m={}\n",
    "for chromosome, intervals in gnomAD_qgr.items():\n",
    "    new_intervals=merge_close_intervals(intervals,10000)\n",
    "    gnomAD_qgr_m[chromosome]=new_intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "id": "c51bea6e-65f1-4920-aad8-2edb2f68fdea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1732"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#estimate, how many queries will be made to gnomAD -  many\n",
    "\n",
    "len([1 for k,v in gnomAD_qgr_m.items() for l in v])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "id": "a04410ea-97eb-4bdb-9e73-d05bddf3b059",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gnomAD_qgr_res={k:{} for k,v in gnomAD_qgr_m.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "id": "3b7a94d0-ef7a-4f35-bdec-05627fe590a7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing: 15, interval 77\n",
      "Saved\n",
      "Processing: 15, interval 78\n",
      "Saved\n",
      "Processing: 15, interval 79\n",
      "Saved\n",
      "Processing: 15, interval 80\n",
      "Saved\n",
      "Processing: 15, interval 81\n",
      "Saved\n",
      "Processing: 15, interval 82\n",
      "Saved\n",
      "Processing: 15, interval 83\n",
      " at 15, interval 83. Retrying...\n",
      "Processing: 15, interval 83\n",
      "Saved\n",
      "Processing: 15, interval 84\n",
      "Saved\n",
      "Processing: 15, interval 85\n",
      "Saved\n",
      "Processing: 15, interval 86\n",
      "Saved\n",
      "Processing: 15, interval 87\n",
      "Saved\n",
      "Processing: 15, interval 88\n",
      "Saved\n",
      "Processing: 15, interval 89\n",
      "Saved\n",
      "Processing: 15, interval 90\n",
      "Saved\n",
      "Processing: 15, interval 91\n",
      "Saved\n",
      "Processing: 15, interval 92\n",
      "Saved\n",
      "Processing: 15, interval 93\n",
      "Saved\n",
      "Processing: 15, interval 94\n",
      "Saved\n",
      "Processing: 15, interval 95\n",
      " at 15, interval 95. Retrying...\n",
      "Processing: 15, interval 95\n",
      "Saved\n",
      "Processing: 15, interval 96\n",
      "Saved\n",
      "Processing: 15, interval 97\n",
      "Saved\n",
      "Processing: 15, interval 98\n",
      "Saved\n",
      "Processing: 15, interval 99\n",
      "Saved\n",
      "Processing: 15, interval 100\n",
      "Saved\n",
      "Processing: 15, interval 101\n",
      "Saved\n",
      "Processing: 15, interval 102\n",
      "Saved\n",
      "Processing: 15, interval 103\n",
      "Saved\n",
      "Processing: 15, interval 104\n",
      "Saved\n",
      "Processing: 15, interval 105\n",
      "Saved\n",
      "Processing: 15, interval 106\n",
      "Saved\n",
      "Processing: 15, interval 107\n",
      "Saved\n",
      "Processing: 15, interval 108\n",
      "Saved\n",
      "Processing: 15, interval 109\n",
      "Saved\n",
      "Processing: 15, interval 110\n",
      "Saved\n",
      "Processing: 15, interval 111\n",
      "Saved\n",
      "Processing: 15, interval 112\n",
      "Saved\n",
      "Processing: 15, interval 113\n",
      "Saved\n",
      "Processing: 15, interval 114\n",
      "Saved\n",
      "Processing: 15, interval 115\n",
      "Saved\n",
      "Processing: 15, interval 116\n",
      "Saved\n",
      "Processing: 15, interval 117\n",
      "Saved\n",
      "Processing: 15, interval 118\n",
      " at 15, interval 118. Retrying...\n",
      "Processing: 15, interval 118\n",
      "Saved\n",
      "Processing: 15, interval 119\n",
      "Saved\n",
      "Processing: 15, interval 120\n",
      "Saved\n",
      "Processing: 15, interval 121\n",
      "Saved\n",
      "Processing: 15, interval 122\n",
      "Saved\n",
      "Processing: 15, interval 123\n",
      "Saved\n",
      "Processing: 15, interval 124\n",
      " at 15, interval 124. Retrying...\n",
      "Processing: 15, interval 124\n",
      "Saved\n",
      "Processing: 15, interval 125\n",
      "Saved\n",
      "Processing: 15, interval 126\n",
      "Saved\n",
      "Processing: 15, interval 127\n",
      "Saved\n",
      "Processing: 15, interval 128\n",
      "Saved\n",
      "Processing: 15, interval 129\n",
      "Saved\n",
      "Processing: 15, interval 130\n",
      "Saved\n",
      "Processing: 15, interval 131\n",
      "Saved\n",
      "Processing: 15, interval 132\n",
      "Saved\n",
      "Processing: 15, interval 133\n",
      "Saved\n",
      "Processing: 15, interval 134\n",
      "Saved\n",
      "Processing: 15, interval 135\n",
      "Saved\n",
      "Processing: 15, interval 136\n",
      "Saved\n",
      "Processing: 15, interval 137\n",
      "Saved\n",
      "Processing: 15, interval 138\n",
      "Saved\n",
      "Processing: 15, interval 139\n",
      "Saved\n",
      "Processing: 15, interval 140\n",
      "Saved\n",
      "Processing: 15, interval 141\n",
      "Saved\n",
      "Processing: 15, interval 142\n",
      "Saved\n",
      "Processing: 15, interval 143\n",
      "Saved\n",
      "Processing: 15, interval 144\n",
      "Saved\n",
      "Processing: 15, interval 145\n",
      "Saved\n",
      "Processing: 15, interval 146\n",
      " at 15, interval 146. Retrying...\n",
      "Processing: 15, interval 146\n",
      "Saved\n",
      "Processing: 15, interval 147\n",
      "Saved\n",
      "Processing: 16, interval 0\n",
      " at 16, interval 0. Retrying...\n",
      "Processing: 16, interval 0\n",
      "Saved\n",
      "Processing: 16, interval 1\n",
      "Saved\n",
      "Processing: 16, interval 2\n",
      "Saved\n",
      "Processing: 16, interval 3\n",
      "Saved\n",
      "Processing: 16, interval 4\n",
      "Saved\n",
      "Processing: 16, interval 5\n",
      "Saved\n",
      "Processing: 16, interval 6\n",
      "Saved\n",
      "Processing: 16, interval 7\n",
      " at 16, interval 7. Retrying...\n",
      "Processing: 16, interval 7\n",
      "Saved\n",
      "Processing: 16, interval 8\n",
      "Saved\n",
      "Processing: 16, interval 9\n",
      "Saved\n",
      "Processing: 16, interval 10\n",
      " at 16, interval 10. Retrying...\n",
      "Processing: 16, interval 10\n",
      "Saved\n",
      "Processing: 16, interval 11\n",
      "Saved\n",
      "Processing: 16, interval 12\n",
      " at 16, interval 12. Retrying...\n",
      "Processing: 16, interval 12\n",
      "Saved\n",
      "Processing: 16, interval 13\n",
      "Saved\n",
      "Processing: 16, interval 14\n",
      "Saved\n",
      "Processing: 16, interval 15\n",
      " at 16, interval 15. Retrying...\n",
      "Processing: 16, interval 15\n",
      "Saved\n",
      "Processing: 16, interval 16\n",
      "Saved\n",
      "Processing: 16, interval 17\n",
      "Saved\n",
      "Processing: 16, interval 18\n",
      "Saved\n",
      "Processing: 16, interval 19\n",
      "Saved\n",
      "Processing: 16, interval 20\n",
      " at 16, interval 20. Retrying...\n",
      "Processing: 16, interval 20\n",
      "Saved\n",
      "Processing: 16, interval 21\n",
      "Saved\n",
      "Processing: 16, interval 22\n",
      "Saved\n",
      "Processing: 16, interval 23\n",
      "Saved\n",
      "Processing: 16, interval 24\n",
      "Saved\n",
      "Processing: 16, interval 25\n",
      "Saved\n",
      "Processing: 16, interval 26\n",
      " at 16, interval 26. Retrying...\n",
      "Processing: 16, interval 26\n",
      "Saved\n",
      "Processing: 16, interval 27\n",
      "Saved\n",
      "Processing: 16, interval 28\n",
      "Saved\n",
      "Processing: 16, interval 29\n",
      "Saved\n",
      "Processing: 16, interval 30\n",
      "Saved\n",
      "Processing: 16, interval 31\n",
      "Saved\n",
      "Processing: 16, interval 32\n",
      "Saved\n",
      "Processing: 16, interval 33\n",
      "Saved\n",
      "Processing: 16, interval 34\n",
      "Saved\n",
      "Processing: 16, interval 35\n",
      "Saved\n",
      "Processing: 16, interval 36\n",
      "Saved\n",
      "Processing: 16, interval 37\n",
      "Saved\n",
      "Processing: 16, interval 38\n",
      "Saved\n",
      "Processing: 16, interval 39\n",
      "Saved\n",
      "Processing: 16, interval 40\n",
      "Saved\n",
      "Processing: 16, interval 41\n",
      "Saved\n",
      "Processing: 16, interval 42\n",
      "Saved\n",
      "Processing: 16, interval 43\n",
      "Saved\n",
      "Processing: 16, interval 44\n",
      "Saved\n",
      "Processing: 16, interval 45\n",
      "Saved\n",
      "Processing: 16, interval 46\n",
      "Saved\n",
      "Processing: 16, interval 47\n",
      "Saved\n",
      "Processing: 16, interval 48\n",
      "Saved\n",
      "Processing: 16, interval 49\n",
      " at 16, interval 49. Retrying...\n",
      "Processing: 16, interval 49\n",
      "Saved\n",
      "Processing: 16, interval 50\n",
      "Saved\n",
      "Processing: 16, interval 51\n",
      "Saved\n",
      "Processing: 16, interval 52\n",
      " at 16, interval 52. Retrying...\n",
      "Processing: 16, interval 52\n",
      "Saved\n",
      "Processing: 16, interval 53\n",
      "Saved\n",
      "Processing: 16, interval 54\n",
      "Saved\n",
      "Processing: 16, interval 55\n",
      "Saved\n",
      "Processing: 16, interval 56\n",
      "Saved\n",
      "Processing: 16, interval 57\n",
      "Saved\n",
      "Processing: 16, interval 58\n",
      "Saved\n",
      "Processing: 16, interval 59\n",
      "Saved\n",
      "Processing: 16, interval 60\n",
      "Saved\n",
      "Processing: 16, interval 61\n",
      "Saved\n",
      "Processing: 16, interval 62\n",
      "Saved\n",
      "Processing: 16, interval 63\n",
      "Saved\n",
      "Processing: 16, interval 64\n",
      "Saved\n",
      "Processing: 16, interval 65\n",
      "Saved\n",
      "Processing: 16, interval 66\n",
      "Saved\n",
      "Processing: 16, interval 67\n",
      "Saved\n",
      "Processing: 16, interval 68\n",
      "Saved\n",
      "Processing: 16, interval 69\n",
      "Saved\n",
      "Processing: 16, interval 70\n",
      "Saved\n",
      "Processing: 16, interval 71\n",
      "Saved\n",
      "Processing: 16, interval 72\n",
      "Saved\n",
      "Processing: 16, interval 73\n",
      "Saved\n",
      "Processing: 16, interval 74\n",
      "Saved\n",
      "Processing: 16, interval 75\n",
      "Saved\n",
      "Processing: 16, interval 76\n",
      "Saved\n",
      "Processing: 16, interval 77\n",
      "Saved\n",
      "Processing: 16, interval 78\n",
      "Saved\n",
      "Processing: 16, interval 79\n",
      "Saved\n",
      "Processing: 16, interval 80\n",
      "Saved\n",
      "Processing: 16, interval 81\n",
      "Saved\n",
      "Processing: 16, interval 82\n",
      "Saved\n",
      "Processing: 16, interval 83\n",
      " at 16, interval 83. Retrying...\n",
      "Processing: 16, interval 83\n",
      "Saved\n",
      "Processing: 16, interval 84\n",
      " at 16, interval 84. Retrying...\n",
      "Processing: 16, interval 84\n",
      "Saved\n",
      "Processing: 16, interval 85\n",
      " at 16, interval 85. Retrying...\n",
      "Processing: 16, interval 85\n",
      "Saved\n",
      "Processing: 16, interval 86\n",
      "Saved\n",
      "Processing: 16, interval 87\n",
      "Saved\n",
      "Processing: 16, interval 88\n",
      "Saved\n",
      "Processing: 16, interval 89\n",
      "Saved\n",
      "Processing: 16, interval 90\n",
      "Saved\n",
      "Processing: 16, interval 91\n",
      "Saved\n",
      "Processing: 16, interval 92\n",
      " at 16, interval 92. Retrying...\n",
      "Processing: 16, interval 92\n",
      "Saved\n",
      "Processing: 16, interval 93\n",
      "Saved\n",
      "Processing: 16, interval 94\n",
      "Saved\n",
      "Processing: 16, interval 95\n",
      "Saved\n",
      "Processing: 16, interval 96\n",
      "Saved\n",
      "Processing: 16, interval 97\n",
      "Saved\n",
      "Processing: 16, interval 98\n",
      "Saved\n",
      "Processing: 16, interval 99\n",
      " at 16, interval 99. Retrying...\n",
      "Processing: 16, interval 99\n",
      "Saved\n",
      "Processing: 16, interval 100\n",
      "Saved\n",
      "Processing: 16, interval 101\n",
      "Saved\n",
      "Processing: 16, interval 102\n",
      "Saved\n",
      "Processing: 16, interval 103\n",
      "Saved\n",
      "Processing: 16, interval 104\n",
      "Saved\n",
      "Processing: 16, interval 105\n",
      "Saved\n",
      "Processing: 16, interval 106\n",
      "Saved\n",
      "Processing: 16, interval 107\n",
      "Saved\n",
      "Processing: 16, interval 108\n",
      "Saved\n",
      "Processing: 16, interval 109\n",
      "Saved\n",
      "Processing: 16, interval 110\n",
      "Saved\n",
      "Processing: 16, interval 111\n",
      "Saved\n",
      "Processing: 16, interval 112\n",
      "Saved\n",
      "Processing: 16, interval 113\n",
      "Saved\n",
      "Processing: 16, interval 114\n",
      "Saved\n",
      "Processing: 16, interval 115\n",
      "Saved\n",
      "Processing: 16, interval 116\n",
      "Saved\n",
      "Processing: 16, interval 117\n",
      "Saved\n",
      "Processing: 16, interval 118\n",
      "Saved\n",
      "Processing: 16, interval 119\n",
      "Saved\n",
      "Processing: 16, interval 120\n",
      "Saved\n",
      "Processing: 16, interval 121\n",
      "Saved\n",
      "Processing: 16, interval 122\n",
      "Saved\n",
      "Processing: 16, interval 123\n",
      "Saved\n",
      "Processing: 16, interval 124\n",
      "Saved\n",
      "Processing: 16, interval 125\n",
      " at 16, interval 125. Retrying...\n",
      "Processing: 16, interval 125\n",
      "Saved\n",
      "Processing: 16, interval 126\n",
      "Saved\n",
      "Processing: 16, interval 127\n",
      "Saved\n",
      "Processing: 16, interval 128\n",
      "Saved\n",
      "Processing: 16, interval 129\n",
      "Saved\n",
      "Processing: 16, interval 130\n",
      "Saved\n",
      "Processing: 16, interval 131\n",
      "Saved\n",
      "Processing: 16, interval 132\n",
      "Saved\n",
      "Processing: 16, interval 133\n",
      "Saved\n",
      "Processing: 16, interval 134\n",
      "Saved\n",
      "Processing: 16, interval 135\n",
      "Saved\n",
      "Processing: 16, interval 136\n",
      "Saved\n",
      "Processing: 16, interval 137\n",
      "Saved\n",
      "Processing: 16, interval 138\n",
      "Saved\n",
      "Processing: 16, interval 139\n",
      "Saved\n",
      "Processing: 16, interval 140\n",
      "Saved\n",
      "Processing: 16, interval 141\n",
      "Saved\n",
      "Processing: 16, interval 142\n",
      "Saved\n",
      "Processing: 16, interval 143\n",
      "Saved\n",
      "Processing: 16, interval 144\n",
      "Saved\n",
      "Processing: 16, interval 145\n",
      "Saved\n",
      "Processing: 16, interval 146\n",
      "Saved\n",
      "Processing: 16, interval 147\n",
      "Saved\n",
      "Processing: 16, interval 148\n",
      "Saved\n",
      "Processing: 16, interval 149\n",
      "Saved\n",
      "Processing: 16, interval 150\n",
      "Saved\n",
      "Processing: 16, interval 151\n",
      "Saved\n",
      "Processing: 16, interval 152\n",
      "Saved\n",
      "Processing: 17, interval 0\n",
      "Saved\n",
      "Processing: 17, interval 1\n",
      "Saved\n",
      "Processing: 17, interval 2\n",
      "Saved\n",
      "Processing: 17, interval 3\n",
      "Saved\n",
      "Processing: 17, interval 4\n",
      "Saved\n",
      "Processing: 17, interval 5\n",
      "Saved\n",
      "Processing: 17, interval 6\n",
      "Saved\n",
      "Processing: 17, interval 7\n",
      "Saved\n",
      "Processing: 17, interval 8\n",
      "Saved\n",
      "Processing: 17, interval 9\n",
      "Saved\n",
      "Processing: 17, interval 10\n",
      "Saved\n",
      "Processing: 17, interval 11\n",
      "Saved\n",
      "Processing: 17, interval 12\n",
      "Saved\n",
      "Processing: 17, interval 13\n",
      "Saved\n",
      "Processing: 17, interval 14\n",
      "Saved\n",
      "Processing: 17, interval 15\n",
      "Saved\n",
      "Processing: 17, interval 16\n",
      "Saved\n",
      "Processing: 17, interval 17\n",
      "Saved\n",
      "Processing: 17, interval 18\n",
      "Saved\n",
      "Processing: 17, interval 19\n",
      "Saved\n",
      "Processing: 17, interval 20\n",
      "Saved\n",
      "Processing: 17, interval 21\n",
      "Saved\n",
      "Processing: 17, interval 22\n",
      "Saved\n",
      "Processing: 17, interval 23\n",
      "Saved\n",
      "Processing: 17, interval 24\n",
      "Saved\n",
      "Processing: 17, interval 25\n",
      "Saved\n",
      "Processing: 17, interval 26\n",
      "Saved\n",
      "Processing: 17, interval 27\n",
      "Saved\n",
      "Processing: 17, interval 28\n",
      "Saved\n",
      "Processing: 17, interval 29\n",
      "Saved\n",
      "Processing: 17, interval 30\n",
      "Saved\n",
      "Processing: 17, interval 31\n",
      "Saved\n",
      "Processing: 17, interval 32\n",
      "Saved\n",
      "Processing: 17, interval 33\n",
      " at 17, interval 33. Retrying...\n",
      "Processing: 17, interval 33\n",
      "Saved\n",
      "Processing: 17, interval 34\n",
      "Saved\n",
      "Processing: 17, interval 35\n",
      "Saved\n",
      "Processing: 17, interval 36\n",
      "Saved\n",
      "Processing: 17, interval 37\n",
      "Saved\n",
      "Processing: 17, interval 38\n",
      "Saved\n",
      "Processing: 17, interval 39\n",
      "Saved\n",
      "Processing: 17, interval 40\n",
      "Saved\n",
      "Processing: 17, interval 41\n",
      " at 17, interval 41. Retrying...\n",
      "Processing: 17, interval 41\n",
      "Saved\n",
      "Processing: 17, interval 42\n",
      "Saved\n",
      "Processing: 17, interval 43\n",
      " at 17, interval 43. Retrying...\n",
      "Processing: 17, interval 43\n",
      "Saved\n",
      "Processing: 17, interval 44\n",
      "Saved\n",
      "Processing: 17, interval 45\n",
      "Saved\n",
      "Processing: 17, interval 46\n",
      "Saved\n",
      "Processing: 17, interval 47\n",
      "Saved\n",
      "Processing: 17, interval 48\n",
      "Saved\n",
      "Processing: 17, interval 49\n",
      "Saved\n",
      "Processing: 17, interval 50\n",
      "Saved\n",
      "Processing: 17, interval 51\n",
      "Saved\n",
      "Processing: 17, interval 52\n",
      "Saved\n",
      "Processing: 17, interval 53\n",
      " at 17, interval 53. Retrying...\n",
      "Processing: 17, interval 53\n",
      "Saved\n",
      "Processing: 17, interval 54\n",
      "Saved\n",
      "Processing: 17, interval 55\n",
      "Saved\n",
      "Processing: 17, interval 56\n",
      "Saved\n",
      "Processing: 17, interval 57\n",
      "Saved\n",
      "Processing: 17, interval 58\n",
      "Saved\n",
      "Processing: 17, interval 59\n",
      " at 17, interval 59. Retrying...\n",
      "Processing: 17, interval 59\n",
      "Saved\n",
      "Processing: 17, interval 60\n",
      " at 17, interval 60. Retrying...\n",
      "Processing: 17, interval 60\n",
      "Saved\n",
      "Processing: 17, interval 61\n",
      "Saved\n",
      "Processing: 17, interval 62\n",
      "Saved\n",
      "Processing: 17, interval 63\n",
      " at 17, interval 63. Retrying...\n",
      "Processing: 17, interval 63\n",
      "Saved\n",
      "Processing: 17, interval 64\n",
      "Saved\n",
      "Processing: 17, interval 65\n",
      "Saved\n",
      "Processing: 17, interval 66\n",
      "Saved\n",
      "Processing: 17, interval 67\n",
      " at 17, interval 67. Retrying...\n",
      "Processing: 17, interval 67\n",
      "Saved\n",
      "Processing: 17, interval 68\n",
      "Saved\n",
      "Processing: 17, interval 69\n",
      "Saved\n",
      "Processing: 17, interval 70\n",
      "Saved\n",
      "Processing: 17, interval 71\n",
      "Saved\n",
      "Processing: 17, interval 72\n",
      " at 17, interval 72. Retrying...\n",
      "Processing: 17, interval 72\n",
      "Saved\n",
      "Processing: 17, interval 73\n",
      "Saved\n",
      "Processing: 17, interval 74\n",
      "Saved\n",
      "Processing: 17, interval 75\n",
      "Saved\n",
      "Processing: 17, interval 76\n",
      "Saved\n",
      "Processing: 17, interval 77\n",
      "Saved\n",
      "Processing: 17, interval 78\n",
      "Saved\n",
      "Processing: 17, interval 79\n",
      "Saved\n",
      "Processing: 17, interval 80\n",
      "Saved\n",
      "Processing: 17, interval 81\n",
      "Saved\n",
      "Processing: 17, interval 82\n",
      "Saved\n",
      "Processing: 17, interval 83\n",
      "Saved\n",
      "Processing: 17, interval 84\n",
      "Saved\n",
      "Processing: 17, interval 85\n",
      "Saved\n",
      "Processing: 17, interval 86\n",
      "Saved\n",
      "Processing: 17, interval 87\n",
      "Saved\n",
      "Processing: 17, interval 88\n",
      "Saved\n",
      "Processing: 17, interval 89\n",
      "Saved\n",
      "Processing: 17, interval 90\n",
      "Saved\n",
      "Processing: 17, interval 91\n",
      "Saved\n",
      "Processing: 17, interval 92\n",
      "Saved\n",
      "Processing: 17, interval 93\n",
      "Saved\n",
      "Processing: 17, interval 94\n",
      "Saved\n",
      "Processing: 17, interval 95\n",
      "Saved\n",
      "Processing: 17, interval 96\n",
      "Saved\n",
      "Processing: 17, interval 97\n",
      "Saved\n",
      "Processing: 17, interval 98\n",
      "Saved\n",
      "Processing: 17, interval 99\n",
      "Saved\n",
      "Processing: 17, interval 100\n",
      "Saved\n",
      "Processing: 17, interval 101\n",
      " at 17, interval 101. Retrying...\n",
      "Processing: 17, interval 101\n",
      "Saved\n",
      "Processing: 17, interval 102\n",
      "Saved\n",
      "Processing: 17, interval 103\n",
      "Saved\n",
      "Processing: 17, interval 104\n",
      "Saved\n",
      "Processing: 17, interval 105\n",
      "Saved\n",
      "Processing: 17, interval 106\n",
      "Saved\n",
      "Processing: 17, interval 107\n",
      "Saved\n",
      "Processing: 17, interval 108\n",
      "Saved\n",
      "Processing: 17, interval 109\n",
      "Saved\n",
      "Processing: 17, interval 110\n",
      "Saved\n",
      "Processing: 17, interval 111\n",
      " at 17, interval 111. Retrying...\n",
      "Processing: 17, interval 111\n",
      "Saved\n",
      "Processing: 17, interval 112\n",
      "Saved\n",
      "Processing: 17, interval 113\n",
      "Saved\n",
      "Processing: 17, interval 114\n",
      "Saved\n",
      "Processing: 17, interval 115\n",
      "Saved\n",
      "Processing: 17, interval 116\n",
      "Saved\n",
      "Processing: 17, interval 117\n",
      "Saved\n",
      "Processing: 17, interval 118\n",
      "Saved\n",
      "Processing: 17, interval 119\n",
      "Saved\n",
      "Processing: 17, interval 120\n",
      "Saved\n",
      "Processing: 17, interval 121\n",
      "Saved\n",
      "Processing: 17, interval 122\n",
      "Saved\n",
      "Processing: 17, interval 123\n",
      "Saved\n",
      "Processing: 17, interval 124\n",
      "Saved\n",
      "Processing: 17, interval 125\n",
      "Saved\n",
      "Processing: 17, interval 126\n",
      "Saved\n",
      "Processing: 17, interval 127\n",
      "Saved\n",
      "Processing: 17, interval 128\n",
      "Saved\n",
      "Processing: 17, interval 129\n",
      "Saved\n",
      "Processing: 17, interval 130\n",
      "Saved\n",
      "Processing: 17, interval 131\n",
      "Saved\n",
      "Processing: 17, interval 132\n",
      "Saved\n",
      "Processing: 17, interval 133\n",
      "Saved\n",
      "Processing: 17, interval 134\n",
      "Saved\n",
      "Processing: 17, interval 135\n",
      "Saved\n",
      "Processing: 17, interval 136\n",
      "Saved\n",
      "Processing: 17, interval 137\n",
      "Saved\n",
      "Processing: 17, interval 138\n",
      "Saved\n",
      "Processing: 17, interval 139\n",
      "Saved\n",
      "Processing: 17, interval 140\n",
      "Saved\n",
      "Processing: 17, interval 141\n",
      "Saved\n",
      "Processing: 17, interval 142\n",
      "Saved\n",
      "Processing: 17, interval 143\n",
      "Saved\n",
      "Processing: 17, interval 144\n",
      "Saved\n",
      "Processing: 17, interval 145\n",
      "Saved\n",
      "Processing: 17, interval 146\n",
      " at 17, interval 146. Retrying...\n",
      "Processing: 17, interval 146\n",
      "Saved\n",
      "Processing: 17, interval 147\n",
      "Saved\n",
      "Processing: 17, interval 148\n",
      "Saved\n",
      "Processing: 17, interval 149\n",
      "Saved\n",
      "Processing: 17, interval 150\n",
      "Saved\n",
      "Processing: 17, interval 151\n",
      "Saved\n",
      "Processing: 17, interval 152\n",
      "Saved\n",
      "Processing: 17, interval 153\n",
      "Saved\n",
      "Processing: 17, interval 154\n",
      "Saved\n",
      "Processing: 17, interval 155\n",
      "Saved\n",
      "Processing: 17, interval 156\n",
      "Saved\n",
      "Processing: 17, interval 157\n",
      "Saved\n",
      "Processing: 17, interval 158\n",
      "Saved\n",
      "Processing: 17, interval 159\n",
      "Saved\n",
      "Processing: 17, interval 160\n",
      "Saved\n",
      "Processing: 17, interval 161\n",
      "Saved\n",
      "Processing: 17, interval 162\n",
      "Saved\n",
      "Processing: 17, interval 163\n",
      "Saved\n",
      "Processing: 17, interval 164\n",
      "Saved\n",
      "Processing: 17, interval 165\n",
      "Saved\n",
      "Processing: 17, interval 166\n",
      "Saved\n",
      "Processing: 17, interval 167\n",
      "Saved\n",
      "Processing: 17, interval 168\n",
      "Saved\n",
      "Processing: 17, interval 169\n",
      "Saved\n",
      "Processing: 17, interval 170\n",
      "Saved\n",
      "Processing: 17, interval 171\n",
      " at 17, interval 171. Retrying...\n",
      "Processing: 17, interval 171\n",
      "Saved\n",
      "Processing: 17, interval 172\n",
      " at 17, interval 172. Retrying...\n",
      "Processing: 17, interval 172\n",
      "Saved\n",
      "Processing: 17, interval 173\n",
      " at 17, interval 173. Retrying...\n",
      "Processing: 17, interval 173\n",
      " at 17, interval 173. Retrying...\n",
      "Processing: 17, interval 173\n",
      "Saved\n",
      "Processing: 17, interval 174\n",
      "Saved\n",
      "Processing: 17, interval 175\n",
      "Saved\n",
      "Processing: 17, interval 176\n",
      "Saved\n",
      "Processing: 17, interval 177\n",
      "Saved\n",
      "Processing: 17, interval 178\n",
      "Saved\n",
      "Processing: 17, interval 179\n",
      "Saved\n",
      "Processing: 17, interval 180\n",
      "Saved\n",
      "Processing: 17, interval 181\n",
      "Saved\n",
      "Processing: 17, interval 182\n",
      "Saved\n",
      "Processing: 17, interval 183\n",
      "Saved\n",
      "Processing: 17, interval 184\n",
      "Saved\n",
      "Processing: 17, interval 185\n",
      "Saved\n",
      "Processing: 17, interval 186\n",
      "Saved\n",
      "Processing: 17, interval 187\n",
      " at 17, interval 187. Retrying...\n",
      "Processing: 17, interval 187\n",
      "Saved\n",
      "Processing: 17, interval 188\n",
      "Saved\n",
      "Processing: 17, interval 189\n",
      "Saved\n",
      "Processing: 17, interval 190\n",
      "Saved\n",
      "Processing: 17, interval 191\n",
      "Saved\n",
      "Processing: 17, interval 192\n",
      "Saved\n",
      "Processing: 17, interval 193\n",
      "Saved\n",
      "Processing: 17, interval 194\n",
      "Saved\n",
      "Processing: 17, interval 195\n",
      "Saved\n",
      "Processing: 17, interval 196\n",
      "Saved\n",
      "Processing: 17, interval 197\n",
      "Saved\n",
      "Processing: 17, interval 198\n",
      "Saved\n",
      "Processing: 17, interval 199\n",
      "Saved\n",
      "Processing: 17, interval 200\n",
      "Saved\n",
      "Processing: 17, interval 201\n",
      "Saved\n",
      "Processing: 17, interval 202\n",
      "Saved\n",
      "Processing: 17, interval 203\n",
      "Saved\n",
      "Processing: 17, interval 204\n",
      "Saved\n",
      "Processing: 17, interval 205\n",
      " at 17, interval 205. Retrying...\n",
      "Processing: 17, interval 205\n",
      "Saved\n",
      "Processing: 17, interval 206\n",
      "Saved\n",
      "Processing: 17, interval 207\n",
      "Saved\n",
      "Processing: 17, interval 208\n",
      "Saved\n",
      "Processing: 17, interval 209\n",
      "Saved\n",
      "Processing: 17, interval 210\n",
      "Saved\n",
      "Processing: 17, interval 211\n",
      "Saved\n",
      "Processing: 17, interval 212\n",
      "Saved\n",
      "Processing: 17, interval 213\n",
      "Saved\n",
      "Processing: 17, interval 214\n",
      "Saved\n",
      "Processing: 17, interval 215\n",
      " at 17, interval 215. Retrying...\n",
      "Processing: 17, interval 215\n",
      "Saved\n",
      "Processing: 17, interval 216\n",
      "Saved\n",
      "Processing: 17, interval 217\n",
      "Saved\n",
      "Processing: 17, interval 218\n",
      "Saved\n",
      "Processing: 17, interval 219\n",
      "Saved\n",
      "Processing: 17, interval 220\n",
      "Saved\n",
      "Processing: 17, interval 221\n",
      "Saved\n",
      "Processing: 17, interval 222\n",
      "Saved\n",
      "Processing: 17, interval 223\n",
      "Saved\n",
      "Processing: 17, interval 224\n",
      "Saved\n",
      "Processing: 17, interval 225\n",
      "Saved\n",
      "Processing: 17, interval 226\n",
      "Saved\n",
      "Processing: 17, interval 227\n",
      "Saved\n",
      "Processing: 17, interval 228\n",
      "Saved\n",
      "Processing: 17, interval 229\n",
      "Saved\n",
      "Processing: 17, interval 230\n",
      "Saved\n",
      "Processing: 17, interval 231\n",
      "Saved\n",
      "Processing: 17, interval 232\n",
      "Saved\n",
      "Processing: 17, interval 233\n",
      "Saved\n",
      "Processing: 17, interval 234\n",
      "Saved\n",
      "Processing: 17, interval 235\n",
      "Saved\n",
      "Processing: 17, interval 236\n",
      "Saved\n",
      "Processing: 17, interval 237\n",
      "Saved\n",
      "Processing: 17, interval 238\n",
      "Saved\n",
      "Processing: 17, interval 239\n",
      "Saved\n",
      "Processing: 17, interval 240\n",
      "Saved\n",
      "Processing: 17, interval 241\n",
      " at 17, interval 241. Retrying...\n",
      "Processing: 17, interval 241\n",
      "Saved\n",
      "Processing: 17, interval 242\n",
      "Saved\n",
      "Processing: 17, interval 243\n",
      "Saved\n",
      "Processing: 17, interval 244\n",
      "Saved\n",
      "Processing: 17, interval 245\n",
      "Saved\n",
      "Processing: 17, interval 246\n",
      "Saved\n",
      "Processing: 17, interval 247\n",
      " at 17, interval 247. Retrying...\n",
      "Processing: 17, interval 247\n",
      "Saved\n",
      "Processing: 17, interval 248\n",
      "Saved\n",
      "Processing: 17, interval 249\n",
      "Saved\n",
      "Processing: 17, interval 250\n",
      "Saved\n",
      "Processing: 17, interval 251\n",
      "Saved\n",
      "Processing: 17, interval 252\n",
      "Saved\n",
      "Processing: 17, interval 253\n",
      " at 17, interval 253. Retrying...\n",
      "Processing: 17, interval 253\n",
      "Saved\n",
      "Processing: 17, interval 254\n",
      "Saved\n",
      "Processing: 17, interval 255\n",
      "Saved\n",
      "Processing: 17, interval 256\n",
      "Saved\n",
      "Processing: 17, interval 257\n",
      "Saved\n",
      "Processing: 17, interval 258\n",
      "Saved\n",
      "Processing: 17, interval 259\n",
      "Saved\n",
      "Processing: 17, interval 260\n",
      "Saved\n",
      "Processing: 17, interval 261\n",
      "Saved\n",
      "Processing: 17, interval 262\n",
      " at 17, interval 262. Retrying...\n",
      "Processing: 17, interval 262\n",
      "Saved\n",
      "Processing: 17, interval 263\n",
      " at 17, interval 263. Retrying...\n",
      "Processing: 17, interval 263\n",
      "Saved\n",
      "Processing: 17, interval 264\n",
      "Saved\n",
      "Processing: 17, interval 265\n",
      " at 17, interval 265. Retrying...\n",
      "Processing: 17, interval 265\n",
      "Saved\n",
      "Processing: 17, interval 266\n",
      "Saved\n",
      "Processing: 17, interval 267\n",
      "Saved\n",
      "Processing: 17, interval 268\n",
      "Saved\n",
      "Processing: 17, interval 269\n",
      "Saved\n",
      "Processing: 17, interval 270\n",
      "Saved\n",
      "Processing: 17, interval 271\n",
      " at 17, interval 271. Retrying...\n",
      "Processing: 17, interval 271\n",
      "Saved\n",
      "Processing: 17, interval 272\n",
      "Saved\n",
      "Processing: 17, interval 273\n",
      " at 17, interval 273. Retrying...\n",
      "Processing: 17, interval 273\n",
      "Saved\n",
      "Processing: 17, interval 274\n",
      "Saved\n",
      "Processing: 17, interval 275\n",
      "Saved\n",
      "Processing: 17, interval 276\n",
      "Saved\n",
      "Processing: 17, interval 277\n",
      "Saved\n",
      "Processing: 17, interval 278\n",
      "Saved\n",
      "Processing: 17, interval 279\n",
      "Saved\n",
      "Processing: 17, interval 280\n",
      "Saved\n",
      "Processing: 17, interval 281\n",
      "Saved\n",
      "Processing: 17, interval 282\n",
      " at 17, interval 282. Retrying...\n",
      "Processing: 17, interval 282\n",
      "Saved\n",
      "Processing: 17, interval 283\n",
      " at 17, interval 283. Retrying...\n",
      "Processing: 17, interval 283\n",
      "Saved\n",
      "Processing: 17, interval 284\n",
      "Saved\n",
      "Processing: 17, interval 285\n",
      "Saved\n",
      "Processing: 17, interval 286\n",
      "Saved\n",
      "Processing: 17, interval 287\n",
      " at 17, interval 287. Retrying...\n",
      "Processing: 17, interval 287\n",
      "Saved\n",
      "Processing: 17, interval 288\n",
      " at 17, interval 288. Retrying...\n",
      "Processing: 17, interval 288\n",
      "Saved\n",
      "Processing: 17, interval 289\n",
      "Saved\n",
      "Processing: 17, interval 290\n",
      "Saved\n",
      "Processing: 17, interval 291\n",
      " at 17, interval 291. Retrying...\n",
      "Processing: 17, interval 291\n",
      "Saved\n",
      "Processing: 17, interval 292\n",
      "Saved\n",
      "Processing: 17, interval 293\n",
      "Saved\n",
      "Processing: 18, interval 0\n",
      "Saved\n",
      "Processing: 18, interval 1\n",
      "Saved\n",
      "Processing: 18, interval 2\n",
      " at 18, interval 2. Retrying...\n",
      "Processing: 18, interval 2\n",
      "Saved\n",
      "Processing: 18, interval 3\n",
      " at 18, interval 3. Retrying...\n",
      "Processing: 18, interval 3\n",
      "Saved\n",
      "Processing: 18, interval 4\n",
      " at 18, interval 4. Retrying...\n",
      "Processing: 18, interval 4\n",
      "Saved\n",
      "Processing: 18, interval 5\n",
      "Saved\n",
      "Processing: 18, interval 6\n",
      " at 18, interval 6. Retrying...\n",
      "Processing: 18, interval 6\n",
      "Saved\n",
      "Processing: 18, interval 7\n",
      " at 18, interval 7. Retrying...\n",
      "Processing: 18, interval 7\n",
      " at 18, interval 7. Retrying...\n",
      "Processing: 18, interval 7\n",
      "Saved\n",
      "Processing: 18, interval 8\n",
      " at 18, interval 8. Retrying...\n",
      "Processing: 18, interval 8\n",
      "Saved\n",
      "Processing: 18, interval 9\n",
      "Saved\n",
      "Processing: 18, interval 10\n",
      " at 18, interval 10. Retrying...\n",
      "Processing: 18, interval 10\n",
      " at 18, interval 10. Retrying...\n",
      "Processing: 18, interval 10\n",
      "Saved\n",
      "Processing: 18, interval 11\n",
      "Saved\n",
      "Processing: 18, interval 12\n",
      " at 18, interval 12. Retrying...\n",
      "Processing: 18, interval 12\n",
      "Saved\n",
      "Processing: 18, interval 13\n",
      "Saved\n",
      "Processing: 18, interval 14\n",
      "Saved\n",
      "Processing: 18, interval 15\n",
      "Saved\n",
      "Processing: 18, interval 16\n",
      "Saved\n",
      "Processing: 18, interval 17\n",
      "Saved\n",
      "Processing: 18, interval 18\n",
      "Saved\n",
      "Processing: 18, interval 19\n",
      "Saved\n",
      "Processing: 18, interval 20\n",
      "Saved\n",
      "Processing: 18, interval 21\n",
      "Saved\n",
      "Processing: 18, interval 22\n",
      "Saved\n",
      "Processing: 18, interval 23\n",
      "Saved\n",
      "Processing: 18, interval 24\n",
      "Saved\n",
      "Processing: 18, interval 25\n",
      "Saved\n",
      "Processing: 18, interval 26\n",
      "Saved\n",
      "Processing: 18, interval 27\n",
      "Saved\n",
      "Processing: 18, interval 28\n",
      "Saved\n",
      "Processing: 18, interval 29\n",
      "Saved\n",
      "Processing: 18, interval 30\n",
      "Saved\n",
      "Processing: 18, interval 31\n",
      "Saved\n",
      "Processing: 18, interval 32\n",
      "Saved\n",
      "Processing: 18, interval 33\n",
      "Saved\n",
      "Processing: 18, interval 34\n",
      "Saved\n",
      "Processing: 18, interval 35\n",
      "Saved\n",
      "Processing: 18, interval 36\n",
      "Saved\n",
      "Processing: 18, interval 37\n",
      "Saved\n",
      "Processing: 18, interval 38\n",
      "Saved\n",
      "Processing: 18, interval 39\n",
      "Saved\n",
      "Processing: 18, interval 40\n",
      "Saved\n",
      "Processing: 18, interval 41\n",
      "Saved\n",
      "Processing: 18, interval 42\n",
      "Saved\n",
      "Processing: 18, interval 43\n",
      "Saved\n",
      "Processing: 18, interval 44\n",
      "Saved\n",
      "Processing: 18, interval 45\n",
      "Saved\n",
      "Processing: 18, interval 46\n",
      "Saved\n",
      "Processing: 18, interval 47\n",
      "Saved\n",
      "Processing: 18, interval 48\n",
      "Saved\n",
      "Processing: 18, interval 49\n",
      "Saved\n",
      "Processing: 18, interval 50\n",
      "Saved\n",
      "Processing: 18, interval 51\n",
      "Saved\n",
      "Processing: 18, interval 52\n",
      "Saved\n",
      "Processing: 18, interval 53\n",
      "Saved\n",
      "Processing: 18, interval 54\n",
      "Saved\n",
      "Processing: 18, interval 55\n",
      "Saved\n",
      "Processing: 18, interval 56\n",
      "Saved\n",
      "Processing: 18, interval 57\n",
      "Saved\n",
      "Processing: 18, interval 58\n",
      "Saved\n",
      "Processing: 18, interval 59\n",
      "Saved\n",
      "Processing: 18, interval 60\n",
      "Saved\n",
      "Processing: 18, interval 61\n",
      "Saved\n",
      "Processing: 18, interval 62\n",
      "Saved\n",
      "Processing: 18, interval 63\n",
      "Saved\n",
      "Processing: 18, interval 64\n",
      "Saved\n",
      "Processing: 18, interval 65\n",
      "Saved\n",
      "Processing: 18, interval 66\n",
      "Saved\n",
      "Processing: 18, interval 67\n",
      "Saved\n",
      "Processing: 18, interval 68\n",
      "Saved\n",
      "Processing: 18, interval 69\n",
      "Saved\n",
      "Processing: 18, interval 70\n",
      "Saved\n",
      "Processing: 18, interval 71\n",
      "Saved\n",
      "Processing: 18, interval 72\n",
      "Saved\n",
      "Processing: 18, interval 73\n",
      "Saved\n",
      "Processing: 18, interval 74\n",
      "Saved\n",
      "Processing: 18, interval 75\n",
      "Saved\n",
      "Processing: 18, interval 76\n",
      "Saved\n",
      "Processing: 18, interval 77\n",
      "Saved\n",
      "Processing: 18, interval 78\n",
      "Saved\n",
      "Processing: 18, interval 79\n",
      "Saved\n",
      "Processing: 18, interval 80\n",
      "Saved\n",
      "Processing: 18, interval 81\n",
      "Saved\n",
      "Processing: 18, interval 82\n",
      "Saved\n",
      "Processing: 18, interval 83\n",
      "Saved\n",
      "Processing: 18, interval 84\n",
      "Saved\n",
      "Processing: 18, interval 85\n",
      " at 18, interval 85. Retrying...\n",
      "Processing: 18, interval 85\n",
      "Saved\n",
      "Processing: 19, interval 0\n",
      "Saved\n",
      "Processing: 19, interval 1\n",
      "Saved\n",
      "Processing: 19, interval 2\n",
      "Saved\n",
      "Processing: 19, interval 3\n",
      "Saved\n",
      "Processing: 19, interval 4\n",
      "Saved\n",
      "Processing: 19, interval 5\n",
      "Saved\n",
      "Processing: 19, interval 6\n",
      "Saved\n",
      "Processing: 19, interval 7\n",
      " at 19, interval 7. Retrying...\n",
      "Processing: 19, interval 7\n",
      "Saved\n",
      "Processing: 19, interval 8\n",
      "Saved\n",
      "Processing: 19, interval 9\n",
      "Saved\n",
      "Processing: 19, interval 10\n",
      "Saved\n",
      "Processing: 19, interval 11\n",
      "Saved\n",
      "Processing: 19, interval 12\n",
      " at 19, interval 12. Retrying...\n",
      "Processing: 19, interval 12\n",
      "Saved\n",
      "Processing: 19, interval 13\n",
      "Saved\n",
      "Processing: 19, interval 14\n",
      " at 19, interval 14. Retrying...\n",
      "Processing: 19, interval 14\n",
      "Saved\n",
      "Processing: 19, interval 15\n",
      "Saved\n",
      "Processing: 19, interval 16\n",
      "Saved\n",
      "Processing: 19, interval 17\n",
      "Saved\n",
      "Processing: 19, interval 18\n",
      "Saved\n",
      "Processing: 19, interval 19\n",
      "Saved\n",
      "Processing: 19, interval 20\n",
      "Saved\n",
      "Processing: 19, interval 21\n",
      "Saved\n",
      "Processing: 19, interval 22\n",
      "Saved\n",
      "Processing: 19, interval 23\n",
      "Saved\n",
      "Processing: 19, interval 24\n",
      "Saved\n",
      "Processing: 19, interval 25\n",
      "Saved\n",
      "Processing: 19, interval 26\n",
      "Saved\n",
      "Processing: 19, interval 27\n",
      "Saved\n",
      "Processing: 19, interval 28\n",
      "Saved\n",
      "Processing: 19, interval 29\n",
      "Saved\n",
      "Processing: 19, interval 30\n",
      "Saved\n",
      "Processing: 19, interval 31\n",
      "Saved\n",
      "Processing: 19, interval 32\n",
      " at 19, interval 32. Retrying...\n",
      "Processing: 19, interval 32\n",
      "Saved\n",
      "Processing: 19, interval 33\n",
      "Saved\n",
      "Processing: 19, interval 34\n",
      "Saved\n",
      "Processing: 19, interval 35\n",
      "Saved\n",
      "Processing: 19, interval 36\n",
      "Saved\n",
      "Processing: 19, interval 37\n",
      "Saved\n",
      "Processing: 19, interval 38\n",
      " at 19, interval 38. Retrying...\n",
      "Processing: 19, interval 38\n",
      "Saved\n",
      "Processing: 19, interval 39\n",
      "Saved\n",
      "Processing: 19, interval 40\n",
      "Saved\n",
      "Processing: 19, interval 41\n",
      " at 19, interval 41. Retrying...\n",
      "Processing: 19, interval 41\n",
      "Saved\n",
      "Processing: 19, interval 42\n",
      " at 19, interval 42. Retrying...\n",
      "Processing: 19, interval 42\n",
      "Saved\n",
      "Processing: 19, interval 43\n",
      "Saved\n",
      "Processing: 19, interval 44\n",
      "Saved\n",
      "Processing: 19, interval 45\n",
      "Saved\n",
      "Processing: 19, interval 46\n",
      "Saved\n",
      "Processing: 19, interval 47\n",
      " at 19, interval 47. Retrying...\n",
      "Processing: 19, interval 47\n",
      "Saved\n",
      "Processing: 19, interval 48\n",
      "Saved\n",
      "Processing: 19, interval 49\n",
      "Saved\n",
      "Processing: 19, interval 50\n",
      "Saved\n",
      "Processing: 19, interval 51\n",
      "Saved\n",
      "Processing: 19, interval 52\n",
      "Saved\n",
      "Processing: 19, interval 53\n",
      "Saved\n",
      "Processing: 19, interval 54\n",
      "Saved\n",
      "Processing: 19, interval 55\n",
      "Saved\n",
      "Processing: 19, interval 56\n",
      "Saved\n",
      "Processing: 19, interval 57\n",
      "Saved\n",
      "Processing: 19, interval 58\n",
      "Saved\n",
      "Processing: 19, interval 59\n",
      "Saved\n",
      "Processing: 19, interval 60\n",
      "Saved\n",
      "Processing: 19, interval 61\n",
      "Saved\n",
      "Processing: 19, interval 62\n",
      "Saved\n",
      "Processing: 19, interval 63\n",
      "Saved\n",
      "Processing: 19, interval 64\n",
      "Saved\n",
      "Processing: 19, interval 65\n",
      "Saved\n",
      "Processing: 19, interval 66\n",
      "Saved\n",
      "Processing: 19, interval 67\n",
      " at 19, interval 67. Retrying...\n",
      "Processing: 19, interval 67\n",
      "Saved\n",
      "Processing: 19, interval 68\n",
      " at 19, interval 68. Retrying...\n",
      "Processing: 19, interval 68\n",
      "Saved\n",
      "Processing: 19, interval 69\n",
      "Saved\n",
      "Processing: 19, interval 70\n",
      "Saved\n",
      "Processing: 19, interval 71\n",
      "Saved\n",
      "Processing: 19, interval 72\n",
      " at 19, interval 72. Retrying...\n",
      "Processing: 19, interval 72\n",
      "Saved\n",
      "Processing: 19, interval 73\n",
      "Saved\n",
      "Processing: 19, interval 74\n",
      "Saved\n",
      "Processing: 19, interval 75\n",
      "Saved\n",
      "Processing: 19, interval 76\n",
      "Saved\n",
      "Processing: 19, interval 77\n",
      "Saved\n",
      "Processing: 19, interval 78\n",
      "Saved\n",
      "Processing: 19, interval 79\n",
      "Saved\n",
      "Processing: 19, interval 80\n",
      "Saved\n",
      "Processing: 19, interval 81\n",
      " at 19, interval 81. Retrying...\n",
      "Processing: 19, interval 81\n",
      "Saved\n",
      "Processing: 19, interval 82\n",
      "Saved\n",
      "Processing: 19, interval 83\n",
      "Saved\n",
      "Processing: 19, interval 84\n",
      "Saved\n",
      "Processing: 19, interval 85\n",
      " at 19, interval 85. Retrying...\n",
      "Processing: 19, interval 85\n",
      "Saved\n",
      "Processing: 19, interval 86\n",
      " at 19, interval 86. Retrying...\n",
      "Processing: 19, interval 86\n",
      "Saved\n",
      "Processing: 19, interval 87\n",
      "Saved\n",
      "Processing: 19, interval 88\n",
      "Saved\n",
      "Processing: 19, interval 89\n",
      "Saved\n",
      "Processing: 19, interval 90\n",
      "Saved\n",
      "Processing: 19, interval 91\n",
      "Saved\n",
      "Processing: 19, interval 92\n",
      "Saved\n",
      "Processing: 19, interval 93\n",
      "Saved\n",
      "Processing: 19, interval 94\n",
      " at 19, interval 94. Retrying...\n",
      "Processing: 19, interval 94\n",
      "Saved\n",
      "Processing: 19, interval 95\n",
      "Saved\n",
      "Processing: 19, interval 96\n",
      "Saved\n",
      "Processing: 19, interval 97\n",
      "Saved\n",
      "Processing: 19, interval 98\n",
      "Saved\n",
      "Processing: 19, interval 99\n",
      "Saved\n",
      "Processing: 19, interval 100\n",
      "Saved\n",
      "Processing: 19, interval 101\n",
      "Saved\n",
      "Processing: 19, interval 102\n",
      "Saved\n",
      "Processing: 19, interval 103\n",
      "Saved\n",
      "Processing: 19, interval 104\n",
      "Saved\n",
      "Processing: 19, interval 105\n",
      "Saved\n",
      "Processing: 19, interval 106\n",
      "Saved\n",
      "Processing: 19, interval 107\n",
      "Saved\n",
      "Processing: 19, interval 108\n",
      "Saved\n",
      "Processing: 19, interval 109\n",
      " at 19, interval 109. Retrying...\n",
      "Processing: 19, interval 109\n",
      "Saved\n",
      "Processing: 19, interval 110\n",
      "Saved\n",
      "Processing: 19, interval 111\n",
      "Saved\n",
      "Processing: 19, interval 112\n",
      " at 19, interval 112. Retrying...\n",
      "Processing: 19, interval 112\n",
      "Saved\n",
      "Processing: 19, interval 113\n",
      "Saved\n",
      "Processing: 19, interval 114\n",
      "Saved\n",
      "Processing: 19, interval 115\n",
      "Saved\n",
      "Processing: 19, interval 116\n",
      " at 19, interval 116. Retrying...\n",
      "Processing: 19, interval 116\n",
      "Saved\n",
      "Processing: 19, interval 117\n",
      " at 19, interval 117. Retrying...\n",
      "Processing: 19, interval 117\n",
      "Saved\n",
      "Processing: 19, interval 118\n",
      " at 19, interval 118. Retrying...\n",
      "Processing: 19, interval 118\n",
      "Saved\n",
      "Processing: 19, interval 119\n",
      "Saved\n",
      "Processing: 19, interval 120\n",
      "Saved\n",
      "Processing: 19, interval 121\n",
      "Saved\n",
      "Processing: 19, interval 122\n",
      "Saved\n",
      "Processing: 19, interval 123\n",
      "Saved\n",
      "Processing: 19, interval 124\n",
      "Saved\n",
      "Processing: 19, interval 125\n",
      "Saved\n",
      "Processing: 19, interval 126\n",
      "Saved\n",
      "Processing: 19, interval 127\n",
      "Saved\n",
      "Processing: 19, interval 128\n",
      "Saved\n",
      "Processing: 19, interval 129\n",
      "Saved\n",
      "Processing: 19, interval 130\n",
      "Saved\n",
      "Processing: 19, interval 131\n",
      "Saved\n",
      "Processing: 19, interval 132\n",
      "Saved\n",
      "Processing: 19, interval 133\n",
      "Saved\n",
      "Processing: 19, interval 134\n",
      "Saved\n",
      "Processing: 19, interval 135\n",
      "Saved\n",
      "Processing: 19, interval 136\n",
      "Saved\n",
      "Processing: 19, interval 137\n",
      "Saved\n",
      "Processing: 19, interval 138\n",
      "Saved\n",
      "Processing: 19, interval 139\n",
      "Saved\n",
      "Processing: 19, interval 140\n",
      "Saved\n",
      "Processing: 19, interval 141\n",
      "Saved\n",
      "Processing: 19, interval 142\n",
      "Saved\n",
      "Processing: 19, interval 143\n",
      " at 19, interval 143. Retrying...\n",
      "Processing: 19, interval 143\n",
      "Saved\n",
      "Processing: 19, interval 144\n",
      "Saved\n",
      "Processing: 19, interval 145\n",
      "Saved\n",
      "Processing: 19, interval 146\n",
      "Saved\n",
      "Processing: 19, interval 147\n",
      "Saved\n",
      "Processing: 19, interval 148\n",
      "Saved\n",
      "Processing: 19, interval 149\n",
      "Saved\n",
      "Processing: 19, interval 150\n",
      "Saved\n",
      "Processing: 19, interval 151\n",
      "Saved\n",
      "Processing: 19, interval 152\n",
      "Saved\n",
      "Processing: 19, interval 153\n",
      "Saved\n",
      "Processing: 19, interval 154\n",
      "Saved\n",
      "Processing: 19, interval 155\n",
      "Saved\n",
      "Processing: 19, interval 156\n",
      "Saved\n",
      "Processing: 19, interval 157\n",
      "Saved\n",
      "Processing: 19, interval 158\n",
      "Saved\n",
      "Processing: 19, interval 159\n",
      "Saved\n",
      "Processing: 19, interval 160\n",
      "Saved\n",
      "Processing: 19, interval 161\n",
      "Saved\n",
      "Processing: 19, interval 162\n",
      "Saved\n",
      "Processing: 19, interval 163\n",
      "Saved\n",
      "Processing: 19, interval 164\n",
      "Saved\n",
      "Processing: 19, interval 165\n",
      "Saved\n",
      "Processing: 19, interval 166\n",
      "Saved\n",
      "Processing: 19, interval 167\n",
      "Saved\n",
      "Processing: 19, interval 168\n",
      "Saved\n",
      "Processing: 19, interval 169\n",
      "Saved\n",
      "Processing: 19, interval 170\n",
      "Saved\n",
      "Processing: 19, interval 171\n",
      "Saved\n",
      "Processing: 19, interval 172\n",
      "Saved\n",
      "Processing: 19, interval 173\n",
      "Saved\n",
      "Processing: 19, interval 174\n",
      " at 19, interval 174. Retrying...\n",
      "Processing: 19, interval 174\n",
      "Saved\n",
      "Processing: 19, interval 175\n",
      "Saved\n",
      "Processing: 19, interval 176\n",
      "Saved\n",
      "Processing: 19, interval 177\n",
      "Saved\n",
      "Processing: 19, interval 178\n",
      "Saved\n",
      "Processing: 19, interval 179\n",
      "Saved\n",
      "Processing: 19, interval 180\n",
      "Saved\n",
      "Processing: 19, interval 181\n",
      "Saved\n",
      "Processing: 19, interval 182\n",
      "Saved\n",
      "Processing: 19, interval 183\n",
      "Saved\n",
      "Processing: 19, interval 184\n",
      "Saved\n",
      "Processing: 19, interval 185\n",
      "Saved\n",
      "Processing: 19, interval 186\n",
      "Saved\n",
      "Processing: 19, interval 187\n",
      "Saved\n",
      "Processing: 19, interval 188\n",
      " at 19, interval 188. Retrying...\n",
      "Processing: 19, interval 188\n",
      "Saved\n",
      "Processing: 19, interval 189\n",
      "Saved\n",
      "Processing: 19, interval 190\n",
      "Saved\n",
      "Processing: 19, interval 191\n",
      "Saved\n",
      "Processing: 19, interval 192\n",
      "Saved\n",
      "Processing: 19, interval 193\n",
      "Saved\n",
      "Processing: 19, interval 194\n",
      " at 19, interval 194. Retrying...\n",
      "Processing: 19, interval 194\n",
      "Saved\n",
      "Processing: 19, interval 195\n",
      " at 19, interval 195. Retrying...\n",
      "Processing: 19, interval 195\n",
      " at 19, interval 195. Retrying...\n",
      "Processing: 19, interval 195\n",
      "Saved\n",
      "Processing: 19, interval 196\n",
      " at 19, interval 196. Retrying...\n",
      "Processing: 19, interval 196\n",
      "Saved\n",
      "Processing: 19, interval 197\n",
      "Saved\n",
      "Processing: 19, interval 198\n",
      " at 19, interval 198. Retrying...\n",
      "Processing: 19, interval 198\n",
      "Saved\n",
      "Processing: 19, interval 199\n",
      "Saved\n",
      "Processing: 19, interval 200\n",
      " at 19, interval 200. Retrying...\n",
      "Processing: 19, interval 200\n",
      "Saved\n",
      "Processing: 19, interval 201\n",
      "Saved\n",
      "Processing: 19, interval 202\n",
      "Saved\n",
      "Processing: 19, interval 203\n",
      "Saved\n",
      "Processing: 19, interval 204\n",
      "Saved\n",
      "Processing: 19, interval 205\n",
      "Saved\n",
      "Processing: 19, interval 206\n",
      "Saved\n",
      "Processing: 19, interval 207\n",
      "Saved\n",
      "Processing: 19, interval 208\n",
      "Saved\n",
      "Processing: 19, interval 209\n",
      "Saved\n",
      "Processing: 19, interval 210\n",
      " at 19, interval 210. Retrying...\n",
      "Processing: 19, interval 210\n",
      "Saved\n",
      "Processing: 19, interval 211\n",
      "Saved\n",
      "Processing: 19, interval 212\n",
      " at 19, interval 212. Retrying...\n",
      "Processing: 19, interval 212\n",
      "Saved\n",
      "Processing: 19, interval 213\n",
      " at 19, interval 213. Retrying...\n",
      "Processing: 19, interval 213\n",
      "Saved\n",
      "Processing: 19, interval 214\n",
      " at 19, interval 214. Retrying...\n",
      "Processing: 19, interval 214\n",
      "Saved\n",
      "Processing: 19, interval 215\n",
      "Saved\n",
      "Processing: 19, interval 216\n",
      "Saved\n",
      "Processing: 19, interval 217\n",
      "Saved\n",
      "Processing: 19, interval 218\n",
      "Saved\n",
      "Processing: 19, interval 219\n",
      "Saved\n",
      "Processing: 19, interval 220\n",
      "Saved\n",
      "Processing: 19, interval 221\n",
      "Saved\n",
      "Processing: 19, interval 222\n",
      "Saved\n",
      "Processing: 19, interval 223\n",
      "Saved\n",
      "Processing: 19, interval 224\n",
      "Saved\n",
      "Processing: 19, interval 225\n",
      "Saved\n",
      "Processing: 19, interval 226\n",
      "Saved\n",
      "Processing: 19, interval 227\n",
      "Saved\n",
      "Processing: 19, interval 228\n",
      "Saved\n",
      "Processing: 19, interval 229\n",
      " at 19, interval 229. Retrying...\n",
      "Processing: 19, interval 229\n",
      "Saved\n",
      "Processing: 19, interval 230\n",
      "Saved\n",
      "Processing: 19, interval 231\n",
      "Saved\n",
      "Processing: 19, interval 232\n",
      "Saved\n",
      "Processing: 19, interval 233\n",
      "Saved\n",
      "Processing: 19, interval 234\n",
      "Saved\n",
      "Processing: 19, interval 235\n",
      "Saved\n",
      "Processing: 19, interval 236\n",
      "Saved\n",
      "Processing: 19, interval 237\n",
      "Saved\n",
      "Processing: 19, interval 238\n",
      "Saved\n",
      "Processing: 19, interval 239\n",
      "Saved\n",
      "Processing: 19, interval 240\n",
      "Saved\n",
      "Processing: 19, interval 241\n",
      "Saved\n",
      "Processing: 19, interval 242\n",
      "Saved\n",
      "Processing: 19, interval 243\n",
      "Saved\n",
      "Processing: 19, interval 244\n",
      "Saved\n",
      "Processing: 19, interval 245\n",
      "Saved\n",
      "Processing: 2, interval 0\n",
      "Saved\n",
      "Processing: 2, interval 1\n",
      "Saved\n",
      "Processing: 2, interval 2\n",
      "Saved\n",
      "Processing: 2, interval 3\n",
      "Saved\n",
      "Processing: 2, interval 4\n",
      "Saved\n",
      "Processing: 2, interval 5\n",
      "Saved\n",
      "Processing: 2, interval 6\n",
      "Saved\n",
      "Processing: 2, interval 7\n",
      "Saved\n",
      "Processing: 2, interval 8\n",
      "Saved\n",
      "Processing: 2, interval 9\n",
      "Saved\n",
      "Processing: 2, interval 10\n",
      "Saved\n",
      "Processing: 2, interval 11\n",
      "Saved\n",
      "Processing: 2, interval 12\n",
      "Saved\n",
      "Processing: 2, interval 13\n",
      "Saved\n",
      "Processing: 2, interval 14\n",
      "Saved\n",
      "Processing: 2, interval 15\n",
      "Saved\n",
      "Processing: 2, interval 16\n",
      "Saved\n",
      "Processing: 2, interval 17\n",
      "Saved\n",
      "Processing: 2, interval 18\n",
      "Saved\n",
      "Processing: 2, interval 19\n",
      "Saved\n",
      "Processing: 2, interval 20\n",
      "Saved\n",
      "Processing: 2, interval 21\n",
      "Saved\n",
      "Processing: 2, interval 22\n",
      "Saved\n",
      "Processing: 2, interval 23\n",
      "Saved\n",
      "Processing: 2, interval 24\n",
      "Saved\n",
      "Processing: 2, interval 25\n",
      "Saved\n",
      "Processing: 2, interval 26\n",
      "Saved\n",
      "Processing: 2, interval 27\n",
      "Saved\n",
      "Processing: 2, interval 28\n",
      "Saved\n",
      "Processing: 2, interval 29\n",
      "Saved\n",
      "Processing: 2, interval 30\n",
      "Saved\n",
      "Processing: 2, interval 31\n",
      "Saved\n",
      "Processing: 2, interval 32\n",
      "Saved\n",
      "Processing: 2, interval 33\n",
      "Saved\n",
      "Processing: 2, interval 34\n",
      " at 2, interval 34. Retrying...\n",
      "Processing: 2, interval 34\n",
      " at 2, interval 34. Retrying...\n",
      "Processing: 2, interval 34\n",
      "Saved\n",
      "Processing: 2, interval 35\n",
      " at 2, interval 35. Retrying...\n",
      "Processing: 2, interval 35\n",
      "Saved\n",
      "Processing: 2, interval 36\n",
      "Saved\n",
      "Processing: 2, interval 37\n",
      "Saved\n",
      "Processing: 2, interval 38\n",
      "Saved\n",
      "Processing: 2, interval 39\n",
      "Saved\n",
      "Processing: 2, interval 40\n",
      "Saved\n",
      "Processing: 2, interval 41\n",
      "Saved\n",
      "Processing: 2, interval 42\n",
      "Saved\n",
      "Processing: 2, interval 43\n",
      "Saved\n",
      "Processing: 2, interval 44\n",
      "Saved\n",
      "Processing: 2, interval 45\n",
      "Saved\n",
      "Processing: 2, interval 46\n",
      "Saved\n",
      "Processing: 2, interval 47\n",
      "Saved\n",
      "Processing: 2, interval 48\n",
      "Saved\n",
      "Processing: 2, interval 49\n",
      "Saved\n",
      "Processing: 2, interval 50\n",
      "Saved\n",
      "Processing: 2, interval 51\n",
      "Saved\n",
      "Processing: 2, interval 52\n",
      "Saved\n",
      "Processing: 2, interval 53\n",
      "Saved\n",
      "Processing: 2, interval 54\n",
      "Saved\n",
      "Processing: 2, interval 55\n",
      "Saved\n",
      "Processing: 2, interval 56\n",
      "Saved\n",
      "Processing: 2, interval 57\n",
      "Saved\n",
      "Processing: 2, interval 58\n",
      "Saved\n",
      "Processing: 2, interval 59\n",
      "Saved\n",
      "Processing: 2, interval 60\n",
      "Saved\n",
      "Processing: 2, interval 61\n",
      "Saved\n",
      "Processing: 2, interval 62\n",
      " at 2, interval 62. Retrying...\n",
      "Processing: 2, interval 62\n",
      "Saved\n",
      "Processing: 2, interval 63\n",
      "Saved\n",
      "Processing: 2, interval 64\n",
      " at 2, interval 64. Retrying...\n",
      "Processing: 2, interval 64\n",
      "Saved\n",
      "Processing: 2, interval 65\n",
      "Saved\n",
      "Processing: 2, interval 66\n",
      "Saved\n",
      "Processing: 2, interval 67\n",
      " at 2, interval 67. Retrying...\n",
      "Processing: 2, interval 67\n",
      "Saved\n",
      "Processing: 2, interval 68\n",
      "Saved\n",
      "Processing: 2, interval 69\n",
      " at 2, interval 69. Retrying...\n",
      "Processing: 2, interval 69\n",
      "Saved\n",
      "Processing: 2, interval 70\n",
      " at 2, interval 70. Retrying...\n",
      "Processing: 2, interval 70\n",
      "Saved\n",
      "Processing: 2, interval 71\n",
      "Saved\n",
      "Processing: 2, interval 72\n",
      " at 2, interval 72. Retrying...\n",
      "Processing: 2, interval 72\n",
      "Saved\n",
      "Processing: 2, interval 73\n",
      "Saved\n",
      "Processing: 2, interval 74\n",
      " at 2, interval 74. Retrying...\n",
      "Processing: 2, interval 74\n",
      "Saved\n",
      "Processing: 2, interval 75\n",
      " at 2, interval 75. Retrying...\n",
      "Processing: 2, interval 75\n",
      "Saved\n",
      "Processing: 2, interval 76\n",
      "Saved\n",
      "Processing: 2, interval 77\n",
      " at 2, interval 77. Retrying...\n",
      "Processing: 2, interval 77\n",
      "Saved\n",
      "Processing: 2, interval 78\n",
      " at 2, interval 78. Retrying...\n",
      "Processing: 2, interval 78\n",
      "Saved\n",
      "Processing: 2, interval 79\n",
      "Saved\n",
      "Processing: 2, interval 80\n",
      "Saved\n",
      "Processing: 2, interval 81\n",
      "Saved\n",
      "Processing: 2, interval 82\n",
      " at 2, interval 82. Retrying...\n",
      "Processing: 2, interval 82\n",
      "Saved\n",
      "Processing: 2, interval 83\n",
      " at 2, interval 83. Retrying...\n",
      "Processing: 2, interval 83\n",
      "Saved\n",
      "Processing: 2, interval 84\n",
      "Saved\n",
      "Processing: 2, interval 85\n",
      "Saved\n",
      "Processing: 2, interval 86\n",
      "Saved\n",
      "Processing: 2, interval 87\n",
      "Saved\n",
      "Processing: 2, interval 88\n",
      " at 2, interval 88. Retrying...\n",
      "Processing: 2, interval 88\n",
      "Saved\n",
      "Processing: 2, interval 89\n",
      "Saved\n",
      "Processing: 2, interval 90\n",
      "Saved\n",
      "Processing: 2, interval 91\n",
      "Saved\n",
      "Processing: 2, interval 92\n",
      "Saved\n",
      "Processing: 2, interval 93\n",
      "Saved\n",
      "Processing: 2, interval 94\n",
      "Saved\n",
      "Processing: 2, interval 95\n",
      "Saved\n",
      "Processing: 2, interval 96\n",
      "Saved\n",
      "Processing: 2, interval 97\n",
      "Saved\n",
      "Processing: 2, interval 98\n",
      "Saved\n",
      "Processing: 2, interval 99\n",
      "Saved\n",
      "Processing: 2, interval 100\n",
      "Saved\n",
      "Processing: 2, interval 101\n",
      "Saved\n",
      "Processing: 2, interval 102\n",
      "Saved\n",
      "Processing: 2, interval 103\n",
      "Saved\n",
      "Processing: 2, interval 104\n",
      "Saved\n",
      "Processing: 2, interval 105\n",
      "Saved\n",
      "Processing: 2, interval 106\n",
      "Saved\n",
      "Processing: 2, interval 107\n",
      "Saved\n",
      "Processing: 2, interval 108\n",
      "Saved\n",
      "Processing: 2, interval 109\n",
      "Saved\n",
      "Processing: 2, interval 110\n",
      "Saved\n",
      "Processing: 2, interval 111\n",
      "Saved\n",
      "Processing: 2, interval 112\n",
      "Saved\n",
      "Processing: 2, interval 113\n",
      "Saved\n",
      "Processing: 2, interval 114\n",
      "Saved\n",
      "Processing: 2, interval 115\n",
      "Saved\n",
      "Processing: 2, interval 116\n",
      "Saved\n",
      "Processing: 2, interval 117\n",
      "Saved\n",
      "Processing: 2, interval 118\n",
      "Saved\n",
      "Processing: 2, interval 119\n",
      "Saved\n",
      "Processing: 2, interval 120\n",
      " at 2, interval 120. Retrying...\n",
      "Processing: 2, interval 120\n",
      " at 2, interval 120. Retrying...\n",
      "Processing: 2, interval 120\n",
      "Saved\n",
      "Processing: 2, interval 121\n",
      "Saved\n",
      "Processing: 2, interval 122\n",
      " at 2, interval 122. Retrying...\n",
      "Processing: 2, interval 122\n",
      "Saved\n",
      "Processing: 2, interval 123\n",
      "Saved\n",
      "Processing: 2, interval 124\n",
      "Saved\n",
      "Processing: 2, interval 125\n",
      "Saved\n",
      "Processing: 2, interval 126\n",
      "Saved\n",
      "Processing: 2, interval 127\n",
      "Saved\n",
      "Processing: 2, interval 128\n",
      "Saved\n",
      "Processing: 2, interval 129\n",
      "Saved\n",
      "Processing: 2, interval 130\n",
      "Saved\n",
      "Processing: 2, interval 131\n",
      "Saved\n",
      "Processing: 2, interval 132\n",
      "Saved\n",
      "Processing: 2, interval 133\n",
      "Saved\n",
      "Processing: 2, interval 134\n",
      "Saved\n",
      "Processing: 2, interval 135\n",
      "Saved\n",
      "Processing: 2, interval 136\n",
      "Saved\n",
      "Processing: 2, interval 137\n",
      "Saved\n",
      "Processing: 2, interval 138\n",
      "Saved\n",
      "Processing: 2, interval 139\n",
      "Saved\n",
      "Processing: 2, interval 140\n",
      "Saved\n",
      "Processing: 2, interval 141\n",
      "Saved\n",
      "Processing: 2, interval 142\n",
      "Saved\n",
      "Processing: 2, interval 143\n",
      "Saved\n",
      "Processing: 2, interval 144\n",
      "Saved\n",
      "Processing: 2, interval 145\n",
      "Saved\n",
      "Processing: 2, interval 146\n",
      "Saved\n",
      "Processing: 2, interval 147\n",
      "Saved\n",
      "Processing: 2, interval 148\n",
      "Saved\n",
      "Processing: 2, interval 149\n",
      "Saved\n",
      "Processing: 2, interval 150\n",
      "Saved\n",
      "Processing: 2, interval 151\n",
      "Saved\n",
      "Processing: 2, interval 152\n",
      "Saved\n",
      "Processing: 2, interval 153\n",
      "Saved\n",
      "Processing: 2, interval 154\n",
      "Saved\n",
      "Processing: 2, interval 155\n",
      "Saved\n",
      "Processing: 2, interval 156\n",
      "Saved\n",
      "Processing: 2, interval 157\n",
      "Saved\n",
      "Processing: 2, interval 158\n",
      "Saved\n",
      "Processing: 2, interval 159\n",
      "Saved\n",
      "Processing: 2, interval 160\n",
      "Saved\n",
      "Processing: 2, interval 161\n",
      "Saved\n",
      "Processing: 2, interval 162\n",
      "Saved\n",
      "Processing: 2, interval 163\n",
      "Saved\n",
      "Processing: 2, interval 164\n",
      "Saved\n",
      "Processing: 2, interval 165\n",
      "Saved\n",
      "Processing: 2, interval 166\n",
      "Saved\n",
      "Processing: 2, interval 167\n",
      "Saved\n",
      "Processing: 2, interval 168\n",
      "Saved\n",
      "Processing: 2, interval 169\n",
      "Saved\n",
      "Processing: 2, interval 170\n",
      "Saved\n",
      "Processing: 2, interval 171\n",
      "Saved\n",
      "Processing: 2, interval 172\n",
      "Saved\n",
      "Processing: 2, interval 173\n",
      "Saved\n",
      "Processing: 2, interval 174\n",
      "Saved\n",
      "Processing: 2, interval 175\n",
      "Saved\n",
      "Processing: 2, interval 176\n",
      "Saved\n",
      "Processing: 2, interval 177\n",
      "Saved\n",
      "Processing: 2, interval 178\n",
      "Saved\n",
      "Processing: 2, interval 179\n",
      "Saved\n",
      "Processing: 2, interval 180\n",
      "Saved\n",
      "Processing: 2, interval 181\n",
      "Saved\n",
      "Processing: 2, interval 182\n",
      "Saved\n",
      "Processing: 2, interval 183\n",
      "Saved\n",
      "Processing: 2, interval 184\n",
      "Saved\n",
      "Processing: 2, interval 185\n",
      "Saved\n",
      "Processing: 2, interval 186\n",
      "Saved\n",
      "Processing: 2, interval 187\n",
      "Saved\n",
      "Processing: 2, interval 188\n",
      "Saved\n",
      "Processing: 2, interval 189\n",
      "Saved\n",
      "Processing: 2, interval 190\n",
      "Saved\n",
      "Processing: 2, interval 191\n",
      "Saved\n",
      "Processing: 2, interval 192\n",
      "Saved\n",
      "Processing: 2, interval 193\n",
      "Saved\n",
      "Processing: 2, interval 194\n",
      "Saved\n",
      "Processing: 2, interval 195\n",
      "Saved\n",
      "Processing: 2, interval 196\n",
      "Saved\n",
      "Processing: 2, interval 197\n",
      "Saved\n",
      "Processing: 2, interval 198\n",
      "Saved\n",
      "Processing: 2, interval 199\n",
      "Saved\n",
      "Processing: 2, interval 200\n",
      "Saved\n",
      "Processing: 2, interval 201\n",
      "Saved\n",
      "Processing: 2, interval 202\n",
      "Saved\n",
      "Processing: 2, interval 203\n",
      "Saved\n",
      "Processing: 2, interval 204\n",
      "Saved\n",
      "Processing: 2, interval 205\n",
      "Saved\n",
      "Processing: 2, interval 206\n",
      "Saved\n",
      "Processing: 2, interval 207\n",
      "Saved\n",
      "Processing: 2, interval 208\n",
      "Saved\n",
      "Processing: 2, interval 209\n",
      "Saved\n",
      "Processing: 2, interval 210\n",
      "Saved\n",
      "Processing: 2, interval 211\n",
      "Saved\n",
      "Processing: 2, interval 212\n",
      "Saved\n",
      "Processing: 2, interval 213\n",
      "Saved\n",
      "Processing: 2, interval 214\n",
      "Saved\n",
      "Processing: 2, interval 215\n",
      "Saved\n",
      "Processing: 2, interval 216\n",
      "Saved\n",
      "Processing: 2, interval 217\n",
      "Saved\n",
      "Processing: 2, interval 218\n",
      "Saved\n",
      "Processing: 2, interval 219\n",
      "Saved\n",
      "Processing: 2, interval 220\n",
      "Saved\n",
      "Processing: 2, interval 221\n",
      "Saved\n",
      "Processing: 2, interval 222\n",
      "Saved\n",
      "Processing: 2, interval 223\n",
      "Saved\n",
      "Processing: 2, interval 224\n",
      "Saved\n",
      "Processing: 2, interval 225\n",
      "Saved\n",
      "Processing: 2, interval 226\n",
      "Saved\n",
      "Processing: 2, interval 227\n",
      " at 2, interval 227. Retrying...\n",
      "Processing: 2, interval 227\n",
      "Saved\n",
      "Processing: 2, interval 228\n",
      " at 2, interval 228. Retrying...\n",
      "Processing: 2, interval 228\n",
      "Saved\n",
      "Processing: 2, interval 229\n",
      "Saved\n",
      "Processing: 2, interval 230\n",
      "Saved\n",
      "Processing: 2, interval 231\n",
      "Saved\n",
      "Processing: 2, interval 232\n",
      "Saved\n",
      "Processing: 2, interval 233\n",
      "Saved\n",
      "Processing: 2, interval 234\n",
      "Saved\n",
      "Processing: 2, interval 235\n",
      "Saved\n",
      "Processing: 2, interval 236\n",
      "Saved\n",
      "Processing: 2, interval 237\n",
      "Saved\n",
      "Processing: 2, interval 238\n",
      "Saved\n",
      "Processing: 2, interval 239\n",
      "Saved\n",
      "Processing: 2, interval 240\n",
      "Saved\n",
      "Processing: 2, interval 241\n",
      "Saved\n",
      "Processing: 2, interval 242\n",
      "Saved\n",
      "Processing: 2, interval 243\n",
      "Saved\n",
      "Processing: 2, interval 244\n",
      "Saved\n",
      "Processing: 2, interval 245\n",
      "Saved\n",
      "Processing: 2, interval 246\n",
      " at 2, interval 246. Retrying...\n",
      "Processing: 2, interval 246\n",
      "Saved\n",
      "Processing: 2, interval 247\n",
      " at 2, interval 247. Retrying...\n",
      "Processing: 2, interval 247\n",
      "Saved\n",
      "Processing: 2, interval 248\n",
      "Saved\n",
      "Processing: 2, interval 249\n",
      "Saved\n",
      "Processing: 2, interval 250\n",
      " at 2, interval 250. Retrying...\n",
      "Processing: 2, interval 250\n",
      "Saved\n",
      "Processing: 2, interval 251\n",
      "Saved\n",
      "Processing: 2, interval 252\n",
      "Saved\n",
      "Processing: 2, interval 253\n",
      "Saved\n",
      "Processing: 2, interval 254\n",
      "Saved\n",
      "Processing: 2, interval 255\n",
      "Saved\n",
      "Processing: 2, interval 256\n",
      "Saved\n",
      "Processing: 2, interval 257\n",
      "Saved\n",
      "Processing: 2, interval 258\n",
      "Saved\n",
      "Processing: 2, interval 259\n",
      " at 2, interval 259. Retrying...\n",
      "Processing: 2, interval 259\n",
      "Saved\n",
      "Processing: 2, interval 260\n",
      "Saved\n",
      "Processing: 2, interval 261\n",
      "Saved\n",
      "Processing: 2, interval 262\n",
      "Saved\n",
      "Processing: 2, interval 263\n",
      "Saved\n",
      "Processing: 2, interval 264\n",
      " at 2, interval 264. Retrying...\n",
      "Processing: 2, interval 264\n",
      "Saved\n",
      "Processing: 2, interval 265\n",
      "Saved\n",
      "Processing: 2, interval 266\n",
      "Saved\n",
      "Processing: 2, interval 267\n",
      "Saved\n",
      "Processing: 2, interval 268\n",
      "Saved\n",
      "Processing: 2, interval 269\n",
      "Saved\n",
      "Processing: 2, interval 270\n",
      "Saved\n",
      "Processing: 2, interval 271\n",
      "Saved\n",
      "Processing: 2, interval 272\n",
      "Saved\n",
      "Processing: 2, interval 273\n",
      "Saved\n",
      "Processing: 2, interval 274\n",
      "Saved\n",
      "Processing: 2, interval 275\n",
      "Saved\n",
      "Processing: 2, interval 276\n",
      "Saved\n",
      "Processing: 2, interval 277\n",
      "Saved\n",
      "Processing: 2, interval 278\n",
      " at 2, interval 278. Retrying...\n",
      "Processing: 2, interval 278\n",
      "Saved\n",
      "Processing: 2, interval 279\n",
      " at 2, interval 279. Retrying...\n",
      "Processing: 2, interval 279\n",
      "Saved\n",
      "Processing: 2, interval 280\n",
      "Saved\n",
      "Processing: 2, interval 281\n",
      "Saved\n",
      "Processing: 2, interval 282\n",
      " at 2, interval 282. Retrying...\n",
      "Processing: 2, interval 282\n",
      "Saved\n",
      "Processing: 2, interval 283\n",
      " at 2, interval 283. Retrying...\n",
      "Processing: 2, interval 283\n",
      "Saved\n",
      "Processing: 2, interval 284\n",
      "Saved\n",
      "Processing: 2, interval 285\n",
      "Saved\n",
      "Processing: 2, interval 286\n",
      "Saved\n",
      "Processing: 2, interval 287\n",
      "Saved\n",
      "Processing: 2, interval 288\n",
      "Saved\n",
      "Processing: 2, interval 289\n",
      "Saved\n",
      "Processing: 2, interval 290\n",
      "Saved\n",
      "Processing: 2, interval 291\n",
      "Saved\n",
      "Processing: 2, interval 292\n",
      "Saved\n",
      "Processing: 2, interval 293\n",
      "Saved\n",
      "Processing: 2, interval 294\n",
      "Saved\n",
      "Processing: 2, interval 295\n",
      "Saved\n",
      "Processing: 2, interval 296\n",
      "Saved\n",
      "Processing: 2, interval 297\n",
      "Saved\n",
      "Processing: 2, interval 298\n",
      "Saved\n",
      "Processing: 2, interval 299\n",
      "Saved\n",
      "Processing: 2, interval 300\n",
      " at 2, interval 300. Retrying...\n",
      "Processing: 2, interval 300\n",
      " at 2, interval 300. Retrying...\n",
      "Processing: 2, interval 300\n",
      "Saved\n",
      "Processing: 2, interval 301\n",
      "Saved\n",
      "Processing: 2, interval 302\n",
      "Saved\n",
      "Processing: 2, interval 303\n",
      "Saved\n",
      "Processing: 2, interval 304\n",
      " at 2, interval 304. Retrying...\n",
      "Processing: 2, interval 304\n",
      "Saved\n",
      "Processing: 2, interval 305\n",
      "Saved\n",
      "Processing: 2, interval 306\n",
      "Saved\n",
      "Processing: 2, interval 307\n",
      "Saved\n",
      "Processing: 2, interval 308\n",
      "Saved\n",
      "Processing: 2, interval 309\n",
      "Saved\n",
      "Processing: 2, interval 310\n",
      "Saved\n",
      "Processing: 2, interval 311\n",
      "Saved\n",
      "Processing: 2, interval 312\n",
      "Saved\n",
      "Processing: 2, interval 313\n",
      "Saved\n",
      "Processing: 2, interval 314\n",
      "Saved\n",
      "Processing: 2, interval 315\n",
      "Saved\n",
      "Processing: 2, interval 316\n",
      "Saved\n",
      "Processing: 2, interval 317\n",
      "Saved\n",
      "Processing: 2, interval 318\n",
      "Saved\n",
      "Processing: 2, interval 319\n",
      "Saved\n",
      "Processing: 2, interval 320\n",
      "Saved\n",
      "Processing: 2, interval 321\n",
      "Saved\n",
      "Processing: 2, interval 322\n",
      "Saved\n",
      "Processing: 2, interval 323\n",
      "Saved\n",
      "Processing: 2, interval 324\n",
      "Saved\n",
      "Processing: 2, interval 325\n",
      "Saved\n",
      "Processing: 2, interval 326\n",
      "Saved\n",
      "Processing: 2, interval 327\n",
      "Saved\n",
      "Processing: 2, interval 328\n",
      "Saved\n",
      "Processing: 2, interval 329\n",
      "Saved\n",
      "Processing: 2, interval 330\n",
      "Saved\n",
      "Processing: 2, interval 331\n",
      "Saved\n",
      "Processing: 2, interval 332\n",
      "Saved\n",
      "Processing: 2, interval 333\n",
      "Saved\n",
      "Processing: 2, interval 334\n",
      "Saved\n",
      "Processing: 2, interval 335\n",
      " at 2, interval 335. Retrying...\n",
      "Processing: 2, interval 335\n",
      "Saved\n",
      "Processing: 2, interval 336\n",
      "Saved\n",
      "Processing: 2, interval 337\n",
      "Saved\n",
      "Processing: 2, interval 338\n",
      "Saved\n",
      "Processing: 2, interval 339\n",
      "Saved\n",
      "Processing: 2, interval 340\n",
      "Saved\n",
      "Processing: 2, interval 341\n",
      "Saved\n",
      "Processing: 2, interval 342\n",
      "Saved\n",
      "Processing: 2, interval 343\n",
      "Saved\n",
      "Processing: 2, interval 344\n",
      "Saved\n",
      "Processing: 2, interval 345\n",
      "Saved\n",
      "Processing: 2, interval 346\n",
      "Saved\n",
      "Processing: 2, interval 347\n",
      "Saved\n",
      "Processing: 2, interval 348\n",
      "Saved\n",
      "Processing: 2, interval 349\n",
      "Saved\n",
      "Processing: 2, interval 350\n",
      "Saved\n",
      "Processing: 2, interval 351\n",
      "Saved\n",
      "Processing: 2, interval 352\n",
      "Saved\n",
      "Processing: 2, interval 353\n",
      "Saved\n",
      "Processing: 2, interval 354\n",
      "Saved\n",
      "Processing: 2, interval 355\n",
      " at 2, interval 355. Retrying...\n",
      "Processing: 2, interval 355\n",
      "Saved\n",
      "Processing: 2, interval 356\n",
      " at 2, interval 356. Retrying...\n",
      "Processing: 2, interval 356\n",
      "Saved\n",
      "Processing: 2, interval 357\n",
      " at 2, interval 357. Retrying...\n",
      "Processing: 2, interval 357\n",
      "Saved\n",
      "Processing: 2, interval 358\n",
      "Saved\n",
      "Processing: 2, interval 359\n",
      "Saved\n",
      "Processing: 2, interval 360\n",
      "Saved\n",
      "Processing: 2, interval 361\n",
      "Saved\n",
      "Processing: 2, interval 362\n",
      " at 2, interval 362. Retrying...\n",
      "Processing: 2, interval 362\n",
      "Saved\n",
      "Processing: 2, interval 363\n",
      " at 2, interval 363. Retrying...\n",
      "Processing: 2, interval 363\n",
      "Saved\n",
      "Processing: 2, interval 364\n",
      " at 2, interval 364. Retrying...\n",
      "Processing: 2, interval 364\n",
      "Saved\n",
      "Processing: 2, interval 365\n",
      "Saved\n",
      "Processing: 2, interval 366\n",
      "Saved\n",
      "Processing: 2, interval 367\n",
      " at 2, interval 367. Retrying...\n",
      "Processing: 2, interval 367\n",
      "Saved\n",
      "Processing: 2, interval 368\n",
      " at 2, interval 368. Retrying...\n",
      "Processing: 2, interval 368\n",
      "Saved\n",
      "Processing: 2, interval 369\n",
      "Saved\n",
      "Processing: 2, interval 370\n",
      " at 2, interval 370. Retrying...\n",
      "Processing: 2, interval 370\n",
      "Saved\n",
      "Processing: 2, interval 371\n",
      "Saved\n",
      "Processing: 2, interval 372\n",
      "Saved\n",
      "Processing: 2, interval 373\n",
      "Saved\n",
      "Processing: 2, interval 374\n",
      "Saved\n",
      "Processing: 2, interval 375\n",
      "Saved\n",
      "Processing: 2, interval 376\n",
      "Saved\n",
      "Processing: 2, interval 377\n",
      "Saved\n",
      "Processing: 2, interval 378\n",
      "Saved\n",
      "Processing: 2, interval 379\n",
      "Saved\n",
      "Processing: 2, interval 380\n",
      "Saved\n",
      "Processing: 2, interval 381\n",
      "Saved\n",
      "Processing: 2, interval 382\n",
      "Saved\n",
      "Processing: 2, interval 383\n",
      "Saved\n",
      "Processing: 2, interval 384\n",
      "Saved\n",
      "Processing: 2, interval 385\n",
      "Saved\n",
      "Processing: 2, interval 386\n",
      "Saved\n",
      "Processing: 2, interval 387\n",
      "Saved\n",
      "Processing: 2, interval 388\n",
      "Saved\n",
      "Processing: 2, interval 389\n",
      "Saved\n",
      "Processing: 2, interval 390\n",
      "Saved\n",
      "Processing: 2, interval 391\n",
      "Saved\n",
      "Processing: 2, interval 392\n",
      "Saved\n",
      "Processing: 2, interval 393\n",
      "Saved\n",
      "Processing: 2, interval 394\n",
      "Saved\n",
      "Processing: 2, interval 395\n",
      " at 2, interval 395. Retrying...\n",
      "Processing: 2, interval 395\n",
      "Saved\n",
      "Processing: 2, interval 396\n",
      " at 2, interval 396. Retrying...\n",
      "Processing: 2, interval 396\n",
      "Saved\n",
      "Processing: 2, interval 397\n",
      " at 2, interval 397. Retrying...\n",
      "Processing: 2, interval 397\n",
      " at 2, interval 397. Retrying...\n",
      "Processing: 2, interval 397\n",
      "Saved\n",
      "Processing: 2, interval 398\n",
      " at 2, interval 398. Retrying...\n",
      "Processing: 2, interval 398\n",
      "Saved\n",
      "Processing: 2, interval 399\n",
      "Saved\n",
      "Processing: 2, interval 400\n",
      "Saved\n",
      "Processing: 2, interval 401\n",
      "Saved\n",
      "Processing: 2, interval 402\n",
      " at 2, interval 402. Retrying...\n",
      "Processing: 2, interval 402\n",
      "Saved\n",
      "Processing: 2, interval 403\n",
      "Saved\n",
      "Processing: 2, interval 404\n",
      "Saved\n",
      "Processing: 2, interval 405\n",
      "Saved\n",
      "Processing: 2, interval 406\n",
      "Saved\n",
      "Processing: 2, interval 407\n",
      "Saved\n",
      "Processing: 2, interval 408\n",
      "Saved\n",
      "Processing: 2, interval 409\n",
      "Saved\n",
      "Processing: 2, interval 410\n",
      "Saved\n",
      "Processing: 20, interval 0\n",
      "Saved\n",
      "Processing: 20, interval 1\n",
      "Saved\n",
      "Processing: 20, interval 2\n",
      "Saved\n",
      "Processing: 20, interval 3\n",
      " at 20, interval 3. Retrying...\n",
      "Processing: 20, interval 3\n",
      "Saved\n",
      "Processing: 20, interval 4\n",
      "Saved\n",
      "Processing: 20, interval 5\n",
      "Saved\n",
      "Processing: 20, interval 6\n",
      "Saved\n",
      "Processing: 20, interval 7\n",
      " at 20, interval 7. Retrying...\n",
      "Processing: 20, interval 7\n",
      "Saved\n",
      "Processing: 20, interval 8\n",
      "Saved\n",
      "Processing: 20, interval 9\n",
      " at 20, interval 9. Retrying...\n",
      "Processing: 20, interval 9\n",
      "Saved\n",
      "Processing: 20, interval 10\n",
      "Saved\n",
      "Processing: 20, interval 11\n",
      "Saved\n",
      "Processing: 20, interval 12\n",
      "Saved\n",
      "Processing: 20, interval 13\n",
      "Saved\n",
      "Processing: 20, interval 14\n",
      "Saved\n",
      "Processing: 20, interval 15\n",
      "Saved\n",
      "Processing: 20, interval 16\n",
      "Saved\n",
      "Processing: 20, interval 17\n",
      "Saved\n",
      "Processing: 20, interval 18\n",
      "Saved\n",
      "Processing: 20, interval 19\n",
      "Saved\n",
      "Processing: 20, interval 20\n",
      "Saved\n",
      "Processing: 20, interval 21\n",
      "Saved\n",
      "Processing: 20, interval 22\n",
      " at 20, interval 22. Retrying...\n",
      "Processing: 20, interval 22\n",
      "Saved\n",
      "Processing: 20, interval 23\n",
      "Saved\n",
      "Processing: 20, interval 24\n",
      "Saved\n",
      "Processing: 20, interval 25\n",
      "Saved\n",
      "Processing: 20, interval 26\n",
      "Saved\n",
      "Processing: 20, interval 27\n",
      "Saved\n",
      "Processing: 20, interval 28\n",
      "Saved\n",
      "Processing: 20, interval 29\n",
      "Saved\n",
      "Processing: 20, interval 30\n",
      "Saved\n",
      "Processing: 20, interval 31\n",
      "Saved\n",
      "Processing: 20, interval 32\n",
      "Saved\n",
      "Processing: 20, interval 33\n",
      "Saved\n",
      "Processing: 20, interval 34\n",
      " at 20, interval 34. Retrying...\n",
      "Processing: 20, interval 34\n",
      "Saved\n",
      "Processing: 20, interval 35\n",
      "Saved\n",
      "Processing: 20, interval 36\n",
      "Saved\n",
      "Processing: 20, interval 37\n",
      "Saved\n",
      "Processing: 20, interval 38\n",
      "Saved\n",
      "Processing: 20, interval 39\n",
      "Saved\n",
      "Processing: 20, interval 40\n",
      "Saved\n",
      "Processing: 20, interval 41\n",
      " at 20, interval 41. Retrying...\n",
      "Processing: 20, interval 41\n",
      "Saved\n",
      "Processing: 20, interval 42\n",
      "Saved\n",
      "Processing: 20, interval 43\n",
      "Saved\n",
      "Processing: 20, interval 44\n",
      "Saved\n",
      "Processing: 20, interval 45\n",
      "Saved\n",
      "Processing: 20, interval 46\n",
      "Saved\n",
      "Processing: 20, interval 47\n",
      "Saved\n",
      "Processing: 20, interval 48\n",
      "Saved\n",
      "Processing: 20, interval 49\n",
      "Saved\n",
      "Processing: 20, interval 50\n",
      "Saved\n",
      "Processing: 20, interval 51\n",
      "Saved\n",
      "Processing: 20, interval 52\n",
      "Saved\n",
      "Processing: 20, interval 53\n",
      "Saved\n",
      "Processing: 20, interval 54\n",
      "Saved\n",
      "Processing: 20, interval 55\n",
      "Saved\n",
      "Processing: 20, interval 56\n",
      "Saved\n",
      "Processing: 20, interval 57\n",
      "Saved\n",
      "Processing: 20, interval 58\n",
      "Saved\n",
      "Processing: 20, interval 59\n",
      "Saved\n",
      "Processing: 20, interval 60\n",
      "Saved\n",
      "Processing: 20, interval 61\n",
      "Saved\n",
      "Processing: 20, interval 62\n",
      "Saved\n",
      "Processing: 20, interval 63\n",
      "Saved\n",
      "Processing: 20, interval 64\n",
      "Saved\n",
      "Processing: 20, interval 65\n",
      "Saved\n",
      "Processing: 20, interval 66\n",
      "Saved\n",
      "Processing: 20, interval 67\n",
      "Saved\n",
      "Processing: 20, interval 68\n",
      "Saved\n",
      "Processing: 20, interval 69\n",
      "Saved\n",
      "Processing: 20, interval 70\n",
      "Saved\n",
      "Processing: 20, interval 71\n",
      "Saved\n",
      "Processing: 20, interval 72\n",
      "Saved\n",
      "Processing: 20, interval 73\n",
      "Saved\n",
      "Processing: 20, interval 74\n",
      "Saved\n",
      "Processing: 20, interval 75\n",
      "Saved\n",
      "Processing: 20, interval 76\n",
      "Saved\n",
      "Processing: 20, interval 77\n",
      "Saved\n",
      "Processing: 20, interval 78\n",
      "Saved\n",
      "Processing: 20, interval 79\n",
      "Saved\n",
      "Processing: 20, interval 80\n",
      "Saved\n",
      "Processing: 20, interval 81\n",
      "Saved\n",
      "Processing: 20, interval 82\n",
      "Saved\n",
      "Processing: 20, interval 83\n",
      "Saved\n",
      "Processing: 20, interval 84\n",
      "Saved\n",
      "Processing: 20, interval 85\n",
      "Saved\n",
      "Processing: 20, interval 86\n",
      "Saved\n",
      "Processing: 20, interval 87\n",
      "Saved\n",
      "Processing: 20, interval 88\n",
      "Saved\n",
      "Processing: 20, interval 89\n",
      "Saved\n",
      "Processing: 20, interval 90\n",
      "Saved\n",
      "Processing: 20, interval 91\n",
      "Saved\n",
      "Processing: 20, interval 92\n",
      "Saved\n",
      "Processing: 20, interval 93\n",
      "Saved\n",
      "Processing: 20, interval 94\n",
      "Saved\n",
      "Processing: 20, interval 95\n",
      "Saved\n",
      "Processing: 20, interval 96\n",
      "Saved\n",
      "Processing: 20, interval 97\n",
      "Saved\n",
      "Processing: 20, interval 98\n",
      "Saved\n",
      "Processing: 20, interval 99\n",
      "Saved\n",
      "Processing: 20, interval 100\n",
      "Saved\n",
      "Processing: 20, interval 101\n",
      "Saved\n",
      "Processing: 20, interval 102\n",
      "Saved\n",
      "Processing: 20, interval 103\n",
      "Saved\n",
      "Processing: 20, interval 104\n",
      "Saved\n",
      "Processing: 20, interval 105\n",
      "Saved\n",
      "Processing: 20, interval 106\n",
      "Saved\n",
      "Processing: 20, interval 107\n",
      "Saved\n",
      "Processing: 20, interval 108\n",
      "Saved\n",
      "Processing: 20, interval 109\n",
      "Saved\n",
      "Processing: 20, interval 110\n",
      "Saved\n",
      "Processing: 20, interval 111\n",
      "Saved\n",
      "Processing: 20, interval 112\n",
      "Saved\n",
      "Processing: 20, interval 113\n",
      "Saved\n",
      "Processing: 20, interval 114\n",
      "Saved\n",
      "Processing: 20, interval 115\n",
      "Saved\n",
      "Processing: 20, interval 116\n",
      "Saved\n",
      "Processing: 20, interval 117\n",
      "Saved\n",
      "Processing: 21, interval 0\n",
      "Saved\n",
      "Processing: 21, interval 1\n",
      "Saved\n",
      "Processing: 21, interval 2\n",
      "Saved\n",
      "Processing: 21, interval 3\n",
      "Saved\n",
      "Processing: 21, interval 4\n",
      "Saved\n",
      "Processing: 21, interval 5\n",
      "Saved\n",
      "Processing: 21, interval 6\n",
      "Saved\n",
      "Processing: 21, interval 7\n",
      "Saved\n",
      "Processing: 21, interval 8\n",
      "Saved\n",
      "Processing: 21, interval 9\n",
      "Saved\n",
      "Processing: 21, interval 10\n",
      "Saved\n",
      "Processing: 21, interval 11\n",
      "Saved\n",
      "Processing: 21, interval 12\n",
      "Saved\n",
      "Processing: 21, interval 13\n",
      "Saved\n",
      "Processing: 21, interval 14\n",
      "Saved\n",
      "Processing: 21, interval 15\n",
      "Saved\n",
      "Processing: 21, interval 16\n",
      "Saved\n",
      "Processing: 21, interval 17\n",
      "Saved\n",
      "Processing: 21, interval 18\n",
      "Saved\n",
      "Processing: 21, interval 19\n",
      "Saved\n",
      "Processing: 21, interval 20\n",
      "Saved\n",
      "Processing: 21, interval 21\n",
      "Saved\n",
      "Processing: 21, interval 22\n",
      "Saved\n",
      "Processing: 21, interval 23\n",
      "Saved\n",
      "Processing: 21, interval 24\n",
      " at 21, interval 24. Retrying...\n",
      "Processing: 21, interval 24\n",
      "Saved\n",
      "Processing: 21, interval 25\n",
      "Saved\n",
      "Processing: 21, interval 26\n",
      "Saved\n",
      "Processing: 21, interval 27\n",
      "Saved\n",
      "Processing: 21, interval 28\n",
      " at 21, interval 28. Retrying...\n",
      "Processing: 21, interval 28\n",
      "Saved\n",
      "Processing: 21, interval 29\n",
      " at 21, interval 29. Retrying...\n",
      "Processing: 21, interval 29\n",
      " at 21, interval 29. Retrying...\n",
      "Processing: 21, interval 29\n",
      "Saved\n",
      "Processing: 21, interval 30\n",
      "Saved\n",
      "Processing: 21, interval 31\n",
      "Saved\n",
      "Processing: 21, interval 32\n",
      "Saved\n",
      "Processing: 21, interval 33\n",
      " at 21, interval 33. Retrying...\n",
      "Processing: 21, interval 33\n",
      "Saved\n",
      "Processing: 21, interval 34\n",
      " at 21, interval 34. Retrying...\n",
      "Processing: 21, interval 34\n",
      "Saved\n",
      "Processing: 21, interval 35\n",
      " at 21, interval 35. Retrying...\n",
      "Processing: 21, interval 35\n",
      "Saved\n",
      "Processing: 21, interval 36\n",
      " at 21, interval 36. Retrying...\n",
      "Processing: 21, interval 36\n",
      "Saved\n",
      "Processing: 21, interval 37\n",
      "Saved\n",
      "Processing: 21, interval 38\n",
      "Saved\n",
      "Processing: 21, interval 39\n",
      "Saved\n",
      "Processing: 21, interval 40\n",
      "Saved\n",
      "Processing: 21, interval 41\n",
      "Saved\n",
      "Processing: 21, interval 42\n",
      " at 21, interval 42. Retrying...\n",
      "Processing: 21, interval 42\n",
      "Saved\n",
      "Processing: 21, interval 43\n",
      "Saved\n",
      "Processing: 21, interval 44\n",
      "Saved\n",
      "Processing: 21, interval 45\n",
      "Saved\n",
      "Processing: 21, interval 46\n",
      "Saved\n",
      "Processing: 21, interval 47\n",
      "Saved\n",
      "Processing: 21, interval 48\n",
      " at 21, interval 48. Retrying...\n",
      "Processing: 21, interval 48\n",
      "Saved\n",
      "Processing: 21, interval 49\n",
      "Saved\n",
      "Processing: 21, interval 50\n",
      "Saved\n",
      "Processing: 21, interval 51\n",
      "Saved\n",
      "Processing: 21, interval 52\n",
      "Saved\n",
      "Processing: 21, interval 53\n",
      "Saved\n",
      "Processing: 21, interval 54\n",
      "Saved\n",
      "Processing: 21, interval 55\n",
      "Saved\n",
      "Processing: 21, interval 56\n",
      "Saved\n",
      "Processing: 21, interval 57\n",
      "Saved\n",
      "Processing: 21, interval 58\n",
      "Saved\n",
      "Processing: 21, interval 59\n",
      "Saved\n",
      "Processing: 21, interval 60\n",
      " at 21, interval 60. Retrying...\n",
      "Processing: 21, interval 60\n",
      "Saved\n",
      "Processing: 21, interval 61\n",
      "Saved\n",
      "Processing: 21, interval 62\n",
      "Saved\n",
      "Processing: 21, interval 63\n",
      " at 21, interval 63. Retrying...\n",
      "Processing: 21, interval 63\n",
      "Saved\n",
      "Processing: 21, interval 64\n",
      "Saved\n",
      "Processing: 21, interval 65\n",
      "Saved\n",
      "Processing: 21, interval 66\n",
      "Saved\n",
      "Processing: 21, interval 67\n",
      "Saved\n",
      "Processing: 21, interval 68\n",
      "Saved\n",
      "Processing: 21, interval 69\n",
      "Saved\n",
      "Processing: 21, interval 70\n",
      "Saved\n",
      "Processing: 21, interval 71\n",
      "Saved\n",
      "Processing: 21, interval 72\n",
      "Saved\n",
      "Processing: 21, interval 73\n",
      "Saved\n",
      "Processing: 21, interval 74\n",
      "Saved\n",
      "Processing: 22, interval 0\n",
      "Saved\n",
      "Processing: 22, interval 1\n",
      "Saved\n",
      "Processing: 22, interval 2\n",
      "Saved\n",
      "Processing: 22, interval 3\n",
      "Saved\n",
      "Processing: 22, interval 4\n",
      "Saved\n",
      "Processing: 22, interval 5\n",
      "Saved\n",
      "Processing: 22, interval 6\n",
      "Saved\n",
      "Processing: 22, interval 7\n",
      "Saved\n",
      "Processing: 22, interval 8\n",
      "Saved\n",
      "Processing: 22, interval 9\n",
      "Saved\n",
      "Processing: 22, interval 10\n",
      "Saved\n",
      "Processing: 22, interval 11\n",
      "Saved\n",
      "Processing: 22, interval 12\n",
      "Saved\n",
      "Processing: 22, interval 13\n",
      "Saved\n",
      "Processing: 22, interval 14\n",
      "Saved\n",
      "Processing: 22, interval 15\n",
      "Saved\n",
      "Processing: 22, interval 16\n",
      "Saved\n",
      "Processing: 22, interval 17\n",
      "Saved\n",
      "Processing: 22, interval 18\n",
      "Saved\n",
      "Processing: 22, interval 19\n",
      "Saved\n",
      "Processing: 22, interval 20\n",
      "Saved\n",
      "Processing: 22, interval 21\n",
      " at 22, interval 21. Retrying...\n",
      "Processing: 22, interval 21\n",
      "Saved\n",
      "Processing: 22, interval 22\n",
      " at 22, interval 22. Retrying...\n",
      "Processing: 22, interval 22\n",
      "Saved\n",
      "Processing: 22, interval 23\n",
      "Saved\n",
      "Processing: 22, interval 24\n",
      "Saved\n",
      "Processing: 22, interval 25\n",
      "Saved\n",
      "Processing: 22, interval 26\n",
      "Saved\n",
      "Processing: 22, interval 27\n",
      "Saved\n",
      "Processing: 22, interval 28\n",
      "Saved\n",
      "Processing: 22, interval 29\n",
      "Saved\n",
      "Processing: 22, interval 30\n",
      "Saved\n",
      "Processing: 22, interval 31\n",
      "Saved\n",
      "Processing: 22, interval 32\n",
      "Saved\n",
      "Processing: 22, interval 33\n",
      "Saved\n",
      "Processing: 22, interval 34\n",
      "Saved\n",
      "Processing: 22, interval 35\n",
      "Saved\n",
      "Processing: 22, interval 36\n",
      "Saved\n",
      "Processing: 22, interval 37\n",
      "Saved\n",
      "Processing: 22, interval 38\n",
      "Saved\n",
      "Processing: 22, interval 39\n",
      "Saved\n",
      "Processing: 22, interval 40\n",
      "Saved\n",
      "Processing: 22, interval 41\n",
      "Saved\n",
      "Processing: 22, interval 42\n",
      "Saved\n",
      "Processing: 22, interval 43\n",
      "Saved\n",
      "Processing: 22, interval 44\n",
      "Saved\n",
      "Processing: 22, interval 45\n",
      "Saved\n",
      "Processing: 22, interval 46\n",
      " at 22, interval 46. Retrying...\n",
      "Processing: 22, interval 46\n",
      "Saved\n",
      "Processing: 22, interval 47\n",
      " at 22, interval 47. Retrying...\n",
      "Processing: 22, interval 47\n",
      "Saved\n",
      "Processing: 22, interval 48\n",
      "Saved\n",
      "Processing: 22, interval 49\n",
      "Saved\n",
      "Processing: 22, interval 50\n",
      "Saved\n",
      "Processing: 22, interval 51\n",
      "Saved\n",
      "Processing: 22, interval 52\n",
      " at 22, interval 52. Retrying...\n",
      "Processing: 22, interval 52\n",
      "Saved\n",
      "Processing: 22, interval 53\n",
      "Saved\n",
      "Processing: 22, interval 54\n",
      "Saved\n",
      "Processing: 22, interval 55\n",
      "Saved\n",
      "Processing: 22, interval 56\n",
      "Saved\n",
      "Processing: 22, interval 57\n",
      "Saved\n",
      "Processing: 22, interval 58\n",
      "Saved\n",
      "Processing: 22, interval 59\n",
      "Saved\n",
      "Processing: 22, interval 60\n",
      "Saved\n",
      "Processing: 22, interval 61\n",
      "Saved\n",
      "Processing: 22, interval 62\n",
      "Saved\n",
      "Processing: 22, interval 63\n",
      "Saved\n",
      "Processing: 22, interval 64\n",
      "Saved\n",
      "Processing: 22, interval 65\n",
      "Saved\n",
      "Processing: 22, interval 66\n",
      "Saved\n",
      "Processing: 22, interval 67\n",
      "Saved\n",
      "Processing: 22, interval 68\n",
      "Saved\n",
      "Processing: 22, interval 69\n",
      "Saved\n",
      "Processing: 22, interval 70\n",
      "Saved\n",
      "Processing: 22, interval 71\n",
      "Saved\n",
      "Processing: 22, interval 72\n",
      "Saved\n",
      "Processing: 22, interval 73\n",
      "Saved\n",
      "Processing: 22, interval 74\n",
      "Saved\n",
      "Processing: 22, interval 75\n",
      "Saved\n",
      "Processing: 22, interval 76\n",
      "Saved\n",
      "Processing: 22, interval 77\n",
      "Saved\n",
      "Processing: 22, interval 78\n",
      " at 22, interval 78. Retrying...\n",
      "Processing: 22, interval 78\n",
      "Saved\n",
      "Processing: 22, interval 79\n",
      "Saved\n",
      "Processing: 22, interval 80\n",
      "Saved\n",
      "Processing: 22, interval 81\n",
      "Saved\n",
      "Processing: 22, interval 82\n",
      "Saved\n",
      "Processing: 22, interval 83\n",
      "Saved\n",
      "Processing: 22, interval 84\n",
      "Saved\n",
      "Processing: 22, interval 85\n",
      "Saved\n",
      "Processing: 22, interval 86\n",
      "Saved\n",
      "Processing: 22, interval 87\n",
      "Saved\n",
      "Processing: 22, interval 88\n",
      "Saved\n",
      "Processing: 22, interval 89\n",
      "Saved\n",
      "Processing: 22, interval 90\n",
      "Saved\n",
      "Processing: 22, interval 91\n",
      "Saved\n",
      "Processing: 22, interval 92\n",
      "Saved\n",
      "Processing: 22, interval 93\n",
      "Saved\n",
      "Processing: 22, interval 94\n",
      "Saved\n",
      "Processing: 22, interval 95\n",
      "Saved\n",
      "Processing: 22, interval 96\n",
      "Saved\n",
      "Processing: 22, interval 97\n",
      " at 22, interval 97. Retrying...\n",
      "Processing: 22, interval 97\n",
      "Saved\n",
      "Processing: 22, interval 98\n",
      " at 22, interval 98. Retrying...\n",
      "Processing: 22, interval 98\n",
      "Saved\n",
      "Processing: 22, interval 99\n",
      "Saved\n",
      "Processing: 22, interval 100\n",
      "Saved\n",
      "Processing: 22, interval 101\n",
      " at 22, interval 101. Retrying...\n",
      "Processing: 22, interval 101\n",
      "Saved\n",
      "Processing: 22, interval 102\n",
      "Saved\n",
      "Processing: 22, interval 103\n",
      "Saved\n",
      "Processing: 22, interval 104\n",
      "Saved\n",
      "Processing: 22, interval 105\n",
      "Saved\n",
      "Processing: 22, interval 106\n",
      "Saved\n",
      "Processing: 22, interval 107\n",
      "Saved\n",
      "Processing: 22, interval 108\n",
      "Saved\n",
      "Processing: 22, interval 109\n",
      "Saved\n",
      "Processing: 22, interval 110\n",
      "Saved\n",
      "Processing: 22, interval 111\n",
      "Saved\n",
      "Processing: 22, interval 112\n",
      "Saved\n",
      "Processing: 22, interval 113\n",
      "Saved\n",
      "Processing: 22, interval 114\n",
      "Saved\n",
      "Processing: 22, interval 115\n",
      "Saved\n",
      "Processing: 22, interval 116\n",
      "Saved\n",
      "Processing: 22, interval 117\n",
      " at 22, interval 117. Retrying...\n",
      "Processing: 22, interval 117\n",
      "Saved\n",
      "Processing: 22, interval 118\n",
      " at 22, interval 118. Retrying...\n",
      "Processing: 22, interval 118\n",
      "Saved\n",
      "Processing: 22, interval 119\n",
      "Saved\n",
      "Processing: 22, interval 120\n",
      "Saved\n",
      "Processing: 22, interval 121\n",
      "Saved\n",
      "Processing: 22, interval 122\n",
      " at 22, interval 122. Retrying...\n",
      "Processing: 22, interval 122\n",
      "Saved\n",
      "Processing: 22, interval 123\n",
      "Saved\n",
      "Processing: 22, interval 124\n",
      " at 22, interval 124. Retrying...\n",
      "Processing: 22, interval 124\n",
      "Saved\n",
      "Processing: 22, interval 125\n",
      " at 22, interval 125. Retrying...\n",
      "Processing: 22, interval 125\n",
      "Saved\n",
      "Processing: 22, interval 126\n",
      " at 22, interval 126. Retrying...\n",
      "Processing: 22, interval 126\n",
      "Saved\n",
      "Processing: 22, interval 127\n",
      " at 22, interval 127. Retrying...\n",
      "Processing: 22, interval 127\n",
      "Saved\n",
      "Processing: 22, interval 128\n",
      "Saved\n",
      "Processing: 22, interval 129\n",
      "Saved\n",
      "Processing: 22, interval 130\n",
      " at 22, interval 130. Retrying...\n",
      "Processing: 22, interval 130\n",
      "Saved\n",
      "Processing: 22, interval 131\n",
      " at 22, interval 131. Retrying...\n",
      "Processing: 22, interval 131\n",
      "Saved\n",
      "Processing: 22, interval 132\n",
      "Saved\n",
      "Processing: 22, interval 133\n",
      "Saved\n",
      "Processing: 22, interval 134\n",
      "Saved\n",
      "Processing: 22, interval 135\n",
      "Saved\n",
      "Processing: 22, interval 136\n",
      " at 22, interval 136. Retrying...\n",
      "Processing: 22, interval 136\n",
      "Saved\n",
      "Processing: 22, interval 137\n",
      "Saved\n",
      "Processing: 22, interval 138\n",
      "Saved\n",
      "Processing: 22, interval 139\n",
      "Saved\n",
      "Processing: 22, interval 140\n",
      "Saved\n",
      "Processing: 22, interval 141\n",
      " at 22, interval 141. Retrying...\n",
      "Processing: 22, interval 141\n",
      " at 22, interval 141. Retrying...\n",
      "Processing: 22, interval 141\n",
      "Saved\n",
      "Processing: 22, interval 142\n",
      "Saved\n",
      "Processing: 3, interval 0\n",
      "Saved\n",
      "Processing: 3, interval 1\n",
      "Saved\n",
      "Processing: 3, interval 2\n",
      "Saved\n",
      "Processing: 3, interval 3\n",
      "Saved\n",
      "Processing: 3, interval 4\n",
      "Saved\n",
      "Processing: 3, interval 5\n",
      "Saved\n",
      "Processing: 3, interval 6\n",
      "Saved\n",
      "Processing: 3, interval 7\n",
      "Saved\n",
      "Processing: 3, interval 8\n",
      " at 3, interval 8. Retrying...\n",
      "Processing: 3, interval 8\n",
      "Saved\n",
      "Processing: 3, interval 9\n",
      " at 3, interval 9. Retrying...\n",
      "Processing: 3, interval 9\n",
      "Saved\n",
      "Processing: 3, interval 10\n",
      " at 3, interval 10. Retrying...\n",
      "Processing: 3, interval 10\n",
      "Saved\n",
      "Processing: 3, interval 11\n",
      " at 3, interval 11. Retrying...\n",
      "Processing: 3, interval 11\n",
      "Saved\n",
      "Processing: 3, interval 12\n",
      "Saved\n",
      "Processing: 3, interval 13\n",
      " at 3, interval 13. Retrying...\n",
      "Processing: 3, interval 13\n",
      "Saved\n",
      "Processing: 3, interval 14\n",
      "Saved\n",
      "Processing: 3, interval 15\n",
      "Saved\n",
      "Processing: 3, interval 16\n",
      "Saved\n",
      "Processing: 3, interval 17\n",
      "Saved\n",
      "Processing: 3, interval 18\n",
      "Saved\n",
      "Processing: 3, interval 19\n",
      "Saved\n",
      "Processing: 3, interval 20\n",
      "Saved\n",
      "Processing: 3, interval 21\n",
      "Saved\n",
      "Processing: 3, interval 22\n",
      "Saved\n",
      "Processing: 3, interval 23\n",
      "Saved\n",
      "Processing: 3, interval 24\n",
      "Saved\n",
      "Processing: 3, interval 25\n",
      "Saved\n",
      "Processing: 3, interval 26\n",
      "Saved\n",
      "Processing: 3, interval 27\n",
      "Saved\n",
      "Processing: 3, interval 28\n",
      "Saved\n",
      "Processing: 3, interval 29\n",
      "Saved\n",
      "Processing: 3, interval 30\n",
      "Saved\n",
      "Processing: 3, interval 31\n",
      "Saved\n",
      "Processing: 3, interval 32\n",
      "Saved\n",
      "Processing: 3, interval 33\n",
      "Saved\n",
      "Processing: 3, interval 34\n",
      "Saved\n",
      "Processing: 3, interval 35\n",
      " at 3, interval 35. Retrying...\n",
      "Processing: 3, interval 35\n",
      "Saved\n",
      "Processing: 3, interval 36\n",
      "Saved\n",
      "Processing: 3, interval 37\n",
      "Saved\n",
      "Processing: 3, interval 38\n",
      "Saved\n",
      "Processing: 3, interval 39\n",
      " at 3, interval 39. Retrying...\n",
      "Processing: 3, interval 39\n",
      "Saved\n",
      "Processing: 3, interval 40\n",
      "Saved\n",
      "Processing: 3, interval 41\n",
      "Saved\n",
      "Processing: 3, interval 42\n",
      "Saved\n",
      "Processing: 3, interval 43\n",
      "Saved\n",
      "Processing: 3, interval 44\n",
      "Saved\n",
      "Processing: 3, interval 45\n",
      "Saved\n",
      "Processing: 3, interval 46\n",
      "Saved\n",
      "Processing: 3, interval 47\n",
      "Saved\n",
      "Processing: 3, interval 48\n",
      "Saved\n",
      "Processing: 3, interval 49\n",
      "Saved\n",
      "Processing: 3, interval 50\n",
      "Saved\n",
      "Processing: 3, interval 51\n",
      "Saved\n",
      "Processing: 3, interval 52\n",
      " at 3, interval 52. Retrying...\n",
      "Processing: 3, interval 52\n",
      "Saved\n",
      "Processing: 3, interval 53\n",
      "Saved\n",
      "Processing: 3, interval 54\n",
      "Saved\n",
      "Processing: 3, interval 55\n",
      " at 3, interval 55. Retrying...\n",
      "Processing: 3, interval 55\n",
      "Saved\n",
      "Processing: 3, interval 56\n",
      " at 3, interval 56. Retrying...\n",
      "Processing: 3, interval 56\n",
      "Saved\n",
      "Processing: 3, interval 57\n",
      " at 3, interval 57. Retrying...\n",
      "Processing: 3, interval 57\n",
      "Saved\n",
      "Processing: 3, interval 58\n",
      "Saved\n",
      "Processing: 3, interval 59\n",
      "Saved\n",
      "Processing: 3, interval 60\n",
      "Saved\n",
      "Processing: 3, interval 61\n",
      "Saved\n",
      "Processing: 3, interval 62\n",
      "Saved\n",
      "Processing: 3, interval 63\n",
      "Saved\n",
      "Processing: 3, interval 64\n",
      "Saved\n",
      "Processing: 3, interval 65\n",
      " at 3, interval 65. Retrying...\n",
      "Processing: 3, interval 65\n",
      "Saved\n",
      "Processing: 3, interval 66\n",
      "Saved\n",
      "Processing: 3, interval 67\n",
      " at 3, interval 67. Retrying...\n",
      "Processing: 3, interval 67\n",
      "Saved\n",
      "Processing: 3, interval 68\n",
      " at 3, interval 68. Retrying...\n",
      "Processing: 3, interval 68\n",
      "Saved\n",
      "Processing: 3, interval 69\n",
      "Saved\n",
      "Processing: 3, interval 70\n",
      "Saved\n",
      "Processing: 3, interval 71\n",
      "Saved\n",
      "Processing: 3, interval 72\n",
      "Saved\n",
      "Processing: 3, interval 73\n",
      "Saved\n",
      "Processing: 3, interval 74\n",
      "Saved\n",
      "Processing: 3, interval 75\n",
      "Saved\n",
      "Processing: 3, interval 76\n",
      "Saved\n",
      "Processing: 3, interval 77\n",
      "Saved\n",
      "Processing: 3, interval 78\n",
      "Saved\n",
      "Processing: 3, interval 79\n",
      "Saved\n",
      "Processing: 3, interval 80\n",
      " at 3, interval 80. Retrying...\n",
      "Processing: 3, interval 80\n",
      "Saved\n",
      "Processing: 3, interval 81\n",
      " at 3, interval 81. Retrying...\n",
      "Processing: 3, interval 81\n",
      "Saved\n",
      "Processing: 3, interval 82\n",
      "502, message='Bad Gateway', url='https://gnomad.broadinstitute.org/api' at 3, interval 82. Retrying...\n",
      "Processing: 3, interval 82\n",
      " at 3, interval 82. Retrying...\n",
      "Processing: 3, interval 82\n",
      "Saved\n",
      "Processing: 3, interval 83\n",
      " at 3, interval 83. Retrying...\n",
      "Processing: 3, interval 83\n",
      "Saved\n",
      "Processing: 3, interval 84\n",
      "Saved\n",
      "Processing: 3, interval 85\n",
      "Saved\n",
      "Processing: 3, interval 86\n",
      "Saved\n",
      "Processing: 3, interval 87\n",
      "Saved\n",
      "Processing: 3, interval 88\n",
      "Saved\n",
      "Processing: 3, interval 89\n",
      "Saved\n",
      "Processing: 3, interval 90\n",
      "Saved\n",
      "Processing: 3, interval 91\n",
      "Saved\n",
      "Processing: 3, interval 92\n",
      "Saved\n",
      "Processing: 3, interval 93\n",
      "Saved\n",
      "Processing: 3, interval 94\n",
      "Saved\n",
      "Processing: 3, interval 95\n",
      "Saved\n",
      "Processing: 3, interval 96\n",
      "Saved\n",
      "Processing: 3, interval 97\n",
      "Saved\n",
      "Processing: 3, interval 98\n",
      " at 3, interval 98. Retrying...\n",
      "Processing: 3, interval 98\n",
      "Saved\n",
      "Processing: 3, interval 99\n",
      "Saved\n",
      "Processing: 3, interval 100\n",
      "Saved\n",
      "Processing: 3, interval 101\n",
      " at 3, interval 101. Retrying...\n",
      "Processing: 3, interval 101\n",
      " at 3, interval 101. Retrying...\n",
      "Processing: 3, interval 101\n",
      "Saved\n",
      "Processing: 3, interval 102\n",
      "Saved\n",
      "Processing: 3, interval 103\n",
      "Saved\n",
      "Processing: 3, interval 104\n",
      "Saved\n",
      "Processing: 3, interval 105\n",
      "Saved\n",
      "Processing: 3, interval 106\n",
      "Saved\n",
      "Processing: 3, interval 107\n",
      "Saved\n",
      "Processing: 3, interval 108\n",
      "Saved\n",
      "Processing: 3, interval 109\n",
      "Saved\n",
      "Processing: 3, interval 110\n",
      "Saved\n",
      "Processing: 3, interval 111\n",
      "Saved\n",
      "Processing: 3, interval 112\n",
      "Saved\n",
      "Processing: 3, interval 113\n",
      "Saved\n",
      "Processing: 3, interval 114\n",
      "Saved\n",
      "Processing: 3, interval 115\n",
      "Saved\n",
      "Processing: 3, interval 116\n",
      "Saved\n",
      "Processing: 3, interval 117\n",
      "Saved\n",
      "Processing: 3, interval 118\n",
      "Saved\n",
      "Processing: 3, interval 119\n",
      "Saved\n",
      "Processing: 3, interval 120\n",
      "Saved\n",
      "Processing: 3, interval 121\n",
      "Saved\n",
      "Processing: 3, interval 122\n",
      "Saved\n",
      "Processing: 3, interval 123\n",
      "Saved\n",
      "Processing: 3, interval 124\n",
      "Saved\n",
      "Processing: 3, interval 125\n",
      "Saved\n",
      "Processing: 3, interval 126\n",
      "Saved\n",
      "Processing: 3, interval 127\n",
      "Saved\n",
      "Processing: 3, interval 128\n",
      "Saved\n",
      "Processing: 3, interval 129\n",
      "Saved\n",
      "Processing: 3, interval 130\n",
      "Saved\n",
      "Processing: 3, interval 131\n",
      "Saved\n",
      "Processing: 3, interval 132\n",
      "Saved\n",
      "Processing: 3, interval 133\n",
      "Saved\n",
      "Processing: 3, interval 134\n",
      "Saved\n",
      "Processing: 3, interval 135\n",
      "Saved\n",
      "Processing: 3, interval 136\n",
      "Saved\n",
      "Processing: 3, interval 137\n",
      "Saved\n",
      "Processing: 3, interval 138\n",
      "Saved\n",
      "Processing: 3, interval 139\n",
      "Saved\n",
      "Processing: 3, interval 140\n",
      "Saved\n",
      "Processing: 3, interval 141\n",
      "Saved\n",
      "Processing: 3, interval 142\n",
      "Saved\n",
      "Processing: 3, interval 143\n",
      "Saved\n",
      "Processing: 3, interval 144\n",
      "Saved\n",
      "Processing: 3, interval 145\n",
      "Saved\n",
      "Processing: 3, interval 146\n",
      "Saved\n",
      "Processing: 3, interval 147\n",
      "Saved\n",
      "Processing: 3, interval 148\n",
      "Saved\n",
      "Processing: 3, interval 149\n",
      " at 3, interval 149. Retrying...\n",
      "Processing: 3, interval 149\n",
      "Saved\n",
      "Processing: 3, interval 150\n",
      "Saved\n",
      "Processing: 3, interval 151\n",
      "Saved\n",
      "Processing: 3, interval 152\n",
      " at 3, interval 152. Retrying...\n",
      "Processing: 3, interval 152\n",
      "Saved\n",
      "Processing: 3, interval 153\n",
      "{'message': 'Service overloaded'} at 3, interval 153. Retrying...\n",
      "Processing: 3, interval 153\n",
      " at 3, interval 153. Retrying...\n",
      "Processing: 3, interval 153\n",
      "Saved\n",
      "Processing: 3, interval 154\n",
      "Saved\n",
      "Processing: 3, interval 155\n",
      "Saved\n",
      "Processing: 3, interval 156\n",
      "Saved\n",
      "Processing: 3, interval 157\n",
      "Saved\n",
      "Processing: 3, interval 158\n",
      "Saved\n",
      "Processing: 3, interval 159\n",
      "Saved\n",
      "Processing: 3, interval 160\n",
      "Saved\n",
      "Processing: 3, interval 161\n",
      "Saved\n",
      "Processing: 3, interval 162\n",
      " at 3, interval 162. Retrying...\n",
      "Processing: 3, interval 162\n",
      "Saved\n",
      "Processing: 3, interval 163\n",
      "Saved\n",
      "Processing: 3, interval 164\n",
      "Saved\n",
      "Processing: 3, interval 165\n",
      "Saved\n",
      "Processing: 3, interval 166\n",
      "Saved\n",
      "Processing: 3, interval 167\n",
      "Saved\n",
      "Processing: 3, interval 168\n",
      "Saved\n",
      "Processing: 3, interval 169\n",
      "Saved\n",
      "Processing: 3, interval 170\n",
      "Saved\n",
      "Processing: 3, interval 171\n",
      "Saved\n",
      "Processing: 3, interval 172\n",
      "Saved\n",
      "Processing: 3, interval 173\n",
      "Saved\n",
      "Processing: 3, interval 174\n",
      "Saved\n",
      "Processing: 3, interval 175\n",
      "Saved\n",
      "Processing: 3, interval 176\n",
      " at 3, interval 176. Retrying...\n",
      "Processing: 3, interval 176\n",
      "Saved\n",
      "Processing: 3, interval 177\n",
      "Saved\n",
      "Processing: 3, interval 178\n",
      " at 3, interval 178. Retrying...\n",
      "Processing: 3, interval 178\n",
      "Saved\n",
      "Processing: 3, interval 179\n",
      "Saved\n",
      "Processing: 3, interval 180\n",
      " at 3, interval 180. Retrying...\n",
      "Processing: 3, interval 180\n",
      "Saved\n",
      "Processing: 3, interval 181\n",
      "Saved\n",
      "Processing: 3, interval 182\n",
      "Saved\n",
      "Processing: 3, interval 183\n",
      " at 3, interval 183. Retrying...\n",
      "Processing: 3, interval 183\n",
      "Saved\n",
      "Processing: 3, interval 184\n",
      "Saved\n",
      "Processing: 3, interval 185\n",
      "Saved\n",
      "Processing: 3, interval 186\n",
      "Saved\n",
      "Processing: 3, interval 187\n",
      "Saved\n",
      "Processing: 3, interval 188\n",
      "Saved\n",
      "Processing: 3, interval 189\n",
      "Saved\n",
      "Processing: 3, interval 190\n",
      "Saved\n",
      "Processing: 3, interval 191\n",
      "Saved\n",
      "Processing: 3, interval 192\n",
      "Saved\n",
      "Processing: 3, interval 193\n",
      "Saved\n",
      "Processing: 3, interval 194\n",
      "Saved\n",
      "Processing: 3, interval 195\n",
      "Saved\n",
      "Processing: 3, interval 196\n",
      "Saved\n",
      "Processing: 3, interval 197\n",
      " at 3, interval 197. Retrying...\n",
      "Processing: 3, interval 197\n",
      "Saved\n",
      "Processing: 3, interval 198\n",
      "Saved\n",
      "Processing: 3, interval 199\n",
      " at 3, interval 199. Retrying...\n",
      "Processing: 3, interval 199\n",
      "Saved\n",
      "Processing: 3, interval 200\n",
      "Saved\n",
      "Processing: 3, interval 201\n",
      "Saved\n",
      "Processing: 3, interval 202\n",
      "Saved\n",
      "Processing: 3, interval 203\n",
      "Saved\n",
      "Processing: 3, interval 204\n",
      "Saved\n",
      "Processing: 3, interval 205\n",
      "Saved\n",
      "Processing: 3, interval 206\n",
      "Saved\n",
      "Processing: 3, interval 207\n",
      "Saved\n",
      "Processing: 3, interval 208\n",
      "Saved\n",
      "Processing: 3, interval 209\n",
      "Saved\n",
      "Processing: 3, interval 210\n",
      "Saved\n",
      "Processing: 3, interval 211\n",
      "Saved\n",
      "Processing: 3, interval 212\n",
      "Saved\n",
      "Processing: 3, interval 213\n",
      "Saved\n",
      "Processing: 3, interval 214\n",
      "Saved\n",
      "Processing: 3, interval 215\n",
      "Saved\n",
      "Processing: 3, interval 216\n",
      "Saved\n",
      "Processing: 3, interval 217\n",
      "Saved\n",
      "Processing: 3, interval 218\n",
      "Saved\n",
      "Processing: 3, interval 219\n",
      "Saved\n",
      "Processing: 3, interval 220\n",
      "Saved\n",
      "Processing: 3, interval 221\n",
      "Saved\n",
      "Processing: 3, interval 222\n",
      " at 3, interval 222. Retrying...\n",
      "Processing: 3, interval 222\n",
      "Saved\n",
      "Processing: 3, interval 223\n",
      " at 3, interval 223. Retrying...\n",
      "Processing: 3, interval 223\n",
      "Saved\n",
      "Processing: 3, interval 224\n",
      "Saved\n",
      "Processing: 3, interval 225\n",
      "Saved\n",
      "Processing: 3, interval 226\n",
      "Saved\n",
      "Processing: 3, interval 227\n",
      "Saved\n",
      "Processing: 3, interval 228\n",
      "Saved\n",
      "Processing: 3, interval 229\n",
      "Saved\n",
      "Processing: 3, interval 230\n",
      "Saved\n",
      "Processing: 3, interval 231\n",
      "Saved\n",
      "Processing: 3, interval 232\n",
      "Saved\n",
      "Processing: 3, interval 233\n",
      "Saved\n",
      "Processing: 3, interval 234\n",
      "Saved\n",
      "Processing: 3, interval 235\n",
      "Saved\n",
      "Processing: 3, interval 236\n",
      "Saved\n",
      "Processing: 3, interval 237\n",
      "Saved\n",
      "Processing: 3, interval 238\n",
      "Saved\n",
      "Processing: 3, interval 239\n",
      "Saved\n",
      "Processing: 3, interval 240\n",
      "Saved\n",
      "Processing: 3, interval 241\n",
      "Saved\n",
      "Processing: 3, interval 242\n",
      " at 3, interval 242. Retrying...\n",
      "Processing: 3, interval 242\n",
      "Saved\n",
      "Processing: 3, interval 243\n",
      "Saved\n",
      "Processing: 3, interval 244\n",
      "Saved\n",
      "Processing: 3, interval 245\n",
      "Saved\n",
      "Processing: 3, interval 246\n",
      "Saved\n",
      "Processing: 3, interval 247\n",
      "Saved\n",
      "Processing: 3, interval 248\n",
      " at 3, interval 248. Retrying...\n",
      "Processing: 3, interval 248\n",
      "Saved\n",
      "Processing: 3, interval 249\n",
      "Saved\n",
      "Processing: 3, interval 250\n",
      "Saved\n",
      "Processing: 3, interval 251\n",
      "Saved\n",
      "Processing: 3, interval 252\n",
      "Saved\n",
      "Processing: 3, interval 253\n",
      "Saved\n",
      "Processing: 3, interval 254\n",
      "Saved\n",
      "Processing: 3, interval 255\n",
      "Saved\n",
      "Processing: 3, interval 256\n",
      "Saved\n",
      "Processing: 3, interval 257\n",
      "Saved\n",
      "Processing: 3, interval 258\n",
      "Saved\n",
      "Processing: 3, interval 259\n",
      "Saved\n",
      "Processing: 3, interval 260\n",
      "Saved\n",
      "Processing: 3, interval 261\n",
      "Saved\n",
      "Processing: 3, interval 262\n",
      "Saved\n",
      "Processing: 3, interval 263\n",
      "Saved\n",
      "Processing: 3, interval 264\n",
      "Saved\n",
      "Processing: 3, interval 265\n",
      "Saved\n",
      "Processing: 3, interval 266\n",
      "Saved\n",
      "Processing: 3, interval 267\n",
      "Saved\n",
      "Processing: 3, interval 268\n",
      "Saved\n",
      "Processing: 3, interval 269\n",
      "Saved\n",
      "Processing: 3, interval 270\n",
      "Saved\n",
      "Processing: 3, interval 271\n",
      "Saved\n",
      "Processing: 3, interval 272\n",
      "Saved\n",
      "Processing: 3, interval 273\n",
      "Saved\n",
      "Processing: 3, interval 274\n",
      "Saved\n",
      "Processing: 3, interval 275\n",
      "Saved\n",
      "Processing: 3, interval 276\n",
      "Saved\n",
      "Processing: 3, interval 277\n",
      "Saved\n",
      "Processing: 3, interval 278\n",
      "Saved\n",
      "Processing: 3, interval 279\n",
      "Saved\n",
      "Processing: 3, interval 280\n",
      "Saved\n",
      "Processing: 3, interval 281\n",
      "Saved\n",
      "Processing: 3, interval 282\n",
      "Saved\n",
      "Processing: 3, interval 283\n",
      "Saved\n",
      "Processing: 3, interval 284\n",
      "Saved\n",
      "Processing: 3, interval 285\n",
      "Saved\n",
      "Processing: 3, interval 286\n",
      "Saved\n",
      "Processing: 3, interval 287\n",
      "Saved\n",
      "Processing: 3, interval 288\n",
      "Saved\n",
      "Processing: 3, interval 289\n",
      "Saved\n",
      "Processing: 3, interval 290\n",
      "Saved\n",
      "Processing: 3, interval 291\n",
      "Saved\n",
      "Processing: 3, interval 292\n",
      "Saved\n",
      "Processing: 3, interval 293\n",
      "Saved\n",
      "Processing: 3, interval 294\n",
      "Saved\n",
      "Processing: 3, interval 295\n",
      "Saved\n",
      "Processing: 3, interval 296\n",
      "Saved\n",
      "Processing: 3, interval 297\n",
      "Saved\n",
      "Processing: 3, interval 298\n",
      "Saved\n",
      "Processing: 3, interval 299\n",
      "Saved\n",
      "Processing: 3, interval 300\n",
      "Saved\n",
      "Processing: 3, interval 301\n",
      "Saved\n",
      "Processing: 3, interval 302\n",
      "Saved\n",
      "Processing: 3, interval 303\n",
      "Saved\n",
      "Processing: 3, interval 304\n",
      "Saved\n",
      "Processing: 3, interval 305\n",
      "Saved\n",
      "Processing: 3, interval 306\n",
      "Saved\n",
      "Processing: 3, interval 307\n",
      "Saved\n",
      "Processing: 3, interval 308\n",
      "Saved\n",
      "Processing: 3, interval 309\n",
      "Saved\n",
      "Processing: 3, interval 310\n",
      "Saved\n",
      "Processing: 3, interval 311\n",
      "Saved\n",
      "Processing: 4, interval 0\n",
      "Saved\n",
      "Processing: 4, interval 1\n",
      "Saved\n",
      "Processing: 4, interval 2\n",
      "Saved\n",
      "Processing: 4, interval 3\n",
      "Saved\n",
      "Processing: 4, interval 4\n",
      "Saved\n",
      "Processing: 4, interval 5\n",
      "Saved\n",
      "Processing: 4, interval 6\n",
      "Saved\n",
      "Processing: 4, interval 7\n",
      "Saved\n",
      "Processing: 4, interval 8\n",
      "Saved\n",
      "Processing: 4, interval 9\n",
      "Saved\n",
      "Processing: 4, interval 10\n",
      "Saved\n",
      "Processing: 4, interval 11\n",
      "Saved\n",
      "Processing: 4, interval 12\n",
      "Saved\n",
      "Processing: 4, interval 13\n",
      "Saved\n",
      "Processing: 4, interval 14\n",
      "Saved\n",
      "Processing: 4, interval 15\n",
      "Saved\n",
      "Processing: 4, interval 16\n",
      "Saved\n",
      "Processing: 4, interval 17\n",
      "Saved\n",
      "Processing: 4, interval 18\n",
      "Saved\n",
      "Processing: 4, interval 19\n",
      "Saved\n",
      "Processing: 4, interval 20\n",
      "Saved\n",
      "Processing: 4, interval 21\n",
      "Saved\n",
      "Processing: 4, interval 22\n",
      "Saved\n",
      "Processing: 4, interval 23\n",
      "Saved\n",
      "Processing: 4, interval 24\n",
      "Saved\n",
      "Processing: 4, interval 25\n",
      "Saved\n",
      "Processing: 4, interval 26\n",
      "Saved\n",
      "Processing: 4, interval 27\n",
      "Saved\n",
      "Processing: 4, interval 28\n",
      "Saved\n",
      "Processing: 4, interval 29\n",
      "Saved\n",
      "Processing: 4, interval 30\n",
      "Saved\n",
      "Processing: 4, interval 31\n",
      "Saved\n",
      "Processing: 4, interval 32\n",
      "Saved\n",
      "Processing: 4, interval 33\n",
      "Saved\n",
      "Processing: 4, interval 34\n",
      "Saved\n",
      "Processing: 4, interval 35\n",
      "Saved\n",
      "Processing: 4, interval 36\n",
      "Saved\n",
      "Processing: 4, interval 37\n",
      "Saved\n",
      "Processing: 4, interval 38\n",
      "Saved\n",
      "Processing: 4, interval 39\n",
      "Saved\n",
      "Processing: 4, interval 40\n",
      "Saved\n",
      "Processing: 4, interval 41\n",
      "Saved\n",
      "Processing: 4, interval 42\n",
      "Saved\n",
      "Processing: 4, interval 43\n",
      "Saved\n",
      "Processing: 4, interval 44\n",
      "Saved\n",
      "Processing: 4, interval 45\n",
      "Saved\n",
      "Processing: 4, interval 46\n",
      "Saved\n",
      "Processing: 4, interval 47\n",
      "Saved\n",
      "Processing: 4, interval 48\n",
      "Saved\n",
      "Processing: 4, interval 49\n",
      "Saved\n",
      "Processing: 4, interval 50\n",
      "Saved\n",
      "Processing: 4, interval 51\n",
      "Saved\n",
      "Processing: 4, interval 52\n",
      "Saved\n",
      "Processing: 4, interval 53\n",
      "Saved\n",
      "Processing: 4, interval 54\n",
      "Saved\n",
      "Processing: 4, interval 55\n",
      "Saved\n",
      "Processing: 4, interval 56\n",
      "Saved\n",
      "Processing: 4, interval 57\n",
      "Saved\n",
      "Processing: 4, interval 58\n",
      "Saved\n",
      "Processing: 4, interval 59\n",
      "Saved\n",
      "Processing: 4, interval 60\n",
      "Saved\n",
      "Processing: 4, interval 61\n",
      "Saved\n",
      "Processing: 4, interval 62\n",
      "Saved\n",
      "Processing: 4, interval 63\n",
      "Saved\n",
      "Processing: 4, interval 64\n",
      "Saved\n",
      "Processing: 4, interval 65\n",
      "Saved\n",
      "Processing: 4, interval 66\n",
      "Saved\n",
      "Processing: 4, interval 67\n",
      "Saved\n",
      "Processing: 4, interval 68\n",
      "Saved\n",
      "Processing: 4, interval 69\n",
      "Saved\n",
      "Processing: 4, interval 70\n",
      "Saved\n",
      "Processing: 4, interval 71\n",
      "Saved\n",
      "Processing: 4, interval 72\n",
      "Saved\n",
      "Processing: 4, interval 73\n",
      "Saved\n",
      "Processing: 4, interval 74\n",
      "Saved\n",
      "Processing: 4, interval 75\n",
      "Saved\n",
      "Processing: 4, interval 76\n",
      "Saved\n",
      "Processing: 4, interval 77\n",
      "Saved\n",
      "Processing: 4, interval 78\n",
      "Saved\n",
      "Processing: 4, interval 79\n",
      "Saved\n",
      "Processing: 4, interval 80\n",
      "Saved\n",
      "Processing: 4, interval 81\n",
      "Saved\n",
      "Processing: 4, interval 82\n",
      "Saved\n",
      "Processing: 4, interval 83\n",
      "Saved\n",
      "Processing: 4, interval 84\n",
      " at 4, interval 84. Retrying...\n",
      "Processing: 4, interval 84\n",
      "Saved\n",
      "Processing: 4, interval 85\n",
      "Saved\n",
      "Processing: 4, interval 86\n",
      "Saved\n",
      "Processing: 4, interval 87\n",
      "Saved\n",
      "Processing: 4, interval 88\n",
      "Saved\n",
      "Processing: 4, interval 89\n",
      "Saved\n",
      "Processing: 4, interval 90\n",
      "Saved\n",
      "Processing: 4, interval 91\n",
      "Saved\n",
      "Processing: 4, interval 92\n",
      "Saved\n",
      "Processing: 4, interval 93\n",
      "Saved\n",
      "Processing: 4, interval 94\n",
      "Saved\n",
      "Processing: 4, interval 95\n",
      "Saved\n",
      "Processing: 4, interval 96\n",
      "Saved\n",
      "Processing: 4, interval 97\n",
      "Saved\n",
      "Processing: 4, interval 98\n",
      "Saved\n",
      "Processing: 4, interval 99\n",
      "Saved\n",
      "Processing: 4, interval 100\n",
      "Saved\n",
      "Processing: 4, interval 101\n",
      "Saved\n",
      "Processing: 4, interval 102\n",
      "Saved\n",
      "Processing: 4, interval 103\n",
      "Saved\n",
      "Processing: 4, interval 104\n",
      "Saved\n",
      "Processing: 4, interval 105\n",
      "Saved\n",
      "Processing: 4, interval 106\n",
      "502, message='Bad Gateway', url='https://gnomad.broadinstitute.org/api' at 4, interval 106. Retrying...\n",
      "Processing: 4, interval 106\n",
      "Saved\n",
      "Processing: 4, interval 107\n",
      "Saved\n",
      "Processing: 4, interval 108\n",
      "Saved\n",
      "Processing: 4, interval 109\n",
      "Saved\n",
      "Processing: 4, interval 110\n",
      "Saved\n",
      "Processing: 4, interval 111\n",
      "Saved\n",
      "Processing: 4, interval 112\n",
      "Saved\n",
      "Processing: 4, interval 113\n",
      "Saved\n",
      "Processing: 4, interval 114\n",
      "Saved\n",
      "Processing: 4, interval 115\n",
      "Saved\n",
      "Processing: 4, interval 116\n",
      "Saved\n",
      "Processing: 4, interval 117\n",
      "Saved\n",
      "Processing: 4, interval 118\n",
      " at 4, interval 118. Retrying...\n",
      "Processing: 4, interval 118\n",
      " at 4, interval 118. Retrying...\n",
      "Processing: 4, interval 118\n",
      "Saved\n",
      "Processing: 4, interval 119\n",
      " at 4, interval 119. Retrying...\n",
      "Processing: 4, interval 119\n",
      "Saved\n",
      "Processing: 4, interval 120\n",
      "Saved\n",
      "Processing: 4, interval 121\n",
      " at 4, interval 121. Retrying...\n",
      "Processing: 4, interval 121\n",
      "Saved\n",
      "Processing: 4, interval 122\n",
      "Saved\n",
      "Processing: 4, interval 123\n",
      "Saved\n",
      "Processing: 4, interval 124\n",
      "Saved\n",
      "Processing: 4, interval 125\n",
      "Saved\n",
      "Processing: 4, interval 126\n",
      "Saved\n",
      "Processing: 4, interval 127\n",
      "Saved\n",
      "Processing: 4, interval 128\n",
      "Saved\n",
      "Processing: 4, interval 129\n",
      "Saved\n",
      "Processing: 4, interval 130\n",
      "Saved\n",
      "Processing: 4, interval 131\n",
      "Saved\n",
      "Processing: 4, interval 132\n",
      "Saved\n",
      "Processing: 4, interval 133\n",
      "Saved\n",
      "Processing: 4, interval 134\n",
      "Saved\n",
      "Processing: 4, interval 135\n",
      "Saved\n",
      "Processing: 4, interval 136\n",
      "Saved\n",
      "Processing: 4, interval 137\n",
      "Saved\n",
      "Processing: 4, interval 138\n",
      "Saved\n",
      "Processing: 4, interval 139\n",
      "Saved\n",
      "Processing: 4, interval 140\n",
      "Saved\n",
      "Processing: 4, interval 141\n",
      "Saved\n",
      "Processing: 4, interval 142\n",
      "Saved\n",
      "Processing: 4, interval 143\n",
      "Saved\n",
      "Processing: 4, interval 144\n",
      "Saved\n",
      "Processing: 4, interval 145\n",
      "Saved\n",
      "Processing: 4, interval 146\n",
      "Saved\n",
      "Processing: 4, interval 147\n",
      "Saved\n",
      "Processing: 4, interval 148\n",
      "Saved\n",
      "Processing: 4, interval 149\n",
      "Saved\n",
      "Processing: 4, interval 150\n",
      "Saved\n",
      "Processing: 4, interval 151\n",
      "Saved\n",
      "Processing: 4, interval 152\n",
      "Saved\n",
      "Processing: 4, interval 153\n",
      "Saved\n",
      "Processing: 4, interval 154\n",
      "Saved\n",
      "Processing: 4, interval 155\n",
      "Saved\n",
      "Processing: 4, interval 156\n",
      "Saved\n",
      "Processing: 4, interval 157\n",
      "Saved\n",
      "Processing: 4, interval 158\n",
      "Saved\n",
      "Processing: 4, interval 159\n",
      "Saved\n",
      "Processing: 4, interval 160\n",
      "Saved\n",
      "Processing: 4, interval 161\n",
      "Saved\n",
      "Processing: 4, interval 162\n",
      "Saved\n",
      "Processing: 4, interval 163\n",
      "Saved\n",
      "Processing: 4, interval 164\n",
      "Saved\n",
      "Processing: 4, interval 165\n",
      "Saved\n",
      "Processing: 4, interval 166\n",
      "Saved\n",
      "Processing: 4, interval 167\n",
      "Saved\n",
      "Processing: 4, interval 168\n",
      "Saved\n",
      "Processing: 4, interval 169\n",
      "Saved\n",
      "Processing: 4, interval 170\n",
      "Saved\n",
      "Processing: 4, interval 171\n",
      "Saved\n",
      "Processing: 4, interval 172\n",
      " at 4, interval 172. Retrying...\n",
      "Processing: 4, interval 172\n",
      "Saved\n",
      "Processing: 4, interval 173\n",
      " at 4, interval 173. Retrying...\n",
      "Processing: 4, interval 173\n",
      "Saved\n",
      "Processing: 4, interval 174\n",
      "Saved\n",
      "Processing: 4, interval 175\n",
      " at 4, interval 175. Retrying...\n",
      "Processing: 4, interval 175\n",
      "Saved\n",
      "Processing: 4, interval 176\n",
      "Saved\n",
      "Processing: 4, interval 177\n",
      "Saved\n",
      "Processing: 4, interval 178\n",
      " at 4, interval 178. Retrying...\n",
      "Processing: 4, interval 178\n",
      "Saved\n",
      "Processing: 4, interval 179\n",
      "Saved\n",
      "Processing: 4, interval 180\n",
      "Saved\n",
      "Processing: 4, interval 181\n",
      " at 4, interval 181. Retrying...\n",
      "Processing: 4, interval 181\n",
      "Saved\n",
      "Processing: 4, interval 182\n",
      "Saved\n",
      "Processing: 4, interval 183\n",
      " at 4, interval 183. Retrying...\n",
      "Processing: 4, interval 183\n",
      "Saved\n",
      "Processing: 4, interval 184\n",
      "Saved\n",
      "Processing: 4, interval 185\n",
      "Saved\n",
      "Processing: 4, interval 186\n",
      "Saved\n",
      "Processing: 4, interval 187\n",
      "Saved\n",
      "Processing: 4, interval 188\n",
      "Saved\n",
      "Processing: 4, interval 189\n",
      "Saved\n",
      "Processing: 4, interval 190\n",
      "Saved\n",
      "Processing: 4, interval 191\n",
      "Saved\n",
      "Processing: 5, interval 0\n",
      "Saved\n",
      "Processing: 5, interval 1\n",
      "Saved\n",
      "Processing: 5, interval 2\n",
      "Saved\n",
      "Processing: 5, interval 3\n",
      "Saved\n",
      "Processing: 5, interval 4\n",
      "Saved\n",
      "Processing: 5, interval 5\n",
      "Saved\n",
      "Processing: 5, interval 6\n",
      "Saved\n",
      "Processing: 5, interval 7\n",
      "Saved\n",
      "Processing: 5, interval 8\n",
      "Saved\n",
      "Processing: 5, interval 9\n",
      "Saved\n",
      "Processing: 5, interval 10\n",
      "Saved\n",
      "Processing: 5, interval 11\n",
      "Saved\n",
      "Processing: 5, interval 12\n",
      "Saved\n",
      "Processing: 5, interval 13\n",
      "Saved\n",
      "Processing: 5, interval 14\n",
      "Saved\n",
      "Processing: 5, interval 15\n",
      "Saved\n",
      "Processing: 5, interval 16\n",
      "Saved\n",
      "Processing: 5, interval 17\n",
      "Saved\n",
      "Processing: 5, interval 18\n",
      "Saved\n",
      "Processing: 5, interval 19\n",
      "Saved\n",
      "Processing: 5, interval 20\n",
      "Saved\n",
      "Processing: 5, interval 21\n",
      "Saved\n",
      "Processing: 5, interval 22\n",
      "Saved\n",
      "Processing: 5, interval 23\n",
      "Saved\n",
      "Processing: 5, interval 24\n",
      "Saved\n",
      "Processing: 5, interval 25\n",
      "Saved\n",
      "Processing: 5, interval 26\n",
      "Saved\n",
      "Processing: 5, interval 27\n",
      "Saved\n",
      "Processing: 5, interval 28\n",
      "Saved\n",
      "Processing: 5, interval 29\n",
      "Saved\n",
      "Processing: 5, interval 30\n",
      "Saved\n",
      "Processing: 5, interval 31\n",
      "Saved\n",
      "Processing: 5, interval 32\n",
      "Saved\n",
      "Processing: 5, interval 33\n",
      "Saved\n",
      "Processing: 5, interval 34\n",
      "Saved\n",
      "Processing: 5, interval 35\n",
      "Saved\n",
      "Processing: 5, interval 36\n",
      "Saved\n",
      "Processing: 5, interval 37\n",
      "Saved\n",
      "Processing: 5, interval 38\n",
      "Saved\n",
      "Processing: 5, interval 39\n",
      "Saved\n",
      "Processing: 5, interval 40\n",
      "Saved\n",
      "Processing: 5, interval 41\n",
      "Saved\n",
      "Processing: 5, interval 42\n",
      "Saved\n",
      "Processing: 5, interval 43\n",
      "Saved\n",
      "Processing: 5, interval 44\n",
      "Saved\n",
      "Processing: 5, interval 45\n",
      "Saved\n",
      "Processing: 5, interval 46\n",
      "Saved\n",
      "Processing: 5, interval 47\n",
      "Saved\n",
      "Processing: 5, interval 48\n",
      "Saved\n",
      "Processing: 5, interval 49\n",
      "Saved\n",
      "Processing: 5, interval 50\n",
      "Saved\n",
      "Processing: 5, interval 51\n",
      "Saved\n",
      "Processing: 5, interval 52\n",
      "Saved\n",
      "Processing: 5, interval 53\n",
      "Saved\n",
      "Processing: 5, interval 54\n",
      "Saved\n",
      "Processing: 5, interval 55\n",
      "Saved\n",
      "Processing: 5, interval 56\n",
      "Saved\n",
      "Processing: 5, interval 57\n",
      "Saved\n",
      "Processing: 5, interval 58\n",
      "Saved\n",
      "Processing: 5, interval 59\n",
      "Saved\n",
      "Processing: 5, interval 60\n",
      "Saved\n",
      "Processing: 5, interval 61\n",
      "Saved\n",
      "Processing: 5, interval 62\n",
      "Saved\n",
      "Processing: 5, interval 63\n",
      " at 5, interval 63. Retrying...\n",
      "Processing: 5, interval 63\n",
      "Saved\n",
      "Processing: 5, interval 64\n",
      "Saved\n",
      "Processing: 5, interval 65\n",
      "Saved\n",
      "Processing: 5, interval 66\n",
      "Saved\n",
      "Processing: 5, interval 67\n",
      "Saved\n",
      "Processing: 5, interval 68\n",
      "Saved\n",
      "Processing: 5, interval 69\n",
      "Saved\n",
      "Processing: 5, interval 70\n",
      "Saved\n",
      "Processing: 5, interval 71\n",
      "Saved\n",
      "Processing: 5, interval 72\n",
      "Saved\n",
      "Processing: 5, interval 73\n",
      "Saved\n",
      "Processing: 5, interval 74\n",
      " at 5, interval 74. Retrying...\n",
      "Processing: 5, interval 74\n",
      "Saved\n",
      "Processing: 5, interval 75\n",
      "Saved\n",
      "Processing: 5, interval 76\n",
      "Saved\n",
      "Processing: 5, interval 77\n",
      "Saved\n",
      "Processing: 5, interval 78\n",
      "Saved\n",
      "Processing: 5, interval 79\n",
      "Saved\n",
      "Processing: 5, interval 80\n",
      "Saved\n",
      "Processing: 5, interval 81\n",
      "Saved\n",
      "Processing: 5, interval 82\n",
      "Saved\n",
      "Processing: 5, interval 83\n",
      "Saved\n",
      "Processing: 5, interval 84\n",
      "Saved\n",
      "Processing: 5, interval 85\n",
      "Saved\n",
      "Processing: 5, interval 86\n",
      "Saved\n",
      "Processing: 5, interval 87\n",
      "Saved\n",
      "Processing: 5, interval 88\n",
      "Saved\n",
      "Processing: 5, interval 89\n",
      "Saved\n",
      "Processing: 5, interval 90\n",
      "Saved\n",
      "Processing: 5, interval 91\n",
      "Saved\n",
      "Processing: 5, interval 92\n",
      "Saved\n",
      "Processing: 5, interval 93\n",
      "Saved\n",
      "Processing: 5, interval 94\n",
      "Saved\n",
      "Processing: 5, interval 95\n",
      "Saved\n",
      "Processing: 5, interval 96\n",
      "Saved\n",
      "Processing: 5, interval 97\n",
      "Saved\n",
      "Processing: 5, interval 98\n",
      "Saved\n",
      "Processing: 5, interval 99\n",
      "Saved\n",
      "Processing: 5, interval 100\n",
      "Saved\n",
      "Processing: 5, interval 101\n",
      "Saved\n",
      "Processing: 5, interval 102\n",
      "Saved\n",
      "Processing: 5, interval 103\n",
      "Saved\n",
      "Processing: 5, interval 104\n",
      "Saved\n",
      "Processing: 5, interval 105\n",
      "Saved\n",
      "Processing: 5, interval 106\n",
      "Saved\n",
      "Processing: 5, interval 107\n",
      "Saved\n",
      "Processing: 5, interval 108\n",
      "Saved\n",
      "Processing: 5, interval 109\n",
      "Saved\n",
      "Processing: 5, interval 110\n",
      "Saved\n",
      "Processing: 5, interval 111\n",
      "Saved\n",
      "Processing: 5, interval 112\n",
      "Saved\n",
      "Processing: 5, interval 113\n",
      "Saved\n",
      "Processing: 5, interval 114\n",
      "Saved\n",
      "Processing: 5, interval 115\n",
      "Saved\n",
      "Processing: 5, interval 116\n",
      "Saved\n",
      "Processing: 5, interval 117\n",
      "Saved\n",
      "Processing: 5, interval 118\n",
      "Saved\n",
      "Processing: 5, interval 119\n",
      "Saved\n",
      "Processing: 5, interval 120\n",
      "Saved\n",
      "Processing: 5, interval 121\n",
      "Saved\n",
      "Processing: 5, interval 122\n",
      "Saved\n",
      "Processing: 5, interval 123\n",
      "Saved\n",
      "Processing: 5, interval 124\n",
      "Saved\n",
      "Processing: 5, interval 125\n",
      "Saved\n",
      "Processing: 5, interval 126\n",
      "Saved\n",
      "Processing: 5, interval 127\n",
      "Saved\n",
      "Processing: 5, interval 128\n",
      "Saved\n",
      "Processing: 5, interval 129\n",
      "Saved\n",
      "Processing: 5, interval 130\n",
      "Saved\n",
      "Processing: 5, interval 131\n",
      "Saved\n",
      "Processing: 5, interval 132\n",
      "Saved\n",
      "Processing: 5, interval 133\n",
      "Saved\n",
      "Processing: 5, interval 134\n",
      "Saved\n",
      "Processing: 5, interval 135\n",
      "Saved\n",
      "Processing: 5, interval 136\n",
      "Saved\n",
      "Processing: 5, interval 137\n",
      "Saved\n",
      "Processing: 5, interval 138\n",
      "Saved\n",
      "Processing: 5, interval 139\n",
      "Saved\n",
      "Processing: 5, interval 140\n",
      "Saved\n",
      "Processing: 5, interval 141\n",
      "Saved\n",
      "Processing: 5, interval 142\n",
      "Saved\n",
      "Processing: 5, interval 143\n",
      "Saved\n",
      "Processing: 5, interval 144\n",
      "Saved\n",
      "Processing: 5, interval 145\n",
      "Saved\n",
      "Processing: 5, interval 146\n",
      "Saved\n",
      "Processing: 5, interval 147\n",
      "Saved\n",
      "Processing: 5, interval 148\n",
      "Saved\n",
      "Processing: 5, interval 149\n",
      "Saved\n",
      "Processing: 5, interval 150\n",
      "Saved\n",
      "Processing: 5, interval 151\n",
      "Saved\n",
      "Processing: 5, interval 152\n",
      "Saved\n",
      "Processing: 5, interval 153\n",
      "Saved\n",
      "Processing: 5, interval 154\n",
      "Saved\n",
      "Processing: 5, interval 155\n",
      "Saved\n",
      "Processing: 5, interval 156\n",
      "Saved\n",
      "Processing: 5, interval 157\n",
      "Saved\n",
      "Processing: 5, interval 158\n",
      "Saved\n",
      "Processing: 5, interval 159\n",
      "Saved\n",
      "Processing: 5, interval 160\n",
      "Saved\n",
      "Processing: 5, interval 161\n",
      "Saved\n",
      "Processing: 5, interval 162\n",
      "Saved\n",
      "Processing: 5, interval 163\n",
      "Saved\n",
      "Processing: 5, interval 164\n",
      "Saved\n",
      "Processing: 5, interval 165\n",
      "Saved\n",
      "Processing: 5, interval 166\n",
      "Saved\n",
      "Processing: 5, interval 167\n",
      "Saved\n",
      "Processing: 5, interval 168\n",
      "Saved\n",
      "Processing: 5, interval 169\n",
      "Saved\n",
      "Processing: 5, interval 170\n",
      "Saved\n",
      "Processing: 5, interval 171\n",
      "Saved\n",
      "Processing: 5, interval 172\n",
      "Saved\n",
      "Processing: 5, interval 173\n",
      "Saved\n",
      "Processing: 5, interval 174\n",
      "Saved\n",
      "Processing: 5, interval 175\n",
      "Saved\n",
      "Processing: 5, interval 176\n",
      "Saved\n",
      "Processing: 5, interval 177\n",
      "Saved\n",
      "Processing: 5, interval 178\n",
      " at 5, interval 178. Retrying...\n",
      "Processing: 5, interval 178\n",
      "Saved\n",
      "Processing: 5, interval 179\n",
      "Saved\n",
      "Processing: 5, interval 180\n",
      "Saved\n",
      "Processing: 5, interval 181\n",
      " at 5, interval 181. Retrying...\n",
      "Processing: 5, interval 181\n",
      "Saved\n",
      "Processing: 5, interval 182\n",
      " at 5, interval 182. Retrying...\n",
      "Processing: 5, interval 182\n",
      "Saved\n",
      "Processing: 5, interval 183\n",
      " at 5, interval 183. Retrying...\n",
      "Processing: 5, interval 183\n",
      "Saved\n",
      "Processing: 5, interval 184\n",
      " at 5, interval 184. Retrying...\n",
      "Processing: 5, interval 184\n",
      "Saved\n",
      "Processing: 5, interval 185\n",
      "Saved\n",
      "Processing: 5, interval 186\n",
      "Saved\n",
      "Processing: 5, interval 187\n",
      "Saved\n",
      "Processing: 5, interval 188\n",
      "Saved\n",
      "Processing: 5, interval 189\n",
      "Saved\n",
      "Processing: 5, interval 190\n",
      "Saved\n",
      "Processing: 5, interval 191\n",
      "Saved\n",
      "Processing: 5, interval 192\n",
      "Saved\n",
      "Processing: 5, interval 193\n",
      "Saved\n",
      "Processing: 5, interval 194\n",
      "Saved\n",
      "Processing: 5, interval 195\n",
      " at 5, interval 195. Retrying...\n",
      "Processing: 5, interval 195\n",
      "Saved\n",
      "Processing: 5, interval 196\n",
      "Saved\n",
      "Processing: 5, interval 197\n",
      "Saved\n",
      "Processing: 5, interval 198\n",
      "Saved\n",
      "Processing: 5, interval 199\n",
      "Saved\n",
      "Processing: 5, interval 200\n",
      "Saved\n",
      "Processing: 5, interval 201\n",
      "Saved\n",
      "Processing: 5, interval 202\n",
      "Saved\n",
      "Processing: 5, interval 203\n",
      " at 5, interval 203. Retrying...\n",
      "Processing: 5, interval 203\n",
      "Saved\n",
      "Processing: 5, interval 204\n",
      "Saved\n",
      "Processing: 5, interval 205\n",
      "Saved\n",
      "Processing: 5, interval 206\n",
      "Saved\n",
      "Processing: 5, interval 207\n",
      "Saved\n",
      "Processing: 5, interval 208\n",
      "Saved\n",
      "Processing: 5, interval 209\n",
      "Saved\n",
      "Processing: 5, interval 210\n",
      "Saved\n",
      "Processing: 5, interval 211\n",
      "Saved\n",
      "Processing: 5, interval 212\n",
      "Saved\n",
      "Processing: 5, interval 213\n",
      "Saved\n",
      "Processing: 5, interval 214\n",
      "Saved\n",
      "Processing: 5, interval 215\n",
      "Saved\n",
      "Processing: 5, interval 216\n",
      "Saved\n",
      "Processing: 5, interval 217\n",
      "Saved\n",
      "Processing: 5, interval 218\n",
      "Saved\n",
      "Processing: 5, interval 219\n",
      "Saved\n",
      "Processing: 5, interval 220\n",
      "Saved\n",
      "Processing: 5, interval 221\n",
      "Saved\n",
      "Processing: 5, interval 222\n",
      "Saved\n",
      "Processing: 5, interval 223\n",
      " at 5, interval 223. Retrying...\n",
      "Processing: 5, interval 223\n",
      "Saved\n",
      "Processing: 5, interval 224\n",
      "Saved\n",
      "Processing: 5, interval 225\n",
      "Saved\n",
      "Processing: 5, interval 226\n",
      " at 5, interval 226. Retrying...\n",
      "Processing: 5, interval 226\n",
      "Saved\n",
      "Processing: 5, interval 227\n",
      " at 5, interval 227. Retrying...\n",
      "Processing: 5, interval 227\n",
      "Saved\n",
      "Processing: 5, interval 228\n",
      "Saved\n",
      "Processing: 5, interval 229\n",
      "Saved\n",
      "Processing: 5, interval 230\n",
      "Saved\n",
      "Processing: 5, interval 231\n",
      "Saved\n",
      "Processing: 5, interval 232\n",
      " at 5, interval 232. Retrying...\n",
      "Processing: 5, interval 232\n",
      "Saved\n",
      "Processing: 5, interval 233\n",
      "Saved\n",
      "Processing: 5, interval 234\n",
      "Saved\n",
      "Processing: 5, interval 235\n",
      "Saved\n",
      "Processing: 5, interval 236\n",
      "Saved\n",
      "Processing: 5, interval 237\n",
      "Saved\n",
      "Processing: 5, interval 238\n",
      "Saved\n",
      "Processing: 5, interval 239\n",
      " at 5, interval 239. Retrying...\n",
      "Processing: 5, interval 239\n",
      "Saved\n",
      "Processing: 5, interval 240\n",
      "Saved\n",
      "Processing: 5, interval 241\n",
      "Saved\n",
      "Processing: 5, interval 242\n",
      "Saved\n",
      "Processing: 5, interval 243\n",
      " at 5, interval 243. Retrying...\n",
      "Processing: 5, interval 243\n",
      "Saved\n",
      "Processing: 5, interval 244\n",
      "Saved\n",
      "Processing: 5, interval 245\n",
      "Saved\n",
      "Processing: 5, interval 246\n",
      "Saved\n",
      "Processing: 5, interval 247\n",
      " at 5, interval 247. Retrying...\n",
      "Processing: 5, interval 247\n",
      "Saved\n",
      "Processing: 5, interval 248\n",
      "Saved\n",
      "Processing: 5, interval 249\n",
      " at 5, interval 249. Retrying...\n",
      "Processing: 5, interval 249\n",
      "Saved\n",
      "Processing: 5, interval 250\n",
      "Saved\n",
      "Processing: 5, interval 251\n",
      "Saved\n",
      "Processing: 5, interval 252\n",
      "Saved\n",
      "Processing: 5, interval 253\n",
      "Saved\n",
      "Processing: 5, interval 254\n",
      " at 5, interval 254. Retrying...\n",
      "Processing: 5, interval 254\n",
      "Saved\n",
      "Processing: 5, interval 255\n",
      "Saved\n",
      "Processing: 5, interval 256\n",
      "Saved\n",
      "Processing: 5, interval 257\n",
      "Saved\n",
      "Processing: 5, interval 258\n",
      "Saved\n",
      "Processing: 5, interval 259\n",
      "Saved\n",
      "Processing: 5, interval 260\n",
      "Saved\n",
      "Processing: 5, interval 261\n",
      "Saved\n",
      "Processing: 5, interval 262\n",
      "Saved\n",
      "Processing: 5, interval 263\n",
      "Saved\n",
      "Processing: 5, interval 264\n",
      "Saved\n",
      "Processing: 5, interval 265\n",
      "Saved\n",
      "Processing: 5, interval 266\n",
      "Saved\n",
      "Processing: 5, interval 267\n",
      "Saved\n",
      "Processing: 5, interval 268\n",
      "Saved\n",
      "Processing: 5, interval 269\n",
      " at 5, interval 269. Retrying...\n",
      "Processing: 5, interval 269\n",
      "Saved\n",
      "Processing: 5, interval 270\n",
      "Saved\n",
      "Processing: 5, interval 271\n",
      "Saved\n",
      "Processing: 5, interval 272\n",
      " at 5, interval 272. Retrying...\n",
      "Processing: 5, interval 272\n",
      "Saved\n",
      "Processing: 5, interval 273\n",
      "Saved\n",
      "Processing: 5, interval 274\n",
      "Saved\n",
      "Processing: 5, interval 275\n",
      "Saved\n",
      "Processing: 5, interval 276\n",
      "Saved\n",
      "Processing: 5, interval 277\n",
      "Saved\n",
      "Processing: 5, interval 278\n",
      "Saved\n",
      "Processing: 5, interval 279\n",
      "Saved\n",
      "Processing: 5, interval 280\n",
      "Saved\n",
      "Processing: 5, interval 281\n",
      "Saved\n",
      "Processing: 5, interval 282\n",
      "Saved\n",
      "Processing: 5, interval 283\n",
      "Saved\n",
      "Processing: 5, interval 284\n",
      "Saved\n",
      "Processing: 5, interval 285\n",
      "Saved\n",
      "Processing: 5, interval 286\n",
      "Saved\n",
      "Processing: 5, interval 287\n",
      "Saved\n",
      "Processing: 5, interval 288\n",
      "Saved\n",
      "Processing: 5, interval 289\n",
      " at 5, interval 289. Retrying...\n",
      "Processing: 5, interval 289\n",
      "Saved\n",
      "Processing: 5, interval 290\n",
      " at 5, interval 290. Retrying...\n",
      "Processing: 5, interval 290\n",
      " at 5, interval 290. Retrying...\n",
      "Processing: 5, interval 290\n",
      "Saved\n",
      "Processing: 5, interval 291\n",
      "{'message': 'Service overloaded'} at 5, interval 291. Retrying...\n",
      "Processing: 5, interval 291\n",
      " at 5, interval 291. Retrying...\n",
      "Processing: 5, interval 291\n",
      "Saved\n",
      "Processing: 5, interval 292\n",
      "Saved\n",
      "Processing: 5, interval 293\n",
      "Saved\n",
      "Processing: 5, interval 294\n",
      "Saved\n",
      "Processing: 5, interval 295\n",
      " at 5, interval 295. Retrying...\n",
      "Processing: 5, interval 295\n",
      " at 5, interval 295. Retrying...\n",
      "Processing: 5, interval 295\n",
      "Saved\n",
      "Processing: 5, interval 296\n",
      "Saved\n",
      "Processing: 5, interval 297\n",
      " at 5, interval 297. Retrying...\n",
      "Processing: 5, interval 297\n",
      "Saved\n",
      "Processing: 5, interval 298\n",
      "Saved\n",
      "Processing: 5, interval 299\n",
      " at 5, interval 299. Retrying...\n",
      "Processing: 5, interval 299\n",
      "Saved\n",
      "Processing: 5, interval 300\n",
      "Saved\n",
      "Processing: 5, interval 301\n",
      "Saved\n",
      "Processing: 5, interval 302\n",
      "Saved\n",
      "Processing: 5, interval 303\n",
      "Saved\n",
      "Processing: 5, interval 304\n",
      "Saved\n",
      "Processing: 5, interval 305\n",
      "Saved\n",
      "Processing: 5, interval 306\n",
      "Saved\n",
      "Processing: 5, interval 307\n",
      "Saved\n",
      "Processing: 5, interval 308\n",
      "Saved\n",
      "Processing: 5, interval 309\n",
      "Saved\n",
      "Processing: 5, interval 310\n",
      "Saved\n",
      "Processing: 5, interval 311\n",
      "Saved\n",
      "Processing: 5, interval 312\n",
      "Saved\n",
      "Processing: 5, interval 313\n",
      "Saved\n",
      "Processing: 5, interval 314\n",
      "Saved\n",
      "Processing: 5, interval 315\n",
      " at 5, interval 315. Retrying...\n",
      "Processing: 5, interval 315\n",
      "Saved\n",
      "Processing: 5, interval 316\n",
      "Saved\n",
      "Processing: 5, interval 317\n",
      "Saved\n",
      "Processing: 5, interval 318\n",
      "Saved\n",
      "Processing: 5, interval 319\n",
      " at 5, interval 319. Retrying...\n",
      "Processing: 5, interval 319\n",
      "Saved\n",
      "Processing: 5, interval 320\n",
      " at 5, interval 320. Retrying...\n",
      "Processing: 5, interval 320\n",
      " at 5, interval 320. Retrying...\n",
      "Processing: 5, interval 320\n",
      "Saved\n",
      "Processing: 5, interval 321\n",
      "Saved\n",
      "Processing: 5, interval 322\n",
      "Saved\n",
      "Processing: 5, interval 323\n",
      "Saved\n",
      "Processing: 5, interval 324\n",
      "Saved\n",
      "Processing: 5, interval 325\n",
      "Saved\n",
      "Processing: 5, interval 326\n",
      "Saved\n",
      "Processing: 6, interval 0\n",
      " at 6, interval 0. Retrying...\n",
      "Processing: 6, interval 0\n",
      "Saved\n",
      "Processing: 6, interval 1\n",
      " at 6, interval 1. Retrying...\n",
      "Processing: 6, interval 1\n",
      "Saved\n",
      "Processing: 6, interval 2\n",
      "Saved\n",
      "Processing: 6, interval 3\n",
      "Saved\n",
      "Processing: 6, interval 4\n",
      "Saved\n",
      "Processing: 6, interval 5\n",
      "Saved\n",
      "Processing: 6, interval 6\n",
      "Saved\n",
      "Processing: 6, interval 7\n",
      "Saved\n",
      "Processing: 6, interval 8\n",
      "Saved\n",
      "Processing: 6, interval 9\n",
      "Saved\n",
      "Processing: 6, interval 10\n",
      "Saved\n",
      "Processing: 6, interval 11\n",
      "Saved\n",
      "Processing: 6, interval 12\n",
      "Saved\n",
      "Processing: 6, interval 13\n",
      "Saved\n",
      "Processing: 6, interval 14\n",
      "Saved\n",
      "Processing: 6, interval 15\n",
      "Saved\n",
      "Processing: 6, interval 16\n",
      "Saved\n",
      "Processing: 6, interval 17\n",
      "Saved\n",
      "Processing: 6, interval 18\n",
      "Saved\n",
      "Processing: 6, interval 19\n",
      "Saved\n",
      "Processing: 6, interval 20\n",
      "Saved\n",
      "Processing: 6, interval 21\n",
      "Saved\n",
      "Processing: 6, interval 22\n",
      "Saved\n",
      "Processing: 6, interval 23\n",
      "Saved\n",
      "Processing: 6, interval 24\n",
      "Saved\n",
      "Processing: 6, interval 25\n",
      "Saved\n",
      "Processing: 6, interval 26\n",
      "Saved\n",
      "Processing: 6, interval 27\n",
      "Saved\n",
      "Processing: 6, interval 28\n",
      "Saved\n",
      "Processing: 6, interval 29\n",
      "Saved\n",
      "Processing: 6, interval 30\n",
      "Saved\n",
      "Processing: 6, interval 31\n",
      "Saved\n",
      "Processing: 6, interval 32\n",
      "Saved\n",
      "Processing: 6, interval 33\n",
      "Saved\n",
      "Processing: 6, interval 34\n",
      "Saved\n",
      "Processing: 6, interval 35\n",
      "Saved\n",
      "Processing: 6, interval 36\n",
      "Saved\n",
      "Processing: 6, interval 37\n",
      "Saved\n",
      "Processing: 6, interval 38\n",
      "Saved\n",
      "Processing: 6, interval 39\n",
      "Saved\n",
      "Processing: 6, interval 40\n",
      "Saved\n",
      "Processing: 6, interval 41\n",
      "Saved\n",
      "Processing: 6, interval 42\n",
      "Saved\n",
      "Processing: 6, interval 43\n",
      "Saved\n",
      "Processing: 6, interval 44\n",
      "Saved\n",
      "Processing: 6, interval 45\n",
      "Saved\n",
      "Processing: 6, interval 46\n",
      "Saved\n",
      "Processing: 6, interval 47\n",
      "Saved\n",
      "Processing: 6, interval 48\n",
      "Saved\n",
      "Processing: 6, interval 49\n",
      "Saved\n",
      "Processing: 6, interval 50\n",
      "Saved\n",
      "Processing: 6, interval 51\n",
      "Saved\n",
      "Processing: 6, interval 52\n",
      "Saved\n",
      "Processing: 6, interval 53\n",
      "Saved\n",
      "Processing: 6, interval 54\n",
      "Saved\n",
      "Processing: 6, interval 55\n",
      "Saved\n",
      "Processing: 6, interval 56\n",
      "Saved\n",
      "Processing: 6, interval 57\n",
      "Saved\n",
      "Processing: 6, interval 58\n",
      "Saved\n",
      "Processing: 6, interval 59\n",
      "Saved\n",
      "Processing: 6, interval 60\n",
      "Saved\n",
      "Processing: 6, interval 61\n",
      "Saved\n",
      "Processing: 6, interval 62\n",
      "Saved\n",
      "Processing: 6, interval 63\n",
      "Saved\n",
      "Processing: 6, interval 64\n",
      "Saved\n",
      "Processing: 6, interval 65\n",
      "Saved\n",
      "Processing: 6, interval 66\n",
      "Saved\n",
      "Processing: 6, interval 67\n",
      "Saved\n",
      "Processing: 6, interval 68\n",
      "Saved\n",
      "Processing: 6, interval 69\n",
      "Saved\n",
      "Processing: 6, interval 70\n",
      "Saved\n",
      "Processing: 6, interval 71\n",
      "Saved\n",
      "Processing: 6, interval 72\n",
      "Saved\n",
      "Processing: 6, interval 73\n",
      "Saved\n",
      "Processing: 6, interval 74\n",
      "Saved\n",
      "Processing: 6, interval 75\n",
      "Saved\n",
      "Processing: 6, interval 76\n",
      "Saved\n",
      "Processing: 6, interval 77\n",
      "Saved\n",
      "Processing: 6, interval 78\n",
      "Saved\n",
      "Processing: 6, interval 79\n",
      "Saved\n",
      "Processing: 6, interval 80\n",
      "Saved\n",
      "Processing: 6, interval 81\n",
      "Saved\n",
      "Processing: 6, interval 82\n",
      "Saved\n",
      "Processing: 6, interval 83\n",
      "Saved\n",
      "Processing: 6, interval 84\n",
      "Saved\n",
      "Processing: 6, interval 85\n",
      "Saved\n",
      "Processing: 6, interval 86\n",
      "Saved\n",
      "Processing: 6, interval 87\n",
      "Saved\n",
      "Processing: 6, interval 88\n",
      "Saved\n",
      "Processing: 6, interval 89\n",
      "Saved\n",
      "Processing: 6, interval 90\n",
      "Saved\n",
      "Processing: 6, interval 91\n",
      "Saved\n",
      "Processing: 6, interval 92\n",
      "Saved\n",
      "Processing: 6, interval 93\n",
      "Saved\n",
      "Processing: 6, interval 94\n",
      "Saved\n",
      "Processing: 6, interval 95\n",
      "Saved\n",
      "Processing: 6, interval 96\n",
      "Saved\n",
      "Processing: 6, interval 97\n",
      "Saved\n",
      "Processing: 6, interval 98\n",
      "Saved\n",
      "Processing: 6, interval 99\n",
      "Saved\n",
      "Processing: 6, interval 100\n",
      "Saved\n",
      "Processing: 6, interval 101\n",
      "Saved\n",
      "Processing: 6, interval 102\n",
      "Saved\n",
      "Processing: 6, interval 103\n",
      "Saved\n",
      "Processing: 6, interval 104\n",
      "Saved\n",
      "Processing: 6, interval 105\n",
      "Saved\n",
      "Processing: 6, interval 106\n",
      "Saved\n",
      "Processing: 6, interval 107\n",
      "Saved\n",
      "Processing: 6, interval 108\n",
      "Saved\n",
      "Processing: 6, interval 109\n",
      " at 6, interval 109. Retrying...\n",
      "Processing: 6, interval 109\n",
      "Saved\n",
      "Processing: 6, interval 110\n",
      " at 6, interval 110. Retrying...\n",
      "Processing: 6, interval 110\n",
      "Saved\n",
      "Processing: 6, interval 111\n",
      "Saved\n",
      "Processing: 6, interval 112\n",
      "Saved\n",
      "Processing: 6, interval 113\n",
      " at 6, interval 113. Retrying...\n",
      "Processing: 6, interval 113\n",
      "Saved\n",
      "Processing: 6, interval 114\n",
      "Saved\n",
      "Processing: 6, interval 115\n",
      "Saved\n",
      "Processing: 6, interval 116\n",
      "Saved\n",
      "Processing: 6, interval 117\n",
      "Saved\n",
      "Processing: 6, interval 118\n",
      "Saved\n",
      "Processing: 6, interval 119\n",
      "Saved\n",
      "Processing: 6, interval 120\n",
      "Saved\n",
      "Processing: 6, interval 121\n",
      "Saved\n",
      "Processing: 6, interval 122\n",
      "Saved\n",
      "Processing: 6, interval 123\n",
      "Saved\n",
      "Processing: 6, interval 124\n",
      "Saved\n",
      "Processing: 6, interval 125\n",
      "Saved\n",
      "Processing: 6, interval 126\n",
      "Saved\n",
      "Processing: 6, interval 127\n",
      "Saved\n",
      "Processing: 6, interval 128\n",
      "Saved\n",
      "Processing: 6, interval 129\n",
      "Saved\n",
      "Processing: 6, interval 130\n",
      "Saved\n",
      "Processing: 6, interval 131\n",
      "Saved\n",
      "Processing: 6, interval 132\n",
      "Saved\n",
      "Processing: 6, interval 133\n",
      "Saved\n",
      "Processing: 6, interval 134\n",
      "Saved\n",
      "Processing: 6, interval 135\n",
      "Saved\n",
      "Processing: 6, interval 136\n",
      "Saved\n",
      "Processing: 6, interval 137\n",
      "Saved\n",
      "Processing: 6, interval 138\n",
      "Saved\n",
      "Processing: 6, interval 139\n",
      "Saved\n",
      "Processing: 6, interval 140\n",
      "Saved\n",
      "Processing: 6, interval 141\n",
      "Saved\n",
      "Processing: 6, interval 142\n",
      "Saved\n",
      "Processing: 6, interval 143\n",
      "Saved\n",
      "Processing: 6, interval 144\n",
      "Saved\n",
      "Processing: 6, interval 145\n",
      "Saved\n",
      "Processing: 6, interval 146\n",
      "Saved\n",
      "Processing: 6, interval 147\n",
      "Saved\n",
      "Processing: 6, interval 148\n",
      "Saved\n",
      "Processing: 6, interval 149\n",
      "Saved\n",
      "Processing: 6, interval 150\n",
      "Saved\n",
      "Processing: 6, interval 151\n",
      "Saved\n",
      "Processing: 6, interval 152\n",
      "Saved\n",
      "Processing: 6, interval 153\n",
      "Saved\n",
      "Processing: 6, interval 154\n",
      "Saved\n",
      "Processing: 6, interval 155\n",
      "Saved\n",
      "Processing: 6, interval 156\n",
      "Saved\n",
      "Processing: 6, interval 157\n",
      "Saved\n",
      "Processing: 6, interval 158\n",
      "Saved\n",
      "Processing: 6, interval 159\n",
      "Saved\n",
      "Processing: 6, interval 160\n",
      "Saved\n",
      "Processing: 6, interval 161\n",
      "Saved\n",
      "Processing: 6, interval 162\n",
      "Saved\n",
      "Processing: 6, interval 163\n",
      "Saved\n",
      "Processing: 6, interval 164\n",
      "Saved\n",
      "Processing: 6, interval 165\n",
      "Saved\n",
      "Processing: 6, interval 166\n",
      "Saved\n",
      "Processing: 6, interval 167\n",
      "Saved\n",
      "Processing: 6, interval 168\n",
      "Saved\n",
      "Processing: 6, interval 169\n",
      "Saved\n",
      "Processing: 6, interval 170\n",
      "Saved\n",
      "Processing: 6, interval 171\n",
      "Saved\n",
      "Processing: 6, interval 172\n",
      "Saved\n",
      "Processing: 6, interval 173\n",
      "Saved\n",
      "Processing: 6, interval 174\n",
      "Saved\n",
      "Processing: 6, interval 175\n",
      "Saved\n",
      "Processing: 6, interval 176\n",
      "Saved\n",
      "Processing: 6, interval 177\n",
      "Saved\n",
      "Processing: 6, interval 178\n",
      "Saved\n",
      "Processing: 6, interval 179\n",
      "Saved\n",
      "Processing: 6, interval 180\n",
      "Saved\n",
      "Processing: 6, interval 181\n",
      "Saved\n",
      "Processing: 6, interval 182\n",
      "Saved\n",
      "Processing: 6, interval 183\n",
      "Saved\n",
      "Processing: 6, interval 184\n",
      "Saved\n",
      "Processing: 6, interval 185\n",
      "Saved\n",
      "Processing: 6, interval 186\n",
      "Saved\n",
      "Processing: 6, interval 187\n",
      "Saved\n",
      "Processing: 6, interval 188\n",
      "Saved\n",
      "Processing: 6, interval 189\n",
      "Saved\n",
      "Processing: 6, interval 190\n",
      "Saved\n",
      "Processing: 6, interval 191\n",
      "Saved\n",
      "Processing: 6, interval 192\n",
      "Saved\n",
      "Processing: 6, interval 193\n",
      "Saved\n",
      "Processing: 6, interval 194\n",
      "Saved\n",
      "Processing: 6, interval 195\n",
      "Saved\n",
      "Processing: 6, interval 196\n",
      "Saved\n",
      "Processing: 6, interval 197\n",
      "Saved\n",
      "Processing: 6, interval 198\n",
      "Saved\n",
      "Processing: 6, interval 199\n",
      "Saved\n",
      "Processing: 6, interval 200\n",
      "Saved\n",
      "Processing: 6, interval 201\n",
      "Saved\n",
      "Processing: 6, interval 202\n",
      "Saved\n",
      "Processing: 6, interval 203\n",
      "Saved\n",
      "Processing: 6, interval 204\n",
      "Saved\n",
      "Processing: 6, interval 205\n",
      "Saved\n",
      "Processing: 6, interval 206\n",
      "Saved\n",
      "Processing: 6, interval 207\n",
      "Saved\n",
      "Processing: 6, interval 208\n",
      "Saved\n",
      "Processing: 6, interval 209\n",
      "Saved\n",
      "Processing: 6, interval 210\n",
      "Saved\n",
      "Processing: 6, interval 211\n",
      "Saved\n",
      "Processing: 6, interval 212\n",
      "Saved\n",
      "Processing: 6, interval 213\n",
      "Saved\n",
      "Processing: 6, interval 214\n",
      "Saved\n",
      "Processing: 6, interval 215\n",
      "Saved\n",
      "Processing: 6, interval 216\n",
      "Saved\n",
      "Processing: 6, interval 217\n",
      "Saved\n",
      "Processing: 6, interval 218\n",
      "Saved\n",
      "Processing: 6, interval 219\n",
      "Saved\n",
      "Processing: 6, interval 220\n",
      "Saved\n",
      "Processing: 6, interval 221\n",
      "Saved\n",
      "Processing: 6, interval 222\n",
      "Saved\n",
      "Processing: 6, interval 223\n",
      "Saved\n",
      "Processing: 6, interval 224\n",
      "Saved\n",
      "Processing: 6, interval 225\n",
      "Saved\n",
      "Processing: 6, interval 226\n",
      "Saved\n",
      "Processing: 6, interval 227\n",
      "Saved\n",
      "Processing: 6, interval 228\n",
      "Saved\n",
      "Processing: 6, interval 229\n",
      "Saved\n",
      "Processing: 6, interval 230\n",
      "Saved\n",
      "Processing: 6, interval 231\n",
      "Saved\n",
      "Processing: 6, interval 232\n",
      "Saved\n",
      "Processing: 6, interval 233\n",
      "Saved\n",
      "Processing: 6, interval 234\n",
      "Saved\n",
      "Processing: 6, interval 235\n",
      "Saved\n",
      "Processing: 6, interval 236\n",
      "Saved\n",
      "Processing: 6, interval 237\n",
      "Saved\n",
      "Processing: 6, interval 238\n",
      "Saved\n",
      "Processing: 6, interval 239\n",
      "Saved\n",
      "Processing: 6, interval 240\n",
      "Saved\n",
      "Processing: 6, interval 241\n",
      "Saved\n",
      "Processing: 6, interval 242\n",
      "Saved\n",
      "Processing: 6, interval 243\n",
      "Saved\n",
      "Processing: 6, interval 244\n",
      "Saved\n",
      "Processing: 6, interval 245\n",
      "Saved\n",
      "Processing: 6, interval 246\n",
      "Saved\n",
      "Processing: 6, interval 247\n",
      "Saved\n",
      "Processing: 6, interval 248\n",
      "Saved\n",
      "Processing: 6, interval 249\n",
      "Saved\n",
      "Processing: 6, interval 250\n",
      "Saved\n",
      "Processing: 6, interval 251\n",
      "Saved\n",
      "Processing: 6, interval 252\n",
      "Saved\n",
      "Processing: 6, interval 253\n",
      "Saved\n",
      "Processing: 6, interval 254\n",
      "Saved\n",
      "Processing: 6, interval 255\n",
      "Saved\n",
      "Processing: 6, interval 256\n",
      "Saved\n",
      "Processing: 6, interval 257\n",
      "Saved\n",
      "Processing: 6, interval 258\n",
      "Saved\n",
      "Processing: 6, interval 259\n",
      "Saved\n",
      "Processing: 6, interval 260\n",
      "Saved\n",
      "Processing: 6, interval 261\n",
      "Saved\n",
      "Processing: 6, interval 262\n",
      "Saved\n",
      "Processing: 6, interval 263\n",
      "Saved\n",
      "Processing: 6, interval 264\n",
      "Saved\n",
      "Processing: 6, interval 265\n",
      "Saved\n",
      "Processing: 6, interval 266\n",
      "Saved\n",
      "Processing: 6, interval 267\n",
      "Saved\n",
      "Processing: 6, interval 268\n",
      "Saved\n",
      "Processing: 6, interval 269\n",
      "Saved\n",
      "Processing: 6, interval 270\n",
      "Saved\n",
      "Processing: 6, interval 271\n",
      "Saved\n",
      "Processing: 6, interval 272\n",
      "Saved\n",
      "Processing: 6, interval 273\n",
      "Saved\n",
      "Processing: 6, interval 274\n",
      "Saved\n",
      "Processing: 6, interval 275\n",
      "Saved\n",
      "Processing: 6, interval 276\n",
      "Saved\n",
      "Processing: 6, interval 277\n",
      "Saved\n",
      "Processing: 6, interval 278\n",
      "Saved\n",
      "Processing: 6, interval 279\n",
      "Saved\n",
      "Processing: 6, interval 280\n",
      "Saved\n",
      "Processing: 6, interval 281\n",
      "Saved\n",
      "Processing: 6, interval 282\n",
      "Saved\n",
      "Processing: 6, interval 283\n",
      "Saved\n",
      "Processing: 6, interval 284\n",
      "Saved\n",
      "Processing: 6, interval 285\n",
      "Saved\n",
      "Processing: 6, interval 286\n",
      "Saved\n",
      "Processing: 6, interval 287\n",
      "Saved\n",
      "Processing: 6, interval 288\n",
      "Saved\n",
      "Processing: 6, interval 289\n",
      "Saved\n",
      "Processing: 6, interval 290\n",
      "Saved\n",
      "Processing: 6, interval 291\n",
      "Saved\n",
      "Processing: 6, interval 292\n",
      "Saved\n",
      "Processing: 6, interval 293\n",
      "Saved\n",
      "Processing: 6, interval 294\n",
      "Saved\n",
      "Processing: 6, interval 295\n",
      "Saved\n",
      "Processing: 6, interval 296\n",
      "Saved\n",
      "Processing: 6, interval 297\n",
      "Saved\n",
      "Processing: 6, interval 298\n",
      "Saved\n",
      "Processing: 6, interval 299\n",
      "Saved\n",
      "Processing: 6, interval 300\n",
      "Saved\n",
      "Processing: 6, interval 301\n",
      "Saved\n",
      "Processing: 6, interval 302\n",
      "Saved\n",
      "Processing: 6, interval 303\n",
      "Saved\n",
      "Processing: 6, interval 304\n",
      "Saved\n",
      "Processing: 6, interval 305\n",
      "Saved\n",
      "Processing: 6, interval 306\n",
      "Saved\n",
      "Processing: 6, interval 307\n",
      "Saved\n",
      "Processing: 6, interval 308\n",
      "Saved\n",
      "Processing: 6, interval 309\n",
      "Saved\n",
      "Processing: 6, interval 310\n",
      "Saved\n",
      "Processing: 6, interval 311\n",
      "Saved\n",
      "Processing: 6, interval 312\n",
      "Saved\n",
      "Processing: 6, interval 313\n",
      "Saved\n",
      "Processing: 6, interval 314\n",
      "Saved\n",
      "Processing: 6, interval 315\n",
      "Saved\n",
      "Processing: 6, interval 316\n",
      "Saved\n",
      "Processing: 6, interval 317\n",
      "Saved\n",
      "Processing: 6, interval 318\n",
      "Saved\n",
      "Processing: 6, interval 319\n",
      "Saved\n",
      "Processing: 6, interval 320\n",
      "Saved\n",
      "Processing: 6, interval 321\n",
      "Saved\n",
      "Processing: 6, interval 322\n",
      "Saved\n",
      "Processing: 6, interval 323\n",
      "Saved\n",
      "Processing: 6, interval 324\n",
      "Saved\n",
      "Processing: 6, interval 325\n",
      "Saved\n",
      "Processing: 6, interval 326\n",
      "Saved\n",
      "Processing: 6, interval 327\n",
      "Saved\n",
      "Processing: 6, interval 328\n",
      "Saved\n",
      "Processing: 6, interval 329\n",
      "Saved\n",
      "Processing: 6, interval 330\n",
      "Saved\n",
      "Processing: 6, interval 331\n",
      "Saved\n",
      "Processing: 6, interval 332\n",
      "Saved\n",
      "Processing: 6, interval 333\n",
      "Saved\n",
      "Processing: 6, interval 334\n",
      "Saved\n",
      "Processing: 6, interval 335\n",
      "Saved\n",
      "Processing: 6, interval 336\n",
      "Saved\n",
      "Processing: 6, interval 337\n",
      "Saved\n",
      "Processing: 6, interval 338\n",
      "Saved\n",
      "Processing: 6, interval 339\n",
      "Saved\n",
      "Processing: 6, interval 340\n",
      "Saved\n",
      "Processing: 6, interval 341\n",
      "Saved\n",
      "Processing: 6, interval 342\n",
      "Saved\n",
      "Processing: 6, interval 343\n",
      "Saved\n",
      "Processing: 6, interval 344\n",
      "Saved\n",
      "Processing: 6, interval 345\n",
      "Saved\n",
      "Processing: 6, interval 346\n",
      "Saved\n",
      "Processing: 6, interval 347\n",
      "Saved\n",
      "Processing: 6, interval 348\n",
      "Saved\n",
      "Processing: 6, interval 349\n",
      "Saved\n",
      "Processing: 6, interval 350\n",
      "Saved\n",
      "Processing: 6, interval 351\n",
      "Saved\n",
      "Processing: 6, interval 352\n",
      "Saved\n",
      "Processing: 6, interval 353\n",
      "Saved\n",
      "Processing: 6, interval 354\n",
      "Saved\n",
      "Processing: 6, interval 355\n",
      "Saved\n",
      "Processing: 6, interval 356\n",
      "Saved\n",
      "Processing: 6, interval 357\n",
      "Saved\n",
      "Processing: 6, interval 358\n",
      "Saved\n",
      "Processing: 6, interval 359\n",
      "Saved\n",
      "Processing: 6, interval 360\n",
      "Saved\n",
      "Processing: 6, interval 361\n",
      "Saved\n",
      "Processing: 6, interval 362\n",
      "Saved\n",
      "Processing: 6, interval 363\n",
      "Saved\n",
      "Processing: 6, interval 364\n",
      "Saved\n",
      "Processing: 6, interval 365\n",
      "Saved\n",
      "Processing: 6, interval 366\n",
      "Saved\n",
      "Processing: 6, interval 367\n",
      "Saved\n",
      "Processing: 6, interval 368\n",
      "Saved\n",
      "Processing: 6, interval 369\n",
      "Saved\n",
      "Processing: 6, interval 370\n",
      "Saved\n",
      "Processing: 6, interval 371\n",
      "Saved\n",
      "Processing: 6, interval 372\n",
      "Saved\n",
      "Processing: 6, interval 373\n",
      "Saved\n",
      "Processing: 6, interval 374\n",
      "Saved\n",
      "Processing: 6, interval 375\n",
      "Saved\n",
      "Processing: 6, interval 376\n",
      "Saved\n",
      "Processing: 6, interval 377\n",
      "Saved\n",
      "Processing: 6, interval 378\n",
      "Saved\n",
      "Processing: 6, interval 379\n",
      "Saved\n",
      "Processing: 6, interval 380\n",
      "Saved\n",
      "Processing: 6, interval 381\n",
      "Saved\n",
      "Processing: 6, interval 382\n",
      "Saved\n",
      "Processing: 6, interval 383\n",
      "Saved\n",
      "Processing: 6, interval 384\n",
      "Saved\n",
      "Processing: 6, interval 385\n",
      " at 6, interval 385. Retrying...\n",
      "Processing: 6, interval 385\n",
      "Saved\n",
      "Processing: 6, interval 386\n",
      " at 6, interval 386. Retrying...\n",
      "Processing: 6, interval 386\n",
      "Saved\n",
      "Processing: 6, interval 387\n",
      "Saved\n",
      "Processing: 6, interval 388\n",
      "Saved\n",
      "Processing: 6, interval 389\n",
      "Saved\n",
      "Processing: 6, interval 390\n",
      " at 6, interval 390. Retrying...\n",
      "Processing: 6, interval 390\n",
      "Saved\n",
      "Processing: 6, interval 391\n",
      "Saved\n",
      "Processing: 6, interval 392\n",
      "Saved\n",
      "Processing: 6, interval 393\n",
      "Saved\n",
      "Processing: 6, interval 394\n",
      "Saved\n",
      "Processing: 6, interval 395\n",
      "Saved\n",
      "Processing: 6, interval 396\n",
      "Saved\n",
      "Processing: 6, interval 397\n",
      " at 6, interval 397. Retrying...\n",
      "Processing: 6, interval 397\n",
      "Saved\n",
      "Processing: 6, interval 398\n",
      "Saved\n",
      "Processing: 6, interval 399\n",
      "Saved\n",
      "Processing: 6, interval 400\n",
      "Saved\n",
      "Processing: 6, interval 401\n",
      " at 6, interval 401. Retrying...\n",
      "Processing: 6, interval 401\n",
      "Saved\n",
      "Processing: 6, interval 402\n",
      "Saved\n",
      "Processing: 6, interval 403\n",
      "Saved\n",
      "Processing: 6, interval 404\n",
      "Saved\n",
      "Processing: 6, interval 405\n",
      "Saved\n",
      "Processing: 6, interval 406\n",
      "Saved\n",
      "Processing: 7, interval 0\n",
      "Saved\n",
      "Processing: 7, interval 1\n",
      "Saved\n",
      "Processing: 7, interval 2\n",
      " at 7, interval 2. Retrying...\n",
      "Processing: 7, interval 2\n",
      "Saved\n",
      "Processing: 7, interval 3\n",
      " at 7, interval 3. Retrying...\n",
      "Processing: 7, interval 3\n",
      "Saved\n",
      "Processing: 7, interval 4\n",
      "Saved\n",
      "Processing: 7, interval 5\n",
      "Saved\n",
      "Processing: 7, interval 6\n",
      "Saved\n",
      "Processing: 7, interval 7\n",
      "Saved\n",
      "Processing: 7, interval 8\n",
      "Saved\n",
      "Processing: 7, interval 9\n",
      "Saved\n",
      "Processing: 7, interval 10\n",
      "Saved\n",
      "Processing: 7, interval 11\n",
      "Saved\n",
      "Processing: 7, interval 12\n",
      "Saved\n",
      "Processing: 7, interval 13\n",
      "Saved\n",
      "Processing: 7, interval 14\n",
      "Saved\n",
      "Processing: 7, interval 15\n",
      "Saved\n",
      "Processing: 7, interval 16\n",
      "Saved\n",
      "Processing: 7, interval 17\n",
      "Saved\n",
      "Processing: 7, interval 18\n",
      "Saved\n",
      "Processing: 7, interval 19\n",
      "Saved\n",
      "Processing: 7, interval 20\n",
      "Saved\n",
      "Processing: 7, interval 21\n",
      "Saved\n",
      "Processing: 7, interval 22\n",
      "Saved\n",
      "Processing: 7, interval 23\n",
      "Saved\n",
      "Processing: 7, interval 24\n",
      "Saved\n",
      "Processing: 7, interval 25\n",
      "Saved\n",
      "Processing: 7, interval 26\n",
      "Saved\n",
      "Processing: 7, interval 27\n",
      "Saved\n",
      "Processing: 7, interval 28\n",
      "Saved\n",
      "Processing: 7, interval 29\n",
      "Saved\n",
      "Processing: 7, interval 30\n",
      "Saved\n",
      "Processing: 7, interval 31\n",
      "Saved\n",
      "Processing: 7, interval 32\n",
      "Saved\n",
      "Processing: 7, interval 33\n",
      "Saved\n",
      "Processing: 7, interval 34\n",
      "Saved\n",
      "Processing: 7, interval 35\n",
      "Saved\n",
      "Processing: 7, interval 36\n",
      "Saved\n",
      "Processing: 7, interval 37\n",
      "Saved\n",
      "Processing: 7, interval 38\n",
      "Saved\n",
      "Processing: 7, interval 39\n",
      "Saved\n",
      "Processing: 7, interval 40\n",
      "Saved\n",
      "Processing: 7, interval 41\n",
      "Saved\n",
      "Processing: 7, interval 42\n",
      "Saved\n",
      "Processing: 7, interval 43\n",
      "Saved\n",
      "Processing: 7, interval 44\n",
      "Saved\n",
      "Processing: 7, interval 45\n",
      " at 7, interval 45. Retrying...\n",
      "Processing: 7, interval 45\n",
      "Saved\n",
      "Processing: 7, interval 46\n",
      " at 7, interval 46. Retrying...\n",
      "Processing: 7, interval 46\n",
      "{'message': 'Service overloaded'} at 7, interval 46. Retrying...\n",
      "Processing: 7, interval 46\n",
      "Saved\n",
      "Processing: 7, interval 47\n",
      "Saved\n",
      "Processing: 7, interval 48\n",
      "Saved\n",
      "Processing: 7, interval 49\n",
      " at 7, interval 49. Retrying...\n",
      "Processing: 7, interval 49\n",
      " at 7, interval 49. Retrying...\n",
      "Processing: 7, interval 49\n",
      "Saved\n",
      "Processing: 7, interval 50\n",
      " at 7, interval 50. Retrying...\n",
      "Processing: 7, interval 50\n",
      "Saved\n",
      "Processing: 7, interval 51\n",
      " at 7, interval 51. Retrying...\n",
      "Processing: 7, interval 51\n",
      "Saved\n",
      "Processing: 7, interval 52\n",
      "Saved\n",
      "Processing: 7, interval 53\n",
      "Saved\n",
      "Processing: 7, interval 54\n",
      "Saved\n",
      "Processing: 7, interval 55\n",
      " at 7, interval 55. Retrying...\n",
      "Processing: 7, interval 55\n",
      "Saved\n",
      "Processing: 7, interval 56\n",
      "Saved\n",
      "Processing: 7, interval 57\n",
      "Saved\n",
      "Processing: 7, interval 58\n",
      "Saved\n",
      "Processing: 7, interval 59\n",
      "Saved\n",
      "Processing: 7, interval 60\n",
      " at 7, interval 60. Retrying...\n",
      "Processing: 7, interval 60\n",
      " at 7, interval 60. Retrying...\n",
      "Processing: 7, interval 60\n",
      "Saved\n",
      "Processing: 7, interval 61\n",
      "Saved\n",
      "Processing: 7, interval 62\n",
      "Saved\n",
      "Processing: 7, interval 63\n",
      "Saved\n",
      "Processing: 7, interval 64\n",
      "Saved\n",
      "Processing: 7, interval 65\n",
      "Saved\n",
      "Processing: 7, interval 66\n",
      "Saved\n",
      "Processing: 7, interval 67\n",
      "Saved\n",
      "Processing: 7, interval 68\n",
      "Saved\n",
      "Processing: 7, interval 69\n",
      "Saved\n",
      "Processing: 7, interval 70\n",
      "Saved\n",
      "Processing: 7, interval 71\n",
      " at 7, interval 71. Retrying...\n",
      "Processing: 7, interval 71\n",
      "Saved\n",
      "Processing: 7, interval 72\n",
      " at 7, interval 72. Retrying...\n",
      "Processing: 7, interval 72\n",
      "Saved\n",
      "Processing: 7, interval 73\n",
      "Saved\n",
      "Processing: 7, interval 74\n",
      "Saved\n",
      "Processing: 7, interval 75\n",
      "Saved\n",
      "Processing: 7, interval 76\n",
      "Saved\n",
      "Processing: 7, interval 77\n",
      "Saved\n",
      "Processing: 7, interval 78\n",
      "Saved\n",
      "Processing: 7, interval 79\n",
      "Saved\n",
      "Processing: 7, interval 80\n",
      "Saved\n",
      "Processing: 7, interval 81\n",
      "Saved\n",
      "Processing: 7, interval 82\n",
      "Saved\n",
      "Processing: 7, interval 83\n",
      "Saved\n",
      "Processing: 7, interval 84\n",
      "Saved\n",
      "Processing: 7, interval 85\n",
      "Saved\n",
      "Processing: 7, interval 86\n",
      "Saved\n",
      "Processing: 7, interval 87\n",
      "Saved\n",
      "Processing: 7, interval 88\n",
      "Saved\n",
      "Processing: 7, interval 89\n",
      "Saved\n",
      "Processing: 7, interval 90\n",
      "Saved\n",
      "Processing: 7, interval 91\n",
      "Saved\n",
      "Processing: 7, interval 92\n",
      "Saved\n",
      "Processing: 7, interval 93\n",
      "Saved\n",
      "Processing: 7, interval 94\n",
      "Saved\n",
      "Processing: 7, interval 95\n",
      "Saved\n",
      "Processing: 7, interval 96\n",
      "Saved\n",
      "Processing: 7, interval 97\n",
      "Saved\n",
      "Processing: 7, interval 98\n",
      "Saved\n",
      "Processing: 7, interval 99\n",
      "Saved\n",
      "Processing: 7, interval 100\n",
      "Saved\n",
      "Processing: 7, interval 101\n",
      "Saved\n",
      "Processing: 7, interval 102\n",
      "Saved\n",
      "Processing: 7, interval 103\n",
      "Saved\n",
      "Processing: 7, interval 104\n",
      " at 7, interval 104. Retrying...\n",
      "Processing: 7, interval 104\n",
      " at 7, interval 104. Retrying...\n",
      "Processing: 7, interval 104\n",
      "Saved\n",
      "Processing: 7, interval 105\n",
      "Saved\n",
      "Processing: 7, interval 106\n",
      "Saved\n",
      "Processing: 7, interval 107\n",
      "Saved\n",
      "Processing: 7, interval 108\n",
      " at 7, interval 108. Retrying...\n",
      "Processing: 7, interval 108\n",
      "Saved\n",
      "Processing: 7, interval 109\n",
      "Saved\n",
      "Processing: 7, interval 110\n",
      "Saved\n",
      "Processing: 7, interval 111\n",
      "Saved\n",
      "Processing: 7, interval 112\n",
      "Saved\n",
      "Processing: 7, interval 113\n",
      "Saved\n",
      "Processing: 7, interval 114\n",
      "Saved\n",
      "Processing: 7, interval 115\n",
      "Saved\n",
      "Processing: 7, interval 116\n",
      "Saved\n",
      "Processing: 7, interval 117\n",
      "Saved\n",
      "Processing: 7, interval 118\n",
      "Saved\n",
      "Processing: 7, interval 119\n",
      "Saved\n",
      "Processing: 7, interval 120\n",
      "Saved\n",
      "Processing: 7, interval 121\n",
      "Saved\n",
      "Processing: 7, interval 122\n",
      "Saved\n",
      "Processing: 7, interval 123\n",
      "Saved\n",
      "Processing: 7, interval 124\n",
      "Saved\n",
      "Processing: 7, interval 125\n",
      "Saved\n",
      "Processing: 7, interval 126\n",
      "Saved\n",
      "Processing: 7, interval 127\n",
      "Saved\n",
      "Processing: 7, interval 128\n",
      "Saved\n",
      "Processing: 7, interval 129\n",
      " at 7, interval 129. Retrying...\n",
      "Processing: 7, interval 129\n",
      "Saved\n",
      "Processing: 7, interval 130\n",
      "Saved\n",
      "Processing: 7, interval 131\n",
      "Saved\n",
      "Processing: 7, interval 132\n",
      " at 7, interval 132. Retrying...\n",
      "Processing: 7, interval 132\n",
      "Saved\n",
      "Processing: 7, interval 133\n",
      " at 7, interval 133. Retrying...\n",
      "Processing: 7, interval 133\n",
      " at 7, interval 133. Retrying...\n",
      "Processing: 7, interval 133\n",
      "Saved\n",
      "Processing: 7, interval 134\n",
      "Saved\n",
      "Processing: 7, interval 135\n",
      "Saved\n",
      "Processing: 7, interval 136\n",
      " at 7, interval 136. Retrying...\n",
      "Processing: 7, interval 136\n",
      "Saved\n",
      "Processing: 7, interval 137\n",
      " at 7, interval 137. Retrying...\n",
      "Processing: 7, interval 137\n",
      "Saved\n",
      "Processing: 7, interval 138\n",
      "Saved\n",
      "Processing: 7, interval 139\n",
      "Saved\n",
      "Processing: 7, interval 140\n",
      "Saved\n",
      "Processing: 7, interval 141\n",
      "Saved\n",
      "Processing: 7, interval 142\n",
      "Saved\n",
      "Processing: 7, interval 143\n",
      " at 7, interval 143. Retrying...\n",
      "Processing: 7, interval 143\n",
      "Saved\n",
      "Processing: 7, interval 144\n",
      " at 7, interval 144. Retrying...\n",
      "Processing: 7, interval 144\n",
      "Saved\n",
      "Processing: 7, interval 145\n",
      " at 7, interval 145. Retrying...\n",
      "Processing: 7, interval 145\n",
      "Saved\n",
      "Processing: 7, interval 146\n",
      "Saved\n",
      "Processing: 7, interval 147\n",
      " at 7, interval 147. Retrying...\n",
      "Processing: 7, interval 147\n",
      "Saved\n",
      "Processing: 7, interval 148\n",
      " at 7, interval 148. Retrying...\n",
      "Processing: 7, interval 148\n",
      "Saved\n",
      "Processing: 7, interval 149\n",
      " at 7, interval 149. Retrying...\n",
      "Processing: 7, interval 149\n",
      "Saved\n",
      "Processing: 7, interval 150\n",
      " at 7, interval 150. Retrying...\n",
      "Processing: 7, interval 150\n",
      "Saved\n",
      "Processing: 7, interval 151\n",
      "Saved\n",
      "Processing: 7, interval 152\n",
      "Saved\n",
      "Processing: 7, interval 153\n",
      "Saved\n",
      "Processing: 7, interval 154\n",
      " at 7, interval 154. Retrying...\n",
      "Processing: 7, interval 154\n",
      "Saved\n",
      "Processing: 7, interval 155\n",
      "Saved\n",
      "Processing: 7, interval 156\n",
      "Saved\n",
      "Processing: 7, interval 157\n",
      "Saved\n",
      "Processing: 7, interval 158\n",
      "Saved\n",
      "Processing: 7, interval 159\n",
      "Saved\n",
      "Processing: 7, interval 160\n",
      "Saved\n",
      "Processing: 7, interval 161\n",
      "Saved\n",
      "Processing: 7, interval 162\n",
      "Saved\n",
      "Processing: 7, interval 163\n",
      "Saved\n",
      "Processing: 7, interval 164\n",
      " at 7, interval 164. Retrying...\n",
      "Processing: 7, interval 164\n",
      " at 7, interval 164. Retrying...\n",
      "Processing: 7, interval 164\n",
      "Saved\n",
      "Processing: 7, interval 165\n",
      " at 7, interval 165. Retrying...\n",
      "Processing: 7, interval 165\n",
      "Saved\n",
      "Processing: 7, interval 166\n",
      " at 7, interval 166. Retrying...\n",
      "Processing: 7, interval 166\n",
      "Saved\n",
      "Processing: 7, interval 167\n",
      " at 7, interval 167. Retrying...\n",
      "Processing: 7, interval 167\n",
      "Saved\n",
      "Processing: 7, interval 168\n",
      "Saved\n",
      "Processing: 7, interval 169\n",
      "Saved\n",
      "Processing: 7, interval 170\n",
      "Saved\n",
      "Processing: 7, interval 171\n",
      "Saved\n",
      "Processing: 7, interval 172\n",
      "Saved\n",
      "Processing: 7, interval 173\n",
      "Saved\n",
      "Processing: 7, interval 174\n",
      "Saved\n",
      "Processing: 7, interval 175\n",
      "Saved\n",
      "Processing: 7, interval 176\n",
      "Saved\n",
      "Processing: 7, interval 177\n",
      "Saved\n",
      "Processing: 7, interval 178\n",
      "Saved\n",
      "Processing: 7, interval 179\n",
      "Saved\n",
      "Processing: 7, interval 180\n",
      "Saved\n",
      "Processing: 7, interval 181\n",
      "Saved\n",
      "Processing: 7, interval 182\n",
      "Saved\n",
      "Processing: 7, interval 183\n",
      "Saved\n",
      "Processing: 7, interval 184\n",
      "Saved\n",
      "Processing: 7, interval 185\n",
      "Saved\n",
      "Processing: 7, interval 186\n",
      "Saved\n",
      "Processing: 7, interval 187\n",
      "Saved\n",
      "Processing: 7, interval 188\n",
      "Saved\n",
      "Processing: 7, interval 189\n",
      "Saved\n",
      "Processing: 7, interval 190\n",
      "Saved\n",
      "Processing: 7, interval 191\n",
      "Saved\n",
      "Processing: 7, interval 192\n",
      "Saved\n",
      "Processing: 7, interval 193\n",
      "Saved\n",
      "Processing: 7, interval 194\n",
      "Saved\n",
      "Processing: 7, interval 195\n",
      "Saved\n",
      "Processing: 7, interval 196\n",
      "Saved\n",
      "Processing: 7, interval 197\n",
      "Saved\n",
      "Processing: 7, interval 198\n",
      "Saved\n",
      "Processing: 7, interval 199\n",
      "Saved\n",
      "Processing: 7, interval 200\n",
      "Saved\n",
      "Processing: 7, interval 201\n",
      "Saved\n",
      "Processing: 7, interval 202\n",
      "Saved\n",
      "Processing: 7, interval 203\n",
      "Saved\n",
      "Processing: 7, interval 204\n",
      "Saved\n",
      "Processing: 7, interval 205\n",
      "Saved\n",
      "Processing: 7, interval 206\n",
      "Saved\n",
      "Processing: 7, interval 207\n",
      "Saved\n",
      "Processing: 7, interval 208\n",
      "Saved\n",
      "Processing: 7, interval 209\n",
      "Saved\n",
      "Processing: 7, interval 210\n",
      "Saved\n",
      "Processing: 7, interval 211\n",
      "Saved\n",
      "Processing: 7, interval 212\n",
      "Saved\n",
      "Processing: 7, interval 213\n",
      "Saved\n",
      "Processing: 7, interval 214\n",
      "Saved\n",
      "Processing: 7, interval 215\n",
      " at 7, interval 215. Retrying...\n",
      "Processing: 7, interval 215\n",
      " at 7, interval 215. Retrying...\n",
      "Processing: 7, interval 215\n",
      "Saved\n",
      "Processing: 7, interval 216\n",
      "Saved\n",
      "Processing: 7, interval 217\n",
      " at 7, interval 217. Retrying...\n",
      "Processing: 7, interval 217\n",
      "Saved\n",
      "Processing: 7, interval 218\n",
      "Saved\n",
      "Processing: 7, interval 219\n",
      "Saved\n",
      "Processing: 7, interval 220\n",
      "Saved\n",
      "Processing: 7, interval 221\n",
      "Saved\n",
      "Processing: 7, interval 222\n",
      "Saved\n",
      "Processing: 7, interval 223\n",
      " at 7, interval 223. Retrying...\n",
      "Processing: 7, interval 223\n",
      "Saved\n",
      "Processing: 7, interval 224\n",
      "Saved\n",
      "Processing: 7, interval 225\n",
      " at 7, interval 225. Retrying...\n",
      "Processing: 7, interval 225\n",
      "Saved\n",
      "Processing: 7, interval 226\n",
      "Saved\n",
      "Processing: 7, interval 227\n",
      "Saved\n",
      "Processing: 7, interval 228\n",
      "Saved\n",
      "Processing: 7, interval 229\n",
      " at 7, interval 229. Retrying...\n",
      "Processing: 7, interval 229\n",
      "Saved\n",
      "Processing: 7, interval 230\n",
      "Saved\n",
      "Processing: 7, interval 231\n",
      "Saved\n",
      "Processing: 8, interval 0\n",
      "Saved\n",
      "Processing: 8, interval 1\n",
      "Saved\n",
      "Processing: 8, interval 2\n",
      "Saved\n",
      "Processing: 8, interval 3\n",
      " at 8, interval 3. Retrying...\n",
      "Processing: 8, interval 3\n",
      "Saved\n",
      "Processing: 8, interval 4\n",
      "Saved\n",
      "Processing: 8, interval 5\n",
      " at 8, interval 5. Retrying...\n",
      "Processing: 8, interval 5\n",
      "Saved\n",
      "Processing: 8, interval 6\n",
      "Saved\n",
      "Processing: 8, interval 7\n",
      "Saved\n",
      "Processing: 8, interval 8\n",
      "Saved\n",
      "Processing: 8, interval 9\n",
      "Saved\n",
      "Processing: 8, interval 10\n",
      "Saved\n",
      "Processing: 8, interval 11\n",
      " at 8, interval 11. Retrying...\n",
      "Processing: 8, interval 11\n",
      " at 8, interval 11. Retrying...\n",
      "Processing: 8, interval 11\n",
      "Saved\n",
      "Processing: 8, interval 12\n",
      " at 8, interval 12. Retrying...\n",
      "Processing: 8, interval 12\n",
      "Saved\n",
      "Processing: 8, interval 13\n",
      " at 8, interval 13. Retrying...\n",
      "Processing: 8, interval 13\n",
      " at 8, interval 13. Retrying...\n",
      "Processing: 8, interval 13\n",
      "Saved\n",
      "Processing: 8, interval 14\n",
      " at 8, interval 14. Retrying...\n",
      "Processing: 8, interval 14\n",
      "Saved\n",
      "Processing: 8, interval 15\n",
      "Saved\n",
      "Processing: 8, interval 16\n",
      "Saved\n",
      "Processing: 8, interval 17\n",
      "Saved\n",
      "Processing: 8, interval 18\n",
      "Saved\n",
      "Processing: 8, interval 19\n",
      "Saved\n",
      "Processing: 8, interval 20\n",
      " at 8, interval 20. Retrying...\n",
      "Processing: 8, interval 20\n",
      "Saved\n",
      "Processing: 8, interval 21\n",
      " at 8, interval 21. Retrying...\n",
      "Processing: 8, interval 21\n",
      "Saved\n",
      "Processing: 8, interval 22\n",
      " at 8, interval 22. Retrying...\n",
      "Processing: 8, interval 22\n",
      "Saved\n",
      "Processing: 8, interval 23\n",
      "Saved\n",
      "Processing: 8, interval 24\n",
      "Saved\n",
      "Processing: 8, interval 25\n",
      "Saved\n",
      "Processing: 8, interval 26\n",
      "Saved\n",
      "Processing: 8, interval 27\n",
      "Saved\n",
      "Processing: 8, interval 28\n",
      "Saved\n",
      "Processing: 8, interval 29\n",
      "Saved\n",
      "Processing: 8, interval 30\n",
      "Saved\n",
      "Processing: 8, interval 31\n",
      "Saved\n",
      "Processing: 8, interval 32\n",
      "Saved\n",
      "Processing: 8, interval 33\n",
      "Saved\n",
      "Processing: 8, interval 34\n",
      "Saved\n",
      "Processing: 8, interval 35\n",
      "Saved\n",
      "Processing: 8, interval 36\n",
      "Saved\n",
      "Processing: 8, interval 37\n",
      "Saved\n",
      "Processing: 8, interval 38\n",
      "Saved\n",
      "Processing: 8, interval 39\n",
      "Saved\n",
      "Processing: 8, interval 40\n",
      "Saved\n",
      "Processing: 8, interval 41\n",
      "Saved\n",
      "Processing: 8, interval 42\n",
      "Saved\n",
      "Processing: 8, interval 43\n",
      "Saved\n",
      "Processing: 8, interval 44\n",
      "Saved\n",
      "Processing: 8, interval 45\n",
      "Saved\n",
      "Processing: 8, interval 46\n",
      "Saved\n",
      "Processing: 8, interval 47\n",
      "Saved\n",
      "Processing: 8, interval 48\n",
      "Saved\n",
      "Processing: 8, interval 49\n",
      "Saved\n",
      "Processing: 8, interval 50\n",
      "Saved\n",
      "Processing: 8, interval 51\n",
      "Saved\n",
      "Processing: 8, interval 52\n",
      "Saved\n",
      "Processing: 8, interval 53\n",
      "Saved\n",
      "Processing: 8, interval 54\n",
      "Saved\n",
      "Processing: 8, interval 55\n",
      "Saved\n",
      "Processing: 8, interval 56\n",
      "Saved\n",
      "Processing: 8, interval 57\n",
      "Saved\n",
      "Processing: 8, interval 58\n",
      "Saved\n",
      "Processing: 8, interval 59\n",
      "Saved\n",
      "Processing: 8, interval 60\n",
      "Saved\n",
      "Processing: 8, interval 61\n",
      "Saved\n",
      "Processing: 8, interval 62\n",
      "Saved\n",
      "Processing: 8, interval 63\n",
      "Saved\n",
      "Processing: 8, interval 64\n",
      "Saved\n",
      "Processing: 8, interval 65\n",
      "Saved\n",
      "Processing: 8, interval 66\n",
      "Saved\n",
      "Processing: 8, interval 67\n",
      "Saved\n",
      "Processing: 8, interval 68\n",
      " at 8, interval 68. Retrying...\n",
      "Processing: 8, interval 68\n",
      "Saved\n",
      "Processing: 8, interval 69\n",
      " at 8, interval 69. Retrying...\n",
      "Processing: 8, interval 69\n",
      "Saved\n",
      "Processing: 8, interval 70\n",
      "Saved\n",
      "Processing: 8, interval 71\n",
      " at 8, interval 71. Retrying...\n",
      "Processing: 8, interval 71\n",
      "Saved\n",
      "Processing: 8, interval 72\n",
      "Saved\n",
      "Processing: 8, interval 73\n",
      "Saved\n",
      "Processing: 8, interval 74\n",
      "Saved\n",
      "Processing: 8, interval 75\n",
      "Saved\n",
      "Processing: 8, interval 76\n",
      "Saved\n",
      "Processing: 8, interval 77\n",
      "Saved\n",
      "Processing: 8, interval 78\n",
      "Saved\n",
      "Processing: 8, interval 79\n",
      "Saved\n",
      "Processing: 8, interval 80\n",
      "Saved\n",
      "Processing: 8, interval 81\n",
      " at 8, interval 81. Retrying...\n",
      "Processing: 8, interval 81\n",
      "Saved\n",
      "Processing: 8, interval 82\n",
      "Saved\n",
      "Processing: 8, interval 83\n",
      " at 8, interval 83. Retrying...\n",
      "Processing: 8, interval 83\n",
      "Saved\n",
      "Processing: 8, interval 84\n",
      "Saved\n",
      "Processing: 8, interval 85\n",
      "Saved\n",
      "Processing: 8, interval 86\n",
      "Saved\n",
      "Processing: 8, interval 87\n",
      " at 8, interval 87. Retrying...\n",
      "Processing: 8, interval 87\n",
      " at 8, interval 87. Retrying...\n",
      "Processing: 8, interval 87\n",
      "Saved\n",
      "Processing: 8, interval 88\n",
      " at 8, interval 88. Retrying...\n",
      "Processing: 8, interval 88\n",
      "Saved\n",
      "Processing: 8, interval 89\n",
      "Saved\n",
      "Processing: 8, interval 90\n",
      " at 8, interval 90. Retrying...\n",
      "Processing: 8, interval 90\n",
      "Saved\n",
      "Processing: 8, interval 91\n",
      "Saved\n",
      "Processing: 8, interval 92\n",
      "Saved\n",
      "Processing: 8, interval 93\n",
      "Saved\n",
      "Processing: 8, interval 94\n",
      " at 8, interval 94. Retrying...\n",
      "Processing: 8, interval 94\n",
      "Saved\n",
      "Processing: 8, interval 95\n",
      "Saved\n",
      "Processing: 8, interval 96\n",
      "Saved\n",
      "Processing: 8, interval 97\n",
      " at 8, interval 97. Retrying...\n",
      "Processing: 8, interval 97\n",
      "Saved\n",
      "Processing: 8, interval 98\n",
      "Saved\n",
      "Processing: 8, interval 99\n",
      "Saved\n",
      "Processing: 8, interval 100\n",
      "Saved\n",
      "Processing: 8, interval 101\n",
      "Saved\n",
      "Processing: 8, interval 102\n",
      " at 8, interval 102. Retrying...\n",
      "Processing: 8, interval 102\n",
      "Saved\n",
      "Processing: 8, interval 103\n",
      " at 8, interval 103. Retrying...\n",
      "Processing: 8, interval 103\n",
      " at 8, interval 103. Retrying...\n",
      "Processing: 8, interval 103\n",
      "Saved\n",
      "Processing: 8, interval 104\n",
      "Saved\n",
      "Processing: 8, interval 105\n",
      "Saved\n",
      "Processing: 8, interval 106\n",
      "Saved\n",
      "Processing: 8, interval 107\n",
      "Saved\n",
      "Processing: 8, interval 108\n",
      "Saved\n",
      "Processing: 8, interval 109\n",
      "Saved\n",
      "Processing: 8, interval 110\n",
      "Saved\n",
      "Processing: 8, interval 111\n",
      "Saved\n",
      "Processing: 8, interval 112\n",
      "Saved\n",
      "Processing: 8, interval 113\n",
      "Saved\n",
      "Processing: 8, interval 114\n",
      "Saved\n",
      "Processing: 8, interval 115\n",
      "Saved\n",
      "Processing: 8, interval 116\n",
      "Saved\n",
      "Processing: 8, interval 117\n",
      "Saved\n",
      "Processing: 8, interval 118\n",
      "Saved\n",
      "Processing: 8, interval 119\n",
      "Saved\n",
      "Processing: 8, interval 120\n",
      "Saved\n",
      "Processing: 8, interval 121\n",
      "Saved\n",
      "Processing: 8, interval 122\n",
      "Saved\n",
      "Processing: 8, interval 123\n",
      " at 8, interval 123. Retrying...\n",
      "Processing: 8, interval 123\n",
      "Saved\n",
      "Processing: 8, interval 124\n",
      " at 8, interval 124. Retrying...\n",
      "Processing: 8, interval 124\n",
      "Saved\n",
      "Processing: 8, interval 125\n",
      "Saved\n",
      "Processing: 8, interval 126\n",
      " at 8, interval 126. Retrying...\n",
      "Processing: 8, interval 126\n",
      " at 8, interval 126. Retrying...\n",
      "Processing: 8, interval 126\n",
      "Saved\n",
      "Processing: 8, interval 127\n",
      "Saved\n",
      "Processing: 8, interval 128\n",
      "Saved\n",
      "Processing: 8, interval 129\n",
      " at 8, interval 129. Retrying...\n",
      "Processing: 8, interval 129\n",
      "Saved\n",
      "Processing: 8, interval 130\n",
      "Saved\n",
      "Processing: 8, interval 131\n",
      "Saved\n",
      "Processing: 8, interval 132\n",
      "Saved\n",
      "Processing: 8, interval 133\n",
      "Saved\n",
      "Processing: 8, interval 134\n",
      "Saved\n",
      "Processing: 8, interval 135\n",
      "Saved\n",
      "Processing: 8, interval 136\n",
      "Saved\n",
      "Processing: 8, interval 137\n",
      "Saved\n",
      "Processing: 8, interval 138\n",
      "Saved\n",
      "Processing: 8, interval 139\n",
      "Saved\n",
      "Processing: 8, interval 140\n",
      "Saved\n",
      "Processing: 8, interval 141\n",
      "Saved\n",
      "Processing: 8, interval 142\n",
      "Saved\n",
      "Processing: 8, interval 143\n",
      "Saved\n",
      "Processing: 8, interval 144\n",
      "Saved\n",
      "Processing: 8, interval 145\n",
      "Saved\n",
      "Processing: 8, interval 146\n",
      " at 8, interval 146. Retrying...\n",
      "Processing: 8, interval 146\n",
      "Saved\n",
      "Processing: 8, interval 147\n",
      "Saved\n",
      "Processing: 8, interval 148\n",
      "Saved\n",
      "Processing: 8, interval 149\n",
      "Saved\n",
      "Processing: 8, interval 150\n",
      "Saved\n",
      "Processing: 8, interval 151\n",
      "Saved\n",
      "Processing: 8, interval 152\n",
      "Saved\n",
      "Processing: 8, interval 153\n",
      "Saved\n",
      "Processing: 8, interval 154\n",
      "Saved\n",
      "Processing: 8, interval 155\n",
      " at 8, interval 155. Retrying...\n",
      "Processing: 8, interval 155\n",
      "Saved\n",
      "Processing: 8, interval 156\n",
      "Saved\n",
      "Processing: 8, interval 157\n",
      "Saved\n",
      "Processing: 8, interval 158\n",
      "Saved\n",
      "Processing: 8, interval 159\n",
      "Saved\n",
      "Processing: 8, interval 160\n",
      " at 8, interval 160. Retrying...\n",
      "Processing: 8, interval 160\n",
      "Saved\n",
      "Processing: 8, interval 161\n",
      "Saved\n",
      "Processing: 8, interval 162\n",
      "Saved\n",
      "Processing: 8, interval 163\n",
      "Saved\n",
      "Processing: 8, interval 164\n",
      "Saved\n",
      "Processing: 8, interval 165\n",
      "Saved\n",
      "Processing: 8, interval 166\n",
      " at 8, interval 166. Retrying...\n",
      "Processing: 8, interval 166\n",
      "Saved\n",
      "Processing: 8, interval 167\n",
      "Saved\n",
      "Processing: 8, interval 168\n",
      "Saved\n",
      "Processing: 8, interval 169\n",
      "Saved\n",
      "Processing: 8, interval 170\n",
      "Saved\n",
      "Processing: 8, interval 171\n",
      "Saved\n",
      "Processing: 8, interval 172\n",
      "Saved\n",
      "Processing: 8, interval 173\n",
      " at 8, interval 173. Retrying...\n",
      "Processing: 8, interval 173\n",
      "Saved\n",
      "Processing: 8, interval 174\n",
      "Saved\n",
      "Processing: 8, interval 175\n",
      "Saved\n",
      "Processing: 8, interval 176\n",
      "Saved\n",
      "Processing: 8, interval 177\n",
      "Saved\n",
      "Processing: 8, interval 178\n",
      "Saved\n",
      "Processing: 8, interval 179\n",
      "Saved\n",
      "Processing: 8, interval 180\n",
      "Saved\n",
      "Processing: 8, interval 181\n",
      "Saved\n",
      "Processing: 8, interval 182\n",
      "Saved\n",
      "Processing: 8, interval 183\n",
      "Saved\n",
      "Processing: 8, interval 184\n",
      "Saved\n",
      "Processing: 8, interval 185\n",
      "Saved\n",
      "Processing: 8, interval 186\n",
      " at 8, interval 186. Retrying...\n",
      "Processing: 8, interval 186\n",
      "Saved\n",
      "Processing: 8, interval 187\n",
      "Saved\n",
      "Processing: 8, interval 188\n",
      "Saved\n",
      "Processing: 8, interval 189\n",
      "Saved\n",
      "Processing: 8, interval 190\n",
      "Saved\n",
      "Processing: 8, interval 191\n",
      "Saved\n",
      "Processing: 8, interval 192\n",
      " at 8, interval 192. Retrying...\n",
      "Processing: 8, interval 192\n",
      "Saved\n",
      "Processing: 8, interval 193\n",
      "Saved\n",
      "Processing: 8, interval 194\n",
      "Saved\n",
      "Processing: 8, interval 195\n",
      "Saved\n",
      "Processing: 8, interval 196\n",
      "Saved\n",
      "Processing: 8, interval 197\n",
      "Saved\n",
      "Processing: 8, interval 198\n",
      "Saved\n",
      "Processing: 8, interval 199\n",
      " at 8, interval 199. Retrying...\n",
      "Processing: 8, interval 199\n",
      " at 8, interval 199. Retrying...\n",
      "Processing: 8, interval 199\n",
      "Saved\n",
      "Processing: 8, interval 200\n",
      "Saved\n",
      "Processing: 8, interval 201\n",
      "Saved\n",
      "Processing: 8, interval 202\n",
      "Saved\n",
      "Processing: 8, interval 203\n",
      " at 8, interval 203. Retrying...\n",
      "Processing: 8, interval 203\n",
      " at 8, interval 203. Retrying...\n",
      "Processing: 8, interval 203\n",
      "Saved\n",
      "Processing: 8, interval 204\n",
      "Saved\n",
      "Processing: 8, interval 205\n",
      " at 8, interval 205. Retrying...\n",
      "Processing: 8, interval 205\n",
      "Saved\n",
      "Processing: 8, interval 206\n",
      "Saved\n",
      "Processing: 8, interval 207\n",
      "Saved\n",
      "Processing: 8, interval 208\n",
      "Saved\n",
      "Processing: 8, interval 209\n",
      "Saved\n",
      "Processing: 8, interval 210\n",
      "Saved\n",
      "Processing: 8, interval 211\n",
      "Saved\n",
      "Processing: 8, interval 212\n",
      "Saved\n",
      "Processing: 8, interval 213\n",
      "Saved\n",
      "Processing: 8, interval 214\n",
      "Saved\n",
      "Processing: 8, interval 215\n",
      "Saved\n",
      "Processing: 8, interval 216\n",
      "Saved\n",
      "Processing: 8, interval 217\n",
      "Saved\n",
      "Processing: 8, interval 218\n",
      "Saved\n",
      "Processing: 8, interval 219\n",
      "Saved\n",
      "Processing: 8, interval 220\n",
      "Saved\n",
      "Processing: 8, interval 221\n",
      "Saved\n",
      "Processing: 8, interval 222\n",
      "Saved\n",
      "Processing: 8, interval 223\n",
      "Saved\n",
      "Processing: 8, interval 224\n",
      "Saved\n",
      "Processing: 8, interval 225\n",
      "Saved\n",
      "Processing: 8, interval 226\n",
      "Saved\n",
      "Processing: 8, interval 227\n",
      "Saved\n",
      "Processing: 9, interval 0\n",
      "Saved\n",
      "Processing: 9, interval 1\n",
      "Saved\n",
      "Processing: 9, interval 2\n",
      "Saved\n",
      "Processing: 9, interval 3\n",
      "Saved\n",
      "Processing: 9, interval 4\n",
      "Saved\n",
      "Processing: 9, interval 5\n",
      "Saved\n",
      "Processing: 9, interval 6\n",
      "Saved\n",
      "Processing: 9, interval 7\n",
      "Saved\n",
      "Processing: 9, interval 8\n",
      "Saved\n",
      "Processing: 9, interval 9\n",
      "Saved\n",
      "Processing: 9, interval 10\n",
      "Saved\n",
      "Processing: 9, interval 11\n",
      "Saved\n",
      "Processing: 9, interval 12\n",
      "Saved\n",
      "Processing: 9, interval 13\n",
      "Saved\n",
      "Processing: 9, interval 14\n",
      "Saved\n",
      "Processing: 9, interval 15\n",
      "Saved\n",
      "Processing: 9, interval 16\n",
      "Saved\n",
      "Processing: 9, interval 17\n",
      "Saved\n",
      "Processing: 9, interval 18\n",
      "Saved\n",
      "Processing: 9, interval 19\n",
      "Saved\n",
      "Processing: 9, interval 20\n",
      "Saved\n",
      "Processing: 9, interval 21\n",
      "Saved\n",
      "Processing: 9, interval 22\n",
      "Saved\n",
      "Processing: 9, interval 23\n",
      "Saved\n",
      "Processing: 9, interval 24\n",
      "Saved\n",
      "Processing: 9, interval 25\n",
      "Saved\n",
      "Processing: 9, interval 26\n",
      "Saved\n",
      "Processing: 9, interval 27\n",
      "Saved\n",
      "Processing: 9, interval 28\n",
      "Saved\n",
      "Processing: 9, interval 29\n",
      "Saved\n",
      "Processing: 9, interval 30\n",
      "Saved\n",
      "Processing: 9, interval 31\n",
      "Saved\n",
      "Processing: 9, interval 32\n",
      "Saved\n",
      "Processing: 9, interval 33\n",
      "Saved\n",
      "Processing: 9, interval 34\n",
      "Saved\n",
      "Processing: 9, interval 35\n",
      " at 9, interval 35. Retrying...\n",
      "Processing: 9, interval 35\n",
      "Saved\n",
      "Processing: 9, interval 36\n",
      "Saved\n",
      "Processing: 9, interval 37\n",
      "Saved\n",
      "Processing: 9, interval 38\n",
      "Saved\n",
      "Processing: 9, interval 39\n",
      "Saved\n",
      "Processing: 9, interval 40\n",
      "Saved\n",
      "Processing: 9, interval 41\n",
      "Saved\n",
      "Processing: 9, interval 42\n",
      "Saved\n",
      "Processing: 9, interval 43\n",
      "Saved\n",
      "Processing: 9, interval 44\n",
      "Saved\n",
      "Processing: 9, interval 45\n",
      "Saved\n",
      "Processing: 9, interval 46\n",
      "Saved\n",
      "Processing: 9, interval 47\n",
      "Saved\n",
      "Processing: 9, interval 48\n",
      "Saved\n",
      "Processing: 9, interval 49\n",
      "Saved\n",
      "Processing: 9, interval 50\n",
      "Saved\n",
      "Processing: 9, interval 51\n",
      "Saved\n",
      "Processing: 9, interval 52\n",
      "Saved\n",
      "Processing: 9, interval 53\n",
      "Saved\n",
      "Processing: 9, interval 54\n",
      "Saved\n",
      "Processing: 9, interval 55\n",
      "Saved\n",
      "Processing: 9, interval 56\n",
      "Saved\n",
      "Processing: 9, interval 57\n",
      "Saved\n",
      "Processing: 9, interval 58\n",
      "Saved\n",
      "Processing: 9, interval 59\n",
      "Saved\n",
      "Processing: 9, interval 60\n",
      "Saved\n",
      "Processing: 9, interval 61\n",
      "Saved\n",
      "Processing: 9, interval 62\n",
      "Saved\n",
      "Processing: 9, interval 63\n",
      "Saved\n",
      "Processing: 9, interval 64\n",
      "Saved\n",
      "Processing: 9, interval 65\n",
      "Saved\n",
      "Processing: 9, interval 66\n",
      "Saved\n",
      "Processing: 9, interval 67\n",
      "Saved\n",
      "Processing: 9, interval 68\n",
      "Saved\n",
      "Processing: 9, interval 69\n",
      "Saved\n",
      "Processing: 9, interval 70\n",
      "Saved\n",
      "Processing: 9, interval 71\n",
      "Saved\n",
      "Processing: 9, interval 72\n",
      "Saved\n",
      "Processing: 9, interval 73\n",
      "Saved\n",
      "Processing: 9, interval 74\n",
      "Saved\n",
      "Processing: 9, interval 75\n",
      "Saved\n",
      "Processing: 9, interval 76\n",
      "Saved\n",
      "Processing: 9, interval 77\n",
      "Saved\n",
      "Processing: 9, interval 78\n",
      "Saved\n",
      "Processing: 9, interval 79\n",
      "Saved\n",
      "Processing: 9, interval 80\n",
      "Saved\n",
      "Processing: 9, interval 81\n",
      "Saved\n",
      "Processing: 9, interval 82\n",
      "Saved\n",
      "Processing: 9, interval 83\n",
      "Saved\n",
      "Processing: 9, interval 84\n",
      "Saved\n",
      "Processing: 9, interval 85\n",
      "Saved\n",
      "Processing: 9, interval 86\n",
      "Saved\n",
      "Processing: 9, interval 87\n",
      "Saved\n",
      "Processing: 9, interval 88\n",
      "Saved\n",
      "Processing: 9, interval 89\n",
      "Saved\n",
      "Processing: 9, interval 90\n",
      "Saved\n",
      "Processing: 9, interval 91\n",
      "Saved\n",
      "Processing: 9, interval 92\n",
      "Saved\n",
      "Processing: 9, interval 93\n",
      "Saved\n",
      "Processing: 9, interval 94\n",
      "Saved\n",
      "Processing: 9, interval 95\n",
      "Saved\n",
      "Processing: 9, interval 96\n",
      "Saved\n",
      "Processing: 9, interval 97\n",
      "Saved\n",
      "Processing: 9, interval 98\n",
      "Saved\n",
      "Processing: 9, interval 99\n",
      "Saved\n",
      "Processing: 9, interval 100\n",
      "Saved\n",
      "Processing: 9, interval 101\n",
      "Saved\n",
      "Processing: 9, interval 102\n",
      " at 9, interval 102. Retrying...\n",
      "Processing: 9, interval 102\n",
      "Saved\n",
      "Processing: 9, interval 103\n",
      "Saved\n",
      "Processing: 9, interval 104\n",
      "Saved\n",
      "Processing: 9, interval 105\n",
      "Saved\n",
      "Processing: 9, interval 106\n",
      "Saved\n",
      "Processing: 9, interval 107\n",
      "Saved\n",
      "Processing: 9, interval 108\n",
      "Saved\n",
      "Processing: 9, interval 109\n",
      "Saved\n",
      "Processing: 9, interval 110\n",
      "Saved\n",
      "Processing: 9, interval 111\n",
      "Saved\n",
      "Processing: 9, interval 112\n",
      "Saved\n",
      "Processing: 9, interval 113\n",
      "Saved\n",
      "Processing: 9, interval 114\n",
      "Saved\n",
      "Processing: 9, interval 115\n",
      "Saved\n",
      "Processing: 9, interval 116\n",
      "Saved\n",
      "Processing: 9, interval 117\n",
      "Saved\n",
      "Processing: 9, interval 118\n",
      "Saved\n",
      "Processing: 9, interval 119\n",
      "Saved\n",
      "Processing: 9, interval 120\n",
      "Saved\n",
      "Processing: 9, interval 121\n",
      "Saved\n",
      "Processing: 9, interval 122\n",
      "Saved\n",
      "Processing: 9, interval 123\n",
      "Saved\n",
      "Processing: 9, interval 124\n",
      "Saved\n",
      "Processing: 9, interval 125\n",
      "Saved\n",
      "Processing: 9, interval 126\n",
      "Saved\n",
      "Processing: 9, interval 127\n",
      "Saved\n",
      "Processing: 9, interval 128\n",
      "Saved\n",
      "Processing: 9, interval 129\n",
      "Saved\n",
      "Processing: 9, interval 130\n",
      "Saved\n",
      "Processing: 9, interval 131\n",
      "Saved\n",
      "Processing: 9, interval 132\n",
      "Saved\n",
      "Processing: 9, interval 133\n",
      "Saved\n",
      "Processing: 9, interval 134\n",
      "Saved\n",
      "Processing: 9, interval 135\n",
      "Saved\n",
      "Processing: 9, interval 136\n",
      "Saved\n",
      "Processing: 9, interval 137\n",
      "Saved\n",
      "Processing: 9, interval 138\n",
      " at 9, interval 138. Retrying...\n",
      "Processing: 9, interval 138\n",
      "Saved\n",
      "Processing: 9, interval 139\n",
      " at 9, interval 139. Retrying...\n",
      "Processing: 9, interval 139\n",
      "Saved\n",
      "Processing: 9, interval 140\n",
      "Saved\n",
      "Processing: 9, interval 141\n",
      " at 9, interval 141. Retrying...\n",
      "Processing: 9, interval 141\n",
      " at 9, interval 141. Retrying...\n",
      "Processing: 9, interval 141\n",
      "Saved\n",
      "Processing: 9, interval 142\n",
      "Saved\n",
      "Processing: 9, interval 143\n",
      " at 9, interval 143. Retrying...\n",
      "Processing: 9, interval 143\n",
      "Saved\n",
      "Processing: 9, interval 144\n",
      "Saved\n",
      "Processing: 9, interval 145\n",
      "Saved\n",
      "Processing: 9, interval 146\n",
      "Saved\n",
      "Processing: 9, interval 147\n",
      "Saved\n",
      "Processing: 9, interval 148\n",
      "Saved\n",
      "Processing: 9, interval 149\n",
      "Saved\n",
      "Processing: 9, interval 150\n",
      "Saved\n",
      "Processing: 9, interval 151\n",
      "Saved\n",
      "Processing: 9, interval 152\n",
      "Saved\n",
      "Processing: 9, interval 153\n",
      "Saved\n",
      "Processing: 9, interval 154\n",
      "Saved\n",
      "Processing: 9, interval 155\n",
      "Saved\n",
      "Processing: 9, interval 156\n",
      "Saved\n",
      "Processing: 9, interval 157\n",
      "Saved\n",
      "Processing: 9, interval 158\n",
      "Saved\n",
      "Processing: 9, interval 159\n",
      "Saved\n",
      "Processing: 9, interval 160\n",
      "Saved\n",
      "Processing: 9, interval 161\n",
      "Saved\n",
      "Processing: 9, interval 162\n",
      "Saved\n",
      "Processing: X, interval 0\n",
      "Saved\n",
      "Processing: X, interval 1\n",
      "Saved\n",
      "Processing: X, interval 2\n",
      "Saved\n",
      "Processing: X, interval 3\n",
      "Saved\n",
      "Processing: X, interval 4\n",
      "Saved\n",
      "Processing: X, interval 5\n",
      "Saved\n",
      "Processing: X, interval 6\n",
      "Saved\n",
      "Processing: X, interval 7\n",
      "Saved\n",
      "Processing: X, interval 8\n",
      "Saved\n",
      "Processing: X, interval 9\n",
      "Saved\n",
      "Processing: X, interval 10\n",
      "Saved\n",
      "Processing: X, interval 11\n",
      "Saved\n",
      "Processing: X, interval 12\n",
      "Saved\n",
      "Processing: X, interval 13\n",
      "Saved\n",
      "Processing: X, interval 14\n",
      "Saved\n",
      "Processing: X, interval 15\n",
      "Saved\n",
      "Processing: X, interval 16\n",
      "Saved\n",
      "Processing: X, interval 17\n",
      "Saved\n",
      "Processing: X, interval 18\n",
      "Saved\n",
      "Processing: X, interval 19\n",
      "Saved\n",
      "Processing: X, interval 20\n",
      "Saved\n",
      "Processing: X, interval 21\n",
      "Saved\n",
      "Processing: X, interval 22\n",
      "Saved\n",
      "Processing: X, interval 23\n",
      "Saved\n",
      "Processing: X, interval 24\n",
      "Saved\n",
      "Processing: X, interval 25\n",
      "Saved\n",
      "Processing: X, interval 26\n",
      "Saved\n",
      "Processing: X, interval 27\n",
      "Saved\n",
      "Processing: X, interval 28\n",
      "Saved\n",
      "Processing: X, interval 29\n",
      "Saved\n",
      "Processing: X, interval 30\n",
      "Saved\n",
      "Processing: X, interval 31\n",
      "Saved\n",
      "Processing: X, interval 32\n",
      "Saved\n",
      "Processing: X, interval 33\n",
      "Saved\n",
      "Processing: X, interval 34\n",
      " at X, interval 34. Retrying...\n",
      "Processing: X, interval 34\n",
      "Saved\n",
      "Processing: X, interval 35\n",
      " at X, interval 35. Retrying...\n",
      "Processing: X, interval 35\n",
      "Saved\n",
      "Processing: X, interval 36\n",
      " at X, interval 36. Retrying...\n",
      "Processing: X, interval 36\n",
      "Saved\n",
      "Processing: X, interval 37\n",
      " at X, interval 37. Retrying...\n",
      "Processing: X, interval 37\n",
      "Saved\n",
      "Processing: X, interval 38\n",
      " at X, interval 38. Retrying...\n",
      "Processing: X, interval 38\n",
      "Saved\n",
      "Processing: X, interval 39\n",
      "Saved\n",
      "Processing: X, interval 40\n",
      "Saved\n",
      "Processing: X, interval 41\n",
      "Saved\n",
      "Processing: X, interval 42\n",
      "Saved\n",
      "Processing: X, interval 43\n",
      "Saved\n",
      "Processing: X, interval 44\n",
      "Saved\n",
      "Processing: X, interval 45\n",
      "Saved\n",
      "Processing: X, interval 46\n",
      "Saved\n",
      "Processing: X, interval 47\n",
      "Saved\n",
      "Processing: X, interval 48\n",
      "Saved\n",
      "Processing: X, interval 49\n",
      " at X, interval 49. Retrying...\n",
      "Processing: X, interval 49\n",
      "Saved\n",
      "Processing: X, interval 50\n",
      "Saved\n",
      "Processing: X, interval 51\n",
      "Saved\n",
      "Processing: X, interval 52\n",
      "Saved\n",
      "Processing: X, interval 53\n",
      "Saved\n",
      "Processing: X, interval 54\n",
      "Saved\n",
      "Processing: X, interval 55\n",
      "Saved\n",
      "Processing: X, interval 56\n",
      "Saved\n",
      "Processing: X, interval 57\n",
      "Saved\n",
      "Processing: X, interval 58\n",
      "Saved\n",
      "Processing: X, interval 59\n",
      "Saved\n",
      "Processing: X, interval 60\n",
      "Saved\n",
      "Processing: X, interval 61\n",
      "Saved\n",
      "Processing: X, interval 62\n",
      "Saved\n",
      "Processing: X, interval 63\n",
      "Saved\n",
      "Processing: X, interval 64\n",
      "Saved\n",
      "Processing: X, interval 65\n",
      "Saved\n",
      "Processing: X, interval 66\n",
      "Saved\n",
      "Processing: X, interval 67\n",
      "Saved\n",
      "Processing: X, interval 68\n",
      "Saved\n",
      "Processing: X, interval 69\n",
      "Saved\n",
      "Processing: X, interval 70\n",
      "Saved\n",
      "Processing: X, interval 71\n",
      "Saved\n",
      "Processing: X, interval 72\n",
      "Saved\n",
      "Processing: X, interval 73\n",
      "Saved\n",
      "Processing: X, interval 74\n",
      "Saved\n",
      "Processing: X, interval 75\n",
      "Saved\n",
      "Processing: X, interval 76\n",
      "Saved\n",
      "Processing: X, interval 77\n",
      "Saved\n",
      "Processing: X, interval 78\n",
      "Saved\n",
      "Processing: X, interval 79\n",
      "Saved\n",
      "Processing: X, interval 80\n",
      "Saved\n",
      "Processing: X, interval 81\n",
      "Saved\n",
      "Processing: X, interval 82\n",
      "Saved\n",
      "Processing: X, interval 83\n",
      "Saved\n",
      "Processing: X, interval 84\n",
      "Saved\n",
      "Processing: X, interval 85\n",
      "Saved\n",
      "Processing: X, interval 86\n",
      "Saved\n",
      "Processing: X, interval 87\n",
      "Saved\n",
      "Processing: X, interval 88\n",
      "Saved\n",
      "Processing: X, interval 89\n",
      "Saved\n",
      "Processing: X, interval 90\n",
      "Saved\n",
      "Processing: X, interval 91\n",
      "Saved\n",
      "Processing: X, interval 92\n",
      "Saved\n",
      "Processing: X, interval 93\n",
      "Saved\n",
      "Processing: X, interval 94\n",
      "Saved\n",
      "Processing: X, interval 95\n",
      "Saved\n",
      "Processing: X, interval 96\n",
      "Saved\n",
      "Processing: X, interval 97\n",
      "Saved\n",
      "Processing: X, interval 98\n",
      "Saved\n",
      "Processing: X, interval 99\n",
      "Saved\n",
      "Processing: X, interval 100\n",
      "Saved\n",
      "Processing: X, interval 101\n",
      "Saved\n",
      "Processing: X, interval 102\n",
      "Saved\n",
      "Processing: X, interval 103\n",
      "Saved\n",
      "Processing: X, interval 104\n",
      "Saved\n",
      "Processing: X, interval 105\n",
      "Saved\n",
      "Processing: X, interval 106\n",
      "Saved\n",
      "Processing: X, interval 107\n",
      "Saved\n",
      "Processing: X, interval 108\n",
      "Saved\n",
      "Processing: X, interval 109\n",
      "Saved\n",
      "Processing: X, interval 110\n",
      "Saved\n",
      "Processing: X, interval 111\n",
      "Saved\n",
      "Processing: X, interval 112\n",
      " at X, interval 112. Retrying...\n",
      "Processing: X, interval 112\n",
      "Saved\n",
      "Processing: X, interval 113\n",
      " at X, interval 113. Retrying...\n",
      "Processing: X, interval 113\n",
      "Saved\n",
      "Processing: X, interval 114\n",
      "Saved\n",
      "Processing: X, interval 115\n",
      " at X, interval 115. Retrying...\n",
      "Processing: X, interval 115\n",
      " at X, interval 115. Retrying...\n",
      "Processing: X, interval 115\n",
      "Saved\n",
      "Processing: X, interval 116\n",
      "Saved\n",
      "Processing: X, interval 117\n",
      "Saved\n",
      "Processing: X, interval 118\n",
      " at X, interval 118. Retrying...\n",
      "Processing: X, interval 118\n",
      "Saved\n",
      "Processing: X, interval 119\n",
      "Saved\n",
      "Processing: X, interval 120\n",
      "Saved\n",
      "Processing: X, interval 121\n",
      " at X, interval 121. Retrying...\n",
      "Processing: X, interval 121\n",
      "Saved\n",
      "Processing: X, interval 122\n",
      "Saved\n",
      "Processing: X, interval 123\n",
      "Saved\n",
      "Processing: X, interval 124\n",
      "Saved\n",
      "Processing: X, interval 125\n",
      "Saved\n",
      "Processing: X, interval 126\n",
      "Saved\n",
      "Processing: X, interval 127\n",
      " at X, interval 127. Retrying...\n",
      "Processing: X, interval 127\n",
      "Saved\n",
      "Processing: X, interval 128\n",
      " at X, interval 128. Retrying...\n",
      "Processing: X, interval 128\n",
      "Saved\n",
      "Processing: X, interval 129\n",
      "Saved\n",
      "Processing: X, interval 130\n",
      "Saved\n",
      "Processing: X, interval 131\n",
      " at X, interval 131. Retrying...\n",
      "Processing: X, interval 131\n",
      "Saved\n",
      "Processing: X, interval 132\n",
      "Saved\n",
      "Processing: X, interval 133\n",
      "Saved\n",
      "Processing: X, interval 134\n",
      "Saved\n",
      "Processing: X, interval 135\n",
      "Saved\n",
      "Processing: X, interval 136\n",
      "Saved\n",
      "Processing: X, interval 137\n",
      " at X, interval 137. Retrying...\n",
      "Processing: X, interval 137\n",
      " at X, interval 137. Retrying...\n",
      "Processing: X, interval 137\n",
      "Saved\n",
      "Processing: X, interval 138\n",
      "Saved\n",
      "Processing: X, interval 139\n",
      " at X, interval 139. Retrying...\n",
      "Processing: X, interval 139\n",
      "Saved\n",
      "Processing: X, interval 140\n",
      "Saved\n",
      "Processing: X, interval 141\n",
      "Saved\n",
      "Processing: X, interval 142\n",
      "Saved\n",
      "Processing: X, interval 143\n",
      "Saved\n",
      "Processing: X, interval 144\n",
      "Saved\n",
      "Processing: X, interval 145\n",
      " at X, interval 145. Retrying...\n",
      "Processing: X, interval 145\n",
      "Saved\n",
      "Processing: X, interval 146\n",
      "Saved\n",
      "Processing: X, interval 147\n",
      "Saved\n",
      "Processing: X, interval 148\n",
      "Saved\n",
      "Processing: X, interval 149\n",
      "Saved\n",
      "Processing: X, interval 150\n",
      "Saved\n",
      "Processing: X, interval 151\n",
      " at X, interval 151. Retrying...\n",
      "Processing: X, interval 151\n",
      "Saved\n",
      "Processing: X, interval 152\n",
      "Saved\n",
      "Processing: X, interval 153\n",
      " at X, interval 153. Retrying...\n",
      "Processing: X, interval 153\n",
      "Saved\n",
      "Processing: X, interval 154\n",
      "Saved\n",
      "Processing: X, interval 155\n",
      "Saved\n",
      "Processing: X, interval 156\n",
      "Saved\n",
      "Processing: X, interval 157\n",
      "Saved\n",
      "Processing: X, interval 158\n",
      "Saved\n",
      "Processing: X, interval 159\n",
      "Saved\n",
      "Processing: X, interval 160\n",
      "Saved\n",
      "Processing: X, interval 161\n",
      "Saved\n",
      "Processing: X, interval 162\n",
      " at X, interval 162. Retrying...\n",
      "Processing: X, interval 162\n",
      "Saved\n",
      "Processing: X, interval 163\n",
      "Saved\n",
      "Processing: X, interval 164\n",
      "Saved\n",
      "Processing: X, interval 165\n",
      "Saved\n",
      "Processing: X, interval 166\n",
      "Saved\n",
      "Processing: X, interval 167\n",
      " at X, interval 167. Retrying...\n",
      "Processing: X, interval 167\n",
      "Saved\n",
      "Processing: X, interval 168\n",
      "Saved\n",
      "Processing: X, interval 169\n",
      " at X, interval 169. Retrying...\n",
      "Processing: X, interval 169\n",
      "Saved\n",
      "Processing: X, interval 170\n",
      "Saved\n",
      "Processing: X, interval 171\n",
      "Saved\n",
      "Processing: X, interval 172\n",
      "Saved\n",
      "Processing: X, interval 173\n",
      "Saved\n",
      "Processing: X, interval 174\n",
      "Saved\n",
      "Processing: X, interval 175\n",
      "Saved\n",
      "Processing: X, interval 176\n",
      "Saved\n",
      "Processing: X, interval 177\n",
      "Saved\n",
      "Processing: X, interval 178\n",
      "Saved\n",
      "Processing: X, interval 179\n",
      "Saved\n",
      "Processing: X, interval 180\n",
      "Saved\n",
      "Processing: X, interval 181\n",
      "Saved\n",
      "Processing: X, interval 182\n",
      "Saved\n",
      "Processing: X, interval 183\n",
      "Saved\n",
      "Processing: X, interval 184\n",
      "Saved\n",
      "Processing: X, interval 185\n",
      "Saved\n",
      "Processing: X, interval 186\n",
      "Saved\n",
      "Processing: X, interval 187\n",
      "Saved\n",
      "Processing: X, interval 188\n",
      "Saved\n",
      "Processing: X, interval 189\n",
      "Saved\n",
      "Processing: X, interval 190\n",
      "Saved\n",
      "Processing: X, interval 191\n",
      "Saved\n",
      "Processing: X, interval 192\n",
      "Saved\n",
      "Processing: X, interval 193\n",
      "Saved\n",
      "Processing: X, interval 194\n",
      "Saved\n",
      "Processing: X, interval 195\n",
      "Saved\n",
      "Processing: X, interval 196\n",
      "Saved\n",
      "Processing: X, interval 197\n",
      "Saved\n",
      "Processing: X, interval 198\n",
      "Saved\n",
      "Processing: X, interval 199\n",
      "Saved\n",
      "Processing: X, interval 200\n",
      "Saved\n",
      "Processing: X, interval 201\n",
      "Saved\n",
      "Processing: X, interval 202\n",
      "Saved\n",
      "Processing: X, interval 203\n",
      "Saved\n",
      "Processing: X, interval 204\n",
      "Saved\n",
      "Processing: X, interval 205\n",
      "Saved\n",
      "Processing: X, interval 206\n",
      "Saved\n",
      "Processing: X, interval 207\n",
      "Saved\n",
      "Processing: X, interval 208\n",
      "Saved\n",
      "Processing: X, interval 209\n",
      "Saved\n",
      "Processing: X, interval 210\n",
      "Saved\n",
      "Processing: X, interval 211\n",
      "Saved\n",
      "Processing: X, interval 212\n",
      "Saved\n",
      "Processing: X, interval 213\n",
      "Saved\n",
      "Processing: X, interval 214\n",
      "Saved\n",
      "Processing: X, interval 215\n",
      "Saved\n",
      "Processing: X, interval 216\n",
      "Saved\n",
      "Processing: X, interval 217\n",
      "Saved\n",
      "Processing: X, interval 218\n",
      "Saved\n",
      "Processing: X, interval 219\n",
      "Saved\n",
      "Processing: X, interval 220\n",
      "Saved\n",
      "Processing: X, interval 221\n",
      "Saved\n",
      "Processing: X, interval 222\n",
      "Saved\n",
      "Processing: X, interval 223\n",
      "Saved\n",
      "Processing: X, interval 224\n",
      "Saved\n",
      "Processing: X, interval 225\n",
      " at X, interval 225. Retrying...\n",
      "Processing: X, interval 225\n",
      "Saved\n",
      "Processing: X, interval 226\n",
      "Saved\n",
      "Processing: X, interval 227\n",
      "Saved\n",
      "Processing: X, interval 228\n",
      "Saved\n",
      "Processing: X, interval 229\n",
      "Saved\n",
      "Processing: X, interval 230\n",
      "Saved\n",
      "Processing: X, interval 231\n",
      "Saved\n",
      "Processing: X, interval 232\n",
      "Saved\n",
      "Processing: X, interval 233\n",
      "Saved\n",
      "Processing: X, interval 234\n",
      "Saved\n",
      "Processing: X, interval 235\n",
      "Saved\n",
      "Processing: X, interval 236\n",
      "Saved\n",
      "Processing: X, interval 237\n",
      "Saved\n",
      "Processing: X, interval 238\n",
      "Saved\n",
      "Processing: X, interval 239\n",
      "Saved\n",
      "Processing: X, interval 240\n",
      "Saved\n",
      "Processing: X, interval 241\n",
      "Saved\n",
      "Processing: X, interval 242\n",
      "Saved\n",
      "Processing: X, interval 243\n",
      "Saved\n",
      "Processing: X, interval 244\n",
      "Saved\n",
      "Processing: X, interval 245\n",
      "Saved\n",
      "Processing: X, interval 246\n",
      "Saved\n",
      "Processing: X, interval 247\n",
      "Saved\n",
      "Processing: X, interval 248\n",
      "Saved\n",
      "Processing: X, interval 249\n",
      "Saved\n",
      "Processing: X, interval 250\n",
      "Saved\n",
      "Processing: X, interval 251\n",
      "Saved\n",
      "Processing: X, interval 252\n",
      "Saved\n",
      "Processing: X, interval 253\n",
      "Saved\n",
      "Processing: X, interval 254\n",
      "Saved\n",
      "Processing: X, interval 255\n",
      "Saved\n",
      "Processing: X, interval 256\n",
      "Saved\n",
      "Processing: X, interval 257\n",
      "Saved\n",
      "Processing: X, interval 258\n",
      "Saved\n",
      "Processing: X, interval 259\n",
      "Saved\n",
      "Processing: X, interval 260\n",
      "Saved\n",
      "Processing: X, interval 261\n",
      "Saved\n",
      "Processing: X, interval 262\n",
      "Saved\n",
      "Processing: X, interval 263\n",
      "Saved\n",
      "Processing: X, interval 264\n",
      "Saved\n",
      "Processing: X, interval 265\n",
      "Saved\n",
      "Processing: X, interval 266\n",
      "Saved\n",
      "Processing: X, interval 267\n",
      "Saved\n",
      "Processing: X, interval 268\n",
      "Saved\n",
      "Processing: X, interval 269\n",
      "Saved\n",
      "Processing: X, interval 270\n",
      "Saved\n",
      "Processing: X, interval 271\n",
      "Saved\n",
      "Processing: X, interval 272\n",
      "Saved\n",
      "Processing: X, interval 273\n",
      "Saved\n",
      "Processing: X, interval 274\n",
      "Saved\n",
      "Processing: X, interval 275\n",
      "Saved\n",
      "Processing: X, interval 276\n",
      "Saved\n",
      "Processing: X, interval 277\n",
      "Saved\n",
      "Processing: X, interval 278\n",
      "Saved\n",
      "Processing: X, interval 279\n",
      "Saved\n",
      "Processing: X, interval 280\n",
      "Saved\n",
      "Processing: X, interval 281\n",
      "Saved\n",
      "Processing: X, interval 282\n",
      "Saved\n",
      "Processing: X, interval 283\n",
      "Saved\n",
      "Processing: X, interval 284\n",
      "Saved\n",
      "Processing: X, interval 285\n",
      "Saved\n",
      "Processing: X, interval 286\n",
      "Saved\n",
      "Processing: X, interval 287\n",
      "Saved\n",
      "Processing: X, interval 288\n",
      "Saved\n",
      "Processing: X, interval 289\n",
      "Saved\n",
      "Processing: X, interval 290\n",
      "Saved\n",
      "Processing: X, interval 291\n",
      "Saved\n",
      "Processing: X, interval 292\n",
      "Saved\n",
      "Processing: Y, interval 0\n",
      "Saved\n",
      "Processing: Y, interval 1\n",
      "Saved\n",
      "Processing: Y, interval 2\n",
      "Saved\n",
      "Processing: Y, interval 3\n",
      "Saved\n",
      "Processing: Y, interval 4\n",
      "Saved\n",
      "Processing: Y, interval 5\n",
      "Saved\n",
      "Processing: Y, interval 6\n",
      "Saved\n",
      "Processing: Y, interval 7\n",
      "Saved\n",
      "Processing: Y, interval 8\n",
      "Saved\n",
      "Processing: Y, interval 9\n",
      "Saved\n",
      "Processing: Y, interval 10\n",
      "Saved\n",
      "Processing: Y, interval 11\n",
      "Saved\n",
      "Processing: Y, interval 12\n",
      "Saved\n",
      "Processing: Y, interval 13\n",
      "Saved\n",
      "Processing: Y, interval 14\n",
      "Saved\n",
      "Processing: Y, interval 15\n",
      "Saved\n",
      "Processing: Y, interval 16\n",
      "Saved\n",
      "Processing: Y, interval 17\n",
      "Saved\n",
      "Processing: Y, interval 18\n",
      "Saved\n",
      "Processing: Y, interval 19\n",
      "Saved\n",
      "Processing: Y, interval 20\n",
      "Saved\n",
      "Processing: Y, interval 21\n",
      "Saved\n",
      "Processing: Y, interval 22\n",
      "Saved\n",
      "Processing: Y, interval 23\n",
      "Saved\n",
      "Processing: Y, interval 24\n",
      "Saved\n",
      "Processing: Y, interval 25\n",
      "Saved\n",
      "Processing: Y, interval 26\n",
      "Saved\n",
      "Processing: Y, interval 27\n",
      "Saved\n",
      "Processing: Y, interval 28\n",
      "Saved\n",
      "Processing: Y, interval 29\n",
      "Saved\n",
      "Processing: Y, interval 30\n",
      "Saved\n",
      "Processing: Y, interval 31\n",
      "Saved\n"
     ]
    }
   ],
   "source": [
    "#query gnomAD, repeat attampts automaticly\n",
    "\n",
    "#for chromosome, intervals in gnomAD_qgr.items():\n",
    "#    for i,interval in enumerate(intervals):\n",
    "#        print(chromosome,i)\n",
    "#        time.sleep(10) #not to overwhelm the server\n",
    "#        result=await check_saavs_bygene(chromosome,list(range(interval[0],interval[1]+1)))\n",
    "#        gnomAD_qgr_res[chromosome].update(result)\n",
    "#        #save dict\n",
    "#        pickle.dump(gnomAD_qgr_res,open(\"gnomAD_qgr_res.pickle\",\"wb\"))\n",
    "#        print(\"saved\")\n",
    "\n",
    "\n",
    "# Load previous results if they exist\n",
    "try:\n",
    "    gnomAD_qgr_res = pickle.load(open(\"gnomAD_qgr_res.pickle\", \"rb\"))\n",
    "except FileNotFoundError:\n",
    "    gnomAD_qgr_res = {chrom: {} for chrom in gnomAD_qgr}\n",
    "\n",
    "# Track progress\n",
    "progress_path = \"gnomAD_qgr_progress.pickle\"\n",
    "if os.path.exists(progress_path):\n",
    "    completed = pickle.load(open(progress_path, \"rb\"))\n",
    "else:\n",
    "    completed = {chrom: set() for chrom in gnomAD_qgr}\n",
    "\n",
    "for chromosome, intervals in gnomAD_qgr.items():\n",
    "    if chromosome not in gnomAD_qgr_res:\n",
    "        gnomAD_qgr_res[chromosome] = {}\n",
    "    if chromosome not in completed:\n",
    "        completed[chromosome] = set()\n",
    "        \n",
    "    i = 0\n",
    "    while i < len(intervals):\n",
    "        if i in completed[chromosome]:\n",
    "            i += 1\n",
    "            continue\n",
    "        interval = intervals[i]\n",
    "        try:\n",
    "            print(f\"Processing: {chromosome}, interval {i}\")\n",
    "            time.sleep(20)  # Not to overwhelm the server\n",
    "            result = await check_saavs_bygene(chromosome, list(range(interval[0], interval[1] + 1)))\n",
    "            gnomAD_qgr_res[chromosome].update(result)\n",
    "\n",
    "            # Save result and progress\n",
    "            pickle.dump(gnomAD_qgr_res, open(\"gnomAD_qgr_res.pickle\", \"wb\"))\n",
    "            completed[chromosome].add(i)\n",
    "            pickle.dump(completed, open(progress_path, \"wb\"))\n",
    "\n",
    "            print(\"Saved\")\n",
    "            i += 1  # Only increment if successful\n",
    "        except Exception as e:\n",
    "            print(f\"{e} at {chromosome}, interval {i}. Retrying...\")\n",
    "            # Will retry same i\n",
    "            continue\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 370,
   "id": "cf52b200-ffd7-4178-b81c-eab08c6465db",
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomAD_qgr_res = pickle.load(open(\"gnomAD_qgr_res.pickle\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 384,
   "id": "4e135efb-564a-4cbc-8122-fef781d92bf1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected: 23049\n",
      "got: 23049\n"
     ]
    }
   ],
   "source": [
    "#quality check: are all chromosomes and positions were searched\n",
    "\n",
    "print(\"expected:\", np.sum([abs(interval[0]-interval[1])+1 for chrom,intervals in gnomAD_qgr.items() for interval in intervals] ))\n",
    "print(\"got:\", np.sum([len(vals) for chrom,vals in gnomAD_qgr_res.items() ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "482c40c2-00e3-4e5f-b8cd-32bcb505d658",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(669, 669, np.int64(2727), 2727)"
      ]
     },
     "execution_count": 383,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(completed[\"1\"]),len(gnomAD_qgr[\"1\"]),np.sum([abs(interval[0]-interval[1])+1  for interval in gnomAD_qgr[\"1\"]]),len(gnomAD_qgr_res[\"1\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 388,
   "id": "044a7a30-53f4-4700-aedd-aee3859b8e5a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "empty positions: 0.29263742461712006\n"
     ]
    }
   ],
   "source": [
    "#how many positions are empty\n",
    "\n",
    "print(\"empty positions:\", np.sum([1 for chrom,vals in gnomAD_qgr_res.items() for v in vals.values() if len(v)==0])/np.sum([len(vals) for chrom,vals in gnomAD_qgr_res.items() ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "id": "79e283b3-690a-4ac6-a624-6f8f3f4a072b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n of variants in total: 27170\n"
     ]
    }
   ],
   "source": [
    "#how many variants at all\n",
    "\n",
    "print(\"n of variants in total:\", np.sum([len(v) for chrom,vals in gnomAD_qgr_res.items() for v in vals.values()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ecf79317-58e3-4d35-88f4-b1faa78ea915",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parce_variantid(x):\n",
    "    \"\"\"i=0 chromosome\n",
    "    i=1 genomic coordinate\n",
    "    i=2 reference bp\n",
    "    i=3 variant bp\"\"\"\n",
    "    return x.split(\"-\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "id": "c147562b-005b-4644-9315-6e8a96541cf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n of indels in total: 1655\n"
     ]
    }
   ],
   "source": [
    "#how many indels - complicated case\n",
    "\n",
    "print(\"n of indels in total:\", np.sum([1 for chrom,vals in gnomAD_qgr_res.items() \n",
    "                                       for v in vals.values() if len(v)!=0\n",
    "                                       for var in v if len(parce_variantid(var['variant_id'])[2])>1 or len(parce_variantid(var['variant_id'])[3])>1 ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 417,
   "id": "3627072b-65a8-4d4e-92bd-460f26449ad5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1484"
      ]
     },
     "execution_count": 417,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many indels by position\n",
    "\n",
    "len(set([k for chrom,vals in gnomAD_qgr_res.items() \n",
    "                                       for k,v in vals.items() if len(v)!=0\n",
    "                                       for var in v if len(parce_variantid(var['variant_id'])[2])>1 or len(parce_variantid(var['variant_id'])[3])>1 ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 418,
   "id": "195951a5-6b4d-4940-ae92-e9a681f8120c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.07180354896090936"
      ]
     },
     "execution_count": 418,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1655/23049"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "id": "64ca077f-3cfd-4258-a934-e724f739a319",
   "metadata": {},
   "outputs": [],
   "source": [
    "#indels=[var['variant_id'] for chrom,vals in gnomAD_qgr_res.items() \n",
    "#                                       for v in vals.values() if len(v)!=0\n",
    "#                                       for var in v if len(parce_variantid(var['variant_id'],2))>1 or len(parce_variantid(var['variant_id'],3))>1 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4b964f1-0a1a-46fc-aea9-b1e1ef29f70f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 454,
   "id": "5368b511-0d99-4dd7-97ff-21476bc5fa11",
   "metadata": {},
   "outputs": [],
   "source": [
    "#facke header, put it manualy in file\n",
    "\"\"\"##fileformat=VCFv4.2\n",
    "##fileDate=20090805\n",
    "##source=myImputationProgramV3.1\n",
    "##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta\n",
    "##contig=<ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species=\"Homo sapiens\",taxonomy=x>\n",
    "##phasing=partial\n",
    "##INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of Samples With Data\">\n",
    "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Total Depth\">\n",
    "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency\">\n",
    "##INFO=<ID=AA,Number=1,Type=String,Description=\"Ancestral Allele\">\n",
    "##INFO=<ID=DB,Number=0,Type=Flag,Description=\"dbSNP membership, build 129\">\n",
    "##INFO=<ID=H2,Number=0,Type=Flag,Description=\"HapMap2 membership\">\n",
    "##FILTER=<ID=q10,Description=\"Quality below 10\">\n",
    "##FILTER=<ID=s50,Description=\"Less than 50% of samples have data\">\n",
    "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
    "##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">\n",
    "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Read Depth\">\n",
    "##FORMAT=<ID=HQ,Number=2,Type=Integer,Description=\"Haplotype Quality\">\n",
    "#CHROM POS ID REF ALT\"\"\"\n",
    "#20 14370 rs6054257 G A\n",
    "VCFdf=pd.DataFrame(columns=[\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\"])\n",
    "for chrom,vals in gnomAD_qgr_res.items():\n",
    "    for position,variants in vals.items():\n",
    "        if len(variants)==0: continue\n",
    "        #alts=[]\n",
    "        for variant in variants:\n",
    "            chrom,pos,ref,alt = parce_variantid(variant['variant_id'])\n",
    "            #alts.append(alt)\n",
    "            VCFdf.loc[len(VCFdf),[\"CHROM\", \"POS\", \"REF\", \"ALT\"]]=[chrom,pos,ref,alt]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 455,
   "id": "909d9b9e-9d82-4922-a076-8c001433befb",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCFdf[\"ID\"]=\".\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 456,
   "id": "93306723-9673-4f5c-b585-4f7ea3087492",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCFdf.to_csv(\"massshift_saav.vcf\",sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 766,
   "id": "2b5a7625-5d1c-4bb8-b30e-048a2e392f51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27170"
      ]
     },
     "execution_count": 766,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(VCFdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 460,
   "id": "681b17c7-5adb-4666-8124-5f15bf2dd421",
   "metadata": {},
   "outputs": [],
   "source": [
    "#group by reference\n",
    "\n",
    "VCFdf_gr=pd.DataFrame(columns=[\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\"])\n",
    "for (chrom,pos,ref), group in VCFdf.groupby([\"CHROM\",\"POS\",\"REF\"]):\n",
    "    VCFdf_gr.loc[len(VCFdf_gr),[\"CHROM\", \"POS\", \"REF\", \"ALT\"]]=[chrom,pos,ref,\",\".join(group.ALT.unique())]\n",
    "VCFdf_gr[\"ID\"]=\".\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 770,
   "id": "b1383e79-6c2b-4b0c-93e3-5f3341fc1968",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17339"
      ]
     },
     "execution_count": 770,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(VCFdf_gr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 504,
   "id": "901e950e-7c05-4504-9905-18a4671077e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[39951299, 39951646]]\n"
     ]
    }
   ],
   "source": [
    "#x=comb_clun_isosub_filt_exp.iloc[0]\n",
    "#if isinstance(x[\"saav_info_triplet\"][0][0],list):\n",
    "#    print(OPcoords[(OPcoords.accessions==x[\"leadprot\"])|(OPcoords.accession==x[\"leadprot\"])].iloc[0].segments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 544,
   "id": "ab5ff209-983e-4be4-9ae1-9484cc6c03d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add segments info\n",
    "\n",
    "comb_clun_isosub_filt_tmp=comb_clun_isosub_filt.copy(deep=True)\n",
    "comb_clun_isosub_filt_tmp[\"segments\"]=comb_clun_isosub_filt_tmp.apply(lambda x: OPcoords[(OPcoords.accessions==x[\"leadprot\"])|(OPcoords.accession==x[\"leadprot\"])].iloc[0].segments\n",
    "                                                                     if isinstance(x[\"saav_info_triplet\"][0][0],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 545,
   "id": "6fe2cdb7-c705-4e45-bf67-51d055791c18",
   "metadata": {},
   "outputs": [],
   "source": [
    "#add cds segments\n",
    "\n",
    "comb_clun_isosub_filt_tmp[\"cds_segments\"]=comb_clun_isosub_filt_tmp.apply(lambda x: OPcoords[(OPcoords.accessions==x[\"leadprot\"])|(OPcoords.accession==x[\"leadprot\"])].iloc[0].cds_segments\n",
    "                                                                     if isinstance(x[\"saav_info_triplet\"][0][0],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 563,
   "id": "ee2d1d14-ace0-488d-a28c-cabca8bc25ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_segment_fromfasta(poss,chromosome,ENSfasta): #strand\n",
    "    if isinstance(poss,float): return np.nan\n",
    "    seq=str(ENSfasta[chromosome].seq)[poss[0]:poss[1]+1]\n",
    "    return seq\n",
    "def get_gencds(x,FASTA):\n",
    "    #print(x.name)\n",
    "    #print(x[\"segments\"])\n",
    "    segments=x[\"segments\"];chromosome=x[\"chromosome\"];strand=x[\"strand\"]\n",
    "    gencds=\"\".join([\"\".join(list(get_segment_fromfasta(segment,chromosome,ENSfasta))[::-1]) \n",
    "                    if strand==-1 else\n",
    "                    get_segment_fromfasta(segment,chromosome,ENSfasta)\n",
    "                    for segment in segments])\n",
    "    #positions=[list(range(segment[0],segment[1]+1)) for segment in segments]\n",
    "    #positions=[e for l in positions for e in l]\n",
    "    #print(positions)\n",
    "    #gencds=get_codon_fromfasta(positions,chromosome,FASTA)\n",
    "    #if strand==-1: #invert\n",
    "    #    gencds=\"\".join(list(gencds)[::-1])\n",
    "    #print(gencds)\n",
    "    return gencds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 564,
   "id": "af28e67d-0186-4b30-837c-04ee566699d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#create cds sequence\n",
    "\n",
    "comb_clun_isosub_filt_tmp[\"chromosome_cds\"]=comb_clun_isosub_filt_tmp.apply(lambda x: get_gencds(x,ENSfasta)\n",
    "                                if isinstance(x[\"segments\"],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 565,
   "id": "e8c7efb0-cb4f-4675-8ccd-812aeb05d169",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub_filt_tmp.to_pickle(\"comb_clun_isosub_filt_tmp.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 774,
   "id": "2c65160e-5650-4d76-9e30-5c04ef04a4ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2490"
      ]
     },
     "execution_count": 774,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(comb_clun_isosub_filt_tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 572,
   "id": "40488e82-0660-4658-9f54-2b5028ac1eb1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#comb_clun_isosub_filt_tmp[[\"chromosome\",\"strand\",\"segments\",\"chromosome_cds\"]].iloc[2].chromosome_cds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 573,
   "id": "f22eefe6-6f19-422d-93c4-56ad1e0d6231",
   "metadata": {},
   "outputs": [],
   "source": [
    "#EXPAND the potential saavss table\n",
    "\n",
    "comb_clun_isosub_filt_exp=comb_clun_isosub_filt_tmp.explode(['peptides_w1sub','saav_info_pep', 'saav_info_prot',\n",
    "                                   'saav_info_triplet','saav_info_refcod','is_expected_aa'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 533,
   "id": "9e471c00-2204-48ff-8220-96b1ce3c303e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#comb_clun_isosub_filt_exp.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 574,
   "id": "a7a6513e-f09b-4de1-a753-19930e2c629c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[('G', 'A')],\n",
       " [('C', 'A,T')],\n",
       " [('G', 'C,GGCA,T,A'), ('GGCA', 'G'), ('GGCAGCA', 'G'), ('GGCAGCAGCA', 'G')]]"
      ]
     },
     "execution_count": 574,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[ list(VCFdf_gr[(VCFdf_gr.CHROM==\"7\")&(VCFdf_gr.POS==str(pos))][[\"REF\",\"ALT\"]].itertuples(index=False, name=None)) for pos in comb_clun_isosub_filt_exp.iloc[0][\"saav_info_triplet\"][0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 575,
   "id": "2ae3304a-fa8e-4980-9a92-49480aa2fdd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#map variants to positions\n",
    "\n",
    "comb_clun_isosub_filt_exp[\"REF_ALT\"]=comb_clun_isosub_filt_exp.apply(lambda x: [ list(VCFdf_gr[(VCFdf_gr.CHROM==x[\"chromosome\"])&(VCFdf_gr.POS==str(pos))][[\"REF\",\"ALT\"]].itertuples(index=False, name=None))\n",
    "                                            for pos in x[\"saav_info_triplet\"][0]] \n",
    "                                if isinstance(x[\"saav_info_triplet\"][0],list) else np.nan, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 576,
   "id": "363a0663-83f1-452f-80bb-884d21535972",
   "metadata": {},
   "outputs": [],
   "source": [
    "#expand by position\n",
    "\n",
    "comb_clun_isosub_filt_exp[\"triplet\"]=comb_clun_isosub_filt_exp.saav_info_triplet.apply(lambda x: x[0] if isinstance(x[0],list) else np.nan)\n",
    "comb_clun_isosub_filt_expvar=comb_clun_isosub_filt_exp.explode([\"triplet\",\"REF_ALT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 609,
   "id": "1246b6cc-6c88-401e-b447-cb7115041cc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cdsbp(segments,strand):\n",
    "    positions=[list(range(segment[0],segment[1]+1))[::-1] if strand==-1\n",
    "               else list(range(segment[0],segment[1]+1))\n",
    "               for segment in segments]\n",
    "    positions=[e for l in positions for e in l]\n",
    "    #print(min(positions),max(positions),segments)\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 605,
   "id": "aa0c2dd0-7930-44e7-b15b-fd9d84bed2ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x=comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.triplet==18833849].iloc[0]#.iloc[22]\n",
    "#get_cdsbp(x[\"segments\"],x[\"strand\"])#.index(x[\"triplet\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 606,
   "id": "38ebe478-c217-4075-a131-797558e03c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x[\"triplet\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 607,
   "id": "699fecac-821e-4190-8af0-935753e92706",
   "metadata": {},
   "outputs": [],
   "source": [
    "#x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 608,
   "id": "df3762c6-f6e7-471f-93ba-9a09c8466fd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OPcoords[OPcoords.accession==\"IP_756098\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 610,
   "id": "8e138d4f-e75e-4248-b771-beb12513c548",
   "metadata": {},
   "outputs": [],
   "source": [
    "#annotate chromosome position to cds position (python-base) 18833849 \n",
    "#triplets are 1-base, but segments are 0-based; as we need index for cds mutation, so transform triplet coods to 0-based (-1)\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"triplet_cds_pos\"]=comb_clun_isosub_filt_expvar.apply(lambda x: get_cdsbp(x[\"segments\"],x[\"strand\"]).index(x[\"triplet\"]-1)\n",
    "                                                                                    if isinstance(x[\"saav_info_triplet\"][0],list) else np.nan,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 612,
   "id": "7310249a-dd9f-48bb-b19d-3fa3cb8ef66b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#expload by reference bp group\n",
    "\n",
    "comb_clun_isosub_filt_expvar=comb_clun_isosub_filt_expvar.explode([\"REF_ALT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 613,
   "id": "8b6b3fd4-2eb7-49d1-83e2-a0fe7bd39152",
   "metadata": {},
   "outputs": [],
   "source": [
    "#expload by alternative variant\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"ALT\"]=comb_clun_isosub_filt_expvar.REF_ALT.apply(lambda x: x[1].split(\",\")\n",
    "                                                                              if isinstance(x,tuple) else np.nan)\n",
    "comb_clun_isosub_filt_expvar=comb_clun_isosub_filt_expvar.explode([\"ALT\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 614,
   "id": "7281d249-d9be-45fb-b027-0c79763429bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub_filt_expvar[\"REF\"]=comb_clun_isosub_filt_expvar.REF_ALT.apply(lambda x: x[0]\n",
    "                                                                              if isinstance(x,tuple) else np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 765,
   "id": "fbebf627-a8d4-4efa-b5e7-19f39541a1a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Noncanon    24885\n",
       "unique_to_Canon       10621\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 765,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many variants per peptide class\n",
    "\n",
    "comb_clun_isosub_filt_expvar[~comb_clun_isosub_filt_expvar.ALT.isna()].peptide_class.value_counts()#.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 788,
   "id": "63d58721-fa8b-4fb5-a7e4-df1507c67c1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Noncanon    18942\n",
       "unique_to_Canon        8228\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 788,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_clun_isosub_filt_expvar[~comb_clun_isosub_filt_expvar.ALT.isna()].drop_duplicates([\"chromosome\",\"triplet\",\"REF\",\"ALT\"]).peptide_class.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 768,
   "id": "59530d06-0ccf-4735-988b-539b4a8c4eca",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1970881/2327696973.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_filt_expvar[~comb_clun_isosub_filt_expvar.ALT.isna()].groupby(\"triplet\").apply(lambda x: \"unique_to_Noncanon\" in x[\"peptide_class\"].unique()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False    16274\n",
       "True        29\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 768,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#is there variant of canon and non-canon that overlap by pos?\n",
    "\n",
    "comb_clun_isosub_filt_expvar[~comb_clun_isosub_filt_expvar.ALT.isna()].groupby(\"triplet\").apply(lambda x: \"unique_to_Noncanon\" in x[\"peptide_class\"].unique()\n",
    "                                                                                               and \"unique_to_Canon\" in x[\"peptide_class\"].unique()).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 621,
   "id": "26e8aabd-57e7-4e3d-98e7-a74f07f90c64",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_cds(x):\n",
    "    cds_seq=list(x[\"chromosome_cds\"]); mut_pos=x[\"triplet_cds_pos\"]; alt=x[\"ALT\"]\n",
    "    cds_seq[mut_pos]=alt\n",
    "    #print(alt)\n",
    "    return \"\".join(cds_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 622,
   "id": "36dfe999-685c-422b-80a4-7d05a37c6b4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mutate sequence\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"chromosome_cds_mutated\"]=comb_clun_isosub_filt_expvar.apply(lambda x: mutate_cds(x)  \n",
    "                                                                                          if not pd.isna(x[\"ALT\"]) else np.nan,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f3ec3ce-8e3b-43c5-aeb3-151e38cfca38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#our target are mutations that do not disrupt the ORF.\n",
    "# so it can be trancation or elongation, but the first aa should be M and the stop codon should be present\n",
    "# whenever the exon junctions are affected (not first end of 1st exon and not 2ed end of last exon) \n",
    "# do not concider this variant in downstream analysis, as we cant say what is a new exon boundary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 636,
   "id": "51fcd5ae-a7ef-415d-87a3-bbde3cf9a225",
   "metadata": {},
   "outputs": [],
   "source": [
    "def if_splicevar(x):\n",
    "    #triplet to 0-base\n",
    "    first_bp=x[\"triplet\"]-1; segments=x[\"segments\"];strand=x[\"strand\"]; ref=x[\"REF\"]; alt=x[\"ALT\"]\n",
    "    n_exon=len(segments)\n",
    "    \n",
    "    #if not indel - ignore\n",
    "    if len(ref)==1 and len(alt)==1:\n",
    "        return False\n",
    "    #length of indel and last affected bp\n",
    "    indel_len=max([len(ref),len(alt)])\n",
    "    last_bp=first_bp+indel_len-1\n",
    "    \n",
    "    #in which exon indel is\n",
    "    i_exon=[i for i,seg in enumerate(segments) if first_bp>=seg[0] and first_bp<=seg[1]][0]\n",
    "    #does it effect any exon boundary?\n",
    "    insertion=len(ref)<len(alt)\n",
    "    delision=len(ref)>len(alt)\n",
    "    #splice var if no last_bp pos in segments (for indels in end of exon) or first_bp is an exon start\n",
    "    is_splvar=((not last_bp in get_cdsbp(segments,strand)) or (first_bp==segments[i_exon][0] or first_bp==segments[i_exon][1]))\n",
    "\n",
    "    ###strand 1: is insertion in 1st_exon_start or last_exon_end - ignor\n",
    "    ###strand -1: is insertion in \"1st_exon_end\" or \"last_exon_start\" - ignor    \n",
    "    if is_splvar:\n",
    "        print(i_exon)\n",
    "        if i_exon==0 or i_exon==n_exon-1:\n",
    "            if strand==1:\n",
    "                if (\n",
    "                    (insertion and i_exon==0 and first_bp==segments[i_exon][0]) or\n",
    "                    (insertion and i_exon==n_exon-1 and first_bp==segments[i_exon][1])\n",
    "                   ):\n",
    "                    return False\n",
    "                else: #inner exon boundary\n",
    "                    print(1)\n",
    "                    return True\n",
    "            if strand==-1:\n",
    "                if (\n",
    "                    (insertion and i_exon==0 and first_bp==segments[i_exon][1]) or\n",
    "                    (insertion and i_exon==n_exon-1 and first_bp==segments[i_exon][0])\n",
    "                   ):\n",
    "                    return False\n",
    "                else: #inner exon boundary\n",
    "                    print(2)\n",
    "                    return True\n",
    "        else:\n",
    "            print(0)\n",
    "            return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 704,
   "id": "2bdcb113-29f1-48c7-a05a-90506adbc566",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "2\n",
      "0\n",
      "1\n",
      "0\n",
      "1\n",
      "0\n",
      "1\n",
      "0\n",
      "1\n",
      "0\n",
      "1\n",
      "0\n",
      "2\n",
      "6\n",
      "2\n",
      "6\n",
      "2\n",
      "0\n",
      "1\n",
      "1\n",
      "2\n",
      "1\n",
      "2\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "38\n",
      "0\n",
      "0\n",
      "2\n",
      "2\n",
      "0\n",
      "6\n",
      "0\n",
      "0\n",
      "1\n",
      "0\n",
      "1\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "1\n",
      "0\n",
      "0\n",
      "2\n",
      "0\n",
      "1\n",
      "0\n",
      "1\n",
      "1\n",
      "0\n",
      "1\n",
      "0\n",
      "0\n",
      "1\n",
      "11\n",
      "0\n",
      "0\n",
      "1\n",
      "6\n",
      "0\n",
      "3\n",
      "0\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n",
      "0\n",
      "2\n"
     ]
    }
   ],
   "source": [
    "#mark idels overlapping with exon junction\n",
    "#ref pos and alt - if last nuc goes beyoond closest exon\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"isSpliceVar\"]=comb_clun_isosub_filt_expvar.apply(lambda x: if_splicevar(x)  \n",
    "                                                                                          if not pd.isna(x[\"ALT\"]) else False,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 705,
   "id": "4f430946-5921-4d52-88b0-d0090b514cd2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isSpliceVar\n",
       "False    43951\n",
       "True        40\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 705,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_clun_isosub_filt_expvar.isSpliceVar.value_counts()#head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 641,
   "id": "11c2c8c8-ba37-4d72-a531-6d0688e96209",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for -1 strand, get seq compliment before translation\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"chromosome_cds_mutated_rev\"]=comb_clun_isosub_filt_expvar.apply(lambda x: \"\".join([reverse_compliment_dna[bp] \n",
    "                                                                                                                 for bp in x[\"chromosome_cds_mutated\"]])  \n",
    "                                                                                          if not pd.isna(x[\"ALT\"]) and x[\"strand\"]==-1 \n",
    "                                                                                              else np.nan,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 653,
   "id": "f499cc98-1ea4-4a88-b814-d8c81cd4f0fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_triplets(cds):\n",
    "    triplets = []\n",
    "    i = 0\n",
    "    while i < len(cds):\n",
    "        if i+3 < len(cds):\n",
    "            triplet = cds[i:i+3]    \n",
    "        else:  # If it's the last triplet and it's not full, extend it\n",
    "            triplet = \"...\" #list(cds[i:])+[\".\"]*3-len(cds[i:])\n",
    "        triplets.append(triplet)\n",
    "        i += 3\n",
    "    return triplets\n",
    "def translate_cds(x):\n",
    "    strand=x[\"strand\"]\n",
    "    cds= x[\"chromosome_cds_mutated_rev\"] if strand==-1 else x[\"chromosome_cds_mutated\"]\n",
    "    #defide to codons\n",
    "    codons=get_triplets(cds)\n",
    "    #print(codons)\n",
    "    protein=\"\".join([aminoacid_table_rev[codon] if not \".\" in codon else codon for codon in codons])\n",
    "    return protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 654,
   "id": "94782c40-135b-4a20-aa9c-1ac0769b9dce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#do translation\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"protein_mutated\"]=comb_clun_isosub_filt_expvar.apply(lambda x: translate_cds(x)  \n",
    "                                                                                          if not pd.isna(x[\"ALT\"]) else np.nan,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 706,
   "id": "e6cb5f92-5dcb-4604-b2e7-0e507e399d5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#mark \n",
    "#if protein does not start from M anymore (disrupted start)\n",
    "#if premature stop codon. If there is an extention of stop codon - we donts know (analysis does not show)\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"isDisrupt\"]=comb_clun_isosub_filt_expvar.apply(lambda x: ((not x[\"protein_mutated\"].startswith(\"M\")) or \n",
    "                                                                                        (\"*\" in x[\"protein_mutated\"] and \n",
    "                                                                                         (not x[\"protein_mutated\"].endswith(\"*\")) )\n",
    "                                                                                       )\n",
    "                                                                                          if not pd.isna(x[\"ALT\"]) else False,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 707,
   "id": "fa5fa324-8698-4dc0-b64f-21b06cb352d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "isDisrupt\n",
       "False    42939\n",
       "True      1052\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 707,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_clun_isosub_filt_expvar.isDisrupt.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 708,
   "id": "f0095e69-fc8e-48da-bb1f-5af8714bcd9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    42909\n",
       "True      1082\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 708,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we are omiting splice mutations and orf disraption\n",
    "#how many?\n",
    "\n",
    "((comb_clun_isosub_filt_expvar.isDisrupt==True)|(comb_clun_isosub_filt_expvar.isSpliceVar==True)).value_counts()#[['peptides_w1sub',\"protein_mutated\"]].iloc[-1]   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1089,
   "id": "6c4efe9e-23a5-48e9-b080-388f6ce324c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.024595940078652453"
      ]
     },
     "execution_count": 1089,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1082/(42909+1082)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 680,
   "id": "c2d78935-2602-40f5-8e44-49c16f2b261a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#find triptic peptide\n",
    "def isTryptic(protein,pos):\n",
    "    return ((protein[pos[0]-1].startswith((\"K\",\"R\"))) and (protein[pos[1]+1].startswith((\"K\",\"R\"))))\n",
    "def isNterm(protein,pos,clip_nM):\n",
    "    if clip_nM:\n",
    "        return (pos[0]==1 or pos[0]==0) and protein[pos[1]+1].startswith((\"K\",\"R\"))\n",
    "    else:\n",
    "        return (pos[0]==0) and protein[pos[1]+1].startswith((\"K\",\"R\"))\n",
    "    \n",
    "def find_tryptic_pep(peptide,protein,clip_nM=True):\n",
    "    pep_pos=hpp.get_pep_positions(protein, peptide )\n",
    "    if len(pep_pos)==0:\n",
    "        return False\n",
    "    return any( (isTryptic(protein,pos) or isNterm(protein,pos,clip_nM)) for pos in pep_pos)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 709,
   "id": "4df968e7-98e4-4a5f-a83e-03e086a38ba9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#find the peptide with possible saav in protein\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"isSAAV\"]=comb_clun_isosub_filt_expvar.apply(lambda x: find_tryptic_pep(x['peptides_w1sub'],x[\"protein_mutated\"])\n",
    "                                                                          if ((not pd.isna(x[\"ALT\"])) \n",
    "                                                                              and ((not x[\"isSpliceVar\"]) \n",
    "                                                                              or (not x[\"isDisrupt\"]))) \n",
    "                                                                          else False,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 710,
   "id": "920d73c8-5a14-4581-9df2-3da830279dbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class       isSAAV\n",
       "unique_to_Noncanon  False     29220\n",
       "unique_to_Canon     False     14592\n",
       "unique_to_Noncanon  True        127\n",
       "unique_to_Canon     True         52\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 710,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many SAAVs we have by peptide class\n",
    "\n",
    "comb_clun_isosub_filt_expvar[[\"peptide_class\",\"isSAAV\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 823,
   "id": "4b9abf65-7a1c-439e-b3d9-a06883ce2618",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count\n",
       "False    165\n",
       "True       7\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 823,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many variants overlap on peptide class (affect both canon and non-canon)\n",
    "\n",
    "(comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.isSAAV==True].groupby([\"chromosome\",\"triplet\",\"REF\",\"ALT\"])[\"peptide_class\"].value_counts()>1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 714,
   "id": "71ea3199-2d45-4900-be61-7053b517310c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class       isSAAV\n",
       "unique_to_Noncanon  True      76\n",
       "unique_to_Canon     True      31\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 714,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many of SAAVs are if group by same peptide? (so different variants leading to potentialy same PSMs)\n",
    "\n",
    "comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.isSAAV].drop_duplicates(\"database_peptide\",keep=\"first\")[[\"peptide_class\",\"isSAAV\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 789,
   "id": "bb394b59-ec01-45fc-aee5-248eb8cfff30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class       isSAAV\n",
       "unique_to_Noncanon  True      74\n",
       "unique_to_Canon     True      27\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 789,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many proteins?\n",
    "\n",
    "comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.isSAAV].drop_duplicates(\"leadprot\",keep=\"first\")[[\"peptide_class\",\"isSAAV\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 715,
   "id": "dec122f9-b0f9-4971-951a-8400e7d0697e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class       isSAAV\n",
       "unique_to_Noncanon  False     1608\n",
       "unique_to_Canon     False      882\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 715,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_clun_isosub_filt_expvar[~comb_clun_isosub_filt_expvar.isSAAV].drop_duplicates(\"database_peptide\",keep=\"first\")[[\"peptide_class\",\"isSAAV\"]].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 712,
   "id": "3a0fec4e-da68-4836-b324-58eeda0d1de6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1970881/722400802.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.isSAAV==True].groupby(\"database_peptide\").apply(lambda x: len(x)).sort_values()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "database_peptide\n",
       "ALLDSVIR                        1\n",
       "DSIPVTELSASGPFESHDLLR           1\n",
       "EAAEQNVEK                       1\n",
       "DYLMSTHFWGPVANWGLPIAAINDMK      1\n",
       "EDFQEAR                         1\n",
       "                               ..\n",
       "HQGVMVSMGQKDSYVGDEAQSK          4\n",
       "AGLPDMINEVDADGNRTDSPEFLTMMAR    4\n",
       "HQGVMVGMGQKDCYVGDEAQSK          5\n",
       "AAEAAAAPAESAAPAAGEEPSKEEGEPK    6\n",
       "SSGGPYGGGYGSGGGSGGYGSR          9\n",
       "Length: 107, dtype: int64"
      ]
     },
     "execution_count": 712,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#what are the largest groups?\n",
    "\n",
    "comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.isSAAV==True].groupby(\"database_peptide\").apply(lambda x: len(x)).sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 719,
   "id": "b7966d80-bdcb-459b-a816-d3254301e4a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#are peptides that have most of n of aa that can be substituted, has most SAAV variants\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"residues_to_sub_count\"]=comb_clun_isosub_filt_expvar.apply(lambda x: \n",
    "                                                                                         {aa:Counter(x[\"database_peptide\"])[aa] for aa in x[\"residues_to_sub\"]}\n",
    "                                                                                        ,axis=1)\n",
    "#comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.database_peptide==\"SSGGPYGGGYGSGGGSGGYGSR\"][['residues_to_sub', 'pos_to_sub', 'sub_rule']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 729,
   "id": "e47ce58d-f56d-4433-a85c-f1a79e482ef7",
   "metadata": {},
   "outputs": [],
   "source": [
    "data=pd.DataFrame(columns=[\"database_peptide\",\"peptide_class\",\"residues_to_sub_count_tot\",\"isSAAV_count\"])\n",
    "for database_peptide, group in comb_clun_isosub_filt_expvar.groupby(\"database_peptide\"):\n",
    "    c1, c2 =np.sum(list(group[\"residues_to_sub_count\"].iloc[0].values())), len(group[group.isSAAV])\n",
    "    data.loc[len(data)]=[database_peptide,group[\"peptide_class\"].iloc[0],c1,c2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 748,
   "id": "46a49c07-b38b-4fa5-925c-95d52993d1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#put a therechols for the Chi2 test residues_to_sub_count_tot>3\n",
    "#coun the number of peptides\n",
    "\n",
    "threshold=3\n",
    "contigency_table=pd.DataFrame(columns=[\"peptide_class\",f\"res_to_sub_>{threshold}\",\"isSAAV_False\",\"isSAAV_True\"])\n",
    "for peptide_class, group in data.groupby(\"peptide_class\"):\n",
    "    contigency_table.loc[len(contigency_table)]=[peptide_class,False,len(group[(group.isSAAV_count==0)&(group.residues_to_sub_count_tot<=threshold)]),\n",
    "                                                len(group[(group.isSAAV_count!=0)&(group.residues_to_sub_count_tot<=threshold)])]\n",
    "    contigency_table.loc[len(contigency_table)]=[peptide_class,True,len(group[(group.isSAAV_count==0)&(group.residues_to_sub_count_tot>threshold)]),\n",
    "                                                len(group[(group.isSAAV_count!=0)&(group.residues_to_sub_count_tot>threshold)])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 749,
   "id": "1a07135f-c950-428d-b9dd-c677891f88a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peptide_class</th>\n",
       "      <th>res_to_sub_&gt;3</th>\n",
       "      <th>isSAAV_False</th>\n",
       "      <th>isSAAV_True</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>unique_to_Canon</td>\n",
       "      <td>False</td>\n",
       "      <td>542</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>unique_to_Canon</td>\n",
       "      <td>True</td>\n",
       "      <td>309</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>False</td>\n",
       "      <td>982</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>True</td>\n",
       "      <td>550</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        peptide_class  res_to_sub_>3  isSAAV_False  isSAAV_True\n",
       "0     unique_to_Canon          False           542           17\n",
       "1     unique_to_Canon           True           309           14\n",
       "2  unique_to_Noncanon          False           982           48\n",
       "3  unique_to_Noncanon           True           550           28"
      ]
     },
     "execution_count": 749,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contigency_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 741,
   "id": "0b88f292-0a7c-43eb-a1d7-d236fc7b2bbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import chi2_contingency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 744,
   "id": "17f8fea1-12ef-40f2-81e3-1bae854bb18e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "unique_to_Canon Chi = 0.66, p = 0.4151\n",
      "unique_to_Noncanon Chi = 0.00, p = 0.9645\n"
     ]
    }
   ],
   "source": [
    "for peptide_class, group in contigency_table.groupby(\"peptide_class\"):\n",
    "    chi2, p, dof, expected = chi2_contingency(group[[\"isSAAV_False\",\"isSAAV_True\"]])\n",
    "    print(peptide_class,f\"Chi = {chi2:.2f}, p = {p:.4f}\") #NS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 751,
   "id": "c0c3f668-12fa-4e56-ac7a-93f33eb7e114",
   "metadata": {},
   "outputs": [],
   "source": [
    "#do non-canonical have more peptides with potential saav residue repeats than of canonical\n",
    "\n",
    "threshold=3\n",
    "contigency_table=pd.DataFrame(columns=[\"peptide_class\",f\"res_to_sub_<={threshold}\",f\"res_to_sub_>{threshold}\"])\n",
    "for peptide_class, group in data.groupby(\"peptide_class\"):\n",
    "    contigency_table.loc[len(contigency_table)]=[peptide_class,len(group[(group.residues_to_sub_count_tot<=threshold)]),\n",
    "                                                len(group[(group.residues_to_sub_count_tot>threshold)])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 752,
   "id": "bed77087-0071-409b-ba21-950febeeeb6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>peptide_class</th>\n",
       "      <th>res_to_sub_&lt;=3</th>\n",
       "      <th>res_to_sub_&gt;3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>unique_to_Canon</td>\n",
       "      <td>559</td>\n",
       "      <td>323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>1030</td>\n",
       "      <td>578</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        peptide_class  res_to_sub_<=3  res_to_sub_>3\n",
       "0     unique_to_Canon             559            323\n",
       "1  unique_to_Noncanon            1030            578"
      ]
     },
     "execution_count": 752,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contigency_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 753,
   "id": "efccf134-d4b7-4bb4-9297-400366027d56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "res_to_sub_>3 btw peptide_classes Chi = 0.09, p = 0.7702\n"
     ]
    }
   ],
   "source": [
    "chi2, p, dof, expected = chi2_contingency(contigency_table[[\"res_to_sub_<=3\",\"res_to_sub_>3\"]])\n",
    "print(\"res_to_sub_>3 btw peptide_classes\",f\"Chi = {chi2:.2f}, p = {p:.4f}\") #NS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "239976af-fa70-4a6a-99e2-669d7e2a24b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hypothesis: peptides with elivated n of aa prone for ptms isobaric to saav, have higher rate of detected saavs\n",
    "# -> aa bias that can 1) bigger probability to bear ptm isobaric to saav 2) genomic variation influence - hinder detection of unique peptides\n",
    "# second point should be checked with additional analysis - are repetetive regions are more prone to mutations? are they are oftem isobaric?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 757,
   "id": "92f5be8b-701a-425f-9a28-91e2d2633852",
   "metadata": {},
   "outputs": [],
   "source": [
    "#just try isSAAV even on disrupt, does anything change? - no\n",
    "\n",
    "comb_clun_isosub_filt_expvar[\"isSAAV_tmp\"]=comb_clun_isosub_filt_expvar.apply(lambda x: find_tryptic_pep(x['peptides_w1sub'],x[\"protein_mutated\"])\n",
    "                                                                          if ((not pd.isna(x[\"ALT\"])) \n",
    "                                                                              and ((not x[\"isSpliceVar\"]))) \n",
    "                                                                          else False,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1045,
   "id": "74424065-5c59-44d9-8a12-53d5c8ee636a",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub_filt_expvar.to_pickle(\"comb_clun_isosub_filt_expvar.pkl\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b02c0a00-379b-4d0c-9501-bc0dce67aa6f",
   "metadata": {},
   "source": [
    "## Get back here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "31deb993-dd5a-47ce-aeb0-4e958b69f3b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gnomAD_qgr_res = pickle.load(open(\"./../gnomAD_qgr_res.pickle\", \"rb\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "dfb3c917-11ef-43bf-8b5e-a3b765e9c41e",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub=pd.read_pickle(\"./../comb_clun_isosub.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "e11c32c4-4201-47bd-ae13-c9c52dec9693",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub_h=comb_clun_isosub[comb_clun_isosub.database_peptide.isin(comb_datasets_clun.database_peptide.unique())]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "9f895b19-df08-46d3-b87b-23bb98692196",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({True: 2134, False: 9})"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many peptides have False is_expected_aa - 10\n",
    "Counter(comb_clun_isosub_h.apply(lambda x: all(x[\"is_expected_aa\"]) if isinstance(x[\"saav_info_refcod\"],list) else True, axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "fc551b06-d60d-40dc-9056-21ffde286c35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.004217432052483599"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "9/2134"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "bdfebee3-b31c-4fb8-81c7-2c64fb6a3afb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "450"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many protein accessions WERE NOT found in OP\n",
    "len(comb_clun_isosub_h[comb_clun_isosub_h.is_expected_aa.isna()].leadprot.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "2fb9f3ff-7f83-47cb-948b-86c7194b90b8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0006584598185870038"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "450/683413"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "101f7475-fd15-4242-837b-0ed1088fa215",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub_filt=comb_clun_isosub[~comb_clun_isosub.is_expected_aa.isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "945d08af-8816-451d-922a-474dcb26678c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1596"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#filter to peptides detected with hybrid filtering\n",
    "comb_clun_isosub_filt_h=comb_clun_isosub_filt[comb_clun_isosub_filt.database_peptide.isin(comb_datasets_clun.database_peptide.unique())]\n",
    "len(comb_clun_isosub_filt_h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "a48656c1-1011-4d73-8662-970de88a2a75",
   "metadata": {},
   "outputs": [],
   "source": [
    "#how many positions were checked with gnomAD\n",
    "\n",
    "gnomAD_qgr_h={}\n",
    "for chromosome, group in comb_clun_isosub_filt_h.groupby(\"chromosome\"):\n",
    "    all_triplets=group.saav_info_triplet.tolist()\n",
    "    all_coordinates=list(set([e for row in all_triplets for l in row for e in l[0]]))\n",
    "    chrom_intervals=get_continuous_intervals(all_coordinates)\n",
    "    gnomAD_qgr_h[chromosome]=chrom_intervals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "ef514cee-828b-4dd8-8c2d-6cf10c3d7c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "#flatten the intervals\n",
    "all_pos_h={chromosome:[] for chromosome in gnomAD_qgr_h}\n",
    "for chromosome, positions in gnomAD_qgr_h.items():\n",
    "    pos_l=[]\n",
    "    for pos_st,pos_end in positions:\n",
    "        pos_l.extend(list(range(pos_st,pos_end+1)))\n",
    "    all_pos_h[chromosome]=pos_l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "103a130b-3eda-489d-b70f-58f9e3799e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter previous gnomAD results with those intervals left in hybrid filtering\n",
    "\n",
    "gnomAD_qgr_res_h={chromosome:{} for chromosome in gnomAD_qgr_h}\n",
    "for chromosome, positions in gnomAD_qgr_res.items():\n",
    "    for pos, variants in positions.items():\n",
    "        if pos in all_pos_h[chromosome]:\n",
    "            gnomAD_qgr_res_h[chromosome].update({pos:variants})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "d0dd5399-e1e7-47af-8af2-482273eb5e64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expected: 16165\n",
      "got: 16165\n"
     ]
    }
   ],
   "source": [
    "print(\"expected:\", np.sum([abs(interval[0]-interval[1])+1 for chrom,intervals in gnomAD_qgr_h.items() for interval in intervals] ))\n",
    "print(\"got:\", np.sum([len(vals) for chrom,vals in gnomAD_qgr_res_h.items() ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "c7c96cbd-0c66-4114-b9af-7593a3d78d7a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "empty positions: 0.2887720383544695\n"
     ]
    }
   ],
   "source": [
    "#how many positions are empty\n",
    "\n",
    "print(\"empty positions:\", np.sum([1 for chrom,vals in gnomAD_qgr_res_h.items() for v in vals.values() if len(v)==0])/np.sum([len(vals) for chrom,vals in gnomAD_qgr_res_h.items() ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "2ffda8cb-f1ed-4fe6-b58b-513484cd7401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7112279616455305"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1-0.2887720383544695"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a22e0f-0de8-4351-b2bc-24f80f2a1f6a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c97ef993-8105-4d4a-92ec-44a766e50810",
   "metadata": {},
   "outputs": [],
   "source": [
    "comb_clun_isosub_filt_expvar=pd.read_pickle(\"./../comb_clun_isosub_filt_expvar.pkl\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c8b08d31-00a5-440d-9b4b-218a16115385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['unique_to_Canon', 'unique_to_Noncanon'], dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_datasets_clun.peptide_class.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "738da72c-73a7-4f87-8c3a-65d99ca2770a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43991"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many eptides were examined before\n",
    "len(comb_clun_isosub_filt_expvar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8b5e0100-68d3-4fe9-b19b-d555eb4f626b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32148"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many now with hybrid filtering\n",
    "comb_clun_isosub_filt_expvar_h=comb_clun_isosub_filt_expvar[comb_clun_isosub_filt_expvar.database_peptide.isin(comb_datasets_clun.database_peptide.unique())]\n",
    "len(comb_clun_isosub_filt_expvar_h)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "26b89463-1e10-442a-a22b-d09628981e4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False    31310\n",
       "True       838\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we are omiting splice mutations and orf disraption\n",
    "#how many?\n",
    "\n",
    "((comb_clun_isosub_filt_expvar_h.isDisrupt==True)|(comb_clun_isosub_filt_expvar_h.isSpliceVar==True)).value_counts()#[['peptides_w1sub',\"protein_mutated\"]].iloc[-1]   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "47928f2f-e322-42ec-ad08-6f03bb393417",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.02606694040064701"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "838/(838+31310)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f3feffb2-275a-4247-a19c-4733c1a1921f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Noncanon    11055\n",
       "unique_to_Canon        8159\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many variant detected in selected pisiotnons by peptide class\n",
    "comb_clun_isosub_filt_expvar_h[~comb_clun_isosub_filt_expvar_h.ALT.isna()].drop_duplicates([\"chromosome\",\"triplet\",\"REF\",\"ALT\"]).peptide_class.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "fd1c788e-fc90-459a-88d1-5f4795073461",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class       isSAAV_tmp\n",
       "unique_to_Noncanon  True          34\n",
       "unique_to_Canon     True          31\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_clun_isosub_filt_expvar_h[comb_clun_isosub_filt_expvar_h.isSAAV_tmp].drop_duplicates(\"database_peptide\",keep=\"first\")[[\"peptide_class\",\"isSAAV_tmp\"]].value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "dab3cf2e-5ed6-47ac-80b0-adff77045a4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#check allele frequency of SAAVs\n",
    "\n",
    "VCFdfaf=pd.DataFrame(columns=[\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\",\"AF\"])\n",
    "for chrom,vals in gnomAD_qgr_res.items():\n",
    "    for position,variants in vals.items():\n",
    "        if len(variants)==0: continue\n",
    "        #alts=[]\n",
    "        for variant in variants:\n",
    "            chrom,pos,ref,alt = parce_variantid(variant['variant_id'])\n",
    "            ex_an=variant[\"exome\"][\"af\"] if isinstance(variant[\"exome\"],dict) else 0\n",
    "            gen_an=variant[\"genome\"][\"af\"] if isinstance(variant[\"genome\"],dict) else 0\n",
    "            af = variant[\"exome\"][\"af\"] if ex_an!=0 else variant[\"genome\"][\"af\"] if isinstance(variant[\"genome\"],dict) else 0\n",
    "            VCFdfaf.loc[len(VCFdfaf),[\"CHROM\", \"POS\", \"REF\", \"ALT\",\"AF\"]]=[chrom,pos,ref,alt,af]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "eeb98b48-b155-4f1e-8e4c-52655560be6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select saavs overlapping with identified proteotypic peptides with potential saav\n",
    "\n",
    "VCFdfaf[\"isSAAV\"]=False\n",
    "VCFdfaf[\"peptide_class\"]=False\n",
    "VCFdfaf[\"database_peptide\"]=[[]]*len(VCFdfaf)\n",
    "VCFdfaf[\"leadprot\"]=[[]]*len(VCFdfaf)\n",
    "issaav_vf=comb_clun_isosub_filt_expvar_h[comb_clun_isosub_filt_expvar_h.isSAAV==True][[\"chromosome\",\"triplet\",\"REF\",\"ALT\",\"peptide_class\",\"database_peptide\",'peptides_w1sub',\"leadprot\"]]\n",
    "for i, row in VCFdfaf.iterrows():\n",
    "    data=issaav_vf[(issaav_vf.chromosome==row[\"CHROM\"])&(issaav_vf.triplet==int(row[\"POS\"]))&(issaav_vf.REF==row[\"REF\"])&(issaav_vf.ALT==row[\"ALT\"])]\n",
    "    if len(data)>0:\n",
    "        VCFdfaf.loc[i,\"isSAAV\"]=True\n",
    "        VCFdfaf.loc[i,\"peptide_class\"]=data.peptide_class.unique()\n",
    "        VCFdfaf.at[i,\"database_peptide\"]=data.database_peptide.unique()\n",
    "        VCFdfaf.at[i,\"leadprot\"]=data.leadprot.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f655af6e-13bb-40cb-9baa-b2e648ccb4c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AF\n",
       "False    64\n",
       "True     52\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(VCFdfaf[VCFdfaf.isSAAV==True][\"AF\"]==0).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0bc79df2-ff25-4e74-af02-20dc3d9d21b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Noncanon    68\n",
       "unique_to_Canon       47\n",
       "unique_to_Canon        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#n of variants\n",
    "\n",
    "VCFdfaf[(VCFdfaf.isSAAV==True)][\"peptide_class\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "99aff365-89a3-470d-94f3-30b0c11b09c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "116"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#total n of variants\n",
    "len(VCFdfaf[(VCFdfaf.isSAAV==True)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "b71b2d42-f13a-4d8e-a383-bb17335f85c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Noncanon    34\n",
       "unique_to_Canon       28\n",
       "unique_to_Canon        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# n of peptides \n",
    "\n",
    "VCFdfaf[(VCFdfaf.isSAAV==True)].drop_duplicates(\"database_peptide\")[\"peptide_class\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a49a54e5-2445-4159-b11e-f3ed9638e9a5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Noncanon    34\n",
       "unique_to_Canon       26\n",
       "unique_to_Canon        1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#n of proteins\n",
    "\n",
    "VCFdfaf[(VCFdfaf.isSAAV==True)].drop_duplicates(\"leadprot\")[\"peptide_class\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "9499e97f-54ab-43b3-8e45-71869476ee7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCFdfaf.columns=['chromosome',\t'triplet', 'ID',\t'REF',\t'ALT'\t,'AF', 'isSAAV', 'peptide_class',\n",
    "       'database_peptide', 'leadprot']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "c035879f-3f57-41d2-ab7e-a95ac41069ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome', 'triplet', 'ID', 'REF', 'ALT', 'AF', 'isSAAV',\n",
       "       'peptide_class', 'database_peptide', 'leadprot'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VCFdfaf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "8c83dc78-ff76-44a6-bd37-a8b4086f02df",
   "metadata": {},
   "outputs": [],
   "source": [
    "VCFdfaf.triplet=VCFdfaf.triplet.astype(int)\n",
    "issaav_vfaf=issaav_vf.merge(VCFdfaf, on=['chromosome',\t'triplet', 'REF', 'ALT'], how=\"outer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "c77754fb-840b-4ae8-8fc3-1b97d9139cda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chromosome', 'triplet', 'REF', 'ALT', 'peptide_class_x',\n",
       "       'database_peptide_x', 'peptides_w1sub', 'leadprot_x', 'ID', 'AF',\n",
       "       'isSAAV', 'peptide_class_y', 'database_peptide_y', 'leadprot_y'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "issaav_vfaf.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6f5c5af5-b002-45b7-93b1-7c4fea8957e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize with appropriate dtypes\n",
    "issaav_vfaf['peptides_w1sub_unique'] = pd.Series(dtype='bool')\n",
    "issaav_vfaf['peptides_w1sub_new'] = pd.Series(dtype='bool')\n",
    "issaav_vfaf['peptides_w1sub_diffprot'] = pd.Series(dtype='bool')\n",
    "issaav_vfaf['peptides_w1sub_newprot'] = pd.Series(dtype='object')\n",
    "\n",
    "for i, row in issaav_vfaf[(issaav_vfaf.isSAAV == True) & (issaav_vfaf[\"AF\"] != 0)].iterrows():\n",
    "    pep = row.peptides_w1sub\n",
    "    prot = row.leadprot_x\n",
    "\n",
    "    try:\n",
    "        unique = is_unique.get(pep, False)\n",
    "        parts = all_peptide_sequences_byprot[pep].split(\"|\")\n",
    "        diff_prot = len(parts) > 1 and parts[1] != prot\n",
    "        new = False\n",
    "        newprot = parts[1] if diff_prot else \"NA\"\n",
    "\n",
    "    except (KeyError, IndexError):\n",
    "        unique = True\n",
    "        diff_prot = \"NA\"\n",
    "        new = True\n",
    "        newprot = \"NA\"\n",
    "\n",
    "    issaav_vfaf.loc[i, ['peptides_w1sub_unique', 'peptides_w1sub_new',\n",
    "                        'peptides_w1sub_diffprot', 'peptides_w1sub_newprot']] = [unique, new, diff_prot, newprot]\n",
    "\n",
    "    #print(unique, diff_prot, new)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "a441ad63-8e83-42db-9955-97b58fe457df",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save saav information\n",
    "save_dir=\"./oui_discovery_gwalk_subfigures/explor-3-hybrid-massshift-saav/\"\n",
    "#select variants that lead to expected saavs and have non-zero allele frequency\n",
    "df=issaav_vfaf[(issaav_vfaf.isSAAV==True)]\n",
    "#select nessesary columns\n",
    "df=df[['chromosome',\t'triplet', 'REF', 'ALT','AF','peptide_class_x','database_peptide_x',\t'peptides_w1sub', 'leadprot_x','isSAAV',\n",
    "      'peptides_w1sub_unique',\t'peptides_w1sub_new',\t'peptides_w1sub_diffprot',\t'peptides_w1sub_newprot']]\n",
    "#rename\n",
    "df.columns=['chromosome',\t'triplet', 'REF', 'ALT','AF','peptide_class','database_peptide',\t'peptides_w1sub', 'leadprot','isSAAV',\n",
    "      'peptides_w1sub_unique',\t'peptides_w1sub_new',\t'peptides_w1sub_diffprot',\t'peptides_w1sub_newprot']\n",
    "df.to_csv(f\"{save_dir}isosaav_saav_gnomad_proteotypic.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8faf9e15-6c7d-4b2b-9330-2f4e2fea4583",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptides_w1sub_unique  peptides_w1sub_new  peptides_w1sub_diffprot\n",
       "True                   True                NA                         70\n",
       "False                  False               True                        6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "issaav_vfaf[(issaav_vfaf.isSAAV==True)].explode(\"database_peptide_y\")[['peptides_w1sub_unique','peptides_w1sub_new','peptides_w1sub_diffprot']] .value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "f06a99cd-d086-40d0-821c-3cb526d0b2dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "peptides_w1sub_unique  peptides_w1sub_new  peptides_w1sub_diffprot  peptides_w1sub_newprot  peptide_class_x   \n",
       "True                   True                NA                       NA                      unique_to_Canon       44\n",
       "                                                                                            unique_to_Noncanon    26\n",
       "False                  False               True                     B0YJ73                  unique_to_Canon        2\n",
       "                                                                    Q0VAC0                  unique_to_Noncanon     1\n",
       "                                                                    A0A2R8Y8A0              unique_to_Noncanon     1\n",
       "                                                                    Q8NFG3                  unique_to_Noncanon     1\n",
       "                                                                    Q53GZ6                  unique_to_Noncanon     1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "issaav_vfaf[(issaav_vfaf.isSAAV==True)].explode(\"database_peptide_y\")[['peptides_w1sub_unique','peptides_w1sub_new','peptides_w1sub_diffprot','peptides_w1sub_newprot',\"peptide_class_x\"]] .value_counts()#.sum()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "6484e91f-96ea-42c0-b0b7-270e281be009",
   "metadata": {},
   "outputs": [],
   "source": [
    "#get the number of PSMs, that support correspondin isoSAAvs in coresponding peptides\n",
    "\n",
    "#in the original set of PSMs, find correspinding database_peptide and mass shift\n",
    "\n",
    "## which substitution rule was used\n",
    "\n",
    "tmp=comb_clun_isosub_filt_expvar_h[comb_clun_isosub_filt_expvar_h.peptides_w1sub.isin(\n",
    "    issaav_vfaf[(issaav_vfaf.isSAAV==True)].peptides_w1sub.tolist())].drop_duplicates(\n",
    "    [\"database_peptide\",\"peptides_w1sub\"])[[\"database_peptide\",\"peptides_w1sub\",'modifications_set','sub_rule',]]\n",
    "\n",
    "#mark which of mass shifts is substitution unimod_df_subs\n",
    "def which_sub(x):\n",
    "    tosub, subto = [(tosub,subto) for tosub, subto in zip(x[\"database_peptide\"],x[\"peptides_w1sub\"]) if tosub!=subto][0]\n",
    "    df=unimod_df_subs[(\n",
    "        unimod_df_subs.residues==tosub)&(\n",
    "        unimod_df_subs.subs_to_residue==subto)&(\n",
    "        unimod_df_subs.mono_mass.isin([mass for _,_,mass in x[\"modifications_set\"]]))]\n",
    "    return df.mono_mass.unique()[0]\n",
    "tmp[\"mass_is_sub\"]=tmp.apply(lambda x:  which_sub(x), axis=1)\n",
    "\n",
    "#ammong peptides with possible isoSAAV, select those with SAAv\n",
    "a=[[str(array)] if not \"[\" in str(array) else list(array) for array in issaav_vfaf[(issaav_vfaf.isSAAV==True)].database_peptide_y.tolist()]\n",
    "a=[e for l in a for e in l]\n",
    "comb_clun_isosub_true=comb_clun_isosub[comb_clun_isosub.database_peptide.isin(set(a))]\n",
    "\n",
    "#explode by \n",
    "comb_clun_isosub_true=comb_clun_isosub_true.explode(['dataset', 'spectrum_title','scan', 'spectrum_file', \n",
    "                                                     'observed_retention_time', 'charge',\n",
    "                                                       'modifications_masses'])\n",
    "#mark which have mass shift that leads to expected SAAV\n",
    "def which_expSAAV(x):\n",
    "    if x[\"modifications_masses\"]==\"Unmodified\":\n",
    "        return False\n",
    "    masses=[mass for _,_,mass in x[\"modifications_masses\"]]\n",
    "    exp_masses=tmp[(tmp.database_peptide==x[\"database_peptide\"])].mass_is_sub.tolist()\n",
    "    return any(mass in exp_masses for mass in masses)\n",
    "comb_clun_isosub_true[\"psm_support_saav\"]=comb_clun_isosub_true.apply(lambda x: which_expSAAV(x), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c954fe96-dd61-4556-9807-82b8833736af",
   "metadata": {},
   "outputs": [],
   "source": [
    "#save summary table\n",
    "#select psms that support saavs\n",
    "save_dir=\"./oui_discovery_gwalk_subfigures/explor-3-hybrid-massshift-saav/\"\n",
    "comb_clun_isosub_true[comb_clun_isosub_true.psm_support_saav].drop('isSubstitution',axis=1).to_csv(f\"{save_dir}isosaav_saav_psm_proteotypic.csv\",index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "743d1a9f-93ce-49cb-9e53-caaff91693d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/2023503077.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: Counter(x[\"psm_support_saav\"])[True]).sum()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "np.int64(223)"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many PSMs in total\n",
    "\n",
    "comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: Counter(x[\"psm_support_saav\"])[True]).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "245ff923-d427-4384-a97a-e8c9d3e7e7e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/3513549675.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: Counter(x[\"psm_support_saav\"])[True]).sum()/len(comb_datasets[comb_datasets.modifications!=\"Unmodified\"])\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "np.float64(0.0019049734328817208)"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#percent comparing to modified peps\n",
    "\n",
    "comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: Counter(x[\"psm_support_saav\"])[True]).sum()/len(comb_datasets[comb_datasets.modifications!=\"Unmodified\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5635acca-f0ac-42c5-80f4-1d62a4fa44a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/1018605230.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_true.groupby(\"peptide_class\").apply(lambda x: (Counter(x[\"psm_support_saav\"])[True]))#.sum()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Canon        62\n",
       "unique_to_Noncanon    161\n",
       "dtype: int64"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many PSMs per peptide_class\n",
    "\n",
    "comb_clun_isosub_true.groupby(\"peptide_class\").apply(lambda x: (Counter(x[\"psm_support_saav\"])[True]))#.sum()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "07666763-1404-44aa-9067-26599869bfae",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/1421114309.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_true.groupby(\"peptide_class\").apply(lambda x: (Counter(x[\"psm_support_saav\"])[True]))/comb_datasets[comb_datasets.modifications!=\"Unmodified\"].peptide_class.value_counts()[[\"unique_to_Canon\",\"unique_to_Noncanon\"]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "peptide_class\n",
       "unique_to_Canon       0.008135\n",
       "unique_to_Noncanon    0.047917\n",
       "dtype: float64"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#percent of PSMs with mass shift that can be explained by SAAV\n",
    "\n",
    "comb_clun_isosub_true.groupby(\"peptide_class\").apply(lambda x: (Counter(x[\"psm_support_saav\"])[True]))/comb_datasets[comb_datasets.modifications!=\"Unmodified\"].peptide_class.value_counts()[[\"unique_to_Canon\",\"unique_to_Noncanon\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "03e960e7-f8e5-4224-a61d-bd76bf05819f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['shared_in_Canon', 'shared_btw_can_noncan', 'unique_to_Canon',\n",
       "       'unique_to_Noncanon', 'shared_in_Noncanon'], dtype=object)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_datasets.peptide_class.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f42501d3-3f0e-4744-978f-8953bac31706",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/1642939173.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: Counter(x[\"psm_support_saav\"])[True]).value_counts()\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "1     31\n",
       "2     13\n",
       "4      6\n",
       "3      4\n",
       "5      3\n",
       "28     1\n",
       "12     1\n",
       "26     1\n",
       "7      1\n",
       "6      1\n",
       "10     1\n",
       "18     1\n",
       "8      1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#how many peptides with isoSAAv are supported by n PSMs\n",
    "\n",
    "comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: Counter(x[\"psm_support_saav\"])[True]).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4df1d880-ebf0-4167-a28e-f9e939df38df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "isoSAAV percent=1: 35\n",
      "isoSAAV >1: 34\n",
      "n_tot=1 , isoSAAV=1 : 20\n",
      "isoSAAV percent<=0.5 : 21\n",
      "isoSAAV percent>0.5 , n_tot>3 : 14\n",
      "isoSAAV percent=1, n_tot>3) : 6\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1044325/239602778.py:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.\n",
      "  a=pd.DataFrame(comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: (Counter(x[\"psm_support_saav\"])[True]/len(x),Counter(x[\"psm_support_saav\"])[True],len(x))))\n"
     ]
    }
   ],
   "source": [
    "#how many peptides with isoSAAv are supported by % PSMs, n, total n\n",
    "\n",
    "a=pd.DataFrame(comb_clun_isosub_true.groupby(\"database_peptide\").apply(lambda x: (Counter(x[\"psm_support_saav\"])[True]/len(x),Counter(x[\"psm_support_saav\"])[True],len(x))))\n",
    "a[\"psm_percent_isoSAAV\"]=a[0].apply(lambda x: x[0])\n",
    "a[\"psm_count_isoSAAV\"]=a[0].apply(lambda x: x[1])\n",
    "a[\"psm_tot\"]=a[0].apply(lambda x: x[2])\n",
    "a.drop(0,inplace=True,axis=1)\n",
    "#plt.scatter(a[\"psm_tot\"],a[\"psm_count_isoSAAV\"]);\n",
    "\n",
    "print(\"isoSAAV percent=1:\",len(a[(a.psm_percent_isoSAAV==1)]))\n",
    "print(\"isoSAAV >1:\",len(a[(a.psm_count_isoSAAV>1)]))\n",
    "\n",
    "\n",
    "#how many n_tot=1 , isoSAAV=1\n",
    "\n",
    "print(\"n_tot=1 , isoSAAV=1 :\",len(a[(a.psm_count_isoSAAV==1)&(a.psm_tot==1)]))\n",
    "\n",
    "#how mnay percent<=0.5\n",
    "\n",
    "print(\"isoSAAV percent<=0.5 :\",len(a[(a.psm_percent_isoSAAV<=0.5)]))\n",
    "\n",
    "#how many n_tot>3 and percent >0.5\n",
    "\n",
    "print(\"isoSAAV percent>0.5 , n_tot>3 :\",len(a[(a.psm_percent_isoSAAV>0.5)&(a.psm_tot>3)]))\n",
    "\n",
    "#how many percent==1 and n_tot>3\n",
    "\n",
    "print(\"isoSAAV percent=1, n_tot>3) :\",len(a[(a.psm_tot>3)&(a.psm_percent_isoSAAV==1)]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "78feb668-29ea-4188-8ba7-f4f11797c5df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "np.int64(223)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a.psm_count_isoSAAV.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "51e4185d-9d08-42a0-8600-131f8e6614f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dataset\n",
       "PXD005833.v0.11.4    101\n",
       "PXD014258.v0.11.4     85\n",
       "PXD002057.v0.11.4     37\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comb_clun_isosub_true[comb_clun_isosub_true[\"psm_support_saav\"]].dataset.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29aa6113-1c56-49f4-8af6-3bd757e15e5d",
   "metadata": {},
   "source": [
    "#### explore peptides with 100% isoSAAv PSMs and >3 total PSMs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "f7ba1400-6798-4142-9046-340475517af9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>psm_percent_isoSAAV</th>\n",
       "      <th>psm_count_isoSAAV</th>\n",
       "      <th>psm_tot</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>database_peptide</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALMLQGIDLLADAVALTMGPKGR</th>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FKNPNAPK</th>\n",
       "      <td>1.0</td>\n",
       "      <td>12</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NQVAMNPTNMVFDAK</th>\n",
       "      <td>1.0</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>QGQETAVAPSLVAPALNTPK</th>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGFGMIYNSLDYAK</th>\n",
       "      <td>1.0</td>\n",
       "      <td>18</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VYNVTQHSVGIVVNK</th>\n",
       "      <td>1.0</td>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         psm_percent_isoSAAV  psm_count_isoSAAV  psm_tot\n",
       "database_peptide                                                        \n",
       "ALMLQGIDLLADAVALTMGPKGR                  1.0                  5        5\n",
       "FKNPNAPK                                 1.0                 12       12\n",
       "NQVAMNPTNMVFDAK                          1.0                  7        7\n",
       "QGQETAVAPSLVAPALNTPK                     1.0                  4        4\n",
       "TTGFGMIYNSLDYAK                          1.0                 18       18\n",
       "VYNVTQHSVGIVVNK                          1.0                  8        8"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#detailed\n",
    "\n",
    "a[(a.psm_tot>3)&(a.psm_percent_isoSAAV==1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "d2e11e38-08f2-4caf-818f-3a92af877a64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>database_peptide</th>\n",
       "      <th>leadprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58776</th>\n",
       "      <td>ALMLQGIDLLADAVALTMGPKGR</td>\n",
       "      <td>IP_761691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>204144</th>\n",
       "      <td>FKNPNAPK</td>\n",
       "      <td>IP_665914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99379</th>\n",
       "      <td>NQVAMNPTNMVFDAK</td>\n",
       "      <td>IP_763499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24247</th>\n",
       "      <td>QGQETAVAPSLVAPALNTPK</td>\n",
       "      <td>IP_569633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2804</th>\n",
       "      <td>TTGFGMIYNSLDYAK</td>\n",
       "      <td>IP_637987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24765</th>\n",
       "      <td>VYNVTQHSVGIVVNK</td>\n",
       "      <td>IP_765292</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               database_peptide   leadprot\n",
       "58776   ALMLQGIDLLADAVALTMGPKGR  IP_761691\n",
       "204144                 FKNPNAPK  IP_665914\n",
       "99379           NQVAMNPTNMVFDAK  IP_763499\n",
       "24247      QGQETAVAPSLVAPALNTPK  IP_569633\n",
       "2804            TTGFGMIYNSLDYAK  IP_637987\n",
       "24765           VYNVTQHSVGIVVNK  IP_765292"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#which proteins?\n",
    "\n",
    "condition=a[(a.psm_tot>3)&(a.psm_percent_isoSAAV==1)].index\n",
    "comb_clun_isosub_true[comb_clun_isosub_true.database_peptide.isin(condition)].drop_duplicates(\"database_peptide\")[[\"database_peptide\",\"leadprot\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1028,
   "id": "d6fb3f1e-e0f9-4fe5-9794-4031fad478ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>triplet</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>peptide_class_x</th>\n",
       "      <th>database_peptide_x</th>\n",
       "      <th>peptides_w1sub</th>\n",
       "      <th>leadprot_x</th>\n",
       "      <th>ID</th>\n",
       "      <th>AF</th>\n",
       "      <th>isSAAV</th>\n",
       "      <th>peptide_class_y</th>\n",
       "      <th>database_peptide_y</th>\n",
       "      <th>leadprot_y</th>\n",
       "      <th>peptides_w1sub_unique</th>\n",
       "      <th>peptides_w1sub_new</th>\n",
       "      <th>peptides_w1sub_diffprot</th>\n",
       "      <th>peptides_w1sub_newprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6847</th>\n",
       "      <td>12</td>\n",
       "      <td>56511143</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>ALMLQGIDLLADAVALTMGPKGR</td>\n",
       "      <td>ALMLQGIDLLSDAVALTMGPKGR</td>\n",
       "      <td>IP_761691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>True</td>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>ALMLQGIDLLADAVALTMGPKGR</td>\n",
       "      <td>IP_761691</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6858</th>\n",
       "      <td>12</td>\n",
       "      <td>56511155</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>ALMLQGIDLLADAVALTMGPKGR</td>\n",
       "      <td>ALMLQGIDLLADAVSLTMGPKGR</td>\n",
       "      <td>IP_761691</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>True</td>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>ALMLQGIDLLADAVALTMGPKGR</td>\n",
       "      <td>IP_761691</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>NA</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chromosome   triplet REF ALT     peptide_class_x  \\\n",
       "6847         12  56511143   G   T  unique_to_Noncanon   \n",
       "6858         12  56511155   G   T  unique_to_Noncanon   \n",
       "\n",
       "           database_peptide_x           peptides_w1sub leadprot_x   ID  \\\n",
       "6847  ALMLQGIDLLADAVALTMGPKGR  ALMLQGIDLLSDAVALTMGPKGR  IP_761691  NaN   \n",
       "6858  ALMLQGIDLLADAVALTMGPKGR  ALMLQGIDLLADAVSLTMGPKGR  IP_761691  NaN   \n",
       "\n",
       "            AF  isSAAV     peptide_class_y       database_peptide_y  \\\n",
       "6847  0.000007    True  unique_to_Noncanon  ALMLQGIDLLADAVALTMGPKGR   \n",
       "6858  0.000001    True  unique_to_Noncanon  ALMLQGIDLLADAVALTMGPKGR   \n",
       "\n",
       "     leadprot_y peptides_w1sub_unique peptides_w1sub_new  \\\n",
       "6847  IP_761691                  True               True   \n",
       "6858  IP_761691                  True               True   \n",
       "\n",
       "     peptides_w1sub_diffprot peptides_w1sub_newprot  \n",
       "6847                      NA                     NA  \n",
       "6858                      NA                     NA  "
      ]
     },
     "execution_count": 1028,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=issaav_vfaf[(issaav_vfaf.isSAAV==True)&(issaav_vfaf[\"AF\"]!=0)].explode(\"database_peptide_y\")\n",
    "\n",
    "a[(a.database_peptide_y==\"ALMLQGIDLLADAVALTMGPKGR\")] #IP_761691 - HSPD1P4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1030,
   "id": "5b5d4c0b-10b0-466e-bd19-3025aa0d9535",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>triplet</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>peptide_class_x</th>\n",
       "      <th>database_peptide_x</th>\n",
       "      <th>peptides_w1sub</th>\n",
       "      <th>leadprot_x</th>\n",
       "      <th>ID</th>\n",
       "      <th>AF</th>\n",
       "      <th>isSAAV</th>\n",
       "      <th>peptide_class_y</th>\n",
       "      <th>database_peptide_y</th>\n",
       "      <th>leadprot_y</th>\n",
       "      <th>peptides_w1sub_unique</th>\n",
       "      <th>peptides_w1sub_new</th>\n",
       "      <th>peptides_w1sub_diffprot</th>\n",
       "      <th>peptides_w1sub_newprot</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14306</th>\n",
       "      <td>2</td>\n",
       "      <td>94912662</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>TTGFGMIYNSLDYAK</td>\n",
       "      <td>TTGFGMIYDSLDYAK</td>\n",
       "      <td>IP_637987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>True</td>\n",
       "      <td>unique_to_Noncanon</td>\n",
       "      <td>TTGFGMIYNSLDYAK</td>\n",
       "      <td>IP_637987</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>A0A2R8Y8A0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      chromosome   triplet REF ALT     peptide_class_x database_peptide_x  \\\n",
       "14306          2  94912662   A   G  unique_to_Noncanon    TTGFGMIYNSLDYAK   \n",
       "\n",
       "        peptides_w1sub leadprot_x   ID        AF  isSAAV     peptide_class_y  \\\n",
       "14306  TTGFGMIYDSLDYAK  IP_637987  NaN  0.000002    True  unique_to_Noncanon   \n",
       "\n",
       "      database_peptide_y leadprot_y peptides_w1sub_unique peptides_w1sub_new  \\\n",
       "14306    TTGFGMIYNSLDYAK  IP_637987                 False              False   \n",
       "\n",
       "      peptides_w1sub_diffprot peptides_w1sub_newprot  \n",
       "14306                    True             A0A2R8Y8A0  "
      ]
     },
     "execution_count": 1030,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[(a.database_peptide_y==\"TTGFGMIYNSLDYAK\")] #IP_637987 - RPS24P6 -> A0A2R8Y8A0 - RPS24 parental gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "253aa595-2c36-4f14-abf3-f8b95ec17d7c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
